---
title: "07-LuskinNMF_projectToSPE"
author: "Bernie Mulvey"
date: "2025-02-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(RcppML)
library(Seurat)
library(SingleCellExperiment)
library(SpatialExperiment)
library(Matrix) # otherwise rcppml gives error 'no item called "package:Matrix" on the search list'
library(dplyr)
library(ggplot2)
# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

## load luskin preprint Seurat object with all cells, convert to SCE
```{r}
mslc <- readRDS("/Users/bmulvey/Desktop/KM Lab/Other Groups' Published Data/Luskin Bruchas mouse periLC snseq publication RDSes/LC_integrated_allcells.rds")

mslc <- as.SingleCellExperiment(mslc)

## load their cluster lookup table. clusters are 0 indexed in coldata but 1 indexed here, so add 1 to mslc$Ident for matching EXCLUDE the vglut1 cluster, which they confirmed was absent from the target dissection area via ish and/or immuno in the mscript
clusts <- fread("/Users/bmulvey/Desktop/KM Lab/Other Groups' Published Data/Luskin Bruchas mouse periLC snseq publication RDSes/luskin_allcellsRDS_snnres03_cluster_lookuptable_013025.txt")
clusts[,Index:=paste0("x",Index)]
clusts[,lab:=cellType_label]
clusts[neuronSubtype_label!="",lab:=neuronSubtype_label]
clusts[,lab:=gsub(lab,pattern="\\/",replacement="_")]

mslc$ident <- as.numeric(as.character(mslc$ident))+1
mslc$ident <- paste0("x",mslc$ident)
stopifnot(all(clusts$Index %in% mslc$ident))

## append labels
tmpcd <- as.data.table(colData(mslc),keep.rownames=T)
# note that the column stim is actually just the sample.

tmpcd <- merge.data.table(tmpcd,clusts[,.(Index,lab)],by.x="ident",by.y="Index",all.x=T)
tmpcd <- DataFrame(tmpcd,row.names=tmpcd$rn)[colnames(mslc),]
tmpcd$rn <- NULL
## return to sce
colData(mslc) <- tmpcd

## drop vglut1
mslc <- mslc[,mslc$lab!="vGLUT1_neurons"]
```

# cleanup
```{r}
rm(clusts)
rm(tmpcd)
gc(full=T)

setDTthreads(1,restore_after_fork = FALSE)
```

## run NMF using Erik Nelson's final parameter set (100 factors, tol 1e-6, 1k max iters, L1 penalty 0.1, diag = T) https://github.com/LieberInstitute/spatial_hpc/blob/main/snRNAseq_hpc/code/nmf/nmf.R
```{r}
mslc.nmf <-RcppML::nmf(assay(mslc,'logcounts'),
    k=100,
    tol = 1e-06,
    maxit = 1000,
    verbose = T,
    L1 = 0.1,
    seed = 42,
    mask_zeros = FALSE,
    diag = TRUE,
    nonneg = TRUE
)
```

## Verbose output: ##
iter |      tol 
---------------
   1 | 8.86e-01
   2 | 1.13e-01
   3 | 3.79e-02
   4 | 1.97e-02
   5 | 1.21e-02
   6 | 8.89e-03
   7 | 7.40e-03
   8 | 6.49e-03
   9 | 5.78e-03
  10 | 5.46e-03
  11 | 5.40e-03
  12 | 5.11e-03
  13 | 4.81e-03
  14 | 4.66e-03
  15 | 4.68e-03
  16 | 4.72e-03
  17 | 4.46e-03
  18 | 3.81e-03
  19 | 2.86e-03
  20 | 1.97e-03
  21 | 1.34e-03
  22 | 9.64e-04
  23 | 7.68e-04
  24 | 6.46e-04
  25 | 5.80e-04
  26 | 5.40e-04
  27 | 5.16e-04
  28 | 5.13e-04
  29 | 5.14e-04
  30 | 5.18e-04
  31 | 5.12e-04
  32 | 5.17e-04
  33 | 5.15e-04
  34 | 5.03e-04
  35 | 4.61e-04
  36 | 3.50e-04
  37 | 3.02e-04
  38 | 2.59e-04
  39 | 2.29e-04
  40 | 2.05e-04
  41 | 1.81e-04
  42 | 1.62e-04
  43 | 1.41e-04
  44 | 1.20e-04
  45 | 1.05e-04
  46 | 9.68e-05
  47 | 8.64e-05
  48 | 7.94e-05
  49 | 7.63e-05
  50 | 7.66e-05
  51 | 7.94e-05
  52 | 8.09e-05
  53 | 8.50e-05
  54 | 8.82e-05
  55 | 9.41e-05
  56 | 9.69e-05
  57 | 9.46e-05
  58 | 8.55e-05
  59 | 7.63e-05
  60 | 6.73e-05
  61 | 6.02e-05
  62 | 5.13e-05
  63 | 4.34e-05
  64 | 3.93e-05
  65 | 3.79e-05
  66 | 3.83e-05
  67 | 4.01e-05
  68 | 4.20e-05
  69 | 4.45e-05
  70 | 4.73e-05
  71 | 4.49e-05
  72 | 4.25e-05
  73 | 3.92e-05
  74 | 3.61e-05
  75 | 3.51e-05
  76 | 3.48e-05
  77 | 3.32e-05
  78 | 3.17e-05
  79 | 2.98e-05
  80 | 2.92e-05
  81 | 2.84e-05
  82 | 2.77e-05
  83 | 2.75e-05
  84 | 2.73e-05
  85 | 2.73e-05
  86 | 2.62e-05
  87 | 2.42e-05
  88 | 2.34e-05
  89 | 2.26e-05
  90 | 2.15e-05
  91 | 1.95e-05
  92 | 1.87e-05
  93 | 1.76e-05
  94 | 1.69e-05
  95 | 1.63e-05
  96 | 1.57e-05
  97 | 1.52e-05
  98 | 1.49e-05
  99 | 1.45e-05
 100 | 1.41e-05
 101 | 1.37e-05
 102 | 1.34e-05
 103 | 1.28e-05
 104 | 1.24e-05
 105 | 1.15e-05
 106 | 1.08e-05
 107 | 1.02e-05
 108 | 9.66e-06
 109 | 9.20e-06
 110 | 8.64e-06
 111 | 8.09e-06
 112 | 7.59e-06
 113 | 7.23e-06
 114 | 6.89e-06
 115 | 6.62e-06
 116 | 6.40e-06
 117 | 6.25e-06
 118 | 6.13e-06
 119 | 6.17e-06
 120 | 5.91e-06
 121 | 5.97e-06
 122 | 6.07e-06
 123 | 6.38e-06
 124 | 6.73e-06
 125 | 7.19e-06
 126 | 7.75e-06
 127 | 8.38e-06
 128 | 9.00e-06
 129 | 9.73e-06
 130 | 1.06e-05
 131 | 1.15e-05
 132 | 1.24e-05
 133 | 1.33e-05
 134 | 1.40e-05
 135 | 1.47e-05
 136 | 1.52e-05
 137 | 1.56e-05
 138 | 1.57e-05
 139 | 1.56e-05
 140 | 1.51e-05
 141 | 1.43e-05
 142 | 1.33e-05
 143 | 1.19e-05
 144 | 1.07e-05
 145 | 9.33e-06
 146 | 8.04e-06
 147 | 6.87e-06
 148 | 5.82e-06
 149 | 4.94e-06
 150 | 4.28e-06
 151 | 3.73e-06
 152 | 3.37e-06
 153 | 3.11e-06
 154 | 2.86e-06
 155 | 2.72e-06
 156 | 2.67e-06
 157 | 2.65e-06
 158 | 2.56e-06
 159 | 2.53e-06
 160 | 2.50e-06
 161 | 2.48e-06
 162 | 2.49e-06
 163 | 2.50e-06
 164 | 2.48e-06
 165 | 2.44e-06
 166 | 2.44e-06
 167 | 2.44e-06
 168 | 2.42e-06
 169 | 2.38e-06
 170 | 2.38e-06
 171 | 2.36e-06
 172 | 2.36e-06
 173 | 2.39e-06
 174 | 2.44e-06
 175 | 2.51e-06
 176 | 2.57e-06
 177 | 2.65e-06
 178 | 2.72e-06
 179 | 2.79e-06
 180 | 2.84e-06
 181 | 2.87e-06
 182 | 2.92e-06
 183 | 2.96e-06
 184 | 3.01e-06
 185 | 3.06e-06
 186 | 3.11e-06
 187 | 3.18e-06
 188 | 3.23e-06
 189 | 3.31e-06
 190 | 3.37e-06
 191 | 3.34e-06
 192 | 3.21e-06
 193 | 3.28e-06
 194 | 3.38e-06
 195 | 3.33e-06
 196 | 3.32e-06
 197 | 3.36e-06
 198 | 3.37e-06
 199 | 3.36e-06
 200 | 3.44e-06
 201 | 3.56e-06
 202 | 3.77e-06
 203 | 4.02e-06
 204 | 4.34e-06
 205 | 4.73e-06
 206 | 5.17e-06
 207 | 5.57e-06
 208 | 6.08e-06
 209 | 6.72e-06
 210 | 7.54e-06
 211 | 8.57e-06
 212 | 9.92e-06
 213 | 1.16e-05
 214 | 1.35e-05
 215 | 1.57e-05
 216 | 1.85e-05
 217 | 2.17e-05
 218 | 2.47e-05
 219 | 2.62e-05
 220 | 2.73e-05
 221 | 2.95e-05
 222 | 3.22e-05
 223 | 3.55e-05
 224 | 4.00e-05
 225 | 4.60e-05
 226 | 5.24e-05
 227 | 6.16e-05
 228 | 6.57e-05
 229 | 6.57e-05
 230 | 5.88e-05
 231 | 4.89e-05
 232 | 4.04e-05
 233 | 3.27e-05
 234 | 2.37e-05
 235 | 1.74e-05
 236 | 1.32e-05
 237 | 9.98e-06
 238 | 7.47e-06
 239 | 5.73e-06
 240 | 4.56e-06
 241 | 3.77e-06
 242 | 3.24e-06
 243 | 2.62e-06
 244 | 2.17e-06
 245 | 1.99e-06
 246 | 1.85e-06
 247 | 1.69e-06
 248 | 1.56e-06
 249 | 1.48e-06
 250 | 1.43e-06
 251 | 1.41e-06
 252 | 1.41e-06
 253 | 1.40e-06
 254 | 1.38e-06
 255 | 1.38e-06
 256 | 1.39e-06
 257 | 1.41e-06
 258 | 1.42e-06
 259 | 1.44e-06
 260 | 1.47e-06
 261 | 1.47e-06
 262 | 1.46e-06
 263 | 1.50e-06
 264 | 1.50e-06
 265 | 1.51e-06
 266 | 1.52e-06
 267 | 1.53e-06
 268 | 1.54e-06
 269 | 1.50e-06
 270 | 1.51e-06
 271 | 1.52e-06
 272 | 1.53e-06
 273 | 1.52e-06
 274 | 1.51e-06
 275 | 1.48e-06
 276 | 1.49e-06
 277 | 1.50e-06
 278 | 1.47e-06
 279 | 1.44e-06
 280 | 1.42e-06
 281 | 1.40e-06
 282 | 1.37e-06
 283 | 1.34e-06
 284 | 1.30e-06
 285 | 1.25e-06
 286 | 1.21e-06
 287 | 1.19e-06
 288 | 1.15e-06
 289 | 1.12e-06
 290 | 1.09e-06
 291 | 1.06e-06
 292 | 1.02e-06
 293 | 9.94e-07
 
# append cell patterns (h, 100 row x 28046 col matrix) to sce 
```{r}
th <- t(mslc.nmf$h)
colData(mslc)<-cbind(colData(mslc),th)

saveRDS(mslc.nmf,"processed-data/08_validitycheck_25hdg75svg_louv1/07a-LuskinMs2024allCells_nmf100_out.RDS")
saveRDS(mslc,"processed-data/08_validitycheck_25hdg75svg_louv1/07b-LuskinMs2024allCells_sce_w_nmf100.RDS")
```

## now to figure out what factor is what in the source data. see jacqui's code at https://github.com/LieberInstitute/spatial_hpc/blob/main/code/revision/plots_nmf-general-specific.R
```{r}
seed1 = as.matrix(colData(mslc)[,paste0("V",1:100)])
seed1 = seed1>0
d1 = cbind.data.frame(authlabel=as.data.frame(colData(mslc))[,"lab"],
                      seed1) %>% 
  group_by(authlabel) %>% add_tally(name="total") %>%
    group_by(authlabel, total) %>%
  summarise_at(paste0("V",1:100), sum) %>%
  tidyr::pivot_longer(paste0("V",1:100), values_to="n", names_to="nmf") %>%
  mutate(prop=n/total)

seed2 = as.matrix(colData(mslc)[,paste0("V",1:100)])
seed2 = apply(seed2, 2, scale)
d2 = cbind.data.frame(authlabel=as.data.frame(colData(mslc))[,"lab"],
                      seed2) %>% 
  group_by(authlabel) %>%
  summarise_at(paste0("V",1:100), mean) %>% 
  tidyr::pivot_longer(paste0("V",1:100), values_to="scaled.avg", names_to="nmf")

pltdat <- left_join(d1[,c("authlabel","nmf","prop")], 
                   d2[,c("authlabel","nmf","scaled.avg")])

nmford <- c("V12",
            "V47",
            "V34","V36","V37",
            "V32","V33","V40",
            "V25",
            "V35","V46",
            "V22",
            "V48",
            "V43",
            "V30",
            "V29",
            "V41",
            "V49",
            "V28",
            "V31",
            "V17","V23","V38","V42",
            "V11",
            "V14","V15","V16","V18",
            "V57",
            "V95",
            "V100",
            "V24","V27","V39")

authlabelord <- rev(c("DBH_neurons",
                    "Chat_neurons",
                    "vGAT_neurons_1",
                    "vGAT_neurons_2",
                    "vGAT_neurons_3",
                    "vGAT_neurons_4",
                    "vGAT_neurons_5",
                    "vGAT_neurons_6",
                    "vGLUT2_neurons_1",
                    "vGLUT2_neurons_2",
                    "vGLUT2_neurons_3",
                    "vGLUT2_neurons_4",
                    "vGLUT2_neurons_5",
                    "vGAT_vGLUT2_neurons",
                    "Microglia",
                    "Astrocyte",
                    "OPC1",
                    "Oligo",
                    "Ependymal",
                    "Mural",
                    "Ambiguous2",
                    "Epithelial",
                    "Ambiguous1"))

## make sure no repeats
stopifnot(length(nmford)!=length(unique(nmford)))

nmfmisc <- paste0("V",c(1:100))
nmford <- c(nmford,nmfmisc[!(nmfmisc %in% nmford)])


## make sure all auth labels accounted for
stopifnot(all(unique(mslc$lab) %in% authlabelord))

# data table-ize for ease of use -- dplyr code is jacqui's almost verbatim
pltdat <- as.data.table(pltdat)

pltdat[,authlabel:=factor(authlabel,levels=authlabelord)]
pltdat[,nmf:=factor(nmf,levels=nmford)]


## save plot
png("plots/08_validitycheck_25hdg75svg_louv1/07a-msLC_NMF100_scaledAvgs_and_props_by_msCluster.png",height=1000,width=1600)
ggplot(pltdat, aes(x=nmf, y=authlabel, size=prop, color=scaled.avg))+
  geom_count()+
  theme_bw()+
  scale_size(range=c(0,3))+
  scale_color_viridis_c(option="F", direction=-1)+
  labs(x="NMF Factor",y="Luskin Bruchas scSeq Label")+
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=.5,size=10),
        axis.text.y=element_text(size=10),
        axis.title.x=element_text(size=12),
        axis.title.y=element_text(size=12))
dev.off()
```

## clean up
```{r}
rm(d1,d2,seed1,seed2,authlabelord,nmfmisc,pltdat)
gc(full=T)
```


##############################################################
###### PART 2: Project mouse single cell into LFF SPE ########
##############################################################

## load SPE and assign hdg_svg_2575_louv_res1 clusterings as autoannotated
```{r}
lc3 <- readRDS("processed-data/06-countsOnly_noImgData_QCed_SPE_split_to_tissSections.RDS")

louvs <- readRDS("processed-data/07_featureSelection_dimred_harmony_clustering/04-Initial_louvLeid_res0.5-1-2_clusterings.RDS")
louvs <- louvs[["HARMONYlmbna_HDG_SVG_2575"]][,.(rn,snnHARMONYlmbna_HDG_SVG_2575_louv_res1)]
setnames(louvs,2,"clusid")

autoan <- readRDS("processed-data/07_featureSelection_dimred_harmony_clustering/05b-init_clusterings_autoannoLCrapheAndGABAergic.RDS")
autoan <- autoan[["snnHARMONYlmbna_HDG_SVG_2575_louv_res1"]]

louvs <- merge.data.table(louvs,autoan,by="clusid")
louvs <- DataFrame(louvs,row.names=louvs$rn)[colnames(lc3),]

colLabels(lc3) <- louvs$autoanno

## calculate log counts (needed for nmf projection)
lc3 <- scater::computeLibraryFactors(lc3)
lc3 <- scater::logNormCounts(lc3)

rm(louvs,autoan)
```


## get everything matched down to 1:1 orthologs between mouse and human. note that, contrary to erik's code cited below, there are NOT rownames on the NMF output w matrix, so we need to keep the pre-filtered rownames of the sce to know what the rows of the w matrix correspond to.
```{r}
nmf.w.rn <- rownames(mslc)

# library(orthogene)
# 
# msg <- as.data.frame(rownames(mslc))
# rownames(msg) <- msg$V1
# gprofiler web service was down--again-- error 301 on queries both from my machine and using JHPCE. manually queried via https://biit.cs.ut.ee/gprofiler_beta/orth
mhorth <- fread("~/Desktop/KM Lab/Other Groups\' Published Data/Luskin Bruchas mouse periLC snseq publication RDSes/gProfiler_mmusculus_hsapiens_1-28-2025_4-42-13 PM.csv",header=T)
## drop multimatches
dropg <- mhorth[,.N,by="initial_alias"][N>1,initial_alias]
mhorth <- mhorth[!(initial_alias %in% dropg)]

dropg <- mhorth[,.N,by="ortholog_ensg"][N>1,ortholog_ensg]
mhorth <- mhorth[!(ortholog_ensg %in% dropg)]

## mouse genes come back all caps there so caps the mslc rownames (and w mat rn vector)
rownames(mslc) <- toupper(rownames(mslc))
nmf.w.rn <- toupper(nmf.w.rn)

## subset to genes whose all caps mouse symbol is in mslc and ensg is in lc3
mhorth <- mhorth[initial_alias %in% rownames(mslc) & ortholog_ensg %in% rownames(lc3)]

## subset SCE, SPE, w rowname vector, and w accordingly
nmf.w.keeprows <- which(nmf.w.rn %in% mhorth$initial_alias)
orthw <- mslc.nmf$w[nmf.w.keeprows,]

mslc <- mslc[mhorth$initial_alias,]
lc3 <- lc3[mhorth$ortholog_ensg,]

stopifnot(nrow(orthw)==nrow(mslc))
## swap mouse LC rownames to be the HUMAN ensg ortholog. since we just reordered both objects by paired mouse-human names, we can pass ortholog_ensg directly and have the appropriate ortholog mappings.
# rownames(mslc) <- mhorth$ortholog_ensg

## clean up
rm(tmprn,dropg,nmf.w.keeprows,nmf.w.rn)
gc(full=T)
```

## use rcppml::project (pretty straightforward but erik's script for posterity and the L2 parameter used). Actually, there is no L2 parameter argument here so idk wtf. also, the argument order has changed for project relative to this code, i think. nonetheless:
https://github.com/LieberInstitute/spatial_hpc/blob/main/code/NMF/project_nmf_patterns.R

Note that Erik's code implies there are rownames on the nmf w matrix, which there aren't. so we also need to re-load the full mouse SPE to determine which genes were there to start with (i.e. are the rows of the w matrix) and which remained after ortholog filtering above.
```{r}
set.seed(42)
msprojinto.lff <- RcppML::project(A=logcounts(lc3),w=orthw)#,L2=0.00005)
msprojinto.lff <- t(msprojinto.lff)

stopifnot(nrow(msprojinto.lff)==ncol(lc3))
rownames(msprojinto.lff) <- colnames(lc3)
colnames(msprojinto.lff) <- paste0("V",c(1:100))
saveRDS(msprojinto.lff,"processed-data/08_validitycheck_25hdg75svg_louv1/07c-LuskinMs2024allCells_nmf100_projected_into_LFF_SPE_mat.RDS")

## append loadings to SPE, only for those factors that were cell type specific in the mouse data, and in the same order for cross-data labeling convenience
colData(lc3) <- cbind(colData(lc3),msprojinto.lff[,nmford])

## make a ms factor- ms cluster name for readability
nmfspeclab <- c("DBH_neurons",
                    "Chat_neurons",
                    rep("vGAT_neurons_1",3),
                    rep("vGAT_neurons_2",3),
                    "vGAT_neurons_3",
                    rep("vGAT_neurons_4",2),
                    "vGAT_neurons_5",
                    "vGAT_neurons_6",
                    "vGLUT2_neurons_1",
                    "vGLUT2_neurons_2",
                    "vGLUT2_neurons_3",
                    "vGLUT2_neurons_4",
                    "vGLUT2_neurons_5",
                    "vGAT_vGLUT2_neurons",
                    "Microglia",
                    rep("Astrocyte",4),
                    "OPC1",
                    rep("Oligo",4),
                    "Ependymal",
                    "Mural",
                    "Epithelial",
                    rep("Ambiguous1",3))

stopifnot(length(nmfspeclab)==length(nmford))
usefulname <- paste0(nmford,"_",nmfspeclab)

colnames(colData(lc3))[c(40:ncol(colData(lc3)))] <- usefulname
```

## back to jacqui's plotting code linked above: https://github.com/LieberInstitute/spatial_hpc/blob/main/code/revision/plots_nmf-general-specific.R 
```{r}
seed3 = as.matrix(colData(lc3)[,usefulname])
seed3 = seed3>0
d3 = cbind.data.frame(LFFclust=as.data.frame(colData(lc3))[,"label"],
                      seed3) %>% 
  group_by(LFFclust) %>% add_tally(name="total") %>%
  group_by(LFFclust, total) %>%
  summarise_at(usefulname, sum) %>%
  tidyr::pivot_longer(usefulname, values_to="n", names_to="nmf") %>%
  mutate(prop=n/total)

seed4 = as.matrix(colData(lc3)[,usefulname])
seed4 = apply(seed4, 2, scale)
d4 = cbind.data.frame(LFFclust=as.data.frame(colData(lc3))[,"label"],
                      seed4) %>% 
  group_by(LFFclust) %>%
  summarise_at(usefulname, mean) %>% 
  tidyr::pivot_longer(usefulname, values_to="scaled.avg", names_to="nmf")


pltdat <- left_join(d3[,c("LFFclust","nmf","prop")], 
                    d4[,c("LFFclust","nmf","scaled.avg")])

# to d.t.
pltdat <- as.data.table(pltdat)
pltdat[,nmf:=factor(nmf,usefulname)]

pltdat[,LFFclust:=factor(LFFclust,rev(c("LC.1","5HT.1","5HT.2","X4","X7","X1","X2","X8","X9")))]

png("plots/08_validitycheck_25hdg75svg_louv1/07b-LuskinMs2024allCells_cellspecNMFs_into_LFFSPE_scaledAvgs_and_props_by_LFFCluster.png",height=750,width=1000)
ggplot(pltdat, aes(x=nmf, y=LFFclust, size=prop, color=scaled.avg))+
  geom_count()+
  theme_bw()+
  scale_size(range=c(0,3))+
  scale_color_viridis_c(option="F", direction=-1)+
  labs(x="Mouse LC NMF Factor and Corresp Cluster",y="LFF Cluster")+
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=.5,size=10),
        axis.text.y=element_text(size=10),
        axis.title.x=element_text(size=12),
        axis.title.y=element_text(size=12))
dev.off()
```

## well THAT's disappointing.

reprod inf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] colorout_1.3-0.2            ggplot2_3.5.1              
 [3] dplyr_1.1.4                 Matrix_1.7-1               
 [5] SpatialExperiment_1.16.0    SingleCellExperiment_1.28.1
 [7] SummarizedExperiment_1.36.0 Biobase_2.66.0             
 [9] GenomicRanges_1.58.0        GenomeInfoDb_1.42.1        
[11] IRanges_2.40.1              S4Vectors_0.44.0           
[13] BiocGenerics_0.52.0         MatrixGenerics_1.18.0      
[15] matrixStats_1.5.0           Seurat_5.2.1               
[17] SeuratObject_5.0.2          sp_2.1-4                   
[19] RcppML_0.3.7                data.table_1.16.4          
[21] rlang_1.1.4                

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3      rstudioapi_0.17.1       jsonlite_1.8.9         
  [4] magrittr_2.0.3          ggbeeswarm_0.7.2        magick_2.8.5           
  [7] spatstat.utils_3.1-2    farver_2.1.2            rmarkdown_2.29         
 [10] zlibbioc_1.52.0         vctrs_0.6.5             ROCR_1.0-11            
 [13] spatstat.explore_3.3-4  S4Arrays_1.6.0          htmltools_0.5.8.1      
 [16] BiocNeighbors_2.0.1     SparseArray_1.6.0       sctransform_0.4.1      
 [19] parallelly_1.42.0       KernSmooth_2.23-24      htmlwidgets_1.6.4      
 [22] ica_1.0-3               plyr_1.8.9              plotly_4.10.4          
 [25] zoo_1.8-12              igraph_2.1.4            mime_0.12              
 [28] lifecycle_1.0.4         pkgconfig_2.0.3         rsvd_1.0.5             
 [31] R6_2.5.1                fastmap_1.2.0           GenomeInfoDbData_1.2.13
 [34] fitdistrplus_1.2-2      future_1.34.0           shiny_1.10.0           
 [37] digest_0.6.37           colorspace_2.1-1        patchwork_1.3.0        
 [40] scater_1.34.0           rprojroot_2.0.4         tensor_1.5             
 [43] RSpectra_0.16-2         irlba_2.3.5.1           beachmat_2.22.0        
 [46] labeling_0.4.3          progressr_0.15.1        spatstat.sparse_3.1-0  
 [49] httr_1.4.7              polyclip_1.10-7         abind_1.4-8            
 [52] compiler_4.4.2          here_1.0.1              withr_3.0.2            
 [55] BiocParallel_1.40.0     viridis_0.6.5           fastDummies_1.7.5      
 [58] MASS_7.3-61             DelayedArray_0.32.0     rjson_0.2.23           
 [61] tools_4.4.2             vipor_0.4.7             lmtest_0.9-40          
 [64] beeswarm_0.4.0          httpuv_1.6.15           future.apply_1.11.3    
 [67] goftest_1.2-3           glue_1.8.0              nlme_3.1-166           
 [70] promises_1.3.2          grid_4.4.2              Rtsne_0.17             
 [73] cluster_2.1.6           reshape2_1.4.4          generics_0.1.3         
 [76] gtable_0.3.6            spatstat.data_3.1-4     tidyr_1.3.1            
 [79] ScaledMatrix_1.14.0     BiocSingular_1.22.0     XVector_0.46.0         
 [82] spatstat.geom_3.3-5     RcppAnnoy_0.0.22        ggrepel_0.9.6          
 [85] RANN_2.6.2              pillar_1.10.1           stringr_1.5.1          
 [88] spam_2.11-1             RcppHNSW_0.6.0          later_1.4.1            
 [91] splines_4.4.2           lattice_0.22-6          survival_3.7-0         
 [94] deldir_2.0-4            tidyselect_1.2.1        scuttle_1.16.0         
 [97] miniUI_0.1.1.1          pbapply_1.7-2           knitr_1.49             
[100] gridExtra_2.3           scattermore_1.2         xfun_0.50              
[103] stringi_1.8.4           UCSC.utils_1.2.0        lazyeval_0.2.2         
[106] yaml_2.3.10             evaluate_1.0.3          codetools_0.2-20       
[109] tibble_3.2.1            cli_3.6.3               uwot_0.2.2             
[112] xtable_1.8-4            reticulate_1.40.0       munsell_0.5.1          
[115] Rcpp_1.0.14             globals_0.16.3          spatstat.random_3.3-2  
[118] png_0.1-8               spatstat.univar_3.1-1   parallel_4.4.2         
[121] dotCall64_1.2           listenv_0.9.1           viridisLite_0.4.2      
[124] scales_1.3.0            ggridges_0.5.6          crayon_1.5.3           
[127] purrr_1.0.2             cowplot_1.1.3          
> sessioninfo::session_info()
─ Session info ─────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.2 (2024-10-31)
 os       macOS Sonoma 14.7.1
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-02-23
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ─────────────────────────────────────────────────────────────────────────
 !  package              * version date (UTC) lib source
    abind                  1.4-8   2024-09-12 [1] CRAN (R 4.4.1)
    beachmat               2.22.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    beeswarm               0.4.0   2021-06-01 [1] CRAN (R 4.4.0)
    Biobase              * 2.66.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocNeighbors          2.0.1   2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
    BiocParallel           1.40.0  2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    BiocSingular           1.22.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
 P  cli                    3.6.3   2024-06-21 [2] CRAN (R 4.4.0)
    cluster                2.1.6   2023-12-01 [1] CRAN (R 4.4.2)
    codetools              0.2-20  2024-03-31 [1] CRAN (R 4.4.2)
    colorout             * 1.3-0.2 2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    colorspace             2.1-1   2024-07-26 [1] CRAN (R 4.4.0)
    cowplot                1.1.3   2024-01-22 [1] CRAN (R 4.4.0)
    crayon                 1.5.3   2024-06-20 [1] CRAN (R 4.4.0)
    data.table           * 1.16.4  2024-12-06 [1] CRAN (R 4.4.1)
    DelayedArray           0.32.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    deldir                 2.0-4   2024-02-28 [1] CRAN (R 4.4.0)
    digest                 0.6.37  2024-08-19 [1] CRAN (R 4.4.1)
    dotCall64              1.2     2024-10-04 [1] CRAN (R 4.4.1)
    dplyr                * 1.1.4   2023-11-17 [1] CRAN (R 4.4.0)
    evaluate               1.0.3   2025-01-10 [1] CRAN (R 4.4.1)
    farver                 2.1.2   2024-05-13 [1] CRAN (R 4.4.0)
    fastDummies            1.7.5   2025-01-20 [1] CRAN (R 4.4.1)
    fastmap                1.2.0   2024-05-15 [1] CRAN (R 4.4.0)
    fitdistrplus           1.2-2   2025-01-07 [1] CRAN (R 4.4.1)
    future                 1.34.0  2024-07-29 [1] CRAN (R 4.4.0)
    future.apply           1.11.3  2024-10-27 [1] CRAN (R 4.4.1)
    generics               0.1.3   2022-07-05 [1] CRAN (R 4.4.0)
    GenomeInfoDb         * 1.42.1  2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13  2024-12-12 [1] Bioconductor
    GenomicRanges        * 1.58.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ggbeeswarm             0.7.2   2023-04-29 [1] CRAN (R 4.4.0)
    ggplot2              * 3.5.1   2024-04-23 [1] CRAN (R 4.4.0)
    ggrepel                0.9.6   2024-09-07 [1] CRAN (R 4.4.1)
    ggridges               0.5.6   2024-01-23 [1] CRAN (R 4.4.0)
    globals                0.16.3  2024-03-08 [1] CRAN (R 4.4.0)
    glue                   1.8.0   2024-09-30 [1] CRAN (R 4.4.1)
    goftest                1.2-3   2021-10-07 [1] CRAN (R 4.4.0)
    gridExtra              2.3     2017-09-09 [1] CRAN (R 4.4.0)
    gtable                 0.3.6   2024-10-25 [1] CRAN (R 4.4.1)
    here                   1.0.1   2020-12-13 [1] CRAN (R 4.4.0)
    htmltools              0.5.8.1 2024-04-04 [1] CRAN (R 4.4.0)
    htmlwidgets            1.6.4   2023-12-06 [1] CRAN (R 4.4.0)
    httpuv                 1.6.15  2024-03-26 [1] CRAN (R 4.4.0)
    httr                   1.4.7   2023-08-15 [1] CRAN (R 4.4.0)
    ica                    1.0-3   2022-07-08 [1] CRAN (R 4.4.0)
    igraph                 2.1.4   2025-01-23 [1] CRAN (R 4.4.1)
    IRanges              * 2.40.1  2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    irlba                  2.3.5.1 2022-10-03 [1] CRAN (R 4.4.0)
    jsonlite               1.8.9   2024-09-20 [1] CRAN (R 4.4.1)
    KernSmooth             2.23-24 2024-05-17 [1] CRAN (R 4.4.2)
    knitr                  1.49    2024-11-08 [1] CRAN (R 4.4.1)
    labeling               0.4.3   2023-08-29 [1] CRAN (R 4.4.0)
    later                  1.4.1   2024-11-27 [1] CRAN (R 4.4.1)
    lattice                0.22-6  2024-03-20 [1] CRAN (R 4.4.2)
    lazyeval               0.2.2   2019-03-15 [1] CRAN (R 4.4.0)
    lifecycle              1.0.4   2023-11-07 [1] CRAN (R 4.4.0)
    listenv                0.9.1   2024-01-29 [1] CRAN (R 4.4.0)
    lmtest                 0.9-40  2022-03-21 [1] CRAN (R 4.4.0)
    magick                 2.8.5   2024-09-20 [1] CRAN (R 4.4.1)
    magrittr               2.0.3   2022-03-30 [1] CRAN (R 4.4.0)
    MASS                   7.3-61  2024-06-13 [1] CRAN (R 4.4.2)
    Matrix               * 1.7-1   2024-10-18 [1] CRAN (R 4.4.2)
    MatrixGenerics       * 1.18.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    matrixStats          * 1.5.0   2025-01-07 [1] CRAN (R 4.4.1)
    mime                   0.12    2021-09-28 [1] CRAN (R 4.4.0)
    miniUI                 0.1.1.1 2018-05-18 [1] CRAN (R 4.4.0)
    munsell                0.5.1   2024-04-01 [1] CRAN (R 4.4.0)
    nlme                   3.1-166 2024-08-14 [1] CRAN (R 4.4.2)
    parallelly             1.42.0  2025-01-30 [1] CRAN (R 4.4.1)
    patchwork              1.3.0   2024-09-16 [1] CRAN (R 4.4.1)
    pbapply                1.7-2   2023-06-27 [1] CRAN (R 4.4.0)
    pillar                 1.10.1  2025-01-07 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3   2019-09-22 [1] CRAN (R 4.4.0)
    plotly                 4.10.4  2024-01-13 [1] CRAN (R 4.4.0)
    plyr                   1.8.9   2023-10-02 [1] CRAN (R 4.4.0)
    png                    0.1-8   2022-11-29 [1] CRAN (R 4.4.0)
    polyclip               1.10-7  2024-07-23 [1] CRAN (R 4.4.0)
    progressr              0.15.1  2024-11-22 [1] CRAN (R 4.4.1)
    promises               1.3.2   2024-11-28 [1] CRAN (R 4.4.1)
    purrr                  1.0.2   2023-08-10 [1] CRAN (R 4.4.0)
    R6                     2.5.1   2021-08-19 [1] CRAN (R 4.4.0)
    RANN                   2.6.2   2024-08-25 [1] CRAN (R 4.4.1)
    RColorBrewer           1.1-3   2022-04-03 [1] CRAN (R 4.4.0)
    Rcpp                   1.0.14  2025-01-12 [1] CRAN (R 4.4.1)
    RcppAnnoy              0.0.22  2024-01-23 [1] CRAN (R 4.4.0)
    RcppHNSW               0.6.0   2024-02-04 [1] CRAN (R 4.4.0)
    RcppML               * 0.3.7   2021-09-21 [1] CRAN (R 4.4.0)
    reshape2               1.4.4   2020-04-09 [1] CRAN (R 4.4.0)
    reticulate             1.40.0  2024-11-15 [1] CRAN (R 4.4.1)
    rjson                  0.2.23  2024-09-16 [1] CRAN (R 4.4.1)
 VP rlang                * 1.1.4   2025-01-17 [2] CRAN (R 4.4.1) (on disk 1.1.5)
    rmarkdown              2.29    2024-11-04 [1] CRAN (R 4.4.1)
    ROCR                   1.0-11  2020-05-02 [1] CRAN (R 4.4.0)
    rprojroot              2.0.4   2023-11-05 [1] CRAN (R 4.4.0)
    RSpectra               0.16-2  2024-07-18 [1] CRAN (R 4.4.0)
    rstudioapi             0.17.1  2024-10-22 [1] CRAN (R 4.4.1)
    rsvd                   1.0.5   2021-04-16 [1] CRAN (R 4.4.0)
    Rtsne                  0.17    2023-12-07 [1] CRAN (R 4.4.0)
    S4Arrays               1.6.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ScaledMatrix           1.14.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.3.0   2023-11-28 [1] CRAN (R 4.4.0)
    scater                 1.34.0  2024-10-29 [1] Bioconductor 3.20 (R 4.4.1)
    scattermore            1.2     2023-06-12 [1] CRAN (R 4.4.0)
    sctransform            0.4.1   2023-10-19 [1] CRAN (R 4.4.0)
    scuttle                1.16.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    sessioninfo            1.2.2   2021-12-06 [1] CRAN (R 4.4.0)
    Seurat               * 5.2.1   2025-01-24 [1] CRAN (R 4.4.1)
    SeuratObject         * 5.0.2   2024-05-08 [1] CRAN (R 4.4.0)
    shiny                  1.10.0  2024-12-14 [1] CRAN (R 4.4.1)
    SingleCellExperiment * 1.28.1  2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    sp                   * 2.1-4   2024-04-30 [1] CRAN (R 4.4.0)
    spam                   2.11-1  2025-01-20 [1] CRAN (R 4.4.1)
    SparseArray            1.6.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SpatialExperiment    * 1.16.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    spatstat.data          3.1-4   2024-11-15 [1] CRAN (R 4.4.1)
    spatstat.explore       3.3-4   2025-01-08 [1] CRAN (R 4.4.1)
    spatstat.geom          3.3-5   2025-01-18 [1] CRAN (R 4.4.1)
    spatstat.random        3.3-2   2024-09-18 [1] CRAN (R 4.4.1)
    spatstat.sparse        3.1-0   2024-06-21 [1] CRAN (R 4.4.0)
    spatstat.univar        3.1-1   2024-11-05 [1] CRAN (R 4.4.1)
    spatstat.utils         3.1-2   2025-01-08 [1] CRAN (R 4.4.1)
    stringi                1.8.4   2024-05-06 [1] CRAN (R 4.4.0)
    stringr                1.5.1   2023-11-14 [1] CRAN (R 4.4.0)
    SummarizedExperiment * 1.36.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    survival               3.7-0   2024-06-05 [1] CRAN (R 4.4.2)
    tensor                 1.5     2012-05-05 [1] CRAN (R 4.4.0)
    tibble                 3.2.1   2023-03-20 [1] CRAN (R 4.4.0)
    tidyr                  1.3.1   2024-01-24 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1   2024-03-11 [1] CRAN (R 4.4.0)
    UCSC.utils             1.2.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    uwot                   0.2.2   2024-04-21 [1] CRAN (R 4.4.0)
    vctrs                  0.6.5   2023-12-01 [1] CRAN (R 4.4.0)
    vipor                  0.4.7   2023-12-18 [1] CRAN (R 4.4.0)
    viridis                0.6.5   2024-01-29 [1] CRAN (R 4.4.0)
    viridisLite            0.4.2   2023-05-02 [1] CRAN (R 4.4.0)
    withr                  3.0.2   2024-10-28 [1] CRAN (R 4.4.1)
    xfun                   0.50    2025-01-07 [1] CRAN (R 4.4.1)
    xtable                 1.8-4   2019-04-21 [1] CRAN (R 4.4.0)
    XVector                0.46.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    yaml                   2.3.10  2024-07-26 [1] CRAN (R 4.4.0)
    zlibbioc               1.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    zoo                    1.8-12  2023-04-13 [1] CRAN (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

────────────────────────────────────────────────────────────────────────────────────
