---
title: "02b-Allclust_dropmixclusts_SexDEG_EDA"
author: "Bernie Mulvey"
date: "2025-05-24"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(ggplot2)
library(ggtext)
library(gridExtra)
library(SpatialExperiment)
library(escheR)
library(spatialLIBD)
library(scater)
library(ggrastr)
# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")
library(parallel)


## ggplot defaults
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 9), axis.title.x = element_text(size = 8), axis.text.y = element_text(size = 8), axis.title.y = element_text(size =9), plot.title = element_markdown(size = 11,hjust=0.5), strip.text = element_text(size=11), legend.text = element_text(size=8), legend.title = element_text(size=9,hjust=0.5),plot.title.position="plot"))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```


#####################
QUESTION 1: What's the deal with astrocyte genes being among the LC DEGs?
Hypothesis: Either due to LC-proximal astrocytes performing unique and sex-differential functions, OR due to differences in LC-local astrocyte pop. size 
#####################
```{r}
detabs <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/02b-voomLmFit_objs_and_DE_tables_LabSex_LabE2vE4_LabSingleGenos_LabAnces_LabAncesE4E2_noMixClusts_YRIcontAndCat_JUL2025.RDS")

detabs <- detabs$cont_ances$sexlabel$detabs
lapply(detabs,function(x){nrow(x[adj.P.Val<0.05,])})
# $ACHE_SERT_LC_WM_Sex 1493
# $ACHE_SERT_Npep_Sex 523
# $Astro_Sex 874
# $LC_Sex 1199
# $Oligo_Astro_Sex 846

lcde <- detabs$LC_Sex
astrode <- detabs$Astro_Sex

lc_not_astro <- lcde[adj.P.Val<0.05 & 
                         !(gene_id %in% astrode[adj.P.Val<0.05,gene_id])
                     ]
## 664 genes
```
# what about shared genes
```{r}
sum(lcde[adj.P.Val<0.05,gene_id] %in% astrode[adj.P.Val<0.05,gene_id])
## 535

lc.and.ast <- lcde[adj.P.Val<0.05 & gene_id %in% astrode[adj.P.Val<0.05,gene_id]]
lc.and.ast <- lc.and.ast[!(chr %in% paste0("chr",c("X","Y","MT"))),.(gene_id,gene_name,adj.P.Val,logFC)]
setnames(lc.and.ast,c(3:4),paste0("LCsex_",names(lc.and.ast)[3:4]))
lc.and.ast <- merge.data.table(lc.and.ast,astrode[adj.P.Val<0.05,.(gene_id,logFC,adj.P.Val)],by="gene_id")

setnames(lc.and.ast,c("adj.P.Val","logFC"),paste0("Astrosex_",c("adj.P.Val","logFC")))

ggplot(lc.and.ast,aes(x=LCsex_logFC,y=Astrosex_logFC))+
    geom_point()
# looks like the only divergences of any note at all are at abs(logFC)>1
dev.off()
lc.and.ast <- lc.and.ast[abs(LCsex_logFC)>1|abs(Astrosex_logFC)>1]

ggplot(lc.and.ast,aes(x=LCsex_logFC,y=Astrosex_logFC))+
    geom_point()
# yup
dev.off()
```

Fed the ensembl IDs and gene symbols to gemini pro 2.5 with the following prompt:
"Which of the following genes are considered specific to or enriched in adult astrocytes relative to other adult brain cell types? Consult the attached table for a list of genes for which the question. Each row contains two identifiers for one gene (its ENSEMBL gene identifier in the format of ENSG(number), and its gene name (symbol))."

After 105s, it highlighted 16 genes (plus a couple extra in the cholesterol synthesis pathway):


#### AI BLOCK TEXT START ####
Based on current knowledge of gene expression in adult brain cell types, the following genes from your list are known to be specific to or highly enriched in adult astrocytes compared to other major brain cell types like neurons, oligodendrocytes, microglia, and endothelial cells:

    ALDH1L1 (ENSG00000144908): Aldehyde Dehydrogenase 1 Family Member L1. This is a widely recognized and relatively specific marker for astrocytes in the adult brain.

    SCD (ENSG00000099194): Stearoyl-CoA Desaturase. Involved in fatty acid metabolism and is known to be highly expressed in astrocytes.

    SCD5 (ENSG00000145284): Stearoyl-CoA Desaturase 5. Likely shares similar enrichment in astrocytes due to its function related to SCD.

    MAOB (ENSG00000069535): Monoamine Oxidase B. Classically known to be enriched in astrocytes compared to neurons.

    CKB (ENSG00000166165): Creatine Kinase B. Important for energy metabolism and is highly expressed in astrocytes.

    SLC16A1 (ENSG00000155380): Solute Carrier Family 16 Member 1 (also known as MCT1). A monocarboxylate transporter crucial for lactate transport, highly expressed on astrocytes and involved in the astrocyte-neuron lactate shuttle.

    ME1 (ENSG00000065833): Malic Enzyme 1. Involved in NADPH production, important for astrocytic functions like glutathione recycling and fatty acid synthesis; typically enriched in astrocytes.

    HMGCS1 (ENSG00000112972): 3-Hydroxy-3-Methylglutaryl-CoA Synthase 1. A key enzyme in the cholesterol biosynthesis pathway, a major function of astrocytes in the CNS. Other genes in this pathway on your list like MVK (ENSG00000110921), FDPS (ENSG00000160752), and SC5D (ENSG00000109929) would also be enriched.

    AGL (ENSG00000162688): Amylo-Alpha-1,6-Glucosidase, 4-Alpha-Glucanotransferase. Involved in glycogen metabolism. Astrocytes are the primary cells in the brain that store and metabolize glycogen.

    CIRBP (ENSG00000099622): Cold Inducible RNA Binding Protein. While expressed in other cells, it's significantly expressed in astrocytes and plays roles in circadian rhythms and stress responses.

    PADI2 (ENSG00000117115): Peptidyl Arginine Deiminase 2. Involved in citrullination and is expressed in astrocytes, particularly in contexts of inflammation or disease.

    MLST8 (ENSG00000167965): MTOR Associated Protein, LST8 Homolog. Component of the mTOR signaling pathway, which is active and important in astrocytes; often shows enrichment.

    ASRGL1 (ENSG00000162174): Asparaginase Like 1. Single-cell RNA sequencing studies have shown this gene to be enriched in astrocytes.

    PHB (ENSG00000167085): Prohibitin. While broadly expressed, it's often found at high levels in glial cells, including astrocytes, and is important for mitochondrial integrity and signaling.

    GPR37 (ENSG00000170775): G Protein-Coupled Receptor 37. Primarily expressed in oligodendrocytes but also found in astrocytes.

    PTPRZ1 (ENSG00000106278): Protein Tyrosine Phosphatase Receptor Type Z1. Expressed by astrocyte precursor cells and mature astrocytes, as well as oligodendrocyte precursor cells.

It's important to note that "enriched" means higher expression relative to other cell types, but many of these genes may still have some level of expression in other cells. ALDH1L1 is one of the most specific markers in this list for mature astrocytes. Genes involved in core metabolic pathways like cholesterol synthesis (HMGCS1, MVK, FDPS, SC5D), glycogen metabolism (AGL), and energy buffering (CKB) are characteristically high in astrocytes due to their supportive roles for neurons.

#### AI BLOCK TEXT END ####

So the genes we should plot astrocyte cluster and LC cluster expression of first are (with the exception of MAOB, which we would expect in the LC and which the AI was not privy to the context of):
ENSG00000144908
ENSG00000099194
ENSG00000145284
ENSG00000166165
ENSG00000155380
ENSG00000065833
ENSG00000112972
ENSG00000162688
ENSG00000099622
ENSG00000117115
ENSG00000167965
ENSG00000162174
ENSG00000167085
ENSG00000170775
ENSG00000106278
ENSG00000110921
ENSG00000160752
ENSG00000109929

Slack message contemplating this:
Notably here, several cholesterol synthesis genes are sex DE in LC but not astrocytes at large. The following cholesterol synthesis pathway genes are all sex DE in LC (small logFCs, 0.2-0.3, but significant; 4/5 male-up): SCD (aka SCD1), SCD5, SC5D, MVK (female-up), FDPS

While in the astrocyte cluster itself, the same directional patterns are recapitulated but with about half the magnitude (lfc 0.1-0.15) and FDRs>0.25.
Mayhap male sex -> greater proportion of cholesterol synthesized locally (read: lower reliance on pickup of circulating cholesterol by oh say APOE) -> local cholesterol levels better calibrated to local needs (read: vs. uptake reflective of cholesterol-overrich modern diets ANDOR with less exposure/uptake of everything besides cholesterol that travels along in HDL/LDL and company)?
(Interestingly, if the above hypothesis were true of other male tissues more voluminous than LC or LC-proximal astrocytes, that would also explain the tendency of males to have higher blood lipids/greater cardiovascular disease risk, ie because if relative to tissue X in females, X in males produces a greater proportion of its required cholesterol locally, then male tissue X is presumably pulling less from the boatloads of circulating lipids made available by the modern human diet) 

### load spatial data and plot expression of each in astro and LC clusters simultaneously. do not plot with sex info for blindness
```{r}
plotg <- c("ENSG00000144908", "ENSG00000099194", "ENSG00000145284", "ENSG00000166165", "ENSG00000155380", "ENSG00000065833", "ENSG00000112972", "ENSG00000162688", "ENSG00000099622", "ENSG00000117115", "ENSG00000167965", "ENSG00000162174", "ENSG00000167085", "ENSG00000170775", "ENSG00000106278", "ENSG00000110921", "ENSG00000160752", "ENSG00000109929")

lc3 <- readRDS("processed-data/06-countsOnly_noImgData_QCed_SPE_split_to_tissSections.RDS")
## calc logcounts on the fly and drop samples not considered for DE

lc3 <- lc3[,!(lc3$brnum %in% paste0("Br",c(5517,5276,5712)))]
lc3 <- computeLibraryFactors(lc3)
lc3 <- logNormCounts(lc3)

# labels
louvs <- readRDS("processed-data/07_featureSelection_dimred_harmony_clustering/04-Initial_louvLeid_res0.5-1-2_clusterings.RDS")
louvs <- louvs$HARMONYlmbna_HDG_SVG_2575[,.(rn,snnHARMONYlmbna_HDG_SVG_2575_louv_res1)]
setnames(louvs,2,"clusid")
anno <- fread("processed-data/08_validitycheck_25hdg75svg_louv1/10-25hdg75svg_louv1_annots.txt")
louvs <- merge.data.table(louvs,anno,by="clusid")

colLabels(lc3) <- DataFrame(louvs,row.names=louvs$rn)[colnames(lc3),"anno"]
rm(louvs,anno)
```

```{r}

### helper function to extract a legend and pipe back to a diff plot: from https://statisticsglobe.com/add-common-legend-to-combined-ggplot2-plots-in-r/#example-2-add-shared-legend-to-ggplot2-plots-using-gridextra-package 
extract_legend <- function(my_ggp) {
    step1 <- ggplot_gtable(ggplot_build(my_ggp))
    step2 <- which(sapply(step1$grobs, function(x) x$name) == "guide-box")
    step3 <- step1$grobs[[step2]]
    return(step3)
}
```

escheR per sample per gene in astro, LC spots only. also plot GFAP and AQP4 as negative comparators.
```{r}
plotg <- c(plotg,rowData(lc3)[rowData(lc3)$gene_name %in% c("GFAP","AQP4"),"gene_id"])

samps <- lapply(unique(lc3$sample_id),FUN=function(x){
    subspe <- lc3[plotg,lc3$sample_id==x&lc3$label %in% c("LC","Astro")]
    return(subspe)
})
names(samps) <- unique(lc3$sample_id)

### plots
cluspal <- c(LC="blue",Astro="gold2")

lapply(plotg,FUN=function(g){
    ## get one legend to use for all plots
    tmpdat <- as.data.table(cbind(as.data.frame(t(as.matrix(logcounts(lc3[g,lc3$label %in% c("LC","Astro")])))),
                                  spatialCoords(lc3[,lc3$label %in% c("LC","Astro")]),
                                  lc3[g,lc3$label %in% c("LC","Astro")]$label))
    setnames(tmpdat,c(1,4),c("lct","lab"))
    
    lctmax <- max(tmpdat$lct)
    
    gsymb <- rowData(lc3[g,])$gene_name
    
    tmpplt <- ggplot(tmpdat,aes(x=pxl_col_in_fullres,y=pxl_row_in_fullres,fill=lct,col=lab))+
        geom_point()+
        scale_fill_gradient(low="#cecece",high="#000000",limits=c(0,lctmax),na.value = "#FFFFFF")+
        scale_color_manual(values=cluspal)
    
    leg.plt <- extract_legend(tmpplt)
    dev.off()
    
    ## now make the eschers without legends
    plt <- mapply(s=samps,n=names(samps),lmx=lctmax,SIMPLIFY =FALSE,FUN=function(s,n,lmn,lmx){
        colData(s)$lct <- as.matrix(logcounts(s[g,]))
        
        s$lct[s$lct==min(s$lct)] <- NA # so that our palette starts in a discernible gray range for expressing spots
        
        sp <- make_escheR(s)
        sp <- sp |> add_ground(var = "label",stroke=0.5,point_size = 1.5)
        sp <- sp |> add_fill(var = "lct",point_size = 1.5)
        
        sp <- sp +
            scale_fill_gradient(low="#cecece",high="#000000",limits=c(0,lctmax),na.value = "#FFFFFF")+
            scale_color_manual(values=cluspal) +
            guides(color="none",fill="none") +
            ggtitle(paste0(n," ",rowData(s[g,])$gene_name))+
            theme(title=element_text(size=7))
        
        sp <- rasterize(sp,layers="Points",dpi=300,dev="cairo_png")
        return(sp)
    })
    
    ## save
    pdf(paste0("plots/12_DEanalyses_removedsampsAndFinalNMseg/02b_Allclust_dropmixclusts_SexDEG_EDA/LC_astro_expr_spotplots/",gsymb,".pdf"),height=61,width=20,onefile = FALSE)
    do.call("grid.arrange",list(arrangeGrob(grobs = plt,ncol = 5),leg.plt,ncol=2,widths=c(18.5,1.5)))
    dev.off()
})
```

Practically every LC spot has weak expression of GFAP and AQP4, so there's no great way to discern which cell type is driving this in LC spots. However, the LC NM+ vs - DE tells us that NM+ spots are more LC signal rich (enrichment of basically all LC-specific genes), so we can look there for some hints as to the driving cell type.


```{r}
nmde <- fread("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/04-LCNMpos_vs_LCNMneg_DE.txt")

nmde <- nmde[gene_name %in% c("SCD","SCD5","SC5D","MVK","FDPS","HMGCS1","GFAP","AQP4")]

# positive direction is NM+ enriched
#    gene_name       logFC    adj.P.Val
#       <char>       <num>        <num>
# 1:      GFAP -0.88731561 2.533786e-22
# 2:       SCD -0.60550386 1.238622e-18
# 3:      AQP4 -0.90421258 4.672839e-16
# 4:      SCD5 -0.17925553 3.979000e-06
# 5:       MVK  0.32211566 3.305839e-05
# 6:    HMGCS1  0.17083356 3.444783e-05
# 7:      FDPS  0.11494299 1.686699e-02
# 8:      SC5D  0.04169909 4.615064e-01
```

Interestingly, HMGCS1 (a downstream step from the SCDs), pretty much fills the LC even though its astrocytic expression is variable sparse. But this info doesn't do us a lot of good understanding which is doing what (or if it's some steps in one cell type and some in the other).

#######################################
Exploration 2: Sex DE GSEA results
#######################################

```{r}
rm(list=grep(ls(),pattern="lcde|nmde|astrode|detabs",value=T,invert=T))
gc(full=T)

sexmsig <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/03bi-MulticlusterDE_OneInterxnPerModel_dropMixClusts_GSEAmsigdb.RDS")
sexmsig <- sexmsig$LC_Sex
sexmsig <- sexmsig[padj<0.05]
sexmsig <- sexmsig[,.(pathway,dir,padj,NES,ES,pathway,leadingEdge)]


## msig terms: WP_CHOLESTEROL_METABOLISM, YAO_TEMPORAL_RESPONSE_TO_PROGESTERONE_CLUSTER_1, GOBP_MODULATION_OF_EXCITATORY_POSTSYNAPTIC_POTENTIAL, BHAT_ESR1_TARGETS_VIA_AKT1_UP, GOMF_TRANSFERASE_ACTIVITY_TRANSFERRING_PHOSPHORUS_CONTAINING_GROUPS, REACTOME_AEROBIC_RESPIRATION_AND_RESPIRATORY_ELECTRON_TRANSPORT, REACTOME_REGULATION_OF_EXPRESSION_OF_SLITS_AND_ROBOS, REACTOME_METABOLISM_OF_AMINO_ACIDS_AND_DERIVATIVES, REACTOME_NONSENSE_MEDIATED_DECAY_NMD,ATF5_TARGET_GENES


sextf <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/08-LCNM_APOESexDE_PredomAncesDE_LCnmf_LCnmfLct18Filt_TFtarg_GSEA_filterToClusAndDEgroupExpressed.RDS")
sextf <- sextf$LC_Sex
## tabulate how many times the TF came up as significant in each sex and how many times it came up total (signif or not)
tfcounts <- sextf[,.(pathway,dir,.N),by="tf"]
setnames(tfcounts,"N","N_tests")

sigres <- sextf[padj<0.05,.(pathway,TF_F_Expr,TF_M_Expr,padj,NES,ES,leadingEdge,.N),by=c("dir","tf")]
setnames(sigres,"N","N_sig_for_dir")
# don't need this copy of the tf column now
sigres[,tf:=NULL]

tfcounts <- merge.data.table(tfcounts,sigres,by=c("pathway","dir"))
## we kept pathway on here to see what datasets were responsible for different enrichments, but can push that to the far end. what we want to do now is determine those TFs that were only positive for enrichment in one sex, aand then for the others look at those where there was a clear imbalance in which DE direction showed enrichment 
onesex <- unique(tfcounts[!is.na(dir),.(tf,dir)])[,.N,by=c("tf")][N==1,tf]
bothsex <- unique(tfcounts[!is.na(dir),.(tf,dir)])[,.N,by=c("tf")][N==2,tf]

tfres.onesex <- tfcounts[tf %in% onesex]
tfres.twosex <- tfcounts[tf %in% bothsex]

tfres.twosex.frxns <- tfres.twosex[,prop_sig:=N_sig_for_dir/N_tests]
tfres.twosex.frxns <- unique(tfres.twosex.frxns[,.(tf,dir,prop_sig)])
# hsf1 kmt2a egr1 creb1 runx1 son zeb2 
# ^ are pretty big differences (like, by several hits ea.; all more hits for male-up). (note that HSF1 and EGR1 themselves are female LC upreg; the others show no DE). so let's grab one interesting male-only tf and then the rest female-only from onesex
tfres.onesex <- unique(tfres.onesex[,.(N_tests,N_sig_for_dir,tf,dir)])
tfres.onesex[,propsig:=N_sig_for_dir/N_tests]
View(unique(tfres.onesex[dir=="female_up"&N_sig_for_dir>1,.(N_tests,N_sig_for_dir,tf,propsig)]))
# for males, NR2F2 (also signif upreg in male LC)
# BAZ2A SRCAP USF1 CIC TCF20 NFYB MBNL2 BHLHE22 FOS are the only ones -- and FOS isn't very informative here. TCF20 is the only LC DE in the same direction.

# Genecards re: USF1 "Diseases associated with USF1 include Hyperlipidemia, Familial Combined, 1 and Hyperlipidemia, Familial Combined, 3. Among its related pathways are ESR-mediated signaling and MITF-M-dependent gene expression."

# TCF20: Female LC upreg; uniprot "binds to the regulatory region of MMP3 [promoter] and thereby controls stromelysin expression. It stimulates the activity of various transcriptional activators [proteins] such as JUN, SP1, PAX6 and ETS1"

# MBNL2 (male LC signif up, female target enriched): Alt splicing regulator; Uniprot: "Antagonizes the alternative splicing activity pattern of CELF proteins." actually that's not clear to me whether it means it prevents the splicing of rna ENCODING celfs or prevents CELFs from splicing other genes.
```


#######################################
"Exploration" 3: Using Jaffe cleaningy
#######################################
```{r}
library(jaffelab)

pbulk <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/02b-voomLmFit_objs_and_DE_tables_LabSex_LabE2vE4_LabSingleGenos_LabAnces_LabAncesE4E2_noMixClusts_YRIcontAndCat_JUL2025.RDS")

gmat <- pbulk$cont_ances$SexByLabel$voomLFt$EList$E
# subset to LC
gmat <- gmat[,colnames(gmat) %in% grep(colnames(gmat),pattern="LC_V",value=T)]
# make symbols
rownames(gmat) <- pbulk$SexByLabel$voomLFt$genes$gene_name

# get sample metadata in the same arrangement for LC pbulks
cd <- as.data.frame(pbulk$SexByLabel$voomLFt$targets)[colnames(gmat),]

# model matrix for "single cluster" (the label_sex thing doesn't work here)
modmat <- model.matrix(~Sex+APOE+Age+YRI,data = cd)

testres <- jaffelab::cleaningY(gmat,modmat,2)

testres2 <- testres["HMGCS1",]
testres2 <- as.data.table(cbind(testres2,cd$label_sex,cd$brnum,cd$Sex))
setnames(testres2,c("cleanedxpr","labsex","brnum","sex"))
testres2[,cleanedxpr:=as.numeric(cleanedxpr)]

ggplot(testres2,aes(x=sex,y=cleanedxpr))+
    geom_boxplot()+
    geom_point()

```
