---
title: "01-INITIAL-glm-SNN10-sexDE-analyses"
author: "Bernie Mulvey"
date: "2024-12-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
setDTthreads(1,restore_after_fork = FALSE)
library(SpatialExperiment)
library(dreamlet)
library(limma)
library(edgeR)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure.
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")
library(parallel)
```

```{r}
## load spe
lc2 <- readRDS("processed-data/05_QC/04-lffLC_noEdges_noLocOutliers_remainder1pctileUMIorGene-removed.RDS")

### remove components we don't need here: 10x dimreds, img data, logcount mat
lc2 <- SpatialExperiment(assays=list(counts=counts(lc2)),rowData=rowData(lc2),colData=colData(lc2),sample_id = colData(lc2)$sample_id,spatialCoords = spatialCoords(lc2))

assayNames(lc2)
# clean up
gc(full=T)

## get rowdata to annotate genes from DE analysis later
tmprd <- as.data.table(rowData(lc2))

## append chromosome info to this
chrinf <- fread("raw-data/USEFOR10X_ensg_lut_hg38_fromEns98_111123.txt")
chrinf <- unique(chrinf[,.(ensembl_gene_id,chromosome_name)])
chrinf[,chromosome_name:=paste0("chr",chromosome_name)]
tmprd <- merge.data.table(tmprd,chrinf,by.x="gene_id",by.y="ensembl_gene_id")

# get cluster assignment files
lvcl <- readRDS("processed-data/07_Clustering/01-GLMPCAallgene_SNN10-15-20_leidenAndLouvain_res05-1-2.RDS")
# subset to cluster assignments of interest
lvcl <- as.data.table(lvcl[[1]][,c(1:3)])
lvcl <- list(louv05=lvcl$snn10_louv_res0.5,louv1=lvcl$snn10_louv_res1)
```

## run sex DE using voomLmFit with a simmple (no covariates) DE model.
```{r}

# loop thru one at a time so we can use biocparallel w/ dreamlet and PSOCK clusters for running the DE tests (19 clustering results * 13 to 31 clusters ea.)

sbp <- MulticoreParam(6)
register(sbp)

bsHna60s <- lapply(lvcl,function(b){
  colLabels(lc2) <- b
  
  # run dreamlet only to pseudobulk each cluster or collapsed clusters 
  ag <- dreamlet::aggregateToPseudoBulk(lc2,sample_id="sample_id",cluster_id = "label",BPPARAM = sbp)
  
  # extract colData to feed into DGEList as the "samples" data
  agcd <- as.data.frame(colData(ag))
  
  ## only retain clusters for which at least 7 samples (66% of smaller group size) are available after pseudobulking+filtering through dreamlet:  
  ag <- as.list(assays(ag))
  agc <- lapply(ag,FUN=function(a){ncol(a[,colSums(a)>0])})
  ag <- ag[!(agc<7)]
  
  ## drop clusters where there are 7+ samples, but only 0 or 1 samples for one of the sexes:
  agsc <- lapply(ag,FUN=function(a){
    cn <- colnames(a[,colSums(a)>0])
    tmp <- as.data.table(agcd[cn,])
    if(nrow(tmp[,.N,by="Sex"][N<2])>0){
      return(0)
    } else {return(1)}
  })
  
  ag <- ag[agsc==1]
  
  ## make the DGELists per remaining cluster:
  dgels <- lapply(ag,FUN=function(a){
    X <- a[,colSums(a)>0]
    s <- copy(agcd)[colnames(X),]
    if(length(unique(s$Sex))<2){
      return()
    } else {
    
      dge <- DGEList(counts=as.matrix(X),samples=s)
      dge <- normLibSizes(dge,method = "TMM")
      
      ## find the smaller group for this cluster
      smolgrp <- table(dge$samples$Sex)[which.min(table(dge$samples$Sex))]
      
      ## for the smallest group, determine low expression cutoff
      rowmn.m <- as.numeric(rowMeans(cpm(dge[,(dge$samples$Sex==names(smolgrp)[1])],log=TRUE)))
      
      ## get histogram distribution with 10 bins at the low end of the logcpm space
      h <- hist(rowmn.m[which(rowmn.m>min(rowmn.m)&rowmn.m<(4+min(rowmn.m)))],breaks=10,plot=FALSE)
      ## get the minimum in this low logcpm range and its corresponding bin (x axis value, i.e. log cpm)
      minc <- h$breaks[which(h$counts==min(h$counts))]
      
      ## get the corresponding smallest-library count value for this minimum^
      smollib <- min(dge$samples[(dge$samples$Sex==names(smolgrp)[1]),"lib.size"])
      countthresh <- (minc^2)*(as.numeric(smollib)/10^6)
      
      ## and convert that smallest count to the logcpm value in said smallest library to use as a low-expression threshold
      cpmthresh <- log2(countthresh/(as.numeric(smollib)*10^-6))
      
      ## filter for genes with at least thresh counts in at least (smaller sex group size) samples
      keep <- rowSums(cpm(dge,log=T)>cpmthresh) >= as.numeric(smolgrp)
      
      ## recalc lib factors after dropping low-expressed genes
      dge <- dge[keep,,keep.lib.sizes=FALSE]
      dge <- calcNormFactors(dge,method = "TMM")
      
      return(dge)
    }
  })
  
  ### run voomLmFit for each pseudobulked cluster
  clus <- makeCluster(10,type = "PSOCK")
    fulldgelres <- clusterApply(x=dgels,cl = clus,fun=function(X){
    
    library(data.table)
    setDTthreads(1,restore_after_fork = FALSE)
    library(SpatialExperiment)
    library(limma)
    library(edgeR)
    
    # prevent resource competition, though we aren't explicitly using DT at all here:
    setDTthreads(1) 
    
    dge <- X
    des <- model.matrix(~0+Sex, data = dge$samples)
    # INITIALIZE the voomwithqualityweights model, equivalent to pseudobulkDGE(method="voom",qualities=T)
    sexdif <- makeContrasts(SexM-SexF,levels=des)
    
    # run voomLmFit for the pseudobulked data, referring donor to duplicateCorrelation; using an adaptive span (number of genes, based on the number of genes in the dge) for smoothing the mean-variance trend
    v.swt <- voomLmFit(dge,design = des,block = as.factor(dge$samples$brnum),adaptive.span = T,sample.weights = T)
    ## get the Elist object to also output for plotting purposes later
    v.swt.elist <- as.data.table(v.swt$EList$E,keep.rownames=T)
    
    v.swt.fit <- contrasts.fit(v.swt,sexdif)
    
    v.swt.fit.e <- eBayes(v.swt.fit)
    
    v.swt.e.tt <- topTable(v.swt.fit.e,coef = 1,number=Inf,adjust.method = "BH")
    
    retlist <- list(elist=v.swt.elist,de=v.swt.e.tt)
    
    return(retlist)
    })
  stopCluster(clus)
  # make sure the output list has the corresponding cell cluster names
  names(fulldgelres) <- names(ag)
  return(fulldgelres)
})

i<-1
for (i in c(1:length(bsHna60s))){d
    j<-1
    for (j in c(1:length(bsHna60s[[i]]))){
        tmp <- as.data.table(bsHna60s[[i]][[j]][[2]],keep.rownames=T)
        tmp <- merge.data.table(tmp,tmprd,by.x="rn",by.y="gene_id")
        bsHna60s[[i]][[j]][[2]] <- tmp
    }
}

saveRDS(bsHna60s,"processed-data/08_DE_preflights/03a-glmPCA-snn10-louvs_sexDE_sexOnlyFactor.RDS")

```

### now do this again, but with age and E4 carrier/noncarrier controlled for
```{r}
lc2$e4 <- gsub(lc2$APOE,pattern="^(.*E4.*)$",replacement="yes")
lc2$e4[lc2$e4!="yes"] <- "no"
# make sure age is a continuous variable
lc2$Age <- as.numeric(lc2$Age)

sbp <- MulticoreParam(6)
register(sbp)
bsHna60cov <- lapply(lvcl,function(b){
  colLabels(lc2) <- b
  
  # run dreamlet only to pseudobulk each cluster or collapsed clusters 
  ag <- dreamlet::aggregateToPseudoBulk(lc2,sample_id="sample_id",cluster_id = "label",BPPARAM = sbp)
  
  # extract colData to feed into DGEList as the "samples" data
  agcd <- as.data.frame(colData(ag))
  
  ## only retain clusters for which at least 7 samples (66% of smaller group size) are available after pseudobulking+filtering through dreamlet:  
  ag <- as.list(assays(ag))
  agc <- lapply(ag,FUN=function(a){ncol(a[,colSums(a)>0])})
  ag <- ag[!(agc<7)]
  
  ## drop clusters where there are 7+ samples, but only 0 or 1 samples for one of the sexes:
  agsc <- lapply(ag,FUN=function(a){
    cn <- colnames(a[,colSums(a)>0])
    tmp <- as.data.table(agcd[cn,])
    if(nrow(tmp[,.N,by="Sex"][N<2])>0){
      return(0)
    } else {return(1)}
  })
  
  ag <- ag[agsc==1]
  
  ## make the DGELists per remaining cluster:
  dgels <- lapply(ag,FUN=function(a){
    X <- a[,colSums(a)>0]
    s <- copy(agcd)[colnames(X),]
    if(length(unique(s$Sex))<2){
      return()
    } else {
    
      dge <- DGEList(counts=as.matrix(X),samples=s)
      dge <- normLibSizes(dge,method = "TMM")
      
      ## find the smaller group for this cluster
      smolgrp <- table(dge$samples$Sex)[which.min(table(dge$samples$Sex))]
      
      ## for the smallest group, determine low expression cutoff
      rowmn.m <- as.numeric(rowMeans(cpm(dge[,(dge$samples$Sex==names(smolgrp)[1])],log=TRUE)))
      
      ## get histogram distribution with 10 bins at the low end of the logcpm space
      h <- hist(rowmn.m[which(rowmn.m>min(rowmn.m)&rowmn.m<(4+min(rowmn.m)))],breaks=10,plot=FALSE)
      ## get the minimum in this low logcpm range and its corresponding bin (x axis value, i.e. log cpm)
      minc <- h$breaks[which(h$counts==min(h$counts))]
      
      ## get the corresponding smallest-library count value for this minimum^
      smollib <- min(dge$samples[(dge$samples$Sex==names(smolgrp)[1]),"lib.size"])
      countthresh <- (minc^2)*(as.numeric(smollib)/10^6)
      
      ## and convert that smallest count to the logcpm value in said smallest library to use as a low-expression threshold
      cpmthresh <- log2(countthresh/(as.numeric(smollib)*10^-6))
      
      ## filter for genes with at least thresh counts in at least (smaller sex group size) samples
      keep <- rowSums(cpm(dge,log=T)>cpmthresh) >= as.numeric(smolgrp)
      
      ## recalc lib factors after dropping low-expressed genes
      dge <- dge[keep,,keep.lib.sizes=FALSE]
      dge <- calcNormFactors(dge,method = "TMM")
      
      return(dge)
    }
  })
  
  ### run voomLmFit for each pseudobulked cluster
  clus <- makeCluster(10,type = "PSOCK")
    fulldgelres <- clusterApply(x=dgels,cl = clus,fun=function(X){
    
    library(data.table)
    setDTthreads(1,restore_after_fork = FALSE)
    library(SpatialExperiment)
    library(limma)
    library(edgeR)
    
    # prevent resource competition, though we aren't explicitly using DT at all here:
    setDTthreads(1) 
    
    dge <- X
    des <- model.matrix(~0+Sex+e4+Age, data = dge$samples)
    # INITIALIZE the voomwithqualityweights model, equivalent to pseudobulkDGE(method="voom",qualities=T)
    sexdif <- makeContrasts(SexM-SexF,levels=des)
    
    # run voomLmFit for the pseudobulked data, referring donor to duplicateCorrelation; using an adaptive span (number of genes, based on the number of genes in the dge) for smoothing the mean-variance trend
    v.swt <- voomLmFit(dge,design = des,block = as.factor(dge$samples$brnum),adaptive.span = T,sample.weights = T)
    ## get the Elist object to also output for plotting purposes later
    v.swt.elist <- as.data.table(v.swt$EList$E,keep.rownames=T)
    
    v.swt.fit <- contrasts.fit(v.swt,sexdif)
    
    v.swt.fit.e <- eBayes(v.swt.fit)
    
    v.swt.e.tt <- topTable(v.swt.fit.e,coef = 1,number=Inf,adjust.method = "BH")
    
    retlist <- list(elist=v.swt.elist,de=v.swt.e.tt)
    
    return(retlist)
    })
  stopCluster(clus)
  # make sure the output list has the corresponding cell cluster names
  names(fulldgelres) <- names(ag)
  return(fulldgelres)
})

i<-1
for (i in c(1:length(bsHna60cov))){
    j<-1
    for (j in c(1:length(bsHna60cov[[i]]))){
        tmp <- as.data.table(bsHna60cov[[i]][[j]][[2]],keep.rownames=T)
        tmp <- merge.data.table(tmp,tmprd,by.x="rn",by.y="gene_id")
        bsHna60cov[[i]][[j]][[2]] <- tmp
    }
}

saveRDS(bsHna60cov,"processed-data/08_DE_preflights/03b-glmPCA-snn10-louvs_sexDE_ageAPOEcovar.RDS")

```

reprod inf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods  
[9] base     

other attached packages:
 [1] parallelly_1.40.1           colorout_1.3-0.2           
 [3] edgeR_4.4.1                 dreamlet_1.4.1             
 [5] variancePartition_1.36.2    BiocParallel_1.40.0        
 [7] limma_3.62.1                ggplot2_3.5.1              
 [9] data.table_1.16.4           SpatialExperiment_1.16.0   
[11] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0
[13] Biobase_2.66.0              GenomicRanges_1.58.0       
[15] GenomeInfoDb_1.42.1         IRanges_2.40.1             
[17] S4Vectors_0.44.0            BiocGenerics_0.52.0        
[19] MatrixGenerics_1.18.0       matrixStats_1.4.1          
[21] rlang_1.1.4                

loaded via a namespace (and not attached):
  [1] splines_4.4.1             bitops_1.0-9              tibble_3.2.1             
  [4] graph_1.84.0              XML_3.99-0.17             lifecycle_1.0.4          
  [7] Rdpack_2.6.2              mixsqp_0.3-54             rprojroot_2.0.4          
 [10] lattice_0.22-6            MASS_7.3-61               backports_1.5.0          
 [13] magrittr_2.0.3            metafor_4.6-0             rmarkdown_2.29           
 [16] yaml_2.3.10               sessioninfo_1.2.2         DBI_1.2.3                
 [19] minqa_1.2.8               abind_1.4-8               zlibbioc_1.52.0          
 [22] EnvStats_3.0.0            purrr_1.0.2               rmeta_3.0                
 [25] msigdbr_7.5.1             RCurl_1.98-1.16           GenomeInfoDbData_1.2.13  
 [28] ggrepel_0.9.6             pbkrtest_0.5.3            irlba_2.3.5.1            
 [31] annotate_1.84.0           DelayedMatrixStats_1.28.0 codetools_0.2-20         
 [34] DelayedArray_0.32.0       tidyselect_1.2.1          UCSC.utils_1.2.0         
 [37] lme4_1.1-35.5             mathjaxr_1.6-0            jsonlite_1.8.9           
 [40] iterators_1.0.14          tools_4.4.1               progress_1.2.3           
 [43] Rcpp_1.0.13-1             zenith_1.8.0              glue_1.8.0               
 [46] SparseArray_1.6.0         xfun_0.49                 here_1.0.1               
 [49] dplyr_1.1.4               withr_3.0.2               numDeriv_2016.8-1.1      
 [52] fastmap_1.2.0             boot_1.3-31               fansi_1.0.6              
 [55] caTools_1.18.3            digest_0.6.37             truncnorm_1.0-9          
 [58] R6_2.5.1                  colorspace_2.1-1          scattermore_1.2          
 [61] gtools_3.9.5              RSQLite_2.3.9             RhpcBLASctl_0.23-42      
 [64] utf8_1.2.4                tidyr_1.3.1               generics_0.1.3           
 [67] corpcor_1.6.10            prettyunits_1.2.0         httr_1.4.7               
 [70] S4Arrays_1.6.0            pkgconfig_2.0.3           gtable_0.3.6             
 [73] blob_1.2.4                XVector_0.46.0            remaCor_0.0.18           
 [76] htmltools_0.5.8.1         GSEABase_1.68.0           scales_1.3.0             
 [79] png_0.1-8                 fANCOVA_0.6-1             ashr_2.2-63              
 [82] knitr_1.49                rstudioapi_0.17.1         reshape2_1.4.4           
 [85] rjson_0.2.23              nlme_3.1-166              nloptr_2.1.1             
 [88] cachem_1.1.0              stringr_1.5.1             KernSmooth_2.23-24       
 [91] vipor_0.4.7               metadat_1.2-0             RcppZiggurat_0.1.6       
 [94] AnnotationDbi_1.68.0      pillar_1.9.0              grid_4.4.1               
 [97] vctrs_0.6.5               gplots_3.2.0              mashr_0.2.79             
[100] xtable_1.8-4              beeswarm_0.4.0            Rgraphviz_2.50.0         
[103] evaluate_1.0.1            KEGGgraph_1.66.0          invgamma_1.1             
[106] magick_2.8.5              mvtnorm_1.3-2             cli_3.6.3                
[109] locfit_1.5-9.10           compiler_4.4.1            crayon_1.5.3             
[112] SQUAREM_2021.1            plyr_1.8.9                ggbeeswarm_0.7.2         
[115] stringi_1.8.4             assertthat_0.2.1          babelgene_22.9           
[118] lmerTest_3.1-3            munsell_0.5.1             Biostrings_2.74.0        
[121] aod_1.3.3                 Matrix_1.7-1              hms_1.1.3                
[124] sparseMatrixStats_1.18.0  bit64_4.5.2               KEGGREST_1.46.0          
[127] statmod_1.5.0             rbibutils_2.3             Rfast_2.1.0              
[130] broom_1.0.7               memoise_2.0.1             RcppParallel_5.1.9       
[133] bit_4.5.0.1               EnrichmentBrowser_2.36.0 
> sessioninfo::session_info()
─ Session info ──────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.7.1
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-12-13
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ──────────────────────────────────────────────────────────────────────
 ! package              * version    date (UTC) lib source
   abind                  1.4-8      2024-09-12 [1] CRAN (R 4.4.1)
   annotate               1.84.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   AnnotationDbi          1.68.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   aod                    1.3.3      2023-12-13 [1] CRAN (R 4.4.0)
   ashr                   2.2-63     2023-08-21 [1] CRAN (R 4.4.0)
   assertthat             0.2.1      2019-03-21 [1] CRAN (R 4.4.0)
   babelgene              22.9       2022-09-29 [1] CRAN (R 4.4.0)
   backports              1.5.0      2024-05-23 [1] CRAN (R 4.4.0)
   beeswarm               0.4.0      2021-06-01 [1] CRAN (R 4.4.0)
   Biobase              * 2.66.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   BiocGenerics         * 0.52.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   BiocParallel         * 1.40.0     2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
   Biostrings             2.74.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   bit                    4.5.0.1    2024-12-03 [1] CRAN (R 4.4.1)
   bit64                  4.5.2      2024-09-22 [1] CRAN (R 4.4.1)
   bitops                 1.0-9      2024-10-03 [1] CRAN (R 4.4.1)
   blob                   1.2.4      2023-03-17 [1] CRAN (R 4.4.0)
   boot                   1.3-31     2024-08-28 [1] CRAN (R 4.4.1)
   broom                  1.0.7      2024-09-26 [1] CRAN (R 4.4.1)
   cachem                 1.1.0      2024-05-16 [1] CRAN (R 4.4.0)
   caTools                1.18.3     2024-09-04 [1] CRAN (R 4.4.1)
 P cli                    3.6.3      2024-06-21 [2] CRAN (R 4.4.0)
   codetools              0.2-20     2024-03-31 [1] CRAN (R 4.4.1)
   colorout             * 1.3-0.2    2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace             2.1-1      2024-07-26 [1] CRAN (R 4.4.0)
   corpcor                1.6.10     2021-09-16 [1] CRAN (R 4.4.0)
   crayon                 1.5.3      2024-06-20 [1] CRAN (R 4.4.0)
   data.table           * 1.16.4     2024-12-06 [1] CRAN (R 4.4.1)
   DBI                    1.2.3      2024-06-02 [1] CRAN (R 4.4.0)
   DelayedArray           0.32.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   DelayedMatrixStats     1.28.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   digest                 0.6.37     2024-08-19 [1] CRAN (R 4.4.1)
   dplyr                  1.1.4      2023-11-17 [1] CRAN (R 4.4.0)
   dreamlet             * 1.4.1      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   edgeR                * 4.4.1      2024-12-02 [1] Bioconductor 3.20 (R 4.4.2)
   EnrichmentBrowser      2.36.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   EnvStats               3.0.0      2024-08-24 [1] CRAN (R 4.4.1)
   evaluate               1.0.1      2024-10-10 [1] CRAN (R 4.4.1)
   fANCOVA                0.6-1      2020-11-13 [1] CRAN (R 4.4.0)
   fansi                  1.0.6      2023-12-08 [1] CRAN (R 4.4.0)
   fastmap                1.2.0      2024-05-15 [1] CRAN (R 4.4.0)
   generics               0.1.3      2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb         * 1.42.1     2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
   GenomeInfoDbData       1.2.13     2024-12-12 [1] Bioconductor
   GenomicRanges        * 1.58.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   ggbeeswarm             0.7.2      2023-04-29 [1] CRAN (R 4.4.0)
   ggplot2              * 3.5.1      2024-04-23 [1] CRAN (R 4.4.0)
   ggrepel                0.9.6      2024-09-07 [1] CRAN (R 4.4.1)
   glue                   1.8.0      2024-09-30 [1] CRAN (R 4.4.1)
   gplots                 3.2.0      2024-10-05 [1] CRAN (R 4.4.1)
   graph                  1.84.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   GSEABase               1.68.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   gtable                 0.3.6      2024-10-25 [1] CRAN (R 4.4.1)
   gtools                 3.9.5      2023-11-20 [1] CRAN (R 4.4.0)
   here                   1.0.1      2020-12-13 [1] CRAN (R 4.4.0)
   hms                    1.1.3      2023-03-21 [1] CRAN (R 4.4.0)
   htmltools              0.5.8.1    2024-04-04 [1] CRAN (R 4.4.0)
   httr                   1.4.7      2023-08-15 [1] CRAN (R 4.4.0)
   invgamma               1.1        2017-05-07 [1] CRAN (R 4.4.0)
   IRanges              * 2.40.1     2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
   irlba                  2.3.5.1    2022-10-03 [1] CRAN (R 4.4.0)
   iterators              1.0.14     2022-02-05 [1] CRAN (R 4.4.0)
   jsonlite               1.8.9      2024-09-20 [1] CRAN (R 4.4.1)
   KEGGgraph              1.66.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   KEGGREST               1.46.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   KernSmooth             2.23-24    2024-05-17 [1] CRAN (R 4.4.1)
   knitr                  1.49       2024-11-08 [1] CRAN (R 4.4.1)
   lattice                0.22-6     2024-03-20 [1] CRAN (R 4.4.1)
   lifecycle              1.0.4      2023-11-07 [1] CRAN (R 4.4.0)
   limma                * 3.62.1     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   lme4                   1.1-35.5   2024-07-03 [1] CRAN (R 4.4.1)
   lmerTest               3.1-3      2020-10-23 [1] CRAN (R 4.4.0)
   locfit                 1.5-9.10   2024-06-24 [1] CRAN (R 4.4.0)
   magick                 2.8.5      2024-09-20 [1] CRAN (R 4.4.1)
   magrittr               2.0.3      2022-03-30 [1] CRAN (R 4.4.0)
   mashr                  0.2.79     2023-10-18 [1] CRAN (R 4.4.0)
   MASS                   7.3-61     2024-06-13 [1] CRAN (R 4.4.0)
   mathjaxr               1.6-0      2022-02-28 [1] CRAN (R 4.4.0)
   Matrix                 1.7-1      2024-10-18 [1] CRAN (R 4.4.1)
   MatrixGenerics       * 1.18.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   matrixStats          * 1.4.1      2024-09-08 [1] CRAN (R 4.4.1)
   memoise                2.0.1      2021-11-26 [1] CRAN (R 4.4.0)
   metadat                1.2-0      2022-04-06 [1] CRAN (R 4.4.0)
   metafor                4.6-0      2024-03-28 [1] CRAN (R 4.4.0)
   minqa                  1.2.8      2024-08-17 [1] CRAN (R 4.4.0)
   mixsqp                 0.3-54     2023-12-20 [1] CRAN (R 4.4.0)
   msigdbr                7.5.1      2022-03-30 [1] CRAN (R 4.4.0)
   munsell                0.5.1      2024-04-01 [1] CRAN (R 4.4.0)
   mvtnorm                1.3-2      2024-11-04 [1] CRAN (R 4.4.1)
   nlme                   3.1-166    2024-08-14 [1] CRAN (R 4.4.0)
   nloptr                 2.1.1      2024-06-25 [1] CRAN (R 4.4.0)
   numDeriv               2016.8-1.1 2019-06-06 [1] CRAN (R 4.4.0)
   parallelly           * 1.40.1     2024-12-04 [1] CRAN (R 4.4.1)
   pbkrtest               0.5.3      2024-06-26 [1] CRAN (R 4.4.0)
   pillar                 1.9.0      2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig              2.0.3      2019-09-22 [1] CRAN (R 4.4.0)
   plyr                   1.8.9      2023-10-02 [1] CRAN (R 4.4.0)
   png                    0.1-8      2022-11-29 [1] CRAN (R 4.4.0)
   prettyunits            1.2.0      2023-09-24 [1] CRAN (R 4.4.0)
   progress               1.2.3      2023-12-06 [1] CRAN (R 4.4.0)
   purrr                  1.0.2      2023-08-10 [1] CRAN (R 4.4.0)
   R6                     2.5.1      2021-08-19 [1] CRAN (R 4.4.0)
   rbibutils              2.3        2024-10-04 [1] CRAN (R 4.4.1)
   Rcpp                   1.0.13-1   2024-11-02 [1] CRAN (R 4.4.1)
   RcppParallel           5.1.9      2024-08-19 [1] CRAN (R 4.4.1)
   RcppZiggurat           0.1.6      2020-10-20 [1] CRAN (R 4.4.0)
   RCurl                  1.98-1.16  2024-07-11 [1] CRAN (R 4.4.0)
   Rdpack                 2.6.2      2024-11-15 [1] CRAN (R 4.4.1)
   remaCor                0.0.18     2024-02-08 [1] CRAN (R 4.4.0)
   reshape2               1.4.4      2020-04-09 [1] CRAN (R 4.4.0)
   Rfast                  2.1.0      2023-11-09 [1] CRAN (R 4.4.0)
   Rgraphviz              2.50.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   RhpcBLASctl            0.23-42    2023-02-11 [1] CRAN (R 4.4.0)
   rjson                  0.2.23     2024-09-16 [1] CRAN (R 4.4.1)
 P rlang                * 1.1.4      2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown              2.29       2024-11-04 [1] CRAN (R 4.4.1)
   rmeta                  3.0        2018-03-20 [1] CRAN (R 4.4.0)
   rprojroot              2.0.4      2023-11-05 [1] CRAN (R 4.4.0)
   RSQLite                2.3.9      2024-12-03 [1] CRAN (R 4.4.1)
   rstudioapi             0.17.1     2024-10-22 [1] CRAN (R 4.4.1)
   S4Arrays               1.6.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   S4Vectors            * 0.44.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   scales                 1.3.0      2023-11-28 [1] CRAN (R 4.4.0)
   scattermore            1.2        2023-06-12 [1] CRAN (R 4.4.0)
   sessioninfo            1.2.2      2021-12-06 [1] CRAN (R 4.4.0)
   SingleCellExperiment * 1.28.1     2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
   SparseArray            1.6.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   sparseMatrixStats      1.18.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   SpatialExperiment    * 1.16.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   SQUAREM                2021.1     2021-01-13 [1] CRAN (R 4.4.0)
   statmod                1.5.0      2023-01-06 [1] CRAN (R 4.4.0)
   stringi                1.8.4      2024-05-06 [1] CRAN (R 4.4.0)
   stringr                1.5.1      2023-11-14 [1] CRAN (R 4.4.0)
   SummarizedExperiment * 1.36.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   tibble                 3.2.1      2023-03-20 [1] CRAN (R 4.4.0)
   tidyr                  1.3.1      2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect             1.2.1      2024-03-11 [1] CRAN (R 4.4.0)
   truncnorm              1.0-9      2023-03-20 [1] CRAN (R 4.4.0)
   UCSC.utils             1.2.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   utf8                   1.2.4      2023-10-22 [1] CRAN (R 4.4.0)
   variancePartition    * 1.36.2     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   vctrs                  0.6.5      2023-12-01 [1] CRAN (R 4.4.0)
   vipor                  0.4.7      2023-12-18 [1] CRAN (R 4.4.0)
   withr                  3.0.2      2024-10-28 [1] CRAN (R 4.4.1)
   xfun                   0.49       2024-10-31 [1] CRAN (R 4.4.1)
   XML                    3.99-0.17  2024-06-25 [1] CRAN (R 4.3.3)
   xtable                 1.8-4      2019-04-21 [1] CRAN (R 4.4.0)
   XVector                0.46.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   yaml                   2.3.10     2024-07-26 [1] CRAN (R 4.4.0)
   zenith                 1.8.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   zlibbioc               1.52.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

─────────────────────────────────────────────────────────────────────────────────
