---
title: "06-split_spe_to_singleTissueSections"
author: "Bernie Mulvey"
date: "2024-12-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(ggplot2)
library(gridExtra)
library(SpatialExperiment)
# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")


## ggplot defaults
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 9), axis.title.x = element_text(size = 8), axis.text.y = element_text(size = 8), axis.title.y = element_text(size = 9), plot.title = element_text(size = 11,hjust=0.5), strip.text = element_text(size=11), legend.text = element_text(size=8), legend.title = element_text(size=9,hjust=0.5)))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```


### load standard form of SPE (two sections per capture area)
### load the table of raw SPE unique spot ids annotated as each section of each capture area
```{r}
lc2 <- readRDS("processed-data/04_QC/04-lffLC_noEdges_noLocOutliers_remainder1pctileUMIorGene-removed.RDS")

lc2splitcoord <- fread("/Users/bmulvey/Desktop/KM Lab/local_lffLC/LFF_spatial_LC/processed-data/04_QC/06c-raw_SPE_spots_assigned_to_tissue-section-1or2_for_all_capture_areas.txt")
```


get coldata in isolation, subset the section annotations to retained spots after filtering (those in the SPE), reconstruct SPE as single sections 
```{r}
tmpcd <- as.data.table(colData(lc2),keep.rownames=T)

lc2splitcoord <- lc2splitcoord[key %in% tmpcd$rn]
tmpcd <- merge.data.table(tmpcd,lc2splitcoord[,.(key,section)],by.x="rn",by.y="key")

stopifnot(length(unique(tmpcd$rn))==ncol(lc2))
stopifnot(all(tmpcd$rn %in% colnames(lc2)))

# unique(tmpcd$section)
tmpcd[section=="section_1",section:="s1"]
tmpcd[section=="section_2",section:="s2"]
tmpcd[,sectid:=paste0(sample_id,section)]

### for consistency's sake, we'll want the colnames (unique spot ids) of the new SPE to reflect the section id instead of the capture id.
tmpcd[,newrn:=apply(.SD,MARGIN = 1,FUN = function(x){
    gsub(x[1],pattern="^(.*_)(V.*)$",replacement=paste0("\\1",x[2]))}),.SDcols=c("rn","sectid")]

### check that the ids are the same minus the appended section info at the end of the new rn
tmpcd[sample(seq(1,nrow(tmpcd)),size=20,replace=F),.(rn,newrn,sectid)]
# nice.

### set the section id to be the SPE-required "sample_id" column values, but retain the capture area ID (ie original sample id) for sanity checks
setnames(tmpcd,"sample_id","capture_id")
setnames(tmpcd,"sectid","sample_id")

# get the new coldata into the same row order as the old SPEs col order (i.e., using the original SPE's rownames)
tmpcd <- DataFrame(tmpcd,row.names=tmpcd$rn)[colnames(lc2),]
stopifnot(all(colnames(lc2)==rownames(tmpcd)))

### annoyingly as shit, spatialExperiment doesn't propagate colname assignments into the rownames of spatialCoords. to make sure these don't get scrambled around, we also need to make the rownames actually unique for spatialCoords and then, later, systematically replace with the new section-spot ids
rownames(spatialCoords(lc2)) <- colnames(lc2)
```

### restruct. SPE accordingly. spatialexperiment in BioC 3.20 has introduced some changes to requirements on identifiers in the spatialCoords and spatial images slots, i think, so probably best to just rebuild bottom up.

### since AFAIK the actual loupe-aligned capture area images haven't been divied into their component tissue sections, we'll just put the same capture area image for both sections for now. and since we don't immediately need the highres images for anything on the R side, we just won't carry the highres images into the new SPE at all.

```{r}
### the biggest lift here will be rebuilding the @int_metadata$imgData slot. which lives in SpatialExp as DataFrame after we've made it. everything actual is in the slot lc2@int_metadata$imgData@listData though so extract that to begin with

imdat <- lc2@int_metadata$imgData@listData

# the structure of this ^  is:
# $sample_id: vector of sample_id (capture area id), repeated for every variety of image_id in the list
# $image_id: lowres, hires, detected, aligned. we don't want the hires anymore.
# $data: the actual LoadedSpatialImage objects
# $scaleFactor: idk, but it differs for each resolution in image_id and has previously further varied by the microscope used for capturing those images. so play it safe and draw using the corresponding sample•image index.

### so, iterate through each capture area and for each, build a new imagedata list with slots for each SECTION and its parent capture area's non hires images 
### having a data.table of the new coldata will make this a bit easier too. 

tmpcd.dt <- as.data.table(tmpcd)

### initialize new list. "sample_id" is the name spatialExp requires, but it's going to become our section identifier now.
newimdat <- list(sample_id=NULL,image_id=NULL,data=list(),scaleFactor=NULL)

i<-1
for (i in c(1:length(unique(imdat$sample_id)))){
    # for grabbing from the original spe's image data, the sample_id will be what is more recently the capture id:
    curcap <- unique(imdat$sample_id)[i]
    capinds <- which(imdat$sample_id==curcap)
    imginds <- which(imdat$image_id %in% c("lowres","detected","aligned"))
    imginds <- capinds[which(capinds %in% imginds)]
    # the image resolutions/alignments always occur in the order lowres, hires, detected, aligned, so their indices will also come back in that order in the above line. and then we can just use that triplet of indices to extract the corresponding image types without bothering with matching strings and shit
    
    j<-1
    # recall that the new coldata in progress has the section-specific ids already in the sample_id column
    for (j in c(1:length(unique(tmpcd.dt[capture_id==curcap]$sample_id)))){
        cursect <- unique(tmpcd.dt[capture_id==curcap,sample_id])[j]
        ## add a new triplet of lowres detected aligned
        newimdat$sample_id <- c(newimdat$sample_id,rep(cursect,length(imginds)))
        newimdat$image_id <- c(newimdat$image_id,c("lowres","detected","aligned"))
        
        ## as explained above, imginds[1] will always be the index of lowres, so that goes to 
        newimdat$data[(length(newimdat$data)+1)] <- imdat$data[imginds[1]]
        newimdat$data[(length(newimdat$data)+1)] <- imdat$data[imginds[2]]
        newimdat$data[(length(newimdat$data)+1)] <- imdat$data[imginds[3]]
        
        newimdat$scaleFactor <- c(newimdat$scaleFactor,imdat$scaleFactor[imginds])
        
        stopifnot(length(newimdat$data)==length(newimdat$sample_id))
    }
    rm(j,cursect)
}
stopifnot(all(newimdat$sample_id %in% tmpcd.dt$sample_id))
rm(i,imginds,curcap,capinds,tmpcd.dt)

```

### now we need to extract the counts matrix, match the old unique ids to the new ones in tmpcd$newrn, swap the rownames of the new coldata (tmpcd) to tmpcd$newrn, double-check they're in the same order, and then pass to SPE the following for the single-section SPE:
the new images slot
renamed counts matrix
renamed logcounts matrix
new coldata
existing rowdata
existing spatialCoords slot i think

```{r}
newct <- counts(lc2)
stopifnot(all(colnames(newct)==rownames(tmpcd)))
colnames(newct) <- tmpcd$newrn

newlct <- logcounts(lc2)
stopifnot(all(colnames(newlct)==rownames(tmpcd)))
colnames(newlct) <- tmpcd$newrn

## now reformat the new colData itself
rownames(tmpcd) <- tmpcd$newrn
colnames(tmpcd)[which(colnames(tmpcd)=="rn")] <- "original_spot_id" 
# this ^ is redundant with $key so dorp that
tmpcd$key <- NULL

# store the new spot identifiers in a column with an interpretable name
colnames(tmpcd)[which(colnames(tmpcd)=="newrn")] <- "section_spot_id" 

## reorder coldata colnames just because its what i'm used to
newcolord <- c("sample_id",
               colnames(tmpcd)[which(colnames(tmpcd)!="sample_id")])

tmpcd2 <- tmpcd[,newcolord]

### check that the spatialCoords row order of the original SPE is still correct (i.e., matches the original spot ids in the order present in tmpcd2). as long as that's the case we can just drop them right in.
newspcoord <- spatialCoords(lc2)
stopifnot(all(rownames(newspcoord)==tmpcd2$original_spot_id))
# phew
rownames(newspcoord) <- tmpcd2$section_spot_id

### checkit build it
stopifnot(all(colnames(newct)==rownames(tmpcd)))
stopifnot(all(colnames(newct)==rownames(newspcoord)))
stopifnot(all(rownames(newct)==rownames(rowData(lc2))))

## since we have the listData part of the image data, just let SPE make a blank DF in the imgData slot and then swap the new listData obj in after 
lc2split <- SpatialExperiment(
    assays=list(counts=newct),
    rowData=rowData(lc2),
    colData=tmpcd2,
    sample_id = tmpcd2$sample_id,
    spatialCoords = newspcoord)

lc2split@int_metadata$imgData@listData <- newimdat
## and S4 is more annoying than I thought. lc2split@int_metadata$imgData$nrows is still 0. so we need to update that and the other slots to be consistent with the new listData

## but we can update all the other slots on $imgData quite easily to be certain there's no fuckery later. in the original SPE, imgData's other contents are
# @rownames: NULL
# @nrows: 172 (the length of each slot's content in listData)
# @elementType: 'ANY'
# @elementMetadata: NULL
# @metadata: list() (empty list)

lc2split@int_metadata$imgData@metadata <- list()
# for nrows, we already checked on line 134 that the number of images in the listData and the number of identifiers are the same. so we can do this safely:
lc2split@int_metadata$imgData@nrows <- length(lc2split@int_metadata$imgData$sample_id)
lc2split@int_metadata$imgData@elementType <- "ANY"
lc2split@int_metadata$imgData@elementMetadata <- NULL
lc2split@int_metadata$imgData@rownames <- NULL
```

boom save it
```{r}
saveRDS(lc2split,"processed-data/06-QCed_SPE_split_to_tissSections.RDS")
```

### integrity test: make a bigol spatialLIBD vis_grid_gene (whatever gene) to make sure there arent
A) singleton spots that were misannotated in the tissue splitting
B) wonky internal things about the imgData, spatialcoords, etc that would break plotting functions

```{r}
# test from the point of loading spe
rm(list=ls())
gc(full=T)
# 

library(spatialLIBD)
library(ggrastr)
library(Cairo)

lc2s <- readRDS("processed-data/06-QCed_SPE_split_to_tissSections.RDS")

# switch rownames to symbols
rownames(lc2s) <- rowData(lc2s)$gene_name

plt <- vis_grid_gene(lc2s,geneid = "DBH",image_id = "lowres",assayname = "counts",return_plots = T,spatial=T,alpha=0.65,auto_crop = T,na_color = "orange",point_size = 1.25)

plt <- lapply(plt,function(p){
    p <- p+
        theme(axis.text.x=element_blank(),axis.text.y=element_blank(),axis.title.x=element_blank(),axis.title.y=element_blank(),title=element_text(size=9,margin = margin(0,0,0.25,0,unit = "in"),hjust=0.5),plot.margin = margin(0,0,0.2,0,unit="in"))+
        guides(color="none",fill="none")
    q <- rasterize(p,layers="Points",dev="cairo_png",dpi=450)
    return(q)
})

pdf("plots/06-split_SPE_check.pdf",height=11,width=8.5,onefile = T)
i<-1
for (i in seq(1,length(plt),9)){
    nplt <- sum(!is.na(names(plt[c(i:(i+8))])))
    do.call("grid.arrange",c(plt[c(i:(i+nplt-1))]))
}
dev.off()
```

reprod info
```{r}
sessionInfo()
sessioninfo::session_info()
```

R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] Cairo_1.6-2                 ggrastr_1.0.2              
 [3] spatialLIBD_1.18.0          BiocParallel_1.40.0        
 [5] parallelly_1.40.1           colorout_1.3-0.2           
 [7] SpatialExperiment_1.16.0    SingleCellExperiment_1.28.1
 [9] SummarizedExperiment_1.36.0 Biobase_2.66.0             
[11] GenomicRanges_1.58.0        GenomeInfoDb_1.42.1        
[13] IRanges_2.40.1              S4Vectors_0.44.0           
[15] BiocGenerics_0.52.0         MatrixGenerics_1.18.0      
[17] matrixStats_1.4.1           gridExtra_2.3              
[19] ggplot2_3.5.1               data.table_1.16.4          
[21] rlang_1.1.4                

loaded via a namespace (and not attached):
  [1] later_1.4.1              BiocIO_1.16.0            bitops_1.0-9            
  [4] filelock_1.0.3           fields_16.3              tibble_3.2.1            
  [7] XML_3.99-0.17            lifecycle_1.0.4          edgeR_4.4.1             
 [10] doParallel_1.0.17        rprojroot_2.0.4          lattice_0.22-6          
 [13] magrittr_2.0.3           limma_3.62.1             plotly_4.10.4           
 [16] sass_0.4.9               rmarkdown_2.29           jquerylib_0.1.4         
 [19] yaml_2.3.10              httpuv_1.6.15            spam_2.11-0             
 [22] sessioninfo_1.2.2        cowplot_1.1.3            DBI_1.2.3               
 [25] RColorBrewer_1.1-3       golem_0.5.1              maps_3.4.2.1            
 [28] abind_1.4-8              pkgload_1.4.0            zlibbioc_1.52.0         
 [31] purrr_1.0.2              RCurl_1.98-1.16          rappdirs_0.3.3          
 [34] GenomeInfoDbData_1.2.13  ggrepel_0.9.6            irlba_2.3.5.1           
 [37] codetools_0.2-20         DelayedArray_0.32.0      DT_0.33                 
 [40] scuttle_1.16.0           tidyselect_1.2.1         UCSC.utils_1.2.0        
 [43] farver_2.1.2             ScaledMatrix_1.14.0      viridis_0.6.5           
 [46] shinyWidgets_0.8.7       BiocFileCache_2.14.0     GenomicAlignments_1.42.0
 [49] jsonlite_1.8.9           BiocNeighbors_2.0.1      scater_1.34.0           
 [52] iterators_1.0.14         foreach_1.5.2            tools_4.4.1             
 [55] Rcpp_1.0.13-1            glue_1.8.0               SparseArray_1.6.0       
 [58] xfun_0.49                here_1.0.1               dplyr_1.1.4             
 [61] withr_3.0.2              BiocManager_1.30.25      fastmap_1.2.0           
 [64] fansi_1.0.6              digest_0.6.37            rsvd_1.0.5              
 [67] R6_2.5.1                 mime_0.12                colorspace_2.1-1        
 [70] RSQLite_2.3.9            config_0.3.2             utf8_1.2.4              
 [73] tidyr_1.3.1              generics_0.1.3           rtracklayer_1.66.0      
 [76] httr_1.4.7               htmlwidgets_1.6.4        S4Arrays_1.6.0          
 [79] pkgconfig_2.0.3          gtable_0.3.6             blob_1.2.4              
 [82] XVector_0.46.0           htmltools_0.5.8.1        dotCall64_1.2           
 [85] scales_1.3.0             png_0.1-8                attempt_0.3.1           
 [88] knitr_1.49               rstudioapi_0.17.1        rjson_0.2.23            
 [91] curl_6.0.1               cachem_1.1.0             BiocVersion_3.20.0      
 [94] parallel_4.4.1           vipor_0.4.7              AnnotationDbi_1.68.0    
 [97] restfulr_0.0.15          pillar_1.9.0             grid_4.4.1              
[100] vctrs_0.6.5              promises_1.3.2           BiocSingular_1.22.0     
[103] dbplyr_2.5.0             beachmat_2.22.0          xtable_1.8-4            
[106] beeswarm_0.4.0           paletteer_1.6.0          evaluate_1.0.1          
[109] magick_2.8.5             cli_3.6.3                locfit_1.5-9.10         
[112] compiler_4.4.1           Rsamtools_2.22.0         crayon_1.5.3            
[115] labeling_0.4.3           rematch2_2.1.2           ggbeeswarm_0.7.2        
[118] viridisLite_0.4.2        munsell_0.5.1            Biostrings_2.74.0       
[121] lazyeval_0.2.2           Matrix_1.7-1             ExperimentHub_2.14.0    
[124] benchmarkme_1.0.8        bit64_4.5.2              KEGGREST_1.46.0         
[127] statmod_1.5.0            shiny_1.9.1              AnnotationHub_3.14.0    
[130] memoise_2.0.1            bslib_0.8.0              benchmarkmeData_1.0.4   
[133] bit_4.5.0.1             
> sessioninfo::session_info()
─ Session info ──────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.7.1
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-12-17
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ──────────────────────────────────────────────────────────────────────
 ! package              * version   date (UTC) lib source
   abind                  1.4-8     2024-09-12 [1] CRAN (R 4.4.1)
   AnnotationDbi          1.68.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   AnnotationHub          3.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   attempt                0.3.1     2020-05-03 [1] CRAN (R 4.4.0)
   beachmat               2.22.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   beeswarm               0.4.0     2021-06-01 [1] CRAN (R 4.4.0)
   benchmarkme            1.0.8     2022-06-12 [1] CRAN (R 4.4.0)
   benchmarkmeData        1.0.4     2020-04-23 [1] CRAN (R 4.4.0)
   Biobase              * 2.66.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   BiocFileCache          2.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   BiocGenerics         * 0.52.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   BiocIO                 1.16.0    2024-12-12 [1] Github (Bioconductor/BiocIO@8e80624)
   BiocManager            1.30.25   2024-08-28 [1] CRAN (R 4.4.1)
   BiocNeighbors          2.0.1     2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
   BiocParallel         * 1.40.0    2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
   BiocSingular           1.22.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   BiocVersion            3.20.0    2024-05-04 [1] Bioconductor 3.20 (R 4.4.0)
   Biostrings             2.74.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   bit                    4.5.0.1   2024-12-03 [1] CRAN (R 4.4.1)
   bit64                  4.5.2     2024-09-22 [1] CRAN (R 4.4.1)
   bitops                 1.0-9     2024-10-03 [1] CRAN (R 4.4.1)
   blob                   1.2.4     2023-03-17 [1] CRAN (R 4.4.0)
   bslib                  0.8.0     2024-07-29 [1] CRAN (R 4.4.0)
   cachem                 1.1.0     2024-05-16 [1] CRAN (R 4.4.0)
   Cairo                * 1.6-2     2023-11-28 [1] CRAN (R 4.4.0)
 P cli                    3.6.3     2024-06-21 [2] CRAN (R 4.4.0)
   codetools              0.2-20    2024-03-31 [1] CRAN (R 4.4.1)
   colorout             * 1.3-0.2   2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace             2.1-1     2024-07-26 [1] CRAN (R 4.4.0)
   config                 0.3.2     2023-08-30 [1] CRAN (R 4.4.0)
   cowplot                1.1.3     2024-01-22 [1] CRAN (R 4.4.0)
   crayon                 1.5.3     2024-06-20 [1] CRAN (R 4.4.0)
   curl                   6.0.1     2024-11-14 [1] CRAN (R 4.4.1)
   data.table           * 1.16.4    2024-12-06 [1] CRAN (R 4.4.1)
   DBI                    1.2.3     2024-06-02 [1] CRAN (R 4.4.0)
   dbplyr                 2.5.0     2024-03-19 [1] CRAN (R 4.4.0)
   DelayedArray           0.32.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   digest                 0.6.37    2024-08-19 [1] CRAN (R 4.4.1)
   doParallel             1.0.17    2022-02-07 [1] CRAN (R 4.4.1)
   dotCall64              1.2       2024-10-04 [1] CRAN (R 4.4.1)
   dplyr                  1.1.4     2023-11-17 [1] CRAN (R 4.4.0)
   DT                     0.33      2024-04-04 [1] CRAN (R 4.4.0)
   edgeR                  4.4.1     2024-12-02 [1] Bioconductor 3.20 (R 4.4.2)
   evaluate               1.0.1     2024-10-10 [1] CRAN (R 4.4.1)
   ExperimentHub          2.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   fansi                  1.0.6     2023-12-08 [1] CRAN (R 4.4.0)
   farver                 2.1.2     2024-05-13 [1] CRAN (R 4.4.0)
   fastmap                1.2.0     2024-05-15 [1] CRAN (R 4.4.0)
   fields                 16.3      2024-09-30 [1] CRAN (R 4.4.1)
   filelock               1.0.3     2023-12-11 [1] CRAN (R 4.4.0)
   foreach                1.5.2     2022-02-02 [1] CRAN (R 4.4.0)
   generics               0.1.3     2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb         * 1.42.1    2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
   GenomeInfoDbData       1.2.13    2024-12-12 [1] Bioconductor
   GenomicAlignments      1.42.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   GenomicRanges        * 1.58.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   ggbeeswarm             0.7.2     2023-04-29 [1] CRAN (R 4.4.0)
   ggplot2              * 3.5.1     2024-04-23 [1] CRAN (R 4.4.0)
   ggrastr              * 1.0.2     2023-06-01 [1] CRAN (R 4.4.0)
   ggrepel                0.9.6     2024-09-07 [1] CRAN (R 4.4.1)
   glue                   1.8.0     2024-09-30 [1] CRAN (R 4.4.1)
   golem                  0.5.1     2024-08-27 [1] CRAN (R 4.4.1)
   gridExtra            * 2.3       2017-09-09 [1] CRAN (R 4.4.0)
   gtable                 0.3.6     2024-10-25 [1] CRAN (R 4.4.1)
   here                   1.0.1     2020-12-13 [1] CRAN (R 4.4.0)
   htmltools              0.5.8.1   2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets            1.6.4     2023-12-06 [1] CRAN (R 4.4.0)
   httpuv                 1.6.15    2024-03-26 [1] CRAN (R 4.4.0)
   httr                   1.4.7     2023-08-15 [1] CRAN (R 4.4.0)
   IRanges              * 2.40.1    2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
   irlba                  2.3.5.1   2022-10-03 [1] CRAN (R 4.4.0)
   iterators              1.0.14    2022-02-05 [1] CRAN (R 4.4.0)
   jquerylib              0.1.4     2021-04-26 [1] CRAN (R 4.4.0)
   jsonlite               1.8.9     2024-09-20 [1] CRAN (R 4.4.1)
   KEGGREST               1.46.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   knitr                  1.49      2024-11-08 [1] CRAN (R 4.4.1)
   labeling               0.4.3     2023-08-29 [1] CRAN (R 4.4.0)
   later                  1.4.1     2024-11-27 [1] CRAN (R 4.4.1)
   lattice                0.22-6    2024-03-20 [1] CRAN (R 4.4.1)
   lazyeval               0.2.2     2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle              1.0.4     2023-11-07 [1] CRAN (R 4.4.0)
   limma                  3.62.1    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   locfit                 1.5-9.10  2024-06-24 [1] CRAN (R 4.4.0)
   magick                 2.8.5     2024-09-20 [1] CRAN (R 4.4.1)
   magrittr               2.0.3     2022-03-30 [1] CRAN (R 4.4.0)
   maps                   3.4.2.1   2024-11-10 [1] CRAN (R 4.4.1)
   Matrix                 1.7-1     2024-10-18 [1] CRAN (R 4.4.1)
   MatrixGenerics       * 1.18.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   matrixStats          * 1.4.1     2024-09-08 [1] CRAN (R 4.4.1)
   memoise                2.0.1     2021-11-26 [1] CRAN (R 4.4.0)
   mime                   0.12      2021-09-28 [1] CRAN (R 4.4.0)
   munsell                0.5.1     2024-04-01 [1] CRAN (R 4.4.0)
   paletteer              1.6.0     2024-01-21 [1] CRAN (R 4.4.0)
   parallelly           * 1.40.1    2024-12-04 [1] CRAN (R 4.4.1)
   pillar                 1.9.0     2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig              2.0.3     2019-09-22 [1] CRAN (R 4.4.0)
   pkgload                1.4.0     2024-06-28 [1] CRAN (R 4.4.1)
   plotly                 4.10.4    2024-01-13 [1] CRAN (R 4.4.0)
   png                    0.1-8     2022-11-29 [1] CRAN (R 4.4.0)
   promises               1.3.2     2024-11-28 [1] CRAN (R 4.4.1)
   purrr                  1.0.2     2023-08-10 [1] CRAN (R 4.4.0)
   R6                     2.5.1     2021-08-19 [1] CRAN (R 4.4.0)
   rappdirs               0.3.3     2021-01-31 [1] CRAN (R 4.4.0)
   RColorBrewer           1.1-3     2022-04-03 [1] CRAN (R 4.4.0)
   Rcpp                   1.0.13-1  2024-11-02 [1] CRAN (R 4.4.1)
   RCurl                  1.98-1.16 2024-07-11 [1] CRAN (R 4.4.0)
   rematch2               2.1.2     2020-05-01 [1] CRAN (R 4.4.0)
   restfulr               0.0.15    2022-06-16 [1] CRAN (R 4.4.0)
   rjson                  0.2.23    2024-09-16 [1] CRAN (R 4.4.1)
 P rlang                * 1.1.4     2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown              2.29      2024-11-04 [1] CRAN (R 4.4.1)
   rprojroot              2.0.4     2023-11-05 [1] CRAN (R 4.4.0)
   Rsamtools              2.22.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   RSQLite                2.3.9     2024-12-03 [1] CRAN (R 4.4.1)
   rstudioapi             0.17.1    2024-10-22 [1] CRAN (R 4.4.1)
   rsvd                   1.0.5     2021-04-16 [1] CRAN (R 4.4.0)
   rtracklayer            1.66.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   S4Arrays               1.6.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   S4Vectors            * 0.44.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   sass                   0.4.9     2024-03-15 [1] CRAN (R 4.4.0)
   ScaledMatrix           1.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   scales                 1.3.0     2023-11-28 [1] CRAN (R 4.4.0)
   scater                 1.34.0    2024-10-29 [1] Bioconductor 3.20 (R 4.4.1)
   scuttle                1.16.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   sessioninfo            1.2.2     2021-12-06 [1] CRAN (R 4.4.0)
   shiny                  1.9.1     2024-08-01 [1] CRAN (R 4.4.0)
   shinyWidgets           0.8.7     2024-09-23 [1] CRAN (R 4.4.1)
   SingleCellExperiment * 1.28.1    2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
   spam                   2.11-0    2024-10-03 [1] CRAN (R 4.4.1)
   SparseArray            1.6.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   SpatialExperiment    * 1.16.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   spatialLIBD          * 1.18.0    2024-11-07 [1] Bioconductor 3.20 (R 4.4.1)
   statmod                1.5.0     2023-01-06 [1] CRAN (R 4.4.0)
   SummarizedExperiment * 1.36.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   tibble                 3.2.1     2023-03-20 [1] CRAN (R 4.4.0)
   tidyr                  1.3.1     2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect             1.2.1     2024-03-11 [1] CRAN (R 4.4.0)
   UCSC.utils             1.2.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   utf8                   1.2.4     2023-10-22 [1] CRAN (R 4.4.0)
   vctrs                  0.6.5     2023-12-01 [1] CRAN (R 4.4.0)
   vipor                  0.4.7     2023-12-18 [1] CRAN (R 4.4.0)
   viridis                0.6.5     2024-01-29 [1] CRAN (R 4.4.0)
   viridisLite            0.4.2     2023-05-02 [1] CRAN (R 4.4.0)
   withr                  3.0.2     2024-10-28 [1] CRAN (R 4.4.1)
   xfun                   0.49      2024-10-31 [1] CRAN (R 4.4.1)
   XML                    3.99-0.17 2024-06-25 [1] CRAN (R 4.3.3)
   xtable                 1.8-4     2019-04-21 [1] CRAN (R 4.4.0)
   XVector                0.46.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   yaml                   2.3.10    2024-07-26 [1] CRAN (R 4.4.0)
   zlibbioc               1.52.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

─────────────────────────────────────────────────────────────────────────────────
