---
title: "01a-Make_Allcluster_DGE_withMinSpots_reqmnt"
author: "Bernie Mulvey"
date: "2025-05-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(SpatialExperiment)
library(dreamlet)
library(limma)
library(edgeR)
library(gridExtra)
library(ggplot2)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")
library(parallel) # use PSOCK clusters for best performance here

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

 
## load SPE and assign hdg_svg_2575_louv_res1 clusterings as autoannotated
```{r}
lc3 <- readRDS("processed-data/06-countsOnly_noImgData_QCed_SPE_split_to_tissSections.RDS")

louvs <- readRDS("processed-data/07_featureSelection_dimred_harmony_clustering/04-Initial_louvLeid_res0.5-1-2_clusterings.RDS")
louvs <- louvs[["HARMONYlmbna_HDG_SVG_2575"]][,.(rn,snnHARMONYlmbna_HDG_SVG_2575_louv_res1)]
setnames(louvs,2,"clusid")

finalann <- fread("processed-data/08_validitycheck_25hdg75svg_louv1/10-25hdg75svg_louv1_annots.txt")

louvs <- merge.data.table(louvs,finalann,by="clusid")
louvs <- DataFrame(louvs,row.names=louvs$rn)[colnames(lc3),]

colLabels(lc3) <- louvs$anno
```

### Append global ancestry estimates -- we need them later anyhow, and it seems edgeR retains differing numbers of genes depending how you specify the design matrix, so we'll just do it here.
```{r}
lc3$brnum[lc3$brnum=="Br6119(re-dis)"] <- "Br6119"
flares <- fread("raw-data/ancestryAndHaplotypes/FLARE/global_ancestry_estimates.txt")

flares <- flares[ID %in% lc3$brnum,.(ID,YRI)]
tmpcd <- as.data.table(colData(lc3),keep.rownames=T)
tmpcd <- merge.data.table(tmpcd,flares,by.x="brnum",by.y="ID",all.x=T)
tmpcd <- DataFrame(tmpcd,row.names=tmpcd$rn)[colnames(lc3),]

colData(lc3) <- tmpcd
rm(tmpcd,flares)
```

## Drop Br5517 (unknown anatomy/orientation), Br5276 (LC has been depleted by QC), Br5712 (unknown anatomy/orientation). Keep 6297; no LC but anatomy is right in vicinity.
```{r}
lc3 <- lc3[,!(lc3$brnum %in% paste0("Br",c(5517,5276,5712)))]
```

### to model all data jointly, we will ultimately need a cluster-sex-genotype grouping variable. then to run sex DE for one cluster we would do e.g. (LC_E2hetMale + LC_E2homMale + LC_E4hetMale + LC_E4homMale) - (LC_E2hetFem + LC_E2homFem + LC_E4hetFem + LC_E4homFem). we can get most of this made now
```{r}
# simplify APOE labels and make sexapo groups
lc3$APOE <- gsub(lc3$APOE,pattern="\\/ ",replacement="_")
lc3$APOE[lc3$APOE=="E4_E4"] <- "E4hom"
lc3$APOE[lc3$APOE=="E3_E4"] <- "E4het"
lc3$APOE[lc3$APOE=="E2_E3"] <- "E2het"
lc3$APOE[lc3$APOE=="E2_E2"] <- "E2hom"

lc3$geno_group <- paste0(lc3$Sex,"_",lc3$APOE)
```

## for plotting pseudobulk MDS by Visium slide, we need to edit the colData(spe)$Visium_slide to JUST be the slide (it still has the slide capture area ID appended to it)
```{r}
colData(lc3)$Visium_slide <- gsub(colData(lc3)$Visium_slide,pattern="^(V.....-...)_..$",replacement="\\1")

# also, things are going to get weird if we have dashes in our sample names, so remove those too
colData(lc3)$sample_id <- gsub(colData(lc3)$sample_id,pattern="-",replacement="_")

rm(louvs,finalann)
```

## pre-checks: are there clusters with <10 spots in over half the samples or over half of the smaller sex group? (ie clusters to exclude?)
```{r}
## only retain clusters for which at least 43 samples (of 85 total, i.e over half) of â‰¥10 spots each are available. do we need to drop any?
check <- as.data.table(table(colData(lc3)[,c("label","sample_id")]))
check <- check[N>=10]
check[,.N,by="label"][N<43,label]
# nope

## drop clusters where there are 20+ samples, but under 18 samples for one of the sexes (i.e., < half of the samples from the smaller group--females w 35 sections total). are there any?
check <- as.data.table(table(colData(lc3)[,c("label","Sex","sample_id")]))
check <- check[Sex=="F"&N>=10]
check[,.N,by="label"][N<18,label]
# nope
rm(check)
```

#### TOP LEVEL CLUSTER DGELists ####

# make pseudobulk object (dreamlet is v. fast at this) by sample-cluster
```{r}
ag <- dreamlet::aggregateToPseudoBulk(lc3,sample_id="sample_id",cluster_id = "label",BPPARAM = MulticoreParam(6))

# dreamlet outputs shared metadata for the samples - grab this to build one big metadata table from
tmpcd <- as.data.table(copy(colData(ag)),keep.rownames=T)
setnames(tmpcd,"rn","sample_id")

# extract number of spots in a pseudobulk sample for filtering. this is in the inverse of the format we want, list entries are sample and subentries are cluster, so rearrange to cluster listed by sample
agn <- ag@int_colData$n_cells
agn <- lapply(agn,function(x){as.data.frame(t(x))})
agn <- rbindlist(agn,idcol="samp")
agn <- melt(agn,id.vars="samp",variable.name ="label")
agn[,samp:=paste0(label,"_",samp)]
agn <- agn[value>=10]
## ^ keep this to append to metadata

## extract pseudobulk matrices
agm <- as.list(assays(ag))
agm <- mapply(a=agm,n=names(agm),SIMPLIFY=FALSE,function(a,n){
  colnames(a) <- paste0(n,"_",colnames(a))
  a <- a[,colnames(a) %in% agn$samp]
  a <- a[,colSums(a)>0]
  return(a)
})


i<-1
for (i in c(1:length(agm))){
  if(i==1){
    megamat <- copy(as.data.table(as.data.frame(as.matrix(agm[[i]])),keep.rownames=T))
  }
  else{
    tmpmat <- copy(as.data.table(as.data.frame(as.matrix(agm[[i]])),keep.rownames=T))
    megamat <- merge.data.table(megamat,tmpmat,by="rn",all=T)
  }
}
rm(i,tmpmat)

megamat <- data.frame(megamat,row.names=megamat$rn)
megamat$rn <- NULL

# extract colData to feed into DGEList as the "samples" data
# important: add column to indicate cluster and append cluster to sample name
cds <- mapply(a=agm,n=names(agm),SIMPLIFY=FALSE,FUN=function(a,n){
  # get pbulk sample names
  cn <- as.data.table(colnames(a))
  setnames(cn,1,"clus_sample_id")
  cn[,label:=n]
  
  ## get orig sample names
  cn[,sample_id:=gsub(clus_sample_id,pattern=paste0(n,"_"),replacement="")]
  tmpcd2 <- copy(tmpcd[sample_id %in% cn$sample_id])
  tmpcd2 <- merge.data.table(tmpcd2,cn,by="sample_id")
  tmpcd2[,clus_geno_group:=paste0(label,"_",geno_group)]
  return(tmpcd2)
})

megacd <- rbindlist(cds)
megacd <- merge.data.table(megacd,agn[,.(samp,value)],by.x="clus_sample_id",by.y="samp")
setnames(megacd,"value","n_spots")

megacd <- data.frame(megacd,row.names=megacd$clus_sample_id)
megacd <- megacd[colnames(megamat),]  

# clean up
rm(cds,tmpcd,agm,agcd,ag,lc3)
gc(full=T)
```


## make the single DGEList; for now, we will filter on a model-wise basis in downstream scripts.
```{r}
## drop genes with 0 counts
keep <- rowSums(megamat)>0
megamat <- megamat[keep,]
gdat <- as.data.frame(rowData(lc3))[rownames(megamat),]
rm(tmprd)


dgel <- DGEList(counts=megamat,samples=megacd,genes=as.data.frame(gdat))

# alternative: filter with grouping by label rather than a specific design (not shown)
```

## save
```{r}
saveRDS(dgel,"processed-data/12_DEanalyses_removedsampsAndFinalNMseg/01a-DGElist_allClust_10spotmin_NOfilt.RDS")
```

## make MDS plots per pseudobulk cluster per potential variable of interest
## need to specify gene selection = "common" or itll try and compare the 600whatever samples pairwise and never get done.
```{r}
## here, use a model of ~0+label for low xpr filtering
des <- model.matrix(~0+label,data=dgel$samples)
keep <- filterByExpr.DGEList(dgel,design = des)
dgel <- dgel[keep,,keep.lib.sizes=FALSE]
dgel <- edgeR::calcNormFactors(dgel,method = "TMM")

## plots for all variables to check
  
## 1,2 ##
covcols <- c("Sex","APOE","Visium_slide","brnum","label","Age","YRI","n_spots","lib.size")
res <- plotMDS(dgel, dim.plot = c(1,2),pch=18)
dev.off()
resdat <- as.data.table(cbind(res$x,res$y,rownames(dgel$samples),dgel$samples[,covcols]))
setnames(resdat,c("PC1","PC2","sample_id",covcols))
resdat[,log2_n_spots:=log2(n_spots)]
resdat[,log2_lib_size:=log2(lib.size)]
covcols2 <- c("Sex","APOE","Visium_slide","brnum","label","Age","YRI","log2_n_spots","log2_lib_size")

lapply(covcols2,FUN=function(cov){
  p <- ggplot(resdat,aes(x=as.numeric(PC1),y=as.numeric(PC2),col=get(cov))) + 
      geom_point() + 
      labs(x=paste0("MDS 1 (",100*round(res[[3]][1],digits=2),"%)"),
          y=paste0("MDS 2 (",100*round(res[[3]][2],digits=2),"%)"),
          col=cov) +
      theme_bw()


  if(is.numeric(resdat[,get(cov)])){
    p <- p+scale_color_viridis_c()}
  
  png(paste0("plots/12_DEanalyses_removedsampsAndFinalNMseg/01_pseudobulk_cluster_MDSes/01a-Allclust-",cov,"_1_2_LabelModelFilt.png"),width=1000,height=800)
  print(p)
  dev.off()
})
  
## 1,3 ##
res <- plotMDS(dgel, dim.plot = c(1,3),pch=18)
dev.off()
resdat <- as.data.table(cbind(res$x,res$y,rownames(dgel$samples),dgel$samples[,covcols]))
setnames(resdat,c("PC1","PC3","sample_id",covcols))
resdat[,log2_n_spots:=log2(n_spots)]
resdat[,log2_lib_size:=log2(lib.size)]

lapply(covcols2,FUN=function(cov){
  p <- ggplot(resdat,aes(x=as.numeric(PC1),y=as.numeric(PC3),col=get(cov))) + 
      geom_point() + 
      labs(x=paste0("MDS 1 (",100*round(res[[3]][1],digits=2),"%)"),
          y=paste0("MDS 3 (",100*round(res[[3]][3],digits=2),"%)"),
          col=cov)+
      theme_bw()
      
  if(is.numeric(resdat[,get(cov)])){
    p <- p+scale_color_viridis_c()}

  png(paste0("plots/12_DEanalyses_removedsampsAndFinalNMseg/01_pseudobulk_cluster_MDSes/01a-Allclust_",cov,"_1_3_LabelModelFilt.png"),width=1000,height=800)
  print(p)
  dev.off()
})

## 2,3 ##
res <- plotMDS(dgel, dim.plot = c(2,3),pch=18)
dev.off()
resdat <- as.data.table(cbind(res$x,res$y,rownames(dgel$samples),dgel$samples[,covcols]))
setnames(resdat,c("PC2","PC3","sample_id",covcols))
resdat[,log2_n_spots:=log2(n_spots)]
resdat[,log2_lib_size:=log2(lib.size)]

lapply(covcols2,FUN=function(cov){
  p <-ggplot(resdat,aes(x=as.numeric(PC2),y=as.numeric(PC3),col=get(cov))) + 
        geom_point() + 
        labs(x=paste0("MDS 2 (",100*round(res[[3]][2],digits=2),"%)"),
            y=paste0("MDS 3 (",100*round(res[[3]][3],digits=2),"%)"),
            col=cov) +
        theme_bw()
  
  if(is.numeric(resdat[,get(cov)])){
      p <- p+scale_color_viridis_c()}

  png(paste0("plots/12_DEanalyses_removedsampsAndFinalNMseg/01_pseudobulk_cluster_MDSes/01a-Allclust-",cov,"_2_3_LabelModelFilt.png"),width=1000,height=800)
  print(p)
  dev.off()
})

gc(full=T)
```

## lib size is a very srs driver of variance, especially for the multi-cell-type clusters (astro_oligo, oligo_astro, etc). 

# rep inf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.2

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  stats4    stats    
[4] graphics  grDevices utils    
[7] methods   base     

other attached packages:
 [1] parallelly_1.43.0          
 [2] colorout_1.3-0.2           
 [3] gridExtra_2.3              
 [4] edgeR_4.4.2                
 [5] dreamlet_1.4.1             
 [6] variancePartition_1.36.3   
 [7] BiocParallel_1.40.0        
 [8] limma_3.62.2               
 [9] ggplot2_3.5.2              
[10] SpatialExperiment_1.16.0   
[11] SingleCellExperiment_1.28.1
[12] SummarizedExperiment_1.36.0
[13] Biobase_2.66.0             
[14] GenomicRanges_1.58.0       
[15] GenomeInfoDb_1.42.3        
[16] IRanges_2.40.1             
[17] S4Vectors_0.44.0           
[18] BiocGenerics_0.52.0        
[19] MatrixGenerics_1.18.1      
[20] matrixStats_1.5.0          
[21] data.table_1.17.0          
[22] rlang_1.1.5                

loaded via a namespace (and not attached):
  [1] splines_4.4.3            
  [2] bitops_1.0-9             
  [3] tibble_3.2.1             
  [4] graph_1.84.1             
  [5] XML_3.99-0.18            
  [6] lifecycle_1.0.4          
  [7] Rdpack_2.6.3             
  [8] mixsqp_0.3-54            
  [9] rprojroot_2.0.4          
 [10] lattice_0.22-6           
 [11] MASS_7.3-65              
 [12] backports_1.5.0          
 [13] magrittr_2.0.3           
 [14] metafor_4.8-0            
 [15] sessioninfo_1.2.3        
 [16] DBI_1.2.3                
 [17] minqa_1.2.8              
 [18] RColorBrewer_1.1-3       
 [19] abind_1.4-8              
 [20] zlibbioc_1.52.0          
 [21] EnvStats_3.0.0           
 [22] purrr_1.0.4              
 [23] rmeta_3.0                
 [24] msigdbr_10.0.1           
 [25] RCurl_1.98-1.17          
 [26] GenomeInfoDbData_1.2.13  
 [27] ggrepel_0.9.6            
 [28] pbkrtest_0.5.3           
 [29] irlba_2.3.5.1            
 [30] annotate_1.84.0          
 [31] DelayedMatrixStats_1.28.1
 [32] codetools_0.2-20         
 [33] DelayedArray_0.32.0      
 [34] tidyselect_1.2.1         
 [35] UCSC.utils_1.2.0         
 [36] farver_2.1.2             
 [37] lme4_1.1-36              
 [38] mathjaxr_1.8-0           
 [39] jsonlite_2.0.0           
 [40] iterators_1.0.14         
 [41] tools_4.4.3              
 [42] progress_1.2.3           
 [43] Rcpp_1.0.14              
 [44] zenith_1.8.0             
 [45] glue_1.8.0               
 [46] SparseArray_1.6.2        
 [47] xfun_0.52                
 [48] here_1.0.1               
 [49] dplyr_1.1.4              
 [50] withr_3.0.2              
 [51] numDeriv_2016.8-1.1      
 [52] fastmap_1.2.0            
 [53] boot_1.3-31              
 [54] caTools_1.18.3           
 [55] digest_0.6.37            
 [56] truncnorm_1.0-9          
 [57] R6_2.6.1                 
 [58] colorspace_2.1-1         
 [59] scattermore_1.2          
 [60] gtools_3.9.5             
 [61] dichromat_2.0-0.1        
 [62] RSQLite_2.3.9            
 [63] RhpcBLASctl_0.23-42      
 [64] tidyr_1.3.1              
 [65] generics_0.1.3           
 [66] corpcor_1.6.10           
 [67] prettyunits_1.2.0        
 [68] httr_1.4.7               
 [69] S4Arrays_1.6.0           
 [70] pkgconfig_2.0.3          
 [71] gtable_0.3.6             
 [72] blob_1.2.4               
 [73] XVector_0.46.0           
 [74] remaCor_0.0.18           
 [75] zigg_0.0.2               
 [76] GSEABase_1.68.0          
 [77] scales_1.4.0             
 [78] png_0.1-8                
 [79] fANCOVA_0.6-1            
 [80] reformulas_0.4.0         
 [81] ashr_2.2-63              
 [82] knitr_1.50               
 [83] reshape2_1.4.4           
 [84] rjson_0.2.23             
 [85] nlme_3.1-167             
 [86] nloptr_2.2.1             
 [87] cachem_1.1.0             
 [88] stringr_1.5.1            
 [89] KernSmooth_2.23-26       
 [90] vipor_0.4.7              
 [91] metadat_1.4-0            
 [92] AnnotationDbi_1.68.0     
 [93] pillar_1.10.2            
 [94] grid_4.4.3               
 [95] vctrs_0.6.5              
 [96] gplots_3.2.0             
 [97] mashr_0.2.79             
 [98] xtable_1.8-4             
 [99] beeswarm_0.4.0           
[100] Rgraphviz_2.50.0         
[101] evaluate_1.0.3           
[102] KEGGgraph_1.66.0         
[103] invgamma_1.1             
[104] magick_2.8.6             
[105] mvtnorm_1.3-3            
[106] cli_3.6.4                
[107] locfit_1.5-9.12          
[108] compiler_4.4.3           
[109] crayon_1.5.3             
[110] SQUAREM_2021.1           
[111] plyr_1.8.9               
[112] ggbeeswarm_0.7.2         
[113] stringi_1.8.7            
[114] assertthat_0.2.1         
[115] babelgene_22.9           
[116] lmerTest_3.1-3           
[117] Biostrings_2.74.1        
[118] aod_1.3.3                
[119] Matrix_1.7-3             
[120] hms_1.1.3                
[121] sparseMatrixStats_1.18.0 
[122] bit64_4.6.0-1            
[123] KEGGREST_1.46.0          
[124] statmod_1.5.0            
[125] rbibutils_2.3            
[126] Rfast_2.1.5.1            
[127] broom_1.0.7              
[128] memoise_2.0.1            
[129] RcppParallel_5.1.10      
[130] bit_4.6.0                
[131] EnrichmentBrowser_2.36.0 
> sessioninfo::session_info()
â”€ Session info â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sequoia 15.3.2
 system   aarch64, darwin20
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-05-13
 pandoc   NA
 quarto   1.6.40 @ /Applications/Positron.app/Contents/Resources/app/quarto/bin/quarto

â”€ Packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
 !  package              * version    date (UTC) lib source
    abind                  1.4-8      2024-09-12 [1] CRAN (R 4.4.1)
    annotate               1.84.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    AnnotationDbi          1.68.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    aod                    1.3.3      2023-12-13 [1] CRAN (R 4.4.0)
    ashr                   2.2-63     2023-08-21 [1] CRAN (R 4.4.0)
    assertthat             0.2.1      2019-03-21 [1] CRAN (R 4.4.0)
    babelgene              22.9       2022-09-29 [1] CRAN (R 4.4.0)
    backports              1.5.0      2024-05-23 [1] CRAN (R 4.4.0)
    beeswarm               0.4.0      2021-06-01 [1] CRAN (R 4.4.0)
    Biobase              * 2.66.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocParallel         * 1.40.0     2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    Biostrings             2.74.1     2024-12-16 [1] Bioconductor 3.20 (R 4.4.2)
    bit                    4.6.0      2025-03-06 [1] CRAN (R 4.4.1)
    bit64                  4.6.0-1    2025-01-16 [1] CRAN (R 4.4.1)
    bitops                 1.0-9      2024-10-03 [1] CRAN (R 4.4.1)
    blob                   1.2.4      2023-03-17 [1] CRAN (R 4.4.0)
    boot                   1.3-31     2024-08-28 [1] CRAN (R 4.4.3)
    broom                  1.0.7      2024-09-26 [1] CRAN (R 4.4.1)
    cachem                 1.1.0      2024-05-16 [1] CRAN (R 4.4.0)
    caTools                1.18.3     2024-09-04 [1] CRAN (R 4.4.1)
 VP cli                    3.6.4      2025-04-23 [2] CRAN (R 4.4.1) (on disk 3.6.5)
    codetools              0.2-20     2024-03-31 [1] CRAN (R 4.4.3)
    colorout             * 1.3-0.2    2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    colorspace             2.1-1      2024-07-26 [1] CRAN (R 4.4.0)
    corpcor                1.6.10     2021-09-16 [1] CRAN (R 4.4.0)
    crayon                 1.5.3      2024-06-20 [1] CRAN (R 4.4.0)
    data.table           * 1.17.0     2025-02-22 [1] CRAN (R 4.4.1)
    DBI                    1.2.3      2024-06-02 [1] CRAN (R 4.4.0)
    DelayedArray           0.32.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    DelayedMatrixStats     1.28.1     2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    dichromat              2.0-0.1    2022-05-02 [1] CRAN (R 4.4.0)
    digest                 0.6.37     2024-08-19 [1] CRAN (R 4.4.1)
    dplyr                  1.1.4      2023-11-17 [1] CRAN (R 4.4.0)
    dreamlet             * 1.4.1      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    edgeR                * 4.4.2      2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    EnrichmentBrowser      2.36.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    EnvStats               3.0.0      2024-08-24 [1] CRAN (R 4.4.1)
    evaluate               1.0.3      2025-01-10 [1] CRAN (R 4.4.1)
    fANCOVA                0.6-1      2020-11-13 [1] CRAN (R 4.4.0)
    farver                 2.1.2      2024-05-13 [1] CRAN (R 4.4.0)
    fastmap                1.2.0      2024-05-15 [1] CRAN (R 4.4.0)
    generics               0.1.3      2022-07-05 [1] CRAN (R 4.4.0)
    GenomeInfoDb         * 1.42.3     2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13     2024-12-12 [1] Bioconductor
    GenomicRanges        * 1.58.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ggbeeswarm             0.7.2      2023-04-29 [1] CRAN (R 4.4.0)
    ggplot2              * 3.5.2      2025-04-09 [1] CRAN (R 4.4.1)
    ggrepel                0.9.6      2024-09-07 [1] CRAN (R 4.4.1)
    glue                   1.8.0      2024-09-30 [1] CRAN (R 4.4.1)
    gplots                 3.2.0      2024-10-05 [1] CRAN (R 4.4.1)
    graph                  1.84.1     2025-01-02 [1] Bioconductor 3.20 (R 4.4.2)
    gridExtra            * 2.3        2017-09-09 [1] CRAN (R 4.4.0)
    GSEABase               1.68.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    gtable                 0.3.6      2024-10-25 [1] CRAN (R 4.4.1)
    gtools                 3.9.5      2023-11-20 [1] CRAN (R 4.4.0)
    here                   1.0.1      2020-12-13 [1] CRAN (R 4.4.0)
    hms                    1.1.3      2023-03-21 [1] CRAN (R 4.4.0)
    httr                   1.4.7      2023-08-15 [1] CRAN (R 4.4.0)
    invgamma               1.1        2017-05-07 [1] CRAN (R 4.4.0)
    IRanges              * 2.40.1     2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    irlba                  2.3.5.1    2022-10-03 [1] CRAN (R 4.4.0)
    iterators              1.0.14     2022-02-05 [1] CRAN (R 4.4.0)
    jsonlite               2.0.0      2025-03-27 [1] CRAN (R 4.4.1)
    KEGGgraph              1.66.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    KEGGREST               1.46.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    KernSmooth             2.23-26    2025-01-01 [1] CRAN (R 4.4.3)
    knitr                  1.50       2025-03-16 [1] CRAN (R 4.4.2)
    lattice                0.22-6     2024-03-20 [1] CRAN (R 4.4.3)
    lifecycle              1.0.4      2023-11-07 [1] CRAN (R 4.4.0)
    limma                * 3.62.2     2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    lme4                   1.1-36     2025-01-11 [1] CRAN (R 4.4.1)
    lmerTest               3.1-3      2020-10-23 [1] CRAN (R 4.4.0)
    locfit                 1.5-9.12   2025-03-05 [1] CRAN (R 4.4.1)
    magick                 2.8.6      2025-03-23 [1] CRAN (R 4.4.1)
    magrittr               2.0.3      2022-03-30 [1] CRAN (R 4.4.0)
    mashr                  0.2.79     2023-10-18 [1] CRAN (R 4.4.0)
    MASS                   7.3-65     2025-02-28 [1] CRAN (R 4.4.1)
    mathjaxr               1.8-0      2025-04-30 [1] CRAN (R 4.4.1)
    Matrix                 1.7-3      2025-03-11 [1] CRAN (R 4.4.1)
    MatrixGenerics       * 1.18.1     2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    matrixStats          * 1.5.0      2025-01-07 [1] CRAN (R 4.4.1)
    memoise                2.0.1      2021-11-26 [1] CRAN (R 4.4.0)
    metadat                1.4-0      2025-02-04 [1] CRAN (R 4.4.1)
    metafor                4.8-0      2025-01-28 [1] CRAN (R 4.4.1)
    minqa                  1.2.8      2024-08-17 [1] CRAN (R 4.4.0)
    mixsqp                 0.3-54     2023-12-20 [1] CRAN (R 4.4.0)
    msigdbr                10.0.1     2025-03-19 [1] CRAN (R 4.4.3)
    mvtnorm                1.3-3      2025-01-10 [1] CRAN (R 4.4.1)
    nlme                   3.1-167    2025-01-27 [1] CRAN (R 4.4.3)
    nloptr                 2.2.1      2025-03-17 [1] CRAN (R 4.4.1)
    numDeriv               2016.8-1.1 2019-06-06 [1] CRAN (R 4.4.0)
    parallelly           * 1.43.0     2025-03-24 [1] CRAN (R 4.4.1)
    pbkrtest               0.5.3      2024-06-26 [1] CRAN (R 4.4.0)
    pillar                 1.10.2     2025-04-05 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3      2019-09-22 [1] CRAN (R 4.4.0)
    plyr                   1.8.9      2023-10-02 [1] CRAN (R 4.4.0)
    png                    0.1-8      2022-11-29 [1] CRAN (R 4.4.0)
    prettyunits            1.2.0      2023-09-24 [1] CRAN (R 4.4.0)
    progress               1.2.3      2023-12-06 [1] CRAN (R 4.4.0)
    purrr                  1.0.4      2025-02-05 [1] CRAN (R 4.4.1)
    R6                     2.6.1      2025-02-15 [1] CRAN (R 4.4.1)
    rbibutils              2.3        2024-10-04 [1] CRAN (R 4.4.1)
    RColorBrewer           1.1-3      2022-04-03 [1] CRAN (R 4.4.0)
    Rcpp                   1.0.14     2025-01-12 [1] CRAN (R 4.4.1)
    RcppParallel           5.1.10     2025-01-24 [1] CRAN (R 4.4.1)
    RCurl                  1.98-1.17  2025-03-22 [1] CRAN (R 4.4.1)
    Rdpack                 2.6.3      2025-03-16 [1] CRAN (R 4.4.2)
    reformulas             0.4.0      2024-11-03 [1] CRAN (R 4.4.1)
    remaCor                0.0.18     2024-02-08 [1] CRAN (R 4.4.0)
    reshape2               1.4.4      2020-04-09 [1] CRAN (R 4.4.0)
    Rfast                  2.1.5.1    2025-03-14 [1] CRAN (R 4.4.1)
    Rgraphviz              2.50.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    RhpcBLASctl            0.23-42    2023-02-11 [1] CRAN (R 4.4.0)
    rjson                  0.2.23     2024-09-16 [1] CRAN (R 4.4.1)
 VP rlang                * 1.1.5      2025-04-11 [2] CRAN (R 4.4.1) (on disk 1.1.6)
    rmeta                  3.0        2018-03-20 [1] CRAN (R 4.4.0)
    rprojroot              2.0.4      2023-11-05 [1] CRAN (R 4.4.0)
    RSQLite                2.3.9      2024-12-03 [1] CRAN (R 4.4.1)
    S4Arrays               1.6.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.4.0      2025-04-24 [1] CRAN (R 4.4.1)
    scattermore            1.2        2023-06-12 [1] CRAN (R 4.4.0)
    sessioninfo            1.2.3      2025-02-05 [1] CRAN (R 4.4.1)
    SingleCellExperiment * 1.28.1     2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    SparseArray            1.6.2      2025-02-20 [1] Bioconductor 3.20 (R 4.4.2)
    sparseMatrixStats      1.18.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SpatialExperiment    * 1.16.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SQUAREM                2021.1     2021-01-13 [1] CRAN (R 4.4.0)
    statmod                1.5.0      2023-01-06 [1] CRAN (R 4.4.0)
    stringi                1.8.7      2025-03-27 [1] CRAN (R 4.4.1)
    stringr                1.5.1      2023-11-14 [1] CRAN (R 4.4.0)
    SummarizedExperiment * 1.36.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    tibble                 3.2.1      2023-03-20 [1] CRAN (R 4.4.0)
    tidyr                  1.3.1      2024-01-24 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1      2024-03-11 [1] CRAN (R 4.4.0)
    truncnorm              1.0-9      2023-03-20 [1] CRAN (R 4.4.0)
    UCSC.utils             1.2.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    variancePartition    * 1.36.3     2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    vctrs                  0.6.5      2023-12-01 [1] CRAN (R 4.4.0)
    vipor                  0.4.7      2023-12-18 [1] CRAN (R 4.4.0)
    withr                  3.0.2      2024-10-28 [1] CRAN (R 4.4.1)
    xfun                   0.52       2025-04-02 [1] CRAN (R 4.4.1)
    XML                    3.99-0.18  2025-01-01 [1] CRAN (R 4.4.1)
    xtable                 1.8-4      2019-04-21 [1] CRAN (R 4.4.0)
    XVector                0.46.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    zenith                 1.8.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    zigg                   0.0.2      2025-02-07 [1] CRAN (R 4.4.1)
    zlibbioc               1.52.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * â”€â”€ Packages attached to the search path.
 V â”€â”€ Loaded and on-disk version mismatch.
 P â”€â”€ Loaded and on-disk path mismatch.

â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
