---
title: "01b-Clusterwise_DGELists_withMinSpots_reqmnt"
author: "Bernie Mulvey"
date: "2025-05-26"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(SpatialExperiment)
library(dreamlet)
library(limma)
library(edgeR)
library(gridExtra)
library(ggplot2)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")
library(parallel) # use PSOCK clusters for best performance here

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

 
## load SPE and assign hdg_svg_2575_louv_res1 clusterings as autoannotated
## Drop Br5517 (unknown anatomy/orientation), Br5276 (LC has been depleted by QC), Br5712 (unknown anatomy/orientation). Keep 6297; no LC but anatomy is right in vicinity and there appears to be nm (?!).
```{r}
lc3 <- readRDS("processed-data/06-countsOnly_noImgData_QCed_SPE_split_to_tissSections.RDS")

louvs <- readRDS("processed-data/07_featureSelection_dimred_harmony_clustering/04-Initial_louvLeid_res0.5-1-2_clusterings.RDS")
louvs <- louvs[["HARMONYlmbna_HDG_SVG_2575"]][,.(rn,snnHARMONYlmbna_HDG_SVG_2575_louv_res1)]
setnames(louvs,2,"clusid")

finalann <- fread("processed-data/08_validitycheck_25hdg75svg_louv1/10-25hdg75svg_louv1_annots.txt")

louvs <- merge.data.table(louvs,finalann,by="clusid")
louvs <- DataFrame(louvs,row.names=louvs$rn)[colnames(lc3),]

colLabels(lc3) <- louvs$anno
lc3 <- lc3[,!(lc3$brnum %in% paste0("Br",c(5517,5276,5712)))]
```

### Append global ancestry estimates -- we need them later anyhow, and it seems edgeR retains differing numbers of genes depending how you specify the design matrix, so we'll just do it here.
```{r}
lc3$brnum[lc3$brnum=="Br6119(re-dis)"] <- "Br6119"
flares <- fread("raw-data/ancestryAndHaplotypes/FLARE/global_ancestry_estimates.txt")

flares <- flares[ID %in% lc3$brnum,.(ID,YRI)]
tmpcd <- as.data.table(colData(lc3),keep.rownames=T)
tmpcd <- merge.data.table(tmpcd,flares,by.x="brnum",by.y="ID",all.x=T)

## 05/26 add a predominant ancestry label that'll be used in a parallel analysis only containing 26 samples that meet one of the thresholds (CEU or YRI > 0.75)
tmpcd[YRI>0.75,predomAncest:="YRI"]
tmpcd[YRI<0.25,predomAncest:="CEU"]
tmpcd <- DataFrame(tmpcd,row.names=tmpcd$rn)[colnames(lc3),]

colData(lc3) <- tmpcd
rm(tmpcd,flares)
```


### make APOE encodings factor-friendly
```{r}
# simplify APOE labels and make sexapo groups
lc3$APOE <- gsub(lc3$APOE,pattern="\\/ ",replacement="_")
lc3$APOE[lc3$APOE=="E4_E4"] <- "E4hom"
lc3$APOE[lc3$APOE=="E3_E4"] <- "E4het"
lc3$APOE[lc3$APOE=="E2_E3"] <- "E2het"
lc3$APOE[lc3$APOE=="E2_E2"] <- "E2hom"

```

## for plotting pseudobulk MDS by Visium slide, we need to edit the colData(spe)$Visium_slide to JUST be the slide (it still has the slide capture area ID appended to it)
```{r}
colData(lc3)$Visium_slide <- gsub(colData(lc3)$Visium_slide,pattern="^(V.....-...)_..$",replacement="\\1")

rm(louvs,finalann)
```

#### TOP LEVEL CLUSTER DGELists ####

# make pseudobulk object (dreamlet is v. fast at this)
```{r}
ag <- dreamlet::aggregateToPseudoBulk(lc3,sample_id="sample_id",cluster_id = "label",BPPARAM = MulticoreParam(6))

# extract number of spots in a pseudobulk sample to filter by. this is in the inverse of the format we want, list entries are sample and subentries are cluster, so rearrange to cluster listed by sample
agn <- ag@int_colData$n_cells
agn <- lapply(agn,function(x){as.data.frame(t(x))})
agn <- rbindlist(agn,idcol="samp")

# extract colData to feed into DGEList as the "samples" data
agcd <- as.data.frame(colData(ag))

## only retain clusters for which at least 43 samples (of 85 total, i.e over half) of â‰¥10 spots each are available:  
ag <- as.list(assays(ag))
agc <- mapply(ag,names(ag),SIMPLIFY=FALSE,FUN=function(a,n){
  tenspot <- agn[get(n)>=10,samp]
  ncol(a[,colSums(a)>0][,tenspot])
})

ag <- ag[!(agc<43)]

## drop clusters where there are 20+ samples, but under 18 samples for one of the sexes (i.e., < half of the samples from the smaller group--females w 35 sections total):
agsc <-  mapply(ag,names(ag),SIMPLIFY=FALSE,FUN=function(a,n){
    cn <- colnames(a[,colSums(a)>0]) # any counts
    cn <- cn[cn %in% agn[get(n)>=10,samp]] # 10+ spots included
    tmp <- as.data.table(agcd[cn,])
    if(nrow(tmp[,.N,by="Sex"][N<18])>0){
      return(0)
    } else {return(1)}
})
ag <- ag[agsc==1]


```

## make the DGELists per remaining cluster:
## REMOVED 05/09/25 (added 02/14/25) : store plots of the entire count distribution with the low-expression cutoff marked to check that this is working as expected.
## ADDED 02/26/25: filter out samples with under 10 spots in the current cluster
```{r}
## get rowData as a data.table from the SPE to append to the DGElist--namely so we can make MDS plots of the pseudobulks while excluding mitochondrial and sex chromosomal genes
tmprd <- as.data.table(rowData(lc3)) # rownames are the same as $gene_id

dgels <- mapply(a=ag,n=names(ag),SIMPLIFY=FALSE,FUN=function(a,n){
    cn <- colnames(a[,colSums(a)>0]) # has counts
    
    tenspot <- cn[cn %in% agn[get(n)>=10,samp]] # and has 10+spots contributed
    X <- a[,tenspot]
    
    s <- copy(agcd)[colnames(X),]
    if(length(unique(s$Sex))<2){return()} else 
    {
        dge <- DGEList(counts=as.matrix(X),samples=s,genes=as.data.frame(tmprd,row.names = tmprd$rn)[rownames(X),])
        return(dge)
    }
})

names(dgels) <- names(ag)
```

## save
```{r}
saveRDS(dgels,"processed-data/12_DEanalyses_removedsampsAndFinalNMseg/01b-Clusterwise_DGElist_per_25HDG75SVGlouv1_10spotmin_noFilt.RDS")
```

## make MDS plots per pseudobulk cluster per potential variable of interest
## note that for continuous variables the coordinates are not quite perfectly 1:1 with the visuals output from plotMDS  because of some scaling thing done in a plotting call when the plot is actually rendered by plotMDS 
```{r}
mapply(d=dgels,n=names(dgels),SIMPLIFY=FALSE,FUN=function(d,n){
  setDTthreads(1,restore_after_fork = FALSE)
  covcols <- c("Sex","APOE","Visium_slide","brnum","predomAncest")
  
  lapply(covcols,FUN=function(cov){
    pal <- rainbow(n=length(unique(d$samples[,cov])))
    d$samples[,cov] <- as.factor(d$samples[,cov])
    
    png(paste0("plots/12_DEanalyses_removedsampsAndFinalNMseg/01_pseudobulk_cluster_MDSes/01b-",n,"_",cov,"_1_2_10spotfilt.png"),width=1000,height=800)
    plotMDS(d,col=pal[factor(d$samples[,cov])], dim.plot = c(1,2),pch = 18)
    legend('bottomright', legend=levels(d$samples[,cov]), col=pal, cex=1, pch=18)
    dev.off()
    
    png(paste0("plots/12_DEanalyses_removedsampsAndFinalNMseg/01_pseudobulk_cluster_MDSes/01b-",n,"_",cov,"_1_3_10spotfilt.png"),width=1000,height=800)
    plotMDS(d,col=pal[factor(d$samples[,cov])], dim.plot = c(1,3),pch = 18)
    legend('bottomright', legend=levels(d$samples[,cov]), col=pal, cex=1, pch=18)
    dev.off()
    
    png(paste0("plots/12_DEanalyses_removedsampsAndFinalNMseg/01_pseudobulk_cluster_MDSes/01b-",n,"_",cov,"_2_3_10spotfilt.png"),width=1000,height=800)
    plotMDS(d,col=pal[factor(d$samples[,cov])], dim.plot = c(2,3),pch = 18)
    legend('bottomright', legend=levels(d$samples[,cov]), col=pal, cex=1, pch=18)
    dev.off()
  })
})

mapply(d=dgels,n=names(dgels),SIMPLIFY=FALSE,FUN=function(d,n){
  covcols <- c("Age","YRI")
  lapply(covcols,FUN=function(cov){
    
    ## 1,2 ##
    res <- plotMDS(d,d$samples[,cov], dim.plot = c(1,2),pch=18)
    dev.off()
    resdat <- as.data.table(cbind(as.numeric(res$x),as.numeric(res$y),rownames(d$samples),as.numeric(d$samples[,cov])))
    setnames(resdat,c("V1","V2","V3","V4"),c("PC1","PC2","sample_id",cov))
    p <- ggplot(resdat,aes(x=as.numeric(PC1),y=as.numeric(PC2),col=as.numeric(get(cov)))) + 
        geom_point() + 
        labs(x=paste0("MDS 1 (",100*round(res[[3]][1],digits=2),"%)"),
            y=paste0("MDS 2 (",100*round(res[[3]][2],digits=2),"%)"),
            col=cov) +
        ggtitle(n)+
        theme_bw()
    
    png(paste0("plots/12_DEanalyses_removedsampsAndFinalNMseg/01_pseudobulk_cluster_MDSes/01b-",n,"_",cov,"_1_2_10spotfilt.png"),width=1000,height=800)
    print(p)
    dev.off()
    
    ## 1,3 ##
    res <- plotMDS(d,d$samples[,cov], dim.plot = c(1,3),pch=18)
    dev.off()
    resdat <- as.data.table(cbind(res$x,res$y,rownames(d$samples),d$samples[,cov]))
    setnames(resdat,c("V1","V2","V3","V4"),c("PC1","PC3","sample_id",cov))

    p <- ggplot(resdat,aes(x=as.numeric(PC1),y=as.numeric(PC3),col=as.numeric(get(cov)))) + 
        geom_point() + 
        labs(x=paste0("MDS 1 (",100*round(res[[3]][1],digits=2),"%)"),
            y=paste0("MDS 3 (",100*round(res[[3]][3],digits=2),"%)"),
            col=cov)+
        ggtitle(n)+
        theme_bw()
    
    png(paste0("plots/12_DEanalyses_removedsampsAndFinalNMseg/01_pseudobulk_cluster_MDSes/01b-",n,"_",cov,"_1_3_10spotfilt.png"),width=1000,height=800)
    print(p)
    dev.off()

     ## 2,3 ##
    res <- plotMDS(d,d$samples[,cov], dim.plot = c(2,3),pch=18)
    dev.off()
    resdat <- as.data.table(cbind(res$x,res$y,rownames(d$samples),d$samples[,cov]))
    setnames(resdat,c("V1","V2","V3","V4"),c("PC2","PC3","sample_id",cov))
    
    p <-ggplot(resdat,aes(x=as.numeric(PC2),y=as.numeric(PC3),col=as.numeric(get(cov)))) + 
        geom_point() + 
        labs(x=paste0("MDS 2 (",100*round(res[[3]][2],digits=2),"%)"),
            y=paste0("MDS 3 (",100*round(res[[3]][3],digits=2),"%)"),
            col=cov) +
        ggtitle(n)+
        theme_bw()
    
    png(paste0("plots/12_DEanalyses_removedsampsAndFinalNMseg/01_pseudobulk_cluster_MDSes/01b-",n,"_",cov,"_2_3_10spotfilt.png"),width=1000,height=800)
    print(p)
    dev.off()
    })
  
})
```

# clean up
```{r}
rm(ag,agc,agcd,agsc,dgels)
gc(full=T)
```

#### LCNM-SEGMENTATION-SUBCLUSTER DGELists -- only LC NM+/NM- frxns. ####
## append NM subassignments
```{r}
lcnmclus <- fread("processed-data/10_subclustering/04-LCsubclus-by-NM_otherclusts-by-LCorLCNMadjacency.txt")
lcnmclus[label!="LC",NMseg_subclus:=""]
lcnmclus <- DataFrame(lcnmclus,row.names=lcnmclus$rn)[colnames(lc3),]

lc3$NMseg_subclus <- lcnmclus$NMseg_subclus
```

## # make pseudobulk object (dreamlet is v. fast at this)
```{r}
LCNMadj.ag <- dreamlet::aggregateToPseudoBulk(lc3[,lc3$NMseg_subclus!=""],sample_id="sample_id",cluster_id = "NMseg_subclus",BPPARAM = MulticoreParam(6))

# extract colData to feed into DGEList as the "samples" data
nm.agcd <- as.data.frame(colData(LCNMadj.ag))

# extract number of spots in a pseudobulk sample to filter by. this is in the inverse of the format we want, list entries are sample and subentries are cluster, so rearrange to cluster listed by sample
nmagn <- LCNMadj.ag@int_colData$n_cells
nmagn <- lapply(nmagn,function(x){as.data.frame(t(x))})
nmagn <- rbindlist(nmagn,idcol="samp")

## only retain LCNM+- clusters, and then only if at least 43 samples with 10+ spots of a type (of 85 total, i.e over half) are available after pseudobulking through dreamlet:  

LCNMadj.ag <- as.list(assays(LCNMadj.ag))
LCNMadj.ag <- LCNMadj.ag[c("LC_NMpos","LC_NMneg")]
LCNMadj.agc <- mapply(LCNMadj.ag,names(LCNMadj.ag),SIMPLIFY=FALSE,FUN=function(a,n){
  tenspot <- nmagn[get(n)>=10,samp]
  ncol(a[,colSums(a)>0][,tenspot])
})

LCNMadj.ag <- LCNMadj.ag[!(LCNMadj.agc<43)]

## drop clusters where there are 20+ samples, but under 18 samples for one of the sexes (i.e., < half of the samples from the smaller group--females w 35 sections total):
agsc <- mapply(LCNMadj.ag,n=names(LCNMadj.ag),SIMPLIFY=FALSE,FUN=function(a,n){
    cn <- colnames(a[,colSums(a)>0])
    cn <- cn[cn %in% nmagn[get(n)>=10,samp]] # 10+ spots included

    tmp <- as.data.table(nm.agcd[cn,])
    if(nrow(tmp[,.N,by="Sex"][N<18])>0){
      return(0)
    } else {return(1)}
})

LCNMadj.ag <- LCNMadj.ag[agsc==1]

rm(agscs,LCadj.agc,LCNMadj.agc)
```

## make the DGELists per retained subcluster:
```{r}
## get rowData as a data.table from the SPE to append to the DGElist--namely so we can make MDS plots of the pseudobulks while excluding mitochondrial and sex chromosomal genes
tmprd <- as.data.table(rowData(lc3)) # rownames are the same as $gene_id

nmdgels <- mapply(a=LCNMadj.ag,n=names(LCNMadj.ag),SIMPLIFY=FALSE,FUN=function(a,n){
    cn <- colnames(a[,colSums(a)>0]) # has counts
    
    tenspot <- cn[cn %in% nmagn[get(n)>=10,samp]] # and has 10+spots contributed
    X <- a[,tenspot]
    
    s <- copy(nm.agcd)[colnames(X),]
    if(length(unique(s$Sex))<2){return()} else 
    {

        # specify the ostensible design matrix
        des <- model.matrix(~0+Sex+APOE+Age+YRI, data=s)

        dge <- DGEList(counts=as.matrix(X),samples=s,genes=as.data.frame(tmprd,row.names = tmprd$rn)[rownames(X),])
        
        ## this takes the raw counts, not the normalized counts. (the results are different if you calculate lib sizes first, and the docs say the input is a matrix of counts, so)
        keep <- filterByExpr.DGEList(dge,design = des)
        
        dge <- dge[keep,,keep.lib.sizes=FALSE]
        dge <- edgeR::normLibSizes(dge,method = "TMM")

        return(dge)
    }
})
```

## save
```{r}
saveRDS(nmdgels,"processed-data/12_DEanalyses_removedsampsAndFinalNMseg/01b-DGElist_per_LCNM_subcluster_10spotmin_edgeRdefaultLoExpFilt.RDS")
```

 rep inf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.2

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    parallel  stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] SpatialExperiment_1.16.0    SingleCellExperiment_1.28.1
 [3] SummarizedExperiment_1.36.0 Biobase_2.66.0             
 [5] GenomicRanges_1.58.0        GenomeInfoDb_1.42.3        
 [7] IRanges_2.40.1              S4Vectors_0.44.0           
 [9] BiocGenerics_0.52.0         MatrixGenerics_1.18.1      
[11] matrixStats_1.5.0           BiocParallel_1.40.0        
[13] parallelly_1.43.0           colorout_1.3-0.2           
[15] edgeR_4.4.2                 limma_3.62.2               
[17] data.table_1.17.0           rlang_1.1.5                

loaded via a namespace (and not attached):
  [1] splines_4.4.3             bitops_1.0-9              tibble_3.2.1             
  [4] graph_1.84.1              XML_3.99-0.18             lifecycle_1.0.4          
  [7] Rdpack_2.6.3              mixsqp_0.3-54             rprojroot_2.0.4          
 [10] lattice_0.22-6            MASS_7.3-65               backports_1.5.0          
 [13] magrittr_2.0.3            metafor_4.8-0             rmarkdown_2.29           
 [16] yaml_2.3.10               sessioninfo_1.2.3         DBI_1.2.3                
 [19] minqa_1.2.8               RColorBrewer_1.1-3        abind_1.4-8              
 [22] zlibbioc_1.52.0           EnvStats_3.0.0            purrr_1.0.4              
 [25] rmeta_3.0                 msigdbr_10.0.1            RCurl_1.98-1.17          
 [28] GenomeInfoDbData_1.2.13   ggrepel_0.9.6             pbkrtest_0.5.3           
 [31] irlba_2.3.5.1             annotate_1.84.0           DelayedMatrixStats_1.28.1
 [34] codetools_0.2-20          DelayedArray_0.32.0       tidyselect_1.2.1         
 [37] UCSC.utils_1.2.0          farver_2.1.2              lme4_1.1-36              
 [40] mathjaxr_1.8-0            jsonlite_2.0.0            iterators_1.0.14         
 [43] progress_1.2.3            tools_4.4.3               Rcpp_1.0.14              
 [46] zenith_1.8.0              glue_1.8.0                SparseArray_1.6.2        
 [49] xfun_0.52                 here_1.0.1                dplyr_1.1.4              
 [52] withr_3.0.2               numDeriv_2016.8-1.1       fastmap_1.2.0            
 [55] boot_1.3-31               caTools_1.18.3            digest_0.6.37            
 [58] truncnorm_1.0-9           R6_2.6.1                  colorspace_2.1-1         
 [61] scattermore_1.2           gtools_3.9.5              dichromat_2.0-0.1        
 [64] RSQLite_2.3.9             RhpcBLASctl_0.23-42       tidyr_1.3.1              
 [67] generics_0.1.3            variancePartition_1.36.3  corpcor_1.6.10           
 [70] prettyunits_1.2.0         httr_1.4.7                S4Arrays_1.6.0           
 [73] pkgconfig_2.0.3           gtable_0.3.6              blob_1.2.4               
 [76] XVector_0.46.0            remaCor_0.0.18            htmltools_0.5.8.1        
 [79] dreamlet_1.4.1            zigg_0.0.2                GSEABase_1.68.0          
 [82] scales_1.4.0              png_0.1-8                 fANCOVA_0.6-1            
 [85] reformulas_0.4.0          ashr_2.2-63               knitr_1.50               
 [88] rstudioapi_0.17.1         reshape2_1.4.4            rjson_0.2.23             
 [91] nlme_3.1-167              nloptr_2.2.1              cachem_1.1.0             
 [94] stringr_1.5.1             KernSmooth_2.23-26        vipor_0.4.7              
 [97] metadat_1.4-0             AnnotationDbi_1.68.0      pillar_1.10.2            
[100] grid_4.4.3                vctrs_0.6.5               gplots_3.2.0             
[103] mashr_0.2.79              xtable_1.8-4              beeswarm_0.4.0           
[106] Rgraphviz_2.50.0          evaluate_1.0.3            KEGGgraph_1.66.0         
[109] invgamma_1.1              magick_2.8.6              mvtnorm_1.3-3            
[112] cli_3.6.4                 locfit_1.5-9.12           compiler_4.4.3           
[115] crayon_1.5.3              SQUAREM_2021.1            plyr_1.8.9               
[118] ggbeeswarm_0.7.2          stringi_1.8.7             datasets_4.4.3           
[121] assertthat_0.2.1          babelgene_22.9            lmerTest_3.1-3           
[124] Biostrings_2.74.1         aod_1.3.3                 Matrix_1.7-3             
[127] hms_1.1.3                 sparseMatrixStats_1.18.0  bit64_4.6.0-1            
[130] ggplot2_3.5.2             KEGGREST_1.46.0           statmod_1.5.0            
[133] rbibutils_2.3             Rfast_2.1.5.1             broom_1.0.7              
[136] memoise_2.0.1             RcppParallel_5.1.10       bit_4.6.0                
[139] EnrichmentBrowser_2.36.0 
> sessioninfo::session_info()
â”€ Session info â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sequoia 15.3.2
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-05-26
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)
 quarto   1.4.555 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/quarto

â”€ Packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
 !  package              * version    date (UTC) lib source
    abind                  1.4-8      2024-09-12 [1] CRAN (R 4.4.1)
    annotate               1.84.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    AnnotationDbi          1.68.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    aod                    1.3.3      2023-12-13 [1] CRAN (R 4.4.0)
    ashr                   2.2-63     2023-08-21 [1] CRAN (R 4.4.0)
    assertthat             0.2.1      2019-03-21 [1] CRAN (R 4.4.0)
    babelgene              22.9       2022-09-29 [1] CRAN (R 4.4.0)
    backports              1.5.0      2024-05-23 [1] CRAN (R 4.4.0)
    beeswarm               0.4.0      2021-06-01 [1] CRAN (R 4.4.0)
    Biobase              * 2.66.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocParallel         * 1.40.0     2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    Biostrings             2.74.1     2024-12-16 [1] Bioconductor 3.20 (R 4.4.2)
    bit                    4.6.0      2025-03-06 [1] CRAN (R 4.4.1)
    bit64                  4.6.0-1    2025-01-16 [1] CRAN (R 4.4.1)
    bitops                 1.0-9      2024-10-03 [1] CRAN (R 4.4.1)
    blob                   1.2.4      2023-03-17 [1] CRAN (R 4.4.0)
    boot                   1.3-31     2024-08-28 [1] CRAN (R 4.4.3)
    broom                  1.0.7      2024-09-26 [1] CRAN (R 4.4.1)
    cachem                 1.1.0      2024-05-16 [1] CRAN (R 4.4.0)
    caTools                1.18.3     2024-09-04 [1] CRAN (R 4.4.1)
 VP cli                    3.6.4      2025-04-23 [2] CRAN (R 4.4.1) (on disk 3.6.5)
    codetools              0.2-20     2024-03-31 [1] CRAN (R 4.4.3)
    colorout             * 1.3-0.2    2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    colorspace             2.1-1      2024-07-26 [1] CRAN (R 4.4.0)
    corpcor                1.6.10     2021-09-16 [1] CRAN (R 4.4.0)
    crayon                 1.5.3      2024-06-20 [1] CRAN (R 4.4.0)
    data.table           * 1.17.0     2025-02-22 [1] CRAN (R 4.4.1)
    DBI                    1.2.3      2024-06-02 [1] CRAN (R 4.4.0)
    DelayedArray           0.32.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    DelayedMatrixStats     1.28.1     2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    dichromat              2.0-0.1    2022-05-02 [1] CRAN (R 4.4.0)
    digest                 0.6.37     2024-08-19 [1] CRAN (R 4.4.1)
    dplyr                  1.1.4      2023-11-17 [1] CRAN (R 4.4.0)
    dreamlet               1.4.1      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    edgeR                * 4.4.2      2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    EnrichmentBrowser      2.36.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    EnvStats               3.0.0      2024-08-24 [1] CRAN (R 4.4.1)
    evaluate               1.0.3      2025-01-10 [1] CRAN (R 4.4.1)
    fANCOVA                0.6-1      2020-11-13 [1] CRAN (R 4.4.0)
    farver                 2.1.2      2024-05-13 [1] CRAN (R 4.4.0)
    fastmap                1.2.0      2024-05-15 [1] CRAN (R 4.4.0)
    generics               0.1.3      2022-07-05 [1] CRAN (R 4.4.0)
    GenomeInfoDb         * 1.42.3     2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13     2024-12-12 [1] Bioconductor
    GenomicRanges        * 1.58.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ggbeeswarm             0.7.2      2023-04-29 [1] CRAN (R 4.4.0)
    ggplot2                3.5.2      2025-04-09 [1] CRAN (R 4.4.1)
    ggrepel                0.9.6      2024-09-07 [1] CRAN (R 4.4.1)
    glue                   1.8.0      2024-09-30 [1] CRAN (R 4.4.1)
    gplots                 3.2.0      2024-10-05 [1] CRAN (R 4.4.1)
    graph                  1.84.1     2025-01-02 [1] Bioconductor 3.20 (R 4.4.2)
    GSEABase               1.68.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    gtable                 0.3.6      2024-10-25 [1] CRAN (R 4.4.1)
    gtools                 3.9.5      2023-11-20 [1] CRAN (R 4.4.0)
    here                   1.0.1      2020-12-13 [1] CRAN (R 4.4.0)
    hms                    1.1.3      2023-03-21 [1] CRAN (R 4.4.0)
    htmltools              0.5.8.1    2024-04-04 [1] CRAN (R 4.4.0)
    httr                   1.4.7      2023-08-15 [1] CRAN (R 4.4.0)
    invgamma               1.1        2017-05-07 [1] CRAN (R 4.4.0)
    IRanges              * 2.40.1     2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    irlba                  2.3.5.1    2022-10-03 [1] CRAN (R 4.4.0)
    iterators              1.0.14     2022-02-05 [1] CRAN (R 4.4.0)
    jsonlite               2.0.0      2025-03-27 [1] CRAN (R 4.4.1)
    KEGGgraph              1.66.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    KEGGREST               1.46.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    KernSmooth             2.23-26    2025-01-01 [1] CRAN (R 4.4.3)
    knitr                  1.50       2025-03-16 [1] CRAN (R 4.4.2)
    lattice                0.22-6     2024-03-20 [1] CRAN (R 4.4.3)
    lifecycle              1.0.4      2023-11-07 [1] CRAN (R 4.4.0)
    limma                * 3.62.2     2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    lme4                   1.1-36     2025-01-11 [1] CRAN (R 4.4.1)
    lmerTest               3.1-3      2020-10-23 [1] CRAN (R 4.4.0)
    locfit                 1.5-9.12   2025-03-05 [1] CRAN (R 4.4.1)
    magick                 2.8.6      2025-03-23 [1] CRAN (R 4.4.1)
    magrittr               2.0.3      2022-03-30 [1] CRAN (R 4.4.0)
    mashr                  0.2.79     2023-10-18 [1] CRAN (R 4.4.0)
    MASS                   7.3-65     2025-02-28 [1] CRAN (R 4.4.1)
    mathjaxr               1.8-0      2025-04-30 [1] CRAN (R 4.4.1)
    Matrix                 1.7-3      2025-03-11 [1] CRAN (R 4.4.1)
    MatrixGenerics       * 1.18.1     2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    matrixStats          * 1.5.0      2025-01-07 [1] CRAN (R 4.4.1)
    memoise                2.0.1      2021-11-26 [1] CRAN (R 4.4.0)
    metadat                1.4-0      2025-02-04 [1] CRAN (R 4.4.1)
    metafor                4.8-0      2025-01-28 [1] CRAN (R 4.4.1)
    minqa                  1.2.8      2024-08-17 [1] CRAN (R 4.4.0)
    mixsqp                 0.3-54     2023-12-20 [1] CRAN (R 4.4.0)
    msigdbr                10.0.1     2025-03-19 [1] CRAN (R 4.4.3)
    mvtnorm                1.3-3      2025-01-10 [1] CRAN (R 4.4.1)
    nlme                   3.1-167    2025-01-27 [1] CRAN (R 4.4.3)
    nloptr                 2.2.1      2025-03-17 [1] CRAN (R 4.4.1)
    numDeriv               2016.8-1.1 2019-06-06 [1] CRAN (R 4.4.0)
    parallelly           * 1.43.0     2025-03-24 [1] CRAN (R 4.4.1)
    pbkrtest               0.5.3      2024-06-26 [1] CRAN (R 4.4.0)
    pillar                 1.10.2     2025-04-05 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3      2019-09-22 [1] CRAN (R 4.4.0)
    plyr                   1.8.9      2023-10-02 [1] CRAN (R 4.4.0)
    png                    0.1-8      2022-11-29 [1] CRAN (R 4.4.0)
    prettyunits            1.2.0      2023-09-24 [1] CRAN (R 4.4.0)
    progress               1.2.3      2023-12-06 [1] CRAN (R 4.4.0)
    purrr                  1.0.4      2025-02-05 [1] CRAN (R 4.4.1)
    R6                     2.6.1      2025-02-15 [1] CRAN (R 4.4.1)
    rbibutils              2.3        2024-10-04 [1] CRAN (R 4.4.1)
    RColorBrewer           1.1-3      2022-04-03 [1] CRAN (R 4.4.0)
    Rcpp                   1.0.14     2025-01-12 [1] CRAN (R 4.4.1)
    RcppParallel           5.1.10     2025-01-24 [1] CRAN (R 4.4.1)
    RCurl                  1.98-1.17  2025-03-22 [1] CRAN (R 4.4.1)
    Rdpack                 2.6.3      2025-03-16 [1] CRAN (R 4.4.2)
    reformulas             0.4.0      2024-11-03 [1] CRAN (R 4.4.1)
    remaCor                0.0.18     2024-02-08 [1] CRAN (R 4.4.0)
    reshape2               1.4.4      2020-04-09 [1] CRAN (R 4.4.0)
    Rfast                  2.1.5.1    2025-03-14 [1] CRAN (R 4.4.1)
    Rgraphviz              2.50.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    RhpcBLASctl            0.23-42    2023-02-11 [1] CRAN (R 4.4.0)
    rjson                  0.2.23     2024-09-16 [1] CRAN (R 4.4.1)
 VP rlang                * 1.1.5      2025-04-11 [2] CRAN (R 4.4.1) (on disk 1.1.6)
    rmarkdown              2.29       2024-11-04 [1] CRAN (R 4.4.1)
    rmeta                  3.0        2018-03-20 [1] CRAN (R 4.4.0)
    rprojroot              2.0.4      2023-11-05 [1] CRAN (R 4.4.0)
    RSQLite                2.3.9      2024-12-03 [1] CRAN (R 4.4.1)
    rstudioapi             0.17.1     2024-10-22 [1] CRAN (R 4.4.1)
    S4Arrays               1.6.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.4.0      2025-04-24 [1] CRAN (R 4.4.1)
    scattermore            1.2        2023-06-12 [1] CRAN (R 4.4.0)
    sessioninfo            1.2.3      2025-02-05 [1] CRAN (R 4.4.1)
    SingleCellExperiment * 1.28.1     2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    SparseArray            1.6.2      2025-02-20 [1] Bioconductor 3.20 (R 4.4.2)
    sparseMatrixStats      1.18.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SpatialExperiment    * 1.16.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SQUAREM                2021.1     2021-01-13 [1] CRAN (R 4.4.0)
    statmod                1.5.0      2023-01-06 [1] CRAN (R 4.4.0)
    stringi                1.8.7      2025-03-27 [1] CRAN (R 4.4.1)
    stringr                1.5.1      2023-11-14 [1] CRAN (R 4.4.0)
    SummarizedExperiment * 1.36.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    tibble                 3.2.1      2023-03-20 [1] CRAN (R 4.4.0)
    tidyr                  1.3.1      2024-01-24 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1      2024-03-11 [1] CRAN (R 4.4.0)
    truncnorm              1.0-9      2023-03-20 [1] CRAN (R 4.4.0)
    UCSC.utils             1.2.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    variancePartition      1.36.3     2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    vctrs                  0.6.5      2023-12-01 [1] CRAN (R 4.4.0)
    vipor                  0.4.7      2023-12-18 [1] CRAN (R 4.4.0)
    withr                  3.0.2      2024-10-28 [1] CRAN (R 4.4.1)
    xfun                   0.52       2025-04-02 [1] CRAN (R 4.4.1)
    XML                    3.99-0.18  2025-01-01 [1] CRAN (R 4.4.1)
    xtable                 1.8-4      2019-04-21 [1] CRAN (R 4.4.0)
    XVector                0.46.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    yaml                   2.3.10     2024-07-26 [1] CRAN (R 4.4.0)
    zenith                 1.8.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    zigg                   0.0.2      2025-02-07 [1] CRAN (R 4.4.1)
    zlibbioc               1.52.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * â”€â”€ Packages attached to the search path.
 V â”€â”€ Loaded and on-disk version mismatch.
 P â”€â”€ Loaded and on-disk path mismatch.

â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
