---
title: "02b-OneInteractionPerRun_dropLowCountClusts"
author: "Bernie Mulvey"
date: "2025-05-15"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(edgeR)
library(limma)

# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")
library(parallel)


## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

```{r}
dgel <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/01a-DGElist_allClust_10spotmin_NOfilt.RDS")

### to avoid any issues, replace "Br6119(re-dis)" with "Br6119" in the samples tables
dgel$samples$brnum[dgel$samples$brnum=="Br6119(re-dis)"] <- "Br6119"

### rather than adjust for lib.size, now we will drop the clusters making that necessary because they are of limited interest / mixed cell types: Vascular, Oligo_complex, Oligo, Astro_Oligo
newsamp <- dgel$samples
newsamp <- newsamp[!(newsamp$label %in% c("Vascular","Oligo_Complex","Oligo","Astro_Oligo")),]

newmat <- dgel$counts[,rownames(newsamp)]
newmat <- newmat[rowSums(newmat)>0,]
newg <- dgel$genes[rownames(newmat),]

## remove old dgelist
rm(dgel)

## make new dgelist from the subsetted data above
dgel <- DGEList(counts=newmat,samples=newsamp,genes=newg)
rm(newmat,newsamp,newg)

## edgeR's not going to like the decimals if the label names carried any, so remove them just in case
dgel$samples$label <-  gsub(dgel$samples$label,pattern="\\.",replacement="")
rownames(dgel$samples) <- gsub(rownames(dgel$samples),pattern="\\.",replacement="")
colnames(dgel$counts) <- gsub(colnames(dgel$counts),pattern="\\.",replacement="")
```

### run low expression filtering separately for each model. NO lib size cofactor--we're dropping the clusters that would necessitate it.

define a second dataset with categorical ancestry instead of continuous (comes out to N=26) and run all models/contrasts in both the full dataset (YRI continuous) and subsetted (YRI categorical).
```{r}
dgelYRI <- dgel

## make a label-predom ancestry grouping column
tmpsamp <- as.data.table(dgelYRI$samples,keep.rownames=T)
tmpsamp <- tmpsamp[YRI<0.25|YRI>0.75]
tmpsamp[,newYRI:=ifelse(YRI>0.75,"YRI","CEU")]
tmpsamp[,YRI:=NULL]
setnames(tmpsamp,"newYRI","YRI")

tmpsamp <- data.frame(tmpsamp,row.names=tmpsamp$rn)
tmpsamp$rn <- NULL

tmpcounts <- dgelYRI$counts
tmpcounts <- tmpcounts[,rownames(tmpsamp)]
tmpcounts <- as.data.frame(tmpcounts)

dgelYRI <- DGEList(counts=tmpcounts,samples=tmpsamp,genes=dgelYRI$genes[rownames(tmpcounts),])

rm(tmpsamp,tmpcounts)

dgels <- list(cont_ancest=dgel,cat_ancest=dgelYRI)
rm(dgel,dgelYRI)
```

```{r}
deres <- mcmapply(dgel=dgels,n=names(dgels),SIMPLIFY=FALSE,mc.cores=2,FUN=function(dgel,n){
    ##### MODEL 1: SEX DE (Sex_cluster "interaction" edgeR style #######
    ####################################################################
    ## make a copy of the dgelist for this analysis
    dgelsex <- dgel
    dgelsex$samples$label_sex <- paste0(dgelsex$samples$label,"_",dgel$samples$Sex)
    
    ## full design matrix with colnames matching clus geno group values
    des <- model.matrix(~0 + label_sex + APOE + Age + YRI , data = dgelsex$samples)
    
    ## fix names for readability
    colnames(des) <- gsub(colnames(des),pattern="label_sex",replacement="")
    
    ## now filter expression. we DO want to get normalized lib sizes first, based on readme for the arguments for filterByExpr.DGEList (it will look for normalized lib sizes)
    dgelsex <- edgeR::normLibSizes(dgelsex)
    
    keep <- filterByExpr.DGEList(dgelsex,design = des)
    dgelsex <- dgelsex[keep,,keep.lib.sizes=F]
    
    ## recalc lib sizes
    dgelsex <- edgeR::normLibSizes(dgelsex)
    
    ## make sex contrasts for each cluster
    i<-1
    for (i in c(1:length(unique(dgelsex$samples$label)))){
        x <- unique(dgelsex$samples$label)[i]
        ret <- makeContrasts(
            str2expression(
                paste0(x,"_Sex=",paste0(x,"_M - ",x,"_F"))),
            levels=des)
    
        colnames(ret) <- paste0(x,"_Sex")
    
        if(i==1){
            trastmat <- ret
        } else{
            trastmat <- cbind(trastmat,ret)
        }
    }
    rm(i,x,ret)
    
    v.swt <- voomLmFit(dgelsex,design = des,block = as.factor(dgelsex$samples$brnum),adaptive.span = T,sample.weights = T)

    ## fit M-F contrasts for each label
    v.swt.fit <- contrasts.fit(v.swt,contrasts=trastmat)
    v.swt.fit.e <- eBayes(v.swt.fit,robust = T)
    
    keeplist_sex_libsize <- list(voomLFt=v.swt,
                     trastsFit=v.swt.fit,
                     eBay=v.swt.fit.e,
                     dgelsex=dgelsex,
                     trasts=trastmat)
    
    sexlabel_detabs <- lapply(colnames(v.swt.fit),FUN=function(t){
            restab <- as.data.table(topTable(v.swt.fit.e,coef = t,number=Inf,adjust.method = "BH"))
        return(restab)
    })
    names(sexlabel_detabs) <- colnames(trastmat)
    
    rm(keep,v.swt,v.swt.fit,v.swt.fit.e,dgelsex,des,trastmat)
    gc(full=T)
    
    keeplist_sex_libsize$sexlabel_detabs <- sexlabel_detabs
    saveRDS(keeplist_sex_libsize,paste0("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/02bTMP_keeplist_sexlabel_noMixClusts_",n,".RDS"))
    
    rm(keeplist_sex_libsize,sexlabel_detabs)
    gc(full=T)
    
    
    #####################################################################
    ##### MODEL 2: E2/E4 carrier status "interaction" edgeR style #######
    #####################################################################
    
    ## make a copy of the dgelist for this analysis
    dgelE2E4 <- dgel
    dgelE2E4$samples$APOlargeGrp <- ifelse(dgelE2E4$samples$APOE %in% c("E2hom","E2het"),yes="E2",no="E4")
    stopifnot(length(unique(dgelE2E4$samples$APOlargeGrp))==2)
    
    dgelE2E4$samples$label_E2E4 <- paste0(dgelE2E4$samples$label,"_",dgelE2E4$samples$APOlargeGrp)
    
    ## full design matrix with colnames matching clus geno group values
    des <- model.matrix(~0 + label_E2E4 + Sex + Age + YRI , data = dgelE2E4$samples)
    
    ## fix names for readability
    colnames(des) <- gsub(colnames(des),pattern="label_E2E4",replacement="")
    
    ## now filter expression. we DO want to get normalized lib sizes first, based on readme for the arguments for filterByExpr.DGEList (it will look for normalized lib sizes)
    dgelE2E4 <- edgeR::normLibSizes(dgelE2E4)
    
    keep <- filterByExpr.DGEList(dgelE2E4,design = des)
    dgelE2E4 <- dgelE2E4[keep,,keep.lib.sizes=F]
    
    ## recalc lib sizes
    dgelE2E4 <- edgeR::normLibSizes(dgelE2E4)
    
    ## make E4-E2 contrasts for each cluster
    i<-1
    for (i in c(1:length(unique(dgelE2E4$samples$label)))){
        x <- unique(dgelE2E4$samples$label)[i]
        ret <- makeContrasts(
            str2expression(
                paste0(x,"_E4vE2=",paste0(x,"_E4 - ",x,"_E2"))),
            levels=des)
    
        colnames(ret) <- paste0(x,"_E4vE2")
    
        if(i==1){
            trastmat <- ret
        } else{
            trastmat <- cbind(trastmat,ret)
        }
    }
    rm(i,x,ret,keep)

    v.swt <- voomLmFit(dgelE2E4,design = des,block = as.factor(dgelE2E4$samples$brnum),adaptive.span = T,sample.weights = T)
    
    ## fit E4-E2 contrasts for each label
    v.swt.fit <- contrasts.fit(v.swt,contrasts=trastmat)
    v.swt.fit.e <- eBayes(v.swt.fit,robust = T)
    
    keeplist_E2E4_libsize <- list(voomLFt=v.swt,
                     trastsFit=v.swt.fit,
                     eBay=v.swt.fit.e,
                     dgelE2E4=dgelE2E4,
                     trasts=trastmat)
    
    E2E4label_detabs <- lapply(colnames(v.swt.fit),FUN=function(t){
            restab <- as.data.table(topTable(v.swt.fit.e,coef = t,number=Inf,adjust.method = "BH"))
        return(restab)
    })
    names(E2E4label_detabs) <- colnames(trastmat)
    
    keeplist_E2E4_libsize$E2E4label_detabs <- E2E4label_detabs
    saveRDS(keeplist_E2E4_libsize,paste0("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/02bTMP_keeplist_E2E4label_noMixClusts",n,".RDS"))
    
    rm(keep,v.swt,v.swt.fit,v.swt.fit.e,dgelE2E4,des,trastmat,keeplist_E2E4_libsize,E2E4label_detabs)
    gc(full=T)

    
    
    
    
    ##########################################################################
    #### MODEL 3: APOE alleles (E2hom/het, 4hom/het) "intxns" edgeR style ####
    ##########################################################################
    
    ## make a copy of the dgelist for this analysis
    dgelApoAlleles <- dgel
    dgelApoAlleles$samples$label_apo <- paste0(dgelApoAlleles$samples$label,"_",dgel$samples$APOE)
    
    ## full design matrix with colnames matching clus geno group values
    
    ## full design matrix with colnames matching clus geno group values
    des <- model.matrix(~0 + label_apo + Sex + Age + YRI , data = dgelApoAlleles$samples)

    ## fix names for readability
    colnames(des) <- gsub(colnames(des),pattern="label_apo",replacement="")
    
    ## now filter expression. we DO want to get normalized lib sizes first, based on readme for the arguments for filterByExpr.DGEList (it will look for normalized lib sizes)
    dgelApoAlleles <- edgeR::normLibSizes(dgelApoAlleles)
    keep <- filterByExpr.DGEList(dgelApoAlleles,design = des)
    dgelApoAlleles <- dgelApoAlleles[keep,,keep.lib.sizes=F]
   
    ## recalc lib sizes
    dgelApoAlleles <- edgeR::normLibSizes(dgelApoAlleles)
    
    ## make genotype contrasts for each cluster
    i<-1
    for (i in c(1:length(unique(dgelApoAlleles$samples$label)))){
        x <- unique(dgelApoAlleles$samples$label)[i]
        ret <- makeContrasts(
            str2expression(
                paste0(x,"_E2hom_vs_anyE4=",paste0(x,"_E2hom - (0.5*",x,"_E4het + 0.5*",x,"_E4hom)"))),
            str2expression(
                paste0(x,"_E2het_vs_anyE4=",paste0(x,"_E2het - (0.5*",x,"_E4het + 0.5*",x,"_E4hom)"))),
            str2expression(
                paste0(x,"_E4hom_vs_anyE2=",paste0(x,"_E4hom - (0.5*",x,"_E2het + 0.5*",x,"_E2hom)"))),
            str2expression(
                paste0(x,"_E4het_vs_anyE2=",paste0(x,"_E4het - (0.5*",x,"_E2het + 0.5*",x,"_E2hom)"))),
            str2expression(
                paste0(x,"_E4hom_vs_E4het=",paste0(x,"_E4hom - ",x,"_E4het"))),
            str2expression(
                paste0(x,"_E4hom_vs_others=",paste0(x,"_E4hom - ((",x,"_E4het + ",x,"_E2het + ",x,"_E2hom"),")/3)")),
            levels=des)
    
        colnames(ret) <- paste0(x,c("_E2hom_vs_anyE4","_E2het_vs_anyE4","_E4hom_vs_anyE2","_E4het_vs_anyE2","_E4hom_vs_E4het","_E4hom_vs_others"))
    
        if(i==1){
            trastmat <- ret
        } else{
            trastmat <- cbind(trastmat,ret)
        }
    }
    rm(i,x,ret)

    v.swt <- voomLmFit(dgelApoAlleles,design = des,block = as.factor(dgelApoAlleles$samples$brnum),adaptive.span = T,sample.weights = T)

    ## fit contrasts for each label
    v.swt.fit <- contrasts.fit(v.swt,contrasts=trastmat)
    v.swt.fit.e <- eBayes(v.swt.fit,robust = T)
    
    keeplist_ApoAlleles_libsize <- list(voomLFt=v.swt,
                     trastsFit=v.swt.fit,
                     eBay=v.swt.fit.e,
                     dgelApoAlleles=dgelApoAlleles,
                     trasts=trastmat)
    
    ApoAllelesLabel_detabs <- lapply(colnames(v.swt.fit),FUN=function(t){
            restab <- as.data.table(topTable(v.swt.fit.e,coef = t,number=Inf,adjust.method = "BH"))
        return(restab)
    })
    names(ApoAllelesLabel_detabs) <- colnames(trastmat)
    
    keeplist_ApoAlleles_libsize$ApoAllelesLabel_detabs <- ApoAllelesLabel_detabs
    saveRDS(keeplist_ApoAlleles_libsize,paste0("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/02bTMP_keeplist_ApoAllelesLabel_noMixClusts_",n,".RDS"))

    rm(keep,keeplist_ApoAlleles_libsize,v.swt,v.swt.fit,v.swt.fit.e,dgelApoAlleles,des,trastmat,ApoAllelesLabel_detabs)
    gc(full=T)
    
    
    
    
    #############################################################################
    #### MODEL 4: label*YRI (continuous or categorical) "intxns" edgeR style ####
    #############################################################################
    
    ## make a copy of the dgelist for this analysis
    dgelances <- dgel
    
    ## for the continuous interaction setup
    if(is.numeric(dgelances$samples$YRI)){
        des <- model.matrix(~0+label:YRI+Sex+0+APOE+Age,data=dgelances$samples)
        
        ## make these neater
        colnames(des) <- gsub(x=colnames(des),pattern="^label(.*)$",replacement="\\1")
        colnames(des) <- gsub(x=colnames(des),pattern="^(.*):YRI$",replacement=paste0("\\1","_by_YRI"))
        
        
        ## specify the COEFFICIENTS we'll use -- but with the same variable name for some simplicity downstream
        trastmat <- paste0(unique(dgelances$samples$label),"_by_YRI")
        stopifnot(all(trastmat %in% colnames(des)))
    } else {
        # categorical setup
        dgelances$samples$labAnces <- paste0(dgelances$samples$label,"_",dgelances$samples$YRI)
        
        des <- model.matrix(~0 + labAnces +
                        Sex +
                        APOE +
                        Age,
                    data = dgelances$samples)
        
        for (cl in unique(dgel$samples$label)){
            ret <- makeContrasts(
                str2expression(
                    paste0(cl,"_YRI","=",paste0("labAnces",cl,"_YRI -","labAnces",cl,"_CEU"))),
                levels=des)
        
            colnames(ret) <- paste0(cl,"_YRI")
        
            if(cl==unique(dgelYRI$samples$label)[1]){
                trastmat <- ret
            } else{
                trastmat <- cbind(trastmat,ret)
            }
        }
        rm(ret,cl,keep)
    }
    
    ## now filter expression. we DO want to get normalized lib sizes first, based on readme for the arguments for filterByExpr.DGEList (it will look for normalized lib sizes)
    dgelances <- edgeR::normLibSizes(dgelances)
    
    keep <- filterByExpr.DGEList(dgelances,design = des)
    dgelances <- dgelances[keep,,keep.lib.sizes=F]
    
    ## recalc lib sizes
    dgelances <- edgeR::normLibSizes(dgelances)
    
    ## run
    v.swt <- voomLmFit(dgelances,design = des,block = as.factor(dgelances$samples$brnum),adaptive.span = T,sample.weights = T)
    
    ## fit YRI effect for each label
    if(is.numeric(dgelances$samples$YRI)){
        v.swt.fit <- contrasts.fit(v.swt,coefficients = trastmat)
    } else {
        v.swt.fit <- contrasts.fit(v.swt,contrasts = trastmat)
    }
    
    v.swt.fit.e <- eBayes(v.swt.fit,robust = T)
    
    keeplist_yrilabel <- list(voomLFt=v.swt,
                     trastsFit=v.swt.fit,
                     eBay=v.swt.fit.e,
                     dgelYRI=dgelances,
                     trasts=trastmat)
    
    yrilabel_detabs <- lapply(colnames(v.swt.fit),FUN=function(t){
        restab <- as.data.table(topTable(v.swt.fit.e,coef = t,number=Inf,adjust.method = "BH"))
        return(restab)
    })
    names(yrilabel_detabs) <- colnames(v.swt.fit)
    
    keeplist_yrilabel$AncesGrpBylabel_detabs <- yrilabel_detabs
    saveRDS(keeplist_yrilabel,paste0("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/02bTMP_keeplist_AncesByLabel_noMixClusts_",n,".RDS"))
    
    rm(keep,keeplist_yrilabel,yrilabel_detabs,v.swt,v.swt.fit,v.swt.fit.e,dgelances,des,trastmat,kee3p)
    gc(full=T)
    
    
    
    
    
    #############################################################################
    ####### MODEL 5: label*YRI*E4orE2 (cat-cat-cat) "intxns" edgeR style ########
    #############################################################################
    if(is.numeric(dgel$samples$YRI)){
        x <- "Model 5 is not run for continuous YRI, only categorical YRI"
        saveRDS(x,paste0("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/02bTMP_keeplist_YRIlabel_noMixClusts_",n,".RDS"))
    } else {
        ## make a copy of the dgelist for this analysis
        dgelYRIAPO <- dgel
        
        ## define E2/E4 groups
        dgelYRIAPO$samples$APOlargeGrp <- ifelse(dgelYRIAPO$samples$APOE %in% c("E2hom","E2het"),yes="E2",no="E4")
        stopifnot(length(unique(dgelYRIAPO$samples$APOlargeGrp))==2)
        
        ## make a label-yri-apo grouping column
        tmpsamp <- as.data.table(dgelYRIAPO$samples,keep.rownames=T)
        tmpsamp[,lab_APOE_PredomAnces:=paste0(label,"_",APOlargeGrp,"_",YRI)]
        
        tmpsamp <- data.frame(tmpsamp,row.names=tmpsamp$rn)[colnames(dgelYRIAPO$counts),]
        tmpsamp$rn <- NULL
        
        dgelYRIAPO$samples <- tmpsamp
        rm(tmpsamp)
        
        ## full design matrix with colnames matching clus geno group values
        des <- model.matrix(~0+lab_APOE_PredomAnces+
                                Sex+
                                Age, 
                            data = dgelYRIAPO$samples)
        
        ## clean up colnames
        colnames(des) <- gsub(x=colnames(des),pattern="^lab_APOE_PredomAnces(.*)$",replacement="\\1")
        
        ## now filter expression. we DO want to get normalized lib sizes first, based on readme for the arguments for filterByExpr.DGEList (it will look for normalized lib sizes)
        dgelYRIAPO <- edgeR::normLibSizes(dgelYRIAPO)
        
        keep <- filterByExpr.DGEList(dgelYRIAPO,design = des)
        dgelYRIAPO <- dgelYRIAPO[keep,,keep.lib.sizes=F]
        
        ## recalc lib sizes
        dgelYRIAPO <- edgeR::normLibSizes(dgelYRIAPO)
        
        ## specify contrasts
        i<-1
        for (i in c(1:length(unique(dgelYRIAPO$samples$label)))){
            x <- unique(dgelYRIAPO$samples$label)[i]
            ret <- makeContrasts(
            str2expression(
                paste0(x,"_E4YRI_vs_E4CEU=",paste0(x,"_E4_YRI - ", x, "_E4_CEU"))),
            str2expression(
                paste0(x,"_E2YRI_vs_E2CEU=",paste0(x,"_E2_YRI - ", x, "_E2_CEU"))),
            str2expression(
                paste0(x,"_E4CEU_vs_E2CEU=",paste0(x,"_E4_CEU - ", x, "_E2_CEU"))),
            str2expression(
                paste0(x,"_E4YRI_vs_E2YRI=",paste0(x,"_E4_YRI - ", x, "_E2_YRI"))),
            levels=des)
    
        colnames(ret) <- paste0(x,c("_E4YRI_vs_E4CEU","_E2YRI_vs_E2CEU","_E4CEU_vs_E2CEU","_E4YRI_vs_E2YRI"))
    
        if(i==1){
            trastmat <- ret
        } else{
            trastmat <- cbind(trastmat,ret)
        }
    }
    rm(i,x,ret)
        
        
        ## fits
        v.swt <- voomLmFit(dgelYRIAPO,design = des,block = as.factor(dgelYRIAPO$samples$brnum),adaptive.span = T,sample.weights = T)
        
        ## fit YRI effect for each label
        v.swt.fit <- contrasts.fit(v.swt,contrasts=trastmat)
        v.swt.fit.e <- eBayes(v.swt.fit,robust = T)
        
        keeplist_yriapolabel_libsize <- list(voomLFt=v.swt,
                         trastsFit=v.swt.fit,
                         eBay=v.swt.fit.e,
                         dgelYRIAPO=dgelYRIAPO,
                         trasts=trastmat)
        
        yriapolabel_detabs <- lapply(colnames(v.swt.fit),FUN=function(t){
            restab <- as.data.table(topTable(v.swt.fit.e,coef = t,number=Inf,adjust.method = "BH"))
            return(restab)
        })
        names(yriapolabel_detabs) <- colnames(v.swt.fit)
        keeplist_yriapolabel_libsize$yriapolabel_detabs <- yriapolabel_detabs
        
        saveRDS(keeplist_yriapolabel_libsize,paste0("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/02bTMP_keeplist_YRIAPOlabel_noMixClusts_",n,".RDS"))
        
        rm(keep,keeplist_yriapolabel_libsize,yriapolabel_detabs,v.swt,v.swt.fit,v.swt.fit.e,dgelYRIAPO,des,trastmat)
    }
})
```

## clean up, load the lists and concatenate into one giant result obj
```{r}
rm(list=ls())
gc(full=T)
```

## load the tmp RDSes
## not included in concatenation because not run (properly, yet): YRI as continuous variable (model 5/test 8)
```{r}
sexlabel <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/02bTMP_keeplist_sexlabel_noMixClusts.RDS")
E2E4label <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/02bTMP_keeplist_E2E4label_noMixClusts.RDS")
ApoAllelesLabel <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/02bTMP_keeplist_ApoAllelesLabel_noMixClusts.RDS")
YRIlabel.cat <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/02bTMP_keeplist_AncesGroupByLabel_noMixClusts.RDS")

## combine into one metalist
outlist <- list(SexByLabel=sexlabel,E2E4ByLabel=E2E4label,
                ApoAllelesByLabel=ApoAllelesLabel,AncesGrpByLabel=YRIlabel.cat)

saveRDS(outlist,"processed-data/12_DEanalyses_removedsampsAndFinalNMseg/02b-voomLmFit_objs_and_DE_tables_LabSex_LabE2vE4_LabSingleGenos_LabAncesGrp_noMixClusts.RDS")
```
## can now delete 02bTMP RDSes.

reprod inf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.2

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  stats     graphics  grDevices utils     methods   base     

other attached packages:
[1] BiocParallel_1.40.0 parallelly_1.43.0   colorout_1.3-0.2    edgeR_4.4.2        
[5] limma_3.62.2        data.table_1.17.0   rlang_1.1.5        

loaded via a namespace (and not attached):
 [1] digest_0.6.37     codetools_0.2-20  fastmap_1.2.0     xfun_0.52        
 [5] lattice_0.22-6    knitr_1.50        htmltools_0.5.8.1 rmarkdown_2.29   
 [9] cli_3.6.4         grid_4.4.3        statmod_1.5.0     compiler_4.4.3   
[13] rprojroot_2.0.4   here_1.0.1        rstudioapi_0.17.1 tools_4.4.3      
[17] evaluate_1.0.3    yaml_2.3.10       locfit_1.5-9.12  
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sequoia 15.3.2
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-05-15
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)
 quarto   1.4.555 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/quarto

─ Packages ────────────────────────────────────────────────────────────────────────
 !  package      * version  date (UTC) lib source
    BiocParallel * 1.40.0   2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
 VP cli            3.6.4    2025-04-23 [2] CRAN (R 4.4.1) (on disk 3.6.5)
    codetools      0.2-20   2024-03-31 [1] CRAN (R 4.4.3)
    colorout     * 1.3-0.2  2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    data.table   * 1.17.0   2025-02-22 [1] CRAN (R 4.4.1)
    digest         0.6.37   2024-08-19 [1] CRAN (R 4.4.1)
    edgeR        * 4.4.2    2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    evaluate       1.0.3    2025-01-10 [1] CRAN (R 4.4.1)
    fastmap        1.2.0    2024-05-15 [1] CRAN (R 4.4.0)
    here           1.0.1    2020-12-13 [1] CRAN (R 4.4.0)
    htmltools      0.5.8.1  2024-04-04 [1] CRAN (R 4.4.0)
    knitr          1.50     2025-03-16 [1] CRAN (R 4.4.2)
    lattice        0.22-6   2024-03-20 [1] CRAN (R 4.4.3)
    limma        * 3.62.2   2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    locfit         1.5-9.12 2025-03-05 [1] CRAN (R 4.4.1)
    parallelly   * 1.43.0   2025-03-24 [1] CRAN (R 4.4.1)
 VP rlang        * 1.1.5    2025-04-11 [2] CRAN (R 4.4.1) (on disk 1.1.6)
    rmarkdown      2.29     2024-11-04 [1] CRAN (R 4.4.1)
    rprojroot      2.0.4    2023-11-05 [1] CRAN (R 4.4.0)
    rstudioapi     0.17.1   2024-10-22 [1] CRAN (R 4.4.1)
    sessioninfo    1.2.3    2025-02-05 [1] CRAN (R 4.4.1)
    statmod        1.5.0    2023-01-06 [1] CRAN (R 4.4.0)
    xfun           0.52     2025-04-02 [1] CRAN (R 4.4.1)
    yaml           2.3.10   2024-07-26 [1] CRAN (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * ── Packages attached to the search path.
 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

───────────────────────────────────────────────────────────────────────────────────
