---
title: "12-LCNMpos_vs_NMneg_SCORPION_Generegnet"
author: "Bernie Mulvey"
date: "2025-07-15"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(SpatialExperiment)
library(dorothea)
library(SCORPION)
library(Matrix)
# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")

```

```{r}
lc3 <- readRDS("processed-data/06-countsOnly_noImgData_QCed_SPE_split_to_tissSections.RDS")

## Drop Br5517 (unknown anatomy/orientation), Br5276 (LC has been depleted by QC), Br5712 (unknown anatomy/orientation). Keep 6297; no LC but anatomy is right in vicinity.

lc3 <- lc3[,!(lc3$brnum %in% paste0("Br",c(5517,5276,5712)))]

## subset to LC spots, load LCNM+- annots
lcnm <- fread("processed-data/10_subclustering/04-LCsubclus-by-NM_otherclusts-by-LCorLCNMadjacency.txt")
lcnm <- lcnm[NMseg_subclus %in% paste0("LC_NM",c("pos","neg"))&rn %in% colnames(lc3),.(rn,NMseg_subclus)]

lc3 <- lc3[,colnames(lc3) %in% lcnm$rn]
lcnm <- DataFrame(lcnm,row.names=lcnm$rn)[colnames(lc3),]
lc3$lcnm <- lcnm$NMseg_subclus
rm(lcnm)

colData(lc3)$sample_id <- gsub(colData(lc3)$sample_id,pattern="-",replacement="_")
```

SCORPION
```{r}
tftab <- as.data.frame(dorothea::dorothea_hs) # needs to be a data.frame 
tftab$confidence <- NULL # column 3 needs to be the sign/directionality
load("raw-data/ref_data/hg38_PPI.RData") # from paper repository 
hsaPPI <- as.data.frame(hsaPPI) # needs to be a data.frame 

rownames(lc3) <- rowData(lc3)$gene_name
lcnmct <- counts(lc3[,lc3$lcnm=="LC_NMpos"])
lcct <- counts(lc3[,lc3$lcnm=="LC_NMneg"])

xprmats <- list(NMpos=lcnmct,NMneg=lcct)
rm(lc3,lcnmct,lcct)
gc(full=T)

## re  scorpion arguments not documented (clearly):
## pcnet appears to have been a predecessor method -- they use spearman in the paper
## error in code as written: no applicable method for `@` applied to an object of class "matrix" coming from somewehere ca line 150 in runPANDA -_- 
## add two lines at line 150 in runPANDA:     geneCoreg <- as.matrix(geneCoreg)
    # geneCoreg <- as(geneCoreg, "dgCMatrix")
## trace(SCORPION:::runPANDA,edit=TRUE)

scorpii <- lapply(xprmats,FUN=function(x){
  scorpi <- scorpion(tfMotifs = tftab, # as used by authors
                   gexMatrix = x,
                   ppiNet = hsaPPI, # as used by authors
                   nCores = 8,
                   gammaValue = 6, #  "Graining level of data (proportion of number of single cells in the initial dataset to the number of super-cells in the final dataset)." Default is 10, but since these are high-content multicellular Visium spots rather than single cells, this can be a little smaller than it would be for scseq. 
                   nPC = 20, # default is 25
                   assocMethod = "spearman", # default
                   alphaValue = 0.1, # default
                   hammingValue = 0.001, # default
                   nIter = Inf, # default
                   outNet = c("regulatory","coregulatory","cooperative"), # all networks
                   showProgress = TRUE,
                   randomizationMethod = "None", # default is "none"
                   filterExpr = TRUE,
                   scaleByPresent = FALSE, # default
                   zScaling=TRUE) #default
  return(scorpi)
})

saveRDS(scorpii,"processed-data/12_DEanalyses_removedsampsAndFinalNMseg/12a-LCNMpos_NMneg_scorpii_actual.RDS")
```
── SCORPION ────────────────────────────────────────────────────────────────────────────────────
✔ Initializing and validating
✔ Verified sufficient samples
ℹ Normalizing networks
ℹ Learning Network
ℹ Using tanimoto similarity
✔ Successfully ran SCORPION on 16520 Genes and 603 TFs
ℹ Time elapsed: 5920.98 seconds          
                                         
── SCORPION ────────────────────────────────────────────────────────────────────────────────────
✔ Initializing and validating
✔ Verified sufficient samples
ℹ Normalizing networks
ℹ Learning Network
ℹ Using tanimoto similarity
✔ Successfully ran SCORPION on 16586 Genes and 603 TFs
ℹ Time elapsed: 6147.76 seconds         


## also generate shuffled-genewise networks in case we need them for something comparative later
```{r}
scorpii <- lapply(xprmats,FUN=function(x){
  scorpi <- scorpion(tfMotifs = tftab, # as used by authors
                   gexMatrix = x,
                   ppiNet = hsaPPI, # as used by authors
                   nCores = 8,
                   gammaValue = 6, #  "Graining level of data (proportion of number of single cells in the initial dataset to the number of super-cells in the final dataset)." Default is 10, but since these are high-content multicellular Visium spots rather than single cells, this can be a little smaller than it would be for scseq. 
                   nPC = 20, # default is 25
                   assocMethod = "spearman",
                   alphaValue = 0.1, # default
                   hammingValue = 0.001, # default
                   nIter = Inf, # default
                   outNet = c("regulatory","coregulatory","cooperative"), # all networks
                   showProgress = TRUE,
                   randomizationMethod = "within.gene", # default is "none"
                   filterExpr = TRUE,
                   scaleByPresent = FALSE, # default
                   zScaling=TRUE) #default
  return(scorpi)
})

saveRDS(scorpii,"processed-data/12_DEanalyses_removedsampsAndFinalNMseg/12b-LCNMpos_NMneg_scorpii_RandomizedByGene.RDS")
```
── SCORPION ───────────────────────────────────────────────────────────────────────────────────
✔ Initializing and validating
ℹ Randomizing by reordering each gene's expression
✔ Verified sufficient samples
ℹ Normalizing networks
ℹ Learning Network
ℹ Using tanimoto similarity
✔ Successfully ran SCORPION on 16520 Genes and 603 TFs
ℹ Time elapsed: 4500.32 seconds          
                                         
── SCORPION ───────────────────────────────────────────────────────────────────────────────────
✔ Initializing and validating
ℹ Randomizing by reordering each gene's expression
✔ Verified sufficient samples
ℹ Normalizing networks
ℹ Learning Network
ℹ Using tanimoto similarity
✔ Successfully ran SCORPION on 16586 Genes and 603 TFs
ℹ Time elapsed: 4571.28 seconds         

### To examine differences in GRN activity between conditions/groups, the authors then..
```{r}
scorpii <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/12a-LCNMpos_NMneg_scorpii_actual.RDS")

```


```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.2

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] BiocParallel_1.40.0         parallelly_1.43.0           colorout_1.3-0.2           
 [4] Matrix_1.7-3                SCORPION_1.0.2              dorothea_1.18.0            
 [7] SpatialExperiment_1.16.0    SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0
[10] Biobase_2.66.0              GenomicRanges_1.58.0        GenomeInfoDb_1.42.3        
[13] IRanges_2.40.1              S4Vectors_0.44.0            BiocGenerics_0.52.0        
[16] MatrixGenerics_1.18.1       matrixStats_1.5.0           gridExtra_2.3              
[19] ggtext_0.1.2                ggplot2_3.5.2               data.table_1.17.4          
[22] rlang_1.1.5                

loaded via a namespace (and not attached):
 [1] gtable_0.3.6            rjson_0.2.23            xfun_0.52              
 [4] lattice_0.22-6          vctrs_0.6.5             tools_4.4.3            
 [7] generics_0.1.4          parallel_4.4.3          tibble_3.2.1           
[10] pkgconfig_2.0.3         RColorBrewer_1.1-3      lifecycle_1.0.4        
[13] GenomeInfoDbData_1.2.13 compiler_4.4.3          farver_2.1.2           
[16] codetools_0.2-20        htmltools_0.5.8.1       yaml_2.3.10            
[19] pillar_1.10.2           crayon_1.5.3            DelayedArray_0.32.0    
[22] magick_2.8.6            abind_1.4-8             tidyselect_1.2.1       
[25] digest_0.6.37           dplyr_1.1.4             rprojroot_2.0.4        
[28] fastmap_1.2.0           grid_4.4.3              here_1.0.1             
[31] cli_3.6.4               SparseArray_1.6.2       magrittr_2.0.3         
[34] S4Arrays_1.6.0          dichromat_2.0-0.1       withr_3.0.2            
[37] scales_1.4.0            UCSC.utils_1.2.0        rmarkdown_2.29         
[40] XVector_0.46.0          httr_1.4.7              igraph_2.1.4           
[43] RANN_2.6.2              bcellViper_1.42.0       pbapply_1.7-2          
[46] evaluate_1.0.3          knitr_1.50              irlba_2.3.5.1          
[49] gridtext_0.1.5          Rcpp_1.0.14             glue_1.8.0             
[52] xml2_1.3.8              rstudioapi_0.17.1       jsonlite_2.0.0         
[55] R6_2.6.1                zlibbioc_1.52.0        
> sessioninfo::session_info()
─ Session info ──────────────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sequoia 15.3.2
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-07-15
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)
 quarto   1.4.555 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/quarto

─ Packages ──────────────────────────────────────────────────────────────────────────────────
 !  package              * version date (UTC) lib source
    abind                  1.4-8   2024-09-12 [1] CRAN (R 4.4.1)
    bcellViper             1.42.0  2024-10-31 [1] Bioconductor 3.20 (R 4.4.3)
    Biobase              * 2.66.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocParallel         * 1.40.0  2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
 VP cli                    3.6.4   2025-04-23 [2] CRAN (R 4.4.1) (on disk 3.6.5)
    codetools              0.2-20  2024-03-31 [1] CRAN (R 4.4.3)
    colorout             * 1.3-0.2 2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    crayon                 1.5.3   2024-06-20 [1] CRAN (R 4.4.0)
    data.table           * 1.17.4  2025-05-26 [1] CRAN (R 4.4.3)
    DelayedArray           0.32.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    dichromat              2.0-0.1 2022-05-02 [1] CRAN (R 4.4.0)
    digest                 0.6.37  2024-08-19 [1] CRAN (R 4.4.1)
    dorothea             * 1.18.0  2024-10-31 [1] Bioconductor 3.20 (R 4.4.3)
    dplyr                  1.1.4   2023-11-17 [1] CRAN (R 4.4.0)
    evaluate               1.0.3   2025-01-10 [1] CRAN (R 4.4.1)
    farver                 2.1.2   2024-05-13 [1] CRAN (R 4.4.0)
    fastmap                1.2.0   2024-05-15 [1] CRAN (R 4.4.0)
    generics               0.1.4   2025-05-09 [1] CRAN (R 4.4.1)
    GenomeInfoDb         * 1.42.3  2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13  2024-12-12 [1] Bioconductor
    GenomicRanges        * 1.58.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ggplot2              * 3.5.2   2025-04-09 [1] CRAN (R 4.4.1)
    ggtext               * 0.1.2   2022-09-16 [1] CRAN (R 4.4.0)
    glue                   1.8.0   2024-09-30 [1] CRAN (R 4.4.1)
    gridExtra            * 2.3     2017-09-09 [1] CRAN (R 4.4.0)
    gridtext               0.1.5   2022-09-16 [1] CRAN (R 4.4.0)
    gtable                 0.3.6   2024-10-25 [1] CRAN (R 4.4.1)
    here                   1.0.1   2020-12-13 [1] CRAN (R 4.4.0)
    htmltools              0.5.8.1 2024-04-04 [1] CRAN (R 4.4.0)
    httr                   1.4.7   2023-08-15 [1] CRAN (R 4.4.0)
    igraph                 2.1.4   2025-01-23 [1] CRAN (R 4.4.1)
    IRanges              * 2.40.1  2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    irlba                  2.3.5.1 2022-10-03 [1] CRAN (R 4.4.0)
    jsonlite               2.0.0   2025-03-27 [1] CRAN (R 4.4.1)
    knitr                  1.50    2025-03-16 [1] CRAN (R 4.4.2)
    lattice                0.22-6  2024-03-20 [1] CRAN (R 4.4.3)
    lifecycle              1.0.4   2023-11-07 [1] CRAN (R 4.4.0)
    magick                 2.8.6   2025-03-23 [1] CRAN (R 4.4.1)
    magrittr               2.0.3   2022-03-30 [1] CRAN (R 4.4.0)
    Matrix               * 1.7-3   2025-03-11 [1] CRAN (R 4.4.1)
    MatrixGenerics       * 1.18.1  2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    matrixStats          * 1.5.0   2025-01-07 [1] CRAN (R 4.4.1)
    parallelly           * 1.43.0  2025-03-24 [1] CRAN (R 4.4.1)
    pbapply                1.7-2   2023-06-27 [1] CRAN (R 4.4.0)
    pillar                 1.10.2  2025-04-05 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3   2019-09-22 [1] CRAN (R 4.4.0)
    R6                     2.6.1   2025-02-15 [1] CRAN (R 4.4.1)
    RANN                   2.6.2   2024-08-25 [1] CRAN (R 4.4.1)
    RColorBrewer           1.1-3   2022-04-03 [1] CRAN (R 4.4.0)
    Rcpp                   1.0.14  2025-01-12 [1] CRAN (R 4.4.1)
    rjson                  0.2.23  2024-09-16 [1] CRAN (R 4.4.1)
 VP rlang                * 1.1.5   2025-04-11 [2] CRAN (R 4.4.1) (on disk 1.1.6)
    rmarkdown              2.29    2024-11-04 [1] CRAN (R 4.4.1)
    rprojroot              2.0.4   2023-11-05 [1] CRAN (R 4.4.0)
    rstudioapi             0.17.1  2024-10-22 [1] CRAN (R 4.4.1)
    S4Arrays               1.6.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.4.0   2025-04-24 [1] CRAN (R 4.4.1)
    SCORPION             * 1.0.2   2024-05-29 [1] CRAN (R 4.4.0)
    sessioninfo            1.2.3   2025-02-05 [1] CRAN (R 4.4.1)
    SingleCellExperiment * 1.28.1  2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    SparseArray            1.6.2   2025-02-20 [1] Bioconductor 3.20 (R 4.4.2)
    SpatialExperiment    * 1.16.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SummarizedExperiment * 1.36.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    tibble                 3.2.1   2023-03-20 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1   2024-03-11 [1] CRAN (R 4.4.0)
    UCSC.utils             1.2.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    vctrs                  0.6.5   2023-12-01 [1] CRAN (R 4.4.0)
    withr                  3.0.2   2024-10-28 [1] CRAN (R 4.4.1)
    xfun                   0.52    2025-04-02 [1] CRAN (R 4.4.1)
    xml2                   1.3.8   2025-03-14 [1] CRAN (R 4.4.1)
    XVector                0.46.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    yaml                   2.3.10  2024-07-26 [1] CRAN (R 4.4.0)
    zlibbioc               1.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * ── Packages attached to the search path.
 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

─────────────────────────────────────────────────────────────────────────────────────────────
