---
title: "12a-d-Make_SCORPION_LCNMpos_NMneg_Generegnets"
author: "Bernie Mulvey"
date: "2025-08-31"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(SpatialExperiment)
library(dorothea)
library(SCORPION)
library(Matrix)
# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")

```

## load SPE and add nm annots
```{r}
lc3 <- readRDS("processed-data/06-countsOnly_noImgData_QCed_SPE_split_to_tissSections.RDS")

## Drop Br5517 (unknown anatomy/orientation), Br5276 (LC has been depleted by QC), Br5712 (unknown anatomy/orientation). Keep 6297; no LC but anatomy is right in vicinity.

lc3 <- lc3[,!(lc3$brnum %in% paste0("Br",c(5517,5276,5712)))]

## subset to LC spots, load LCNM+- annots
lcnm <- fread("processed-data/10_subclustering/04-LCsubclus-by-NM_otherclusts-by-LCorLCNMadjacency.txt")
lcnm <- lcnm[NMseg_subclus %in% paste0("LC_NM",c("pos","neg"))&rn %in% colnames(lc3),.(rn,NMseg_subclus)]

lc3 <- lc3[,colnames(lc3) %in% lcnm$rn]
lcnm <- DataFrame(lcnm,row.names=lcnm$rn)[colnames(lc3),]
lc3$lcnm <- lcnm$NMseg_subclus
rm(lcnm)

colData(lc3)$sample_id <- gsub(colData(lc3)$sample_id,pattern="-",replacement="_")
```



#### SCORPION with minimalist filtering #### 
(filtering of TF PPI/motif-target tables to genes in the SPE only, only removing uncounted genes ie rowSums==0).

### WHY WE'RE PREFILTERING THE PPI AND MOTIF-TARG tables:
 missing genes are inferred if they're in one of the inputs but not the other. (e.g., a TF PPI and TF-motif target set are present but one of the TFs in the PPI is not in the expression data will still get an activity value without filtering in advance). teaching materials from the Kuijer lab using SCORPION recommend pre-filtering everything down to shared, expressed genes so we'll try it this way too. 
 
the aforementioned teaching mats: 
https://github.com/kuijjerlab/multi-omics-course/blob/main/multi-omics_workshop_full_code.ipynb ### 
```{r}
rownames(lc3) <- rowData(lc3)$gene_name # to match scorpion TF-TF and TF-target datasetss
lcnmct <- counts(lc3[,lc3$lcnm=="LC_NMpos"])
lcct <- counts(lc3[,lc3$lcnm=="LC_NMneg"])

## drop zero-count rows from each matrix
lcnmct <- lcnmct[rowSums(lcnmct)>0,]
lcct <- lcct[rowSums(lcct)>0,]
## for simplicity, use the SPE for the filtering below before removing
keepg <- unique(c(rownames(lcnmct),rownames(lcct)))
lc3 <- lc3[keepg,]

## get the tf motif - target gene relationships table as used in SCORPION (those in dorothea)
tftab <- as.data.table(dorothea::dorothea_hs)
tftab[,confidence:=NULL] # column 3 needs to be the sign/directionality
tftab <- tftab[tf %in% rownames(lc3)&target %in% rownames(lc3)] # TF and target are in at least one of the two count matrices
tftab <- as.data.frame(tftab) # needs to be a data.frame 

## get PPI from SCORPION paper and filter
load("raw-data/ref_data/hg38_PPI.RData") # from paper repository. no rownames 
hsaPPI <- as.data.table(hsaPPI)
hsaPPI <- hsaPPI[tf1 %in% rownames(lc3)&tf2 %in% rownames(lc3)] # each TF is in at least one of the two count matrices
hsaPPI <- as.data.frame(hsaPPI) # needs to be a data.frame 

xprmats <- list(NMpos=lcnmct,NMneg=lcct)
rm(lc3,lcnmct,lcct)
gc(full=T)

## re  scorpion arguments not documented (clearly):
## pcnet appears to have been a predecessor method -- they use spearman in the paper
## error in code as written: no applicable method for `@` applied to an object of class "matrix" coming from somewehere ca line 150 in runPANDA -_- 
## add two lines between lines 151 and 152 in runPANDA:     geneCoreg <- as.matrix(geneCoreg)
    # NOT COMMENTED, second line: geneCoreg <- as(geneCoreg, "dgCMatrix")##
## this is done by calling trace(SCORPION:::runPANDA,edit=TRUE) in the Rstudio session where this script will be executed 

scorpiia <- lapply(xprmats,FUN=function(x){
  scorpi <- scorpion(tfMotifs = tftab, # as used by authors
                   gexMatrix = x,
                   ppiNet = hsaPPI, # as used by authors
                   nCores = 8,
                   gammaValue = 6, #  "Graining level of data (proportion of number of single cells in the initial dataset to the number of super-cells in the final dataset)." Default is 10, but since these are high-content multicellular Visium spots rather than single cells, this can be a little smaller than it would be for scseq. 
                   nPC = 20, # default is 25
                   assocMethod = "spearman", # default
                   alphaValue = 0.1, # default
                   hammingValue = 0.001, # default
                   nIter = Inf, # default
                   outNet = c("regulatory","coregulatory","cooperative"), # all networks
                   showProgress = TRUE,
                   randomizationMethod = "None", # default is "None"
                   filterExpr = TRUE, # only filters 0-total-count genes
                   scaleByPresent = FALSE, # default
                   zScaling=TRUE) #default
  return(scorpi)
})

saveRDS(scorpii,"processed-data/12_DEanalyses_removedsampsAndFinalNMseg/12a-LCNMpos_NMneg_scorpii_zerofilt_actual.RDS")

rm(scorpiia)
gc(full=T)
```

## NM+ out:
── SCORPION ──────────────────────────────────────────────────────────────────────────────────────
✔ Initializing and validating
✔ Verified sufficient samples
ℹ Normalizing networks
ℹ Learning Network
ℹ Using tanimoto similarity
✔ Successfully ran SCORPION on 16493 Genes and 534 TFs
ℹ Time elapsed: 5686.55 seconds

## NM- out:
── SCORPION ──────────────────────────────────────────────────────────────────────────────────────
✔ Initializing and validating
✔ Verified sufficient samples
ℹ Normalizing networks
ℹ Learning Network
ℹ Using tanimoto similarity
✔ Successfully ran SCORPION on 16556 Genes and 534 TFs
ℹ Time elapsed: 5608.5 seconds  

## also generate shuffled-genewise networks in case we need them for something comparative later; minimalist filtering.
```{r}
scorpiib <- lapply(xprmats,FUN=function(x){
  scorpi <- scorpion(tfMotifs = tftab, # as used by authors
                   gexMatrix = x,
                   ppiNet = hsaPPI, # as used by authors
                   nCores = 8,
                   gammaValue = 6, #  "Graining level of data (proportion of number of single cells in the initial dataset to the number of super-cells in the final dataset)." Default is 10, but since these are high-content multicellular Visium spots rather than single cells, this can be a little smaller than it would be for scseq. 
                   nPC = 20, # default is 25
                   assocMethod = "spearman",
                   alphaValue = 0.1, # default
                   hammingValue = 0.001, # default
                   nIter = Inf, # default
                   outNet = c("regulatory","coregulatory","cooperative"), # all networks
                   showProgress = TRUE,
                   randomizationMethod = "within.gene", # default is "none"
                   filterExpr = TRUE,
                   scaleByPresent = FALSE, # default
                   zScaling=TRUE) #default
  return(scorpi)
})

saveRDS(scorpiib,"processed-data/12_DEanalyses_removedsampsAndFinalNMseg/12b-LCNMpos_NMneg_scorpii_zerofilt_RandomizedByGene.RDS")

rm(list=ls())
gc(full=T)
```
randomized outs:
── SCORPION ───────────────────────────────────────────────────────────────────────────────────
✔ Initializing and validating
ℹ Randomizing by reordering each gene's expression
✔ Verified sufficient samples
ℹ Normalizing networks
ℹ Learning Network
ℹ Using tanimoto similarity
✔ Successfully ran SCORPION on 16493 Genes and 534 TFs
ℹ Time elapsed: 4077.23 seconds          
                                         
── SCORPION ───────────────────────────────────────────────────────────────────────────────────
✔ Initializing and validating
ℹ Randomizing by reordering each gene's expression
✔ Verified sufficient samples
ℹ Normalizing networks
ℹ Learning Network
ℹ Using tanimoto similarity
✔ Successfully ran SCORPION on 16556 Genes and 534 TFs
ℹ Time elapsed: 4192.43 seconds          


#############################################################################################




####### SEPARATELY, PRE-FILTER to genes considered in full-on pseudobulk DE analyses--which should be the highest-confidence network components #######

specifically, filter the expression, TF PPI, and TF-Target matrices rather than using SCORPION's internal filtering arg (which just removes rows that have a total of zero counts and doesn't filter the TF PPI/motif-target tables to genes in the dataset besides).


### For explicitly filtering everything down to shared genes we'll use:
•LC NM pos, NM neg expressed genes: genes passing a DE filter in ANY DE test we've run (NM+ vs -, where a separate filtering step was performed, or any of the DE-model-provided filtering steps used for the multiple-domains-simultaneously)
•Target genes: all genes in the above set
•TFs: all TF genes in the above set

we cleared the workspace for running scorpion, so reload the SPE and annotations here.
```{r}
lc3 <- readRDS("processed-data/06-countsOnly_noImgData_QCed_SPE_split_to_tissSections.RDS")

## Drop Br5517 (unknown anatomy/orientation), Br5276 (LC has been depleted by QC), Br5712 (unknown anatomy/orientation). Keep 6297; no LC but anatomy is right in vicinity.

lc3 <- lc3[,!(lc3$brnum %in% paste0("Br",c(5517,5276,5712)))]

## subset to LC spots, load LCNM+- annots
lcnm <- fread("processed-data/10_subclustering/04-LCsubclus-by-NM_otherclusts-by-LCorLCNMadjacency.txt")
lcnm <- lcnm[NMseg_subclus %in% paste0("LC_NM",c("pos","neg"))&rn %in% colnames(lc3),.(rn,NMseg_subclus)]

lc3 <- lc3[,colnames(lc3) %in% lcnm$rn]
lcnm <- DataFrame(lcnm,row.names=lcnm$rn)[colnames(lc3),]
lc3$lcnm <- lcnm$NMseg_subclus
rm(lcnm)

colData(lc3)$sample_id <- gsub(colData(lc3)$sample_id,pattern="-",replacement="_")
```

## now load NM+/- DE pseudobulk object and other DE test objects to get genes to retain
```{r}
nmpb <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/01b-DGElist_per_LCNM_subcluster_10spotmin_edgeRdefaultLoExpFilt.RDS")
## and the rest of the DE tests which are here:
pb <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/02b-voomLmFit_objs_and_DE_tables_LabSex_LabE2vE4_LabSingleGenos_LabAnces_LabAncesE4E2_noMixClusts_YRIcontAndCat_JUL2025.RDS")

keepg <- unique(
   c(rownames(nmpb$LC_NMpos),
     rownames(nmpb$LC_NMneg), 
     pb$cont_ances$E2E4label$detabs$LC_E4vE2$gene_id,
     pb$cont_ances$sexlabel$detabs$LC_Sex$gene_id,
     pb$cat_ances$AncesByLabel$detabs$LC_YRI$gene_id,
     pb$cat_ances$YRIAPOlabel$detabs$LC_E4YRI_vs_E4CEU$gene_id,
     pb$cat_ances$YRIAPOlabel$detabs$LC_E2YRI_vs_E2CEU$gene_id,
     pb$cat_ances$YRIAPOlabel$detabs$LC_E4CEU_vs_E2CEU$gene_id,
     pb$cat_ances$YRIAPOlabel$detabs$LC_E4YRI_vs_E2YRI$gene_id
   )
)

lc3 <- lc3[keepg,]
# convert to symbols for matching PPI/target tables
rownames(lc3) <- rowData(lc3)$gene_name
lcnmct <- counts(lc3[,lc3$lcnm=="LC_NMpos"])
lcct <- counts(lc3[,lc3$lcnm=="LC_NMneg"])
```

## get the scorpion data fresh and filter to our genes of interest
```{r}
## get TF-target data as used in the SCORPION paper and filter. does not have rownames.
tftab <- as.data.table(dorothea::dorothea_hs) 
tftab[,confidence:=NULL] # column 3 needs to be the sign/directionality (which is the 4th column here because confidence collumn wedged in but not used by SCORPION)
tftab <- tftab[tf %in% rownames(lc3)&target %in% rownames(lc3)]
tftab <- as.data.frame(tftab) # needs to be a data.frame 

## get PPI from SCORPION paper and filter
load("raw-data/ref_data/hg38_PPI.RData") # from paper repository. no rownames 
hsaPPI <- as.data.table(hsaPPI)
hsaPPI <- hsaPPI[tf1 %in% rownames(lc3)&tf2 %in% rownames(lc3)]
hsaPPI <- as.data.frame(hsaPPI) # needs to be a data.frame 

# make list of single-spot expression matrices to apply to (NM+, NM-):
xprmats <- list(NMpos=lcnmct,NMneg=lcct)
rm(lcnmct,lcct,lc3,pb,nmpb,keepg)
gc(full=T)

## re  scorpion arguments not documented (clearly):
## pcnet appears to have been a predecessor method -- they use spearman in the paper
## error in code as written: no applicable method for `@` applied to an object of class "matrix" coming from somewehere ca line 150 in runPANDA -_- 
## add two lines between lines 151 and 152 in runPANDA:     geneCoreg <- as.matrix(geneCoreg)
    # NOT COMMENTED, second line: geneCoreg <- as(geneCoreg, "dgCMatrix")
## do this by calling trace(SCORPION:::runPANDA,edit=TRUE)

scorpiic <- lapply(xprmats,FUN=function(x){
  scorpi <- scorpion(tfMotifs = tftab, # as used by authors
                   gexMatrix = x,
                   ppiNet = hsaPPI, # as used by authors
                   nCores = 8,
                   gammaValue = 6, #  "Graining level of data (proportion of number of single cells in the initial dataset to the number of super-cells in the final dataset)." Default is 10, but since these are high-content multicellular Visium spots rather than single cells, this can be a little smaller than it would be for scseq. 
                   nPC = 20, # default is 25
                   assocMethod = "spearman", # default
                   alphaValue = 0.1, # default
                   hammingValue = 0.001, # default
                   nIter = Inf, # default
                   outNet = c("regulatory","coregulatory","cooperative"), # all networks
                   showProgress = TRUE,
                   randomizationMethod = "None", # default is "none"; other arguments return a randomized effectiely null network
                   filterExpr = FALSE, # default TRUE; we are doing this using our previous pseudobulking and paring down of everything to shared genes
                   scaleByPresent = FALSE, # default
                   zScaling=TRUE) #default
  return(scorpi)
})

saveRDS(scorpiic,"processed-data/12_DEanalyses_removedsampsAndFinalNMseg/12c-LCNMpos_NMneg_scorpii_pbulkGenesPreFilt_actual.RDS")

rm(scorpiic)
gc(full=T)
```
## console output reports:
── SCORPION ──────────────────────────────────────────────────────────────────────────────────────
✔ Initializing and validating
✔ Verified sufficient samples
ℹ Normalizing networks
ℹ Learning Network
ℹ Using tanimoto similarity
✔ Successfully ran SCORPION on 8187 Genes and 165 TFs
ℹ Time elapsed: 507.66 seconds

── SCORPION ──────────────────────────────────────────────────────────────────────────────────────
✔ Initializing and validating
✔ Verified sufficient samples
ℹ Normalizing networks
ℹ Learning Network
ℹ Using tanimoto similarity
✔ Successfully ran SCORPION on 8187 Genes and 165 TFs
ℹ Time elapsed: 495.43 seconds     

## also generate shuffled-genewise networks in case we need them for something comparative later, again using the now-filtered stuff.
```{r}
scorpiid <- lapply(xprmats,FUN=function(x){
  scorpi <- scorpion(tfMotifs = tftab, # as used by authors
                   gexMatrix = x,
                   ppiNet = hsaPPI, # as used by authors
                   nCores = 8,
                   gammaValue = 6, #  "Graining level of data (proportion of number of single cells in the initial dataset to the number of super-cells in the final dataset)." Default is 10, but since these are high-content multicellular Visium spots rather than single cells, this can be a little smaller than it would be for scseq. 
                   nPC = 20, # default is 25
                   assocMethod = "spearman",
                   alphaValue = 0.1, # default
                   hammingValue = 0.001, # default
                   nIter = Inf, # default
                   outNet = c("regulatory","coregulatory","cooperative"), # all networks
                   showProgress = TRUE,
                   randomizationMethod = "within.gene", # default is "none"
                   filterExpr = FALSE, # default TRUE; we pre-did this using our previous pseudobulking and paring down of everything to shared genes. that would have gotten rid of anything with rowSums==0 and quite a lot more than that. (filterExpr here just removes zero-sum rows).
                   scaleByPresent = FALSE, # default
                   zScaling=TRUE) #default
  return(scorpi)
})

saveRDS(scorpiid,"processed-data/12_DEanalyses_removedsampsAndFinalNMseg/12d-LCNMpos_NMneg_scorpii_pbulkGenesPreFilt_RandomizedByGene.RDS")
```
## console output reports: (accidentally deleted)


```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.2

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] BiocParallel_1.40.0         parallelly_1.43.0           colorout_1.3-0.2           
 [4] Matrix_1.7-3                SCORPION_1.0.2              dorothea_1.18.0            
 [7] SpatialExperiment_1.16.0    SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0
[10] Biobase_2.66.0              GenomicRanges_1.58.0        GenomeInfoDb_1.42.3        
[13] IRanges_2.40.1              S4Vectors_0.44.0            BiocGenerics_0.52.0        
[16] MatrixGenerics_1.18.1       matrixStats_1.5.0           data.table_1.17.4          
[19] rlang_1.1.5                

loaded via a namespace (and not attached):
 [1] bcellViper_1.42.0       generics_0.1.4          SparseArray_1.6.2      
 [4] lattice_0.22-6          digest_0.6.37           magrittr_2.0.3         
 [7] evaluate_1.0.3          grid_4.4.3              fastmap_1.2.0          
[10] rprojroot_2.0.4         jsonlite_2.0.0          httr_1.4.7             
[13] UCSC.utils_1.2.0        pbapply_1.7-2           codetools_0.2-20       
[16] abind_1.4-8             cli_3.6.4               crayon_1.5.3           
[19] XVector_0.46.0          RANN_2.6.2              DelayedArray_0.32.0    
[22] yaml_2.3.10             S4Arrays_1.6.0          parallel_4.4.3         
[25] tools_4.4.3             dplyr_1.1.4             GenomeInfoDbData_1.2.13
[28] here_1.0.1              vctrs_0.6.5             R6_2.6.1               
[31] lifecycle_1.0.4         magick_2.8.6            zlibbioc_1.52.0        
[34] irlba_2.3.5.1           pkgconfig_2.0.3         pillar_1.10.2          
[37] glue_1.8.0              Rcpp_1.0.14             tidyselect_1.2.1       
[40] xfun_0.52               tibble_3.2.1            rstudioapi_0.17.1      
[43] knitr_1.50              rjson_0.2.23            igraph_2.1.4           
[46] htmltools_0.5.8.1       rmarkdown_2.29          compiler_4.4.3         
> sessioninfo::session_info()
─ Session info ─────────────────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sequoia 15.3.2
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-09-01
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)
 quarto   1.4.555 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/quarto

─ Packages ─────────────────────────────────────────────────────────────────────────────────────
 !  package              * version date (UTC) lib source
    abind                  1.4-8   2024-09-12 [1] CRAN (R 4.4.1)
    bcellViper             1.42.0  2024-10-31 [1] Bioconductor 3.20 (R 4.4.3)
    Biobase              * 2.66.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocParallel         * 1.40.0  2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
 VP cli                    3.6.4   2025-04-23 [2] CRAN (R 4.4.1) (on disk 3.6.5)
    codetools              0.2-20  2024-03-31 [1] CRAN (R 4.4.3)
    colorout             * 1.3-0.2 2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    crayon                 1.5.3   2024-06-20 [1] CRAN (R 4.4.0)
    data.table           * 1.17.4  2025-05-26 [1] CRAN (R 4.4.3)
    DelayedArray           0.32.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    digest                 0.6.37  2024-08-19 [1] CRAN (R 4.4.1)
    dorothea             * 1.18.0  2024-10-31 [1] Bioconductor 3.20 (R 4.4.3)
    dplyr                  1.1.4   2023-11-17 [1] CRAN (R 4.4.0)
    evaluate               1.0.3   2025-01-10 [1] CRAN (R 4.4.1)
    fastmap                1.2.0   2024-05-15 [1] CRAN (R 4.4.0)
    generics               0.1.4   2025-05-09 [1] CRAN (R 4.4.1)
    GenomeInfoDb         * 1.42.3  2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13  2024-12-12 [1] Bioconductor
    GenomicRanges        * 1.58.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    glue                   1.8.0   2024-09-30 [1] CRAN (R 4.4.1)
    here                   1.0.1   2020-12-13 [1] CRAN (R 4.4.0)
    htmltools              0.5.8.1 2024-04-04 [1] CRAN (R 4.4.0)
    httr                   1.4.7   2023-08-15 [1] CRAN (R 4.4.0)
    igraph                 2.1.4   2025-01-23 [1] CRAN (R 4.4.1)
    IRanges              * 2.40.1  2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    irlba                  2.3.5.1 2022-10-03 [1] CRAN (R 4.4.0)
    jsonlite               2.0.0   2025-03-27 [1] CRAN (R 4.4.1)
    knitr                  1.50    2025-03-16 [1] CRAN (R 4.4.2)
    lattice                0.22-6  2024-03-20 [1] CRAN (R 4.4.3)
    lifecycle              1.0.4   2023-11-07 [1] CRAN (R 4.4.0)
    magick                 2.8.6   2025-03-23 [1] CRAN (R 4.4.1)
    magrittr               2.0.3   2022-03-30 [1] CRAN (R 4.4.0)
    Matrix               * 1.7-3   2025-03-11 [1] CRAN (R 4.4.1)
    MatrixGenerics       * 1.18.1  2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    matrixStats          * 1.5.0   2025-01-07 [1] CRAN (R 4.4.1)
    parallelly           * 1.43.0  2025-03-24 [1] CRAN (R 4.4.1)
    pbapply                1.7-2   2023-06-27 [1] CRAN (R 4.4.0)
    pillar                 1.10.2  2025-04-05 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3   2019-09-22 [1] CRAN (R 4.4.0)
    R6                     2.6.1   2025-02-15 [1] CRAN (R 4.4.1)
    RANN                   2.6.2   2024-08-25 [1] CRAN (R 4.4.1)
    Rcpp                   1.0.14  2025-01-12 [1] CRAN (R 4.4.1)
    rjson                  0.2.23  2024-09-16 [1] CRAN (R 4.4.1)
 VP rlang                * 1.1.5   2025-04-11 [2] CRAN (R 4.4.1) (on disk 1.1.6)
    rmarkdown              2.29    2024-11-04 [1] CRAN (R 4.4.1)
    rprojroot              2.0.4   2023-11-05 [1] CRAN (R 4.4.0)
    rstudioapi             0.17.1  2024-10-22 [1] CRAN (R 4.4.1)
    S4Arrays               1.6.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SCORPION             * 1.0.2   2024-05-29 [1] CRAN (R 4.4.0)
    sessioninfo            1.2.3   2025-02-05 [1] CRAN (R 4.4.1)
    SingleCellExperiment * 1.28.1  2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    SparseArray            1.6.2   2025-02-20 [1] Bioconductor 3.20 (R 4.4.2)
    SpatialExperiment    * 1.16.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SummarizedExperiment * 1.36.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    tibble                 3.2.1   2023-03-20 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1   2024-03-11 [1] CRAN (R 4.4.0)
    UCSC.utils             1.2.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    vctrs                  0.6.5   2023-12-01 [1] CRAN (R 4.4.0)
    xfun                   0.52    2025-04-02 [1] CRAN (R 4.4.1)
    XVector                0.46.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    yaml                   2.3.10  2024-07-26 [1] CRAN (R 4.4.0)
    zlibbioc               1.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * ── Packages attached to the search path.
 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

────────────────────────────────────────────────────────────────────────────────────────────────
