---
title: "13-NMexpression_LMM_hits_enrichr"
format: html
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(enrichR)
library(SpatialExperiment)
library(biomaRt)
library(stringr)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")


## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

We're focusing on intensity here, but we'll run it with the proportion results as well for kicks.

```{r}
nmlm <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/11-LCNMspots_NMmetric_LMMed_by_geneLogcounts_3apoeFactorizations.RDS")

nmlm <- nmlm$E4carrierYN
nmlm <- lapply(nmlm,function(l){
    l[variable %in% c(unique(gene_name),"APOE.x"),bonfP_gene:=p*(length(unique(l$gene_name)))]
    l[bonfP_gene>1,bonfP_gene:=1]
    l[variable %in% c(unique(gene_name),"APOE.x"),fdr:=p.adjust(p,"fdr",n = length(unique(l$gene_name)))] 
})

### we need entrez symbols for enrichr
lc3 <- readRDS("processed-data/06-countsOnly_noImgData_QCed_SPE_split_to_tissSections.RDS")
rddt <- as.data.table(rowData(lc3))
rm(lc3)

### biomart query
biomart <- useMart("ensembl",dataset="hsapiens_gene_ensembl")
lut <- biomaRt::getBM(attributes=c("ensembl_gene_id",
                                   "entrezgene_id",
                                   "entrezgene_accession",
                                   "entrezgene_description"),
                      mart=biomart,
                      filter="ensembl_gene_id",
                      values=unique(rddt$gene_id)) # uh ok "accession" = symbol

lut <- as.data.table(lut)
rddt <- merge.data.table(rddt[,.(gene_id,gene_name,chr)],lut[,.(ensembl_gene_id,entrezgene_accession)],by.x="gene_id",by.y="ensembl_gene_id")

## append entrezes to lmm results
nmlm <- lapply(nmlm,function(l){
    l <- merge.data.table(l,rddt[,.(gene_name,entrezgene_accession)],by="gene_name")
    l[variable==gene_name|variable=="APOE.x",variable:=entrezgene_accession] # change the variable column to match the entrez gene identifier for easier processing here
    return(l)
})

### drop duplicated entrezes because can't enrich analysis properly with such
nmlm <- lapply(nmlm, function(l){
    dupg <- l[entrezgene_accession==variable,.N,by=entrezgene_accession][N>1,entrezgene_accession]
    l <- l[!(entrezgene_accession %in% dupg)]
    return(l)
})

## make background sets 
nmlm.bg <- lapply(nmlm,function(l){
    unique(l$entrezgene_accession)
})

### clean up
rm(lut,dupg,rddt,biomart)
```

### enrichr setup
## rummagenes don't work here unfortunately.
```{r}
checkdbs <- enrichR::listEnrichrDbs()
enrichr_dbs <- grep(checkdbs$libraryName,pattern="Aging|Allen|ARCHS4|Azimuth|BioCarta_2016|BioPlex|CellMarker|ChEA_2022|CORUM|DepMap_CRISPR_GeneDependency_CellLines_2023|Descartes|DGIdb_Drug_Targets_2024|Disease_Perturbations|Disease_Signatures|Drug_Perturbations|DrugMatrix|Elsevier_Pathway|ENCODE_Histone_Modifications_2015|ENCODE_TF_ChIP-seq_2015|Roadmap|FANTOM6|Gene_Perturbations|GO_Biological_Process_2025|GO_Cellular_Component_2025|GO_Molecular_Function_2025|HMS_LINCS|HuBMAP|huMAP|IDG_Drug_Targs_2022|Jensen_DISEASES_Curated_2025|KEGG_2021_Human|Kinase_Perturbations|KOMP2|L1000|Ligand_Perturbations|lncHUB|MAGNET|MGI_Mammalian_Phenotype_Level_4_2024|NIBR_DRUGseq_2025|PerturbAtlas|PFOCR_Pathways_2023|PPI_Hub_Proteins|Proteomics|Reactome_Pathways_2024|SILAC|SubCell|SynGO_2024|Tabula_Muris|Tabula_Sapiens|TF_Perturbations|TF-LOF_Expression|TG_GATES_2020|The_Kinase_Library_2024|Tissue_Protein_Expression|Transcription_Factor_PPIs|TRANSFAC_and_JASPAR|TRRUST|WikiPathways_2024_Human|^DSigDB",value=T)

## filter out the ARCHS4 predictions re: rare disease, NIH PIs--interesting but not useful here
enrichr_dbs <- grep(enrichr_dbs,pattern="AutoRIF|GeneRIF",value=T,invert=T)

## two of these aren't working
enrichr_dbs <- enrichr_dbs[!(enrichr_dbs %in% c("PerturbAtlas_MouseGenePerturbationSigs","HMS_LINCS_KinomeScan"))]
rm(checkdbs)
```

Interestingly, enrichr allows for the gene sets to have a "membership value of 0-1" in the second column. we COULD test this later on.

but for now, let's just feed the genes in as a straight-up vector (against the whole set of genes fed into the lmm run.)

once not considering direction of assoc, the other with runs for each dir.
```{r}
## need to run serially, otherwise server will start blocking requests and/or gene sets requested will get mixed up. plus other unexpected errors
reslists <- list(fdr=list(),bonfP_gene=list())
i <- 1
for (i in (c(1:length(reslists)))){
    curthresh <- names(reslists)[i]
    j<-1
    for (j in c(1:length(unique(nmlm)))){
        d <- copy(nmlm[[j]])

        ## drop any dbs that throw specific errors
        # if(){
        #     cur_enrichr_dbs <- enrichr_dbs[!(enrichr_dbs %in% c("example"))]
        # } else {
        #     cur_enrichr_dbs <- enrichr_dbs
        # }
        
        cur_enrichr_dbs <- enrichr_dbs

        testg <- d[get(curthresh)<0.05,entrezgene_accession] # get the FDR or bonfp column and filter

        res <- enrichr(genes=testg,
                        background = nmlm.bg[[j]],
                        databases=cur_enrichr_dbs)
    
        res.dt <- rbindlist(res,idcol="Enrichr_database")
        res.dt[,NMmetric:=names(nmlm)[j]]
    
        reslists[[i]][[j]] <- res.dt
        names(reslists[[i]])[j] <- names(nmlm)[j]
    
        rm(d,res.dt,res,cur_enrichr_dbs,testg)
        gc(full=T)
    }
    rm(j,curthresh)
}
rm(i)

outlists <- list(nondir=reslists)

rm(reslists)
```

## directional tests
```{r}
## need to run serially, otherwise server will start blocking requests and/or gene sets requested will get mixed up. plus other unexpected errors

ks <- c("pos","neg")
k<-1
for (k in c(1:2)){
    reslists <- list(fdr=list(),bonfP_gene=list())
    curk <- ks[k]

    i <- 1
    for (i in (c(1:length(reslists)))){
        curthresh <- names(reslists)[i]
        j<-1
        for (j in c(1:length(unique(nmlm)))){
            d <- copy(nmlm[[j]])

            ## drop any dbs that throw specific errors
            # if(){
            #     cur_enrichr_dbs <- enrichr_dbs[!(enrichr_dbs %in% c("example"))]
            # } else {
            #     cur_enrichr_dbs <- enrichr_dbs
            # }
            
            cur_enrichr_dbs <- enrichr_dbs

            testd <- d[get(curthresh)<0.05] # get the FDR or bonfp column and filter
            if(curk=="pos"){
                testg <- testd[variable==entrezgene_accession&coef>0,entrezgene_accession]
            } else{
                testg <- testd[variable==entrezgene_accession&coef<0,entrezgene_accession]
            }

            res <- enrichr(genes=testg,
                            background = nmlm.bg[[j]],
                            databases=cur_enrichr_dbs)
        
            res.dt <- rbindlist(res,idcol="Enrichr_database")
            res.dt[,NMmetric:=paste0(names(nmlm)[j],"_",curk,"assoc")]
        
            reslists[[i]][[j]] <- res.dt
            names(reslists[[i]])[j] <- names(nmlm)[j]
        
            rm(d,res.dt,res,cur_enrichr_dbs)
            gc(full=T)
        }
        rm(j,curthresh)
    }
    rm(i)
    outlists[[curk]] <- reslists
    rm(reslists,curk)
}
```

```{r}
saveRDS(outlists,"processed-data/12_DEanalyses_removedsampsAndFinalNMseg/13-enrichr_fdrsig_and_bonfP_genes_NMintensityAndProp-to-geneXpr_lmms.RDS")
###
```

sessioninf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.2

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils    
[6] methods   base     

other attached packages:
 [1] BiocParallel_1.40.0        
 [2] parallelly_1.43.0          
 [3] colorout_1.3-0.2           
 [4] stringr_1.5.1              
 [5] biomaRt_2.62.1             
 [6] SpatialExperiment_1.16.0   
 [7] SingleCellExperiment_1.28.1
 [8] SummarizedExperiment_1.36.0
 [9] Biobase_2.66.0             
[10] GenomicRanges_1.58.0       
[11] GenomeInfoDb_1.42.3        
[12] IRanges_2.40.1             
[13] S4Vectors_0.44.0           
[14] BiocGenerics_0.52.0        
[15] MatrixGenerics_1.18.1      
[16] matrixStats_1.5.0          
[17] enrichR_3.4                
[18] data.table_1.17.4          
[19] rlang_1.1.5                

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.1        dplyr_1.1.4            
 [3] farver_2.1.2            blob_1.2.4             
 [5] filelock_1.0.3          Biostrings_2.74.1      
 [7] fastmap_1.2.0           BiocFileCache_2.14.0   
 [9] digest_0.6.37           lifecycle_1.0.4        
[11] KEGGREST_1.46.0         RSQLite_2.3.9          
[13] magrittr_2.0.3          compiler_4.4.3         
[15] progress_1.2.3          tools_4.4.3            
[17] knitr_1.50              prettyunits_1.2.0      
[19] S4Arrays_1.6.0          bit_4.6.0              
[21] curl_6.2.2              here_1.0.1             
[23] DelayedArray_0.32.0     xml2_1.3.8             
[25] RColorBrewer_1.1-3      abind_1.4-8            
[27] withr_3.0.2             purrr_1.0.4            
[29] grid_4.4.3              ggplot2_3.5.2          
[31] scales_1.4.0            dichromat_2.0-0.1      
[33] cli_3.6.4               crayon_1.5.3           
[35] generics_0.1.4          httr_1.4.7             
[37] rjson_0.2.23            DBI_1.2.3              
[39] cachem_1.1.0            zlibbioc_1.52.0        
[41] parallel_4.4.3          AnnotationDbi_1.68.0   
[43] XVector_0.46.0          WriteXLS_6.7.0         
[45] vctrs_0.6.5             Matrix_1.7-3           
[47] jsonlite_2.0.0          hms_1.1.3              
[49] bit64_4.6.0-1           magick_2.8.6           
[51] glue_1.8.0              codetools_0.2-20       
[53] stringi_1.8.7           gtable_0.3.6           
[55] UCSC.utils_1.2.0        tibble_3.2.1           
[57] pillar_1.10.2           rappdirs_0.3.3         
[59] GenomeInfoDbData_1.2.13 R6_2.6.1               
[61] dbplyr_2.5.0            httr2_1.1.1            
[63] rprojroot_2.0.4         evaluate_1.0.3         
[65] lattice_0.22-6          png_0.1-8              
[67] memoise_2.0.1           Rcpp_1.0.14            
[69] SparseArray_1.6.2       xfun_0.52              
[71] pkgconfig_2.0.3        
> sessioninfo::session_info()
─ Session info ──────────────────────────────────────
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sequoia 15.3.2
 system   aarch64, darwin20
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-08-07
 pandoc   NA
 quarto   1.7.32 @ /Applications/Positron.app/Contents/Resources/app/quarto/bin/quarto

─ Packages ──────────────────────────────────────────
 !  package              * version date (UTC) lib source
    abind                  1.4-8   2024-09-12 [1] CRAN (R 4.4.1)
    AnnotationDbi          1.68.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    Biobase              * 2.66.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocFileCache          2.14.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocParallel         * 1.40.0  2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    biomaRt              * 2.62.1  2025-01-30 [1] Bioconductor 3.20 (R 4.4.2)
    Biostrings             2.74.1  2024-12-16 [1] Bioconductor 3.20 (R 4.4.2)
    bit                    4.6.0   2025-03-06 [1] CRAN (R 4.4.1)
    bit64                  4.6.0-1 2025-01-16 [1] CRAN (R 4.4.1)
    blob                   1.2.4   2023-03-17 [1] CRAN (R 4.4.0)
    cachem                 1.1.0   2024-05-16 [1] CRAN (R 4.4.0)
 VP cli                    3.6.4   2025-04-23 [2] CRAN (R 4.4.1) (on disk 3.6.5)
    codetools              0.2-20  2024-03-31 [1] CRAN (R 4.4.3)
    colorout             * 1.3-0.2 2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    crayon                 1.5.3   2024-06-20 [1] CRAN (R 4.4.0)
    curl                   6.2.2   2025-03-24 [1] CRAN (R 4.4.1)
    data.table           * 1.17.4  2025-05-26 [1] CRAN (R 4.4.3)
    DBI                    1.2.3   2024-06-02 [1] CRAN (R 4.4.0)
    dbplyr                 2.5.0   2024-03-19 [1] CRAN (R 4.4.0)
    DelayedArray           0.32.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    dichromat              2.0-0.1 2022-05-02 [1] CRAN (R 4.4.0)
    digest                 0.6.37  2024-08-19 [1] CRAN (R 4.4.1)
    dplyr                  1.1.4   2023-11-17 [1] CRAN (R 4.4.0)
    enrichR              * 3.4     2025-02-02 [1] CRAN (R 4.4.1)
    evaluate               1.0.3   2025-01-10 [1] CRAN (R 4.4.1)
    farver                 2.1.2   2024-05-13 [1] CRAN (R 4.4.0)
    fastmap                1.2.0   2024-05-15 [1] CRAN (R 4.4.0)
    filelock               1.0.3   2023-12-11 [1] CRAN (R 4.4.0)
    generics               0.1.4   2025-05-09 [1] CRAN (R 4.4.1)
    GenomeInfoDb         * 1.42.3  2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13  2024-12-12 [1] Bioconductor
    GenomicRanges        * 1.58.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ggplot2                3.5.2   2025-04-09 [1] CRAN (R 4.4.1)
    glue                   1.8.0   2024-09-30 [1] CRAN (R 4.4.1)
    gtable                 0.3.6   2024-10-25 [1] CRAN (R 4.4.1)
    here                   1.0.1   2020-12-13 [1] CRAN (R 4.4.0)
    hms                    1.1.3   2023-03-21 [1] CRAN (R 4.4.0)
    httr                   1.4.7   2023-08-15 [1] CRAN (R 4.4.0)
    httr2                  1.1.1   2025-03-08 [1] CRAN (R 4.4.1)
    IRanges              * 2.40.1  2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    jsonlite               2.0.0   2025-03-27 [1] CRAN (R 4.4.1)
    KEGGREST               1.46.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    knitr                  1.50    2025-03-16 [1] CRAN (R 4.4.2)
    lattice                0.22-6  2024-03-20 [1] CRAN (R 4.4.3)
    lifecycle              1.0.4   2023-11-07 [1] CRAN (R 4.4.0)
    magick                 2.8.6   2025-03-23 [1] CRAN (R 4.4.1)
    magrittr               2.0.3   2022-03-30 [1] CRAN (R 4.4.0)
    Matrix                 1.7-3   2025-03-11 [1] CRAN (R 4.4.1)
    MatrixGenerics       * 1.18.1  2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    matrixStats          * 1.5.0   2025-01-07 [1] CRAN (R 4.4.1)
    memoise                2.0.1   2021-11-26 [1] CRAN (R 4.4.0)
    parallelly           * 1.43.0  2025-03-24 [1] CRAN (R 4.4.1)
    pillar                 1.10.2  2025-04-05 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3   2019-09-22 [1] CRAN (R 4.4.0)
    png                    0.1-8   2022-11-29 [1] CRAN (R 4.4.0)
    prettyunits            1.2.0   2023-09-24 [1] CRAN (R 4.4.0)
    progress               1.2.3   2023-12-06 [1] CRAN (R 4.4.0)
    purrr                  1.0.4   2025-02-05 [1] CRAN (R 4.4.1)
    R6                     2.6.1   2025-02-15 [1] CRAN (R 4.4.1)
    rappdirs               0.3.3   2021-01-31 [1] CRAN (R 4.4.0)
    RColorBrewer           1.1-3   2022-04-03 [1] CRAN (R 4.4.0)
    Rcpp                   1.0.14  2025-01-12 [1] CRAN (R 4.4.1)
    rjson                  0.2.23  2024-09-16 [1] CRAN (R 4.4.1)
 VP rlang                * 1.1.5   2025-04-11 [2] CRAN (R 4.4.1) (on disk 1.1.6)
    rprojroot              2.0.4   2023-11-05 [1] CRAN (R 4.4.0)
    RSQLite                2.3.9   2024-12-03 [1] CRAN (R 4.4.1)
    S4Arrays               1.6.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.4.0   2025-04-24 [1] CRAN (R 4.4.1)
    sessioninfo            1.2.3   2025-02-05 [1] CRAN (R 4.4.1)
    SingleCellExperiment * 1.28.1  2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    SparseArray            1.6.2   2025-02-20 [1] Bioconductor 3.20 (R 4.4.2)
    SpatialExperiment    * 1.16.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    stringi                1.8.7   2025-03-27 [1] CRAN (R 4.4.1)
    stringr              * 1.5.1   2023-11-14 [1] CRAN (R 4.4.0)
    SummarizedExperiment * 1.36.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    tibble                 3.2.1   2023-03-20 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1   2024-03-11 [1] CRAN (R 4.4.0)
    UCSC.utils             1.2.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    vctrs                  0.6.5   2023-12-01 [1] CRAN (R 4.4.0)
    withr                  3.0.2   2024-10-28 [1] CRAN (R 4.4.1)
    WriteXLS               6.7.0   2024-07-20 [1] CRAN (R 4.4.0)
    xfun                   0.52    2025-04-02 [1] CRAN (R 4.4.1)
    xml2                   1.3.8   2025-03-14 [1] CRAN (R 4.4.1)
    XVector                0.46.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    zlibbioc               1.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * ── Packages attached to the search path.
 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

─────────────────────────────────────────────────────
