---
title: "Fig2-Clusters, Markers, WeberDivecha registration, LC NMpos vs NMneg, LConly NMF-GSEA"
format: html
date: "07-09-2025"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(ggplot2)
library(ggtext)
library(gridExtra)
library(SpatialExperiment)
library(spatialLIBD)
library(ggrastr)
library(Cairo) # pdf device for ^ to properly save; also used for complexheatmap raster functions
library(circlize) #colorRamp2
library(viridis) # magma
library(ggrepel) # volcano point labeling

## rstudio GUI tweaks
require(colorout)
ColorOut()
options('styler.addins_style_transformer' = 'biocthis::bioc_style()')
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn='')
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost='localhost')


## ggplot defaults
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 9), axis.title.x = element_text(size = 8), axis.text.y = element_text(size = 8), axis.title.y = element_text(size =9), plot.title = element_markdown(size = 11,hjust=0.5), strip.text = element_text(size=11), legend.text = element_text(size=8), legend.title = element_text(size=9,hjust=0.5),plot.title.position='plot'))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace('datasets')
```



Fig 3A: LC NMF GSEA didn't really yield anything. instead, use the enrichr results on factor-specific genes (with background sets fed in as all genes on which the respective NMFs were performed).

Start by grabbing results for the following terms, which popped up in one or more factor-specific-gene set, then see how many we really have 
postsynaptic (GO)
presynaptic (GO)
ubiquitin (GO, other pathway/ontol dbs)
esr1
mitochondrial (GO, other pathway/ontol dbs)
calml3 (protein complexes)
znf365 (TF coexp, TF ppi)
LMTK3 (kinase coexp)
subceruleus (allen atlas regions)
COPS4 (protein complexes? or ppis? don't remember)
BOLA3 (" or " or TF coexp dont recall)
Sncg (mouse scseq type)
ZNF444 (TF coexp)
TCF25 (TF coexp)
Sensory (cell types/tissues)
Endothel (cell types/tissues)
Olig2 (TF coexpression)
```{r}
nmfenrichr <- readRDS("processed-data/10_subclustering/09-LCnmfFilt_factorSpecGenes_enrichr_results.RDS")
## fix something stupid where N_genes came back as a list
nmfenrichr[,N_genes:=as.numeric(N_genes)]

factspec <- fread("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/06b-LCNMFLogCt18Filt_factors_and_factorSpecGenes_min25specGenes.txt")

enrichrpltdat <- nmfenrichr[Term %in% grep(Term,pattern="postsynaptic|presynaptic|ubiquitin|esr1|mitochondr|calml3|znf365|LMTK3|subcoeruleus|COPS4|BOLA3|Sncg up|ZNF444|TCF25|Sensory|Olig2|ATF3 A549 hg19|ATF6 Down|Xbp1 expression of XBP1s, an spliced version of the protein mouse GSE6980 sample 1507",value=T,ignore.case=T)]

rm(nmfenrichr)

## subset to terms of interest first filtering for those signif after correction and matching terms of interest
keepterm <- unique(enrichrpltdat[Adjusted.P.value<0.05,.(Enrichr_database,Term)])
keepterm <- keepterm[Term %in% grep(Term,pattern="Ubiquitin-Dependent Protein Catabolic Process|Ubiquitin-Protein Transferase Activity (GO:0004842)|Ubiquitin E3 ligase \\(CRY|OLIG2 human tf ARCHS4|Mitochondrial Membrane|COPS4|BOLA3|ZNF365|ESR1 knockdown|ESR1 KD human|ESR1 mut human|ESR1 22446102 ChIP|OLIG2 26023283 ChIP-Seq|TCF25 Up|ESR1 Up|LMTK3 Up|Olig2 Down|ESR1 SRP254635|OLIG2 23332759|Regulation Of Postsynaptic|Sensory Perception \\(GO|Sensory neuron|Integral Component Of Presynaptic Active|Voltage-gated|Ligand-gated|PERTURB 5973 KD TCF25|subcoeruleus",value=T)|Term=="Mitochondrial Translation"|Term=="ESR1"|Term=="ATF3 A549 hg19"|(Term=="ATF6 Down"&Enrichr_database=="LINCS_L1000_CRISPR_KO_Consensus_Sigs")|Term=="Xbp1 expression of XBP1s, an spliced version of the protein mouse GSE6980 sample 1507"]


## drop a couple less important ones
keepterm <- keepterm[!(Term %in% grep(Term,pattern="OLIG2 26023283",value=T))] # less relevant cell type
keepterm <- keepterm[!(Term=="ESR1"&Enrichr_database=="PPI_Hub_Proteins")] # redundant with another
keepterm <- keepterm[Term!="ESR1 22446102 ChIP-Seq UTERUS Mouse"] # several human cell line datasets supporting same
keepterm <- keepterm[Term!="ZNF365 Up"] # weak/redundant

## Now clean up/consolidate terms
keepterm[Term %in% grep(Term,pattern="subcoeruleus",value=T),pltterm:="Allen Ms r1/r2 Sub-<br>coeruleus nu."]
keepterm[,pltterm:=gsub(Term,pattern="^(.*) human tf ARCHS4 coexpression",replacement=paste0("*","\\1","*"," Coexp."))]
keepterm[Term %in% c("PERTURB 5973 KD TCF25 SRP271175 DOWN","TCF25 Up"),pltterm:="*TCF25* KD/KO Diff. Expr."]
keepterm[Term=="LMTK3 Up",pltterm:="*LMTK3* KD/KO Upreg."]
keepterm[Term=="OLIG2 23332759 ChIP-Seq OLIGODENDROCYTES Mouse",pltterm:="Ms Olig2 ChIP Targets"]
keepterm[Term=="Mitochondrial Membrane (GO:0031966)",pltterm:="Mitochondrial Memb."]
keepterm[Term=="Regulation Of Postsynaptic Neurotransmitter Receptor Endocytosis (GO:0099149) BP",pltterm:="Regulation of Postsynaptic<br>Recep. Endocytosis"]
keepterm[Term=="Integral Component Of Presynaptic Active Zone Membrane (GO:0099059) CC",pltterm:="Presynaptic Active Zone<br>Memb. Integral Component"]
keepterm[Term %in% c("PERTURB 2483 MUT ESR1 SRP254635 UP","PERTURB 2487 MUT ESR1 SRP254635 UP","ESR1 knockdown 96h HCC515","ESR1 KD human GSE26298 sample 479","ESR1 Up"),pltterm:="*ESR1* KD/KO Upreg."]
keepterm[Term=="Ubiquitin E3 ligase (CRY1, SKP1A, CUL1, FBXL3) (human)",pltterm:="E3 Ligase (CRY1*)"]
keepterm[Term=="Ubiquitin E3 ligase (CRY2, SKP1A, CUL1, FBXL3) (human)",pltterm:="E3 Ligase (CRY2*)"]
keepterm[Term=="Sensory Perception (GO:0007600)",pltterm:="Sensory Perception"]
keepterm[Term=="Ubiquitin-Dependent Protein Catabolic Process (GO:0006511)",pltterm:="Ub-Dependent<br>Protein Catabolism"]
keepterm[Term=="ESR1",pltterm:="*ESR1* Coexp."]
keepterm[Term=="ATF3 A549 hg19",pltterm:="ATF3 ChIP Targets"]
keepterm[Term=="ATF6 Down",pltterm:="*ATF6* KD/KO Downreg."]
keepterm[Term=="Xbp1 expression of XBP1s, an spliced version of the protein mouse GSE6980 sample 1507",pltterm:="Ms *Xbp1s* OE Diff. Expr."]

## for some reason lines keep getting dropped from running all these calls in series, so make sure we end up with 18 terms
stopifnot(length(unique(keepterm$pltterm))==20)

### merge these reduced terms back to the data we're keeping to plot
enrichrpltdat <- merge.data.table(enrichrpltdat[Term %in% keepterm$Term],keepterm,by=c("Enrichr_database","Term"))
stopifnot(all(keepterm$pltterm %in% enrichrpltdat$pltterm))

## make a new table with every factor and every term
pltdat <- unique(expand.grid(unique(enrichrpltdat[,.(LCnmf_filt_factor,pltterm)])))

pltdat <- merge(pltdat,enrichrpltdat[,.(LCnmf_filt_factor,pltterm,Adjusted.P.value,Odds.Ratio,N_genes)],by=c("LCnmf_filt_factor","pltterm"),all.x=T)
pltdat <- as.data.table(pltdat)
pltdat[,`-log10(FDR)`:=-log10(Adjusted.P.value)]
pltdat[Adjusted.P.value>0.2,Odds.Ratio:=NA]
pltdat[Adjusted.P.value>0.2,N_genes:=NA]
## factors 156-160 don't have FDR<0.2 for any of these
pltdat <- pltdat[!(LCnmf_filt_factor %in% paste0("LCnmf",c(156:160)))]



## attach percentage of factor-specific genes that are in the set as an alternative size scaling metric
factspecn <- factspec[,.N,by="factor"]
pltdat <- merge.data.table(pltdat,factspecn,by.x="LCnmf_filt_factor",by.y="factor")
pltdat[,propfactgenes:=100*(N_genes/N)]
## reorder pltterm to make more sense
pltdat[,pltterm:=factor(pltterm,levels=c("Mitochondrial Memb.", 
                                         "Mitochondrial Translation",
                                         "*BOLA3* Coexp.",
                                         "*COPS4* Coexp.", 
                                         "Ub-Dependent<br>Protein Catabolism",
                                         "ATF3 ChIP Targets",
                                         "Sensory Perception", 
                                         "Presynaptic Active Zone<br>Memb. Integral Component", 
                                         "Regulation of Postsynaptic<br>Recep. Endocytosis", 
                                         "*OLIG2* Coexp.", 
                                         "Ms Olig2 ChIP Targets", 
                                         "*ZNF365* KD/KO Upreg.", 
                                         "E3 Ligase (CRY1*)", 
                                         "E3 Ligase (CRY2*)", 
                                         "*ESR1* Coexp.", 
                                         "*ESR1* KD/KO Upreg.",
                                         "*ATF6* KD/KO Downreg.",
                                         "Ms *Xbp1s* OE Diff. Expr.",
                                         "*LMTK3* KD/KO Upreg.",
                                         "*TCF25* KD/KO Diff. Expr.",
                                         "*ZNF365* Coexp."))]
## reorder factor to match
pltdat[,LCnmf_filt_factor:=factor(LCnmf_filt_factor,levels=paste0("LCnmf",c(1,2,3,4,5,6,7,8,13,17,21,103)))]


pdf("plots/manuscript/Fig3/3A-Lowexpfilt_LCnmf_FactorSpecGene_Enrichr_nmfs156-160dropped_maxAdjP0.2.pdf",height=4,width=4.35)
ggplot(pltdat,aes(x=LCnmf_filt_factor,y=pltterm,col=`-log10(FDR)`))+
    geom_point(aes(size=propfactgenes))+
    scale_color_viridis_c(option="magma")+
    #scale_x_discrete(expand = c(0,0)) +
    #scale_y_discrete(expand = c(0,0)) +
    labs(x="LC NMF Factor",y="Enrichr Term",size="% Factor Genes\nin Term")+
    theme(axis.text.x=element_text(angle=45,hjust=1,size=7,colour = "#000000"),
          axis.text.y=element_markdown(size=7,colour = "#000000"),
          axis.title.x=element_text(size=8,colour = "#000000"),
          axis.title.y=element_text(size=8,colour = "#000000", margin=margin(0,0,0,-0.02,unit="in")),
          legend.title=element_text(size=8,hjust=0.5,colour = "#000000"),
          legend.text=element_text(size=7,colour = "#000000"),
          legend.background=element_blank(),
          legend.margin=margin(0,-0.08,0,-0.1,unit="in"))
dev.off()
```

## For the clustering spotplots, we want to use the H&E images and plot by capture area, so we will need to load the big daddy SPE
```{r}
lc2 <- readRDS("processed-data/04_QC/04-lffLC_noEdges_noLocOutliers_remainder1pctileUMIorGene-removed.RDS")

# load clustering and the split-sections coldata, which has a column with pre-split spot ids we can use to match everything back to capture level
louvs <- readRDS("processed-data/07_featureSelection_dimred_harmony_clustering/04-Initial_louvLeid_res0.5-1-2_clusterings.RDS")
louvs <- louvs$HARMONYlmbna_HDG_SVG_2575[,.(rn,snnHARMONYlmbna_HDG_SVG_2575_louv_res1)]
setnames(louvs,2,"clusid")

anno <- fread("processed-data/08_validitycheck_25hdg75svg_louv1/10-25hdg75svg_louv1_annots.txt")
louvs <- merge.data.table(louvs,anno,by="clusid")

splitcd <- fread("processed-data/06-QCed_SPE_splitToTissSect_colData.txt")
splitcd <- splitcd[,.(rowname,original_spot_id)]

louvs <- merge.data.table(louvs,splitcd,by.x="rn",by.y="rowname") # split section ids

## also attach LCnm subsets because we will use them in this figure
lcnm <- fread("processed-data/10_subclustering/04-LCsubclus-by-NM_otherclusts-by-LCorLCNMadjacency.txt")
lcnm <- lcnm[,.(rn,NMseg_subclus)]

louvs <- merge.data.table(louvs,lcnm,by="rn",all.x=T)
louvs$anno[louvs$anno=="Oligo_Complex"] <- "Oligo and Other Types Mixed"
louvs$anno[louvs$anno=="Astro_Oligo"] <- "Astro-Oligo Mixed"
louvs$anno[louvs$anno=="Oligo_Astro"] <- "Oligo-Astro Mixed"
louvs$anno[louvs$anno=="ACHE_SERT_LC_WM"] <- "*ACHE*, *SERT*, LC, Oligo Mixed"
louvs$anno[louvs$anno=="ACHE_SERT_Npep"] <- "*ACHE*, *SERT*, Neuropeptide Mixed"


louvs <- DataFrame(louvs,row.names=louvs$original_spot_id)[colnames(lc2),]
stopifnot(nrow(louvs)==ncol(lc2))

colLabels(lc2) <- louvs$anno
lc2$lcnm_subclus <- louvs$NMseg_subclus

rm(splitcd,lcnm)
gc(full=T)
```

## load cluster 9-color palette
```{r}
pal <- readRDS("plots/finalpal_somuchfornorainbows.RDS")

stopifnot(all(names(pal) %in% unique(lc2$label)))
```

### custom function to extrct legend from one plot to extend to all ###
```{r}
extract_legend <- function(my_ggp) {
    step1 <- ggplot_gtable(ggplot_build(my_ggp))
    step2 <- which(sapply(step1$grobs, function(x) x$name) == "guide-box")
    step3 <- step1$grobs[[step2]]
    return(step3)
}
```

Fig 3B: Show one of the previous samples, now with only LC NM+ and - labeled. V13M06-404_B1 looks like a good sample for illustrating
```{r}
fig2c <- lc2[,lc2$sample_id=="V13M06-404_B1"]
fig2c <- fig2c[,fig2c$lcnm_subclus %in% c("LC_NMpos","LC_NMneg")]
fig2c$lcnm_subclus <- gsub(fig2c$lcnm_subclus,pattern="_",replacement=" ")
fig2c$lcnm_subclus <- gsub(fig2c$lcnm_subclus,pattern="pos",replacement="+")
fig2c$lcnm_subclus <- gsub(fig2c$lcnm_subclus,pattern="neg",replacement="-")

lcnmpal <- c(pal[1],"#9a9a9a")
names(lcnmpal) <- c("LC NM+","LC NM-")

# point size of 0.9175 at this pdf size (2.4 x 3) and theme/legend parameter set gives spots nearly touching; the center-to-center distance between visium spots is 100µm (i.e., 100-(27.5µm radii of two spots)=45µm between spots). So we want to use approx (1-0.45)*0.9175 for point_size to accurately represent dead space (and NM overlap)

## that comes out a little small. use 0.55 straight up

## finally, spatialLIBD's autocrop is still including the vast majority of the image (whereas LC is a much smaller area of it, and despite us throwing out all the spots that aren't LC here. so some post-adjusting in illustrator will be required; triple all the parameters for sizes and themes besides legend so that we can rework it later).

plotfact <- 3
p <- vis_clus(fig2c,sampleid="V13M06-404_B1",spatial = T,na_color = NULL,auto_crop = T,point_size = 0.55*plotfact,alpha=0.45,clustervar = "lcnm_subclus",colors = lcnmpal,image_id = "hires")+
    theme(plot.title = element_text(size = 9, hjust = 0.5,margin=margin(0,0,0,0,unit="in")),
          panel.border = element_blank(),
          plot.margin = margin(-0.15*plotfact,
                               0.0375*plotfact,
                               -0.7125*plotfact,
                               -0.675*plotfact,unit="in"),
          legend.key.height = unit(0.083,"in"),
          legend.key.width = unit(0.156,"in"),
          legend.text=element_text(size=7,margin=margin(0,0,0,-0.025,unit="in")),
          legend.title=element_text(size=8),
          legend.key=element_blank(),
          legend.background = element_blank())+
    scale_fill_manual(values=lcnmpal)+
    labs(fill="LC Spot Type")+
    guides(fill=guide_legend(override.aes=list(size=1.25,alpha=0.45)))

## also plot in higher res than would otherwise be necessary so we have good resolution of the illustrator-cropped out LC zoone
p <- rasterize(p,layers="Points",dpi=600,dev="cairo_png")

pdf("plots/manuscript/Fig3/Fig3B_needs_illust_postproc.pdf",height=plotfact*2.4,width=plotfact*3)
p
dev.off()

rm(fig2c,p,plotfact)
```

Fig 3C: LC NM+- DE volcano
```{r}
nmde <- fread("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/04-LCNMpos_vs_LCNMneg_DE.txt")

nmde[,yval:=-log10(adj.P.Val)]
nmde[,uptype:=ifelse(logFC>0,yes="LC NM+",no="LC NM-")]

## illustrate glial and inhibitory genes for NM- enrichment; LC markers and GPX3 (NM-assoc gene) for NM+ enrichment
nmde[gene_name %in% c("GAD1","STAT3","TCN","GJB6","VCAN","PHOX2B","SLC6A2","ATF4","TH","GPX3"),label:=gene_name]

p <- ggplot(nmde,aes(x=logFC,y=yval,col=uptype,alpha=ifelse(nmde$adj.P.Val<0.05,yes=1,no=0.4)))+
    geom_point(size=0.6,stroke = 0)+
    geom_hline(yintercept=-log10(0.05),linetype="dashed",col="red",linewidth=0.5)+
    labs(x="logFC\n(NM+ vs NM-)",y="-log10 (FDR)",col="LC\nSpot Type")+
    scale_color_manual(values=lcnmpal)+
        geom_text_repel(seed = 23942,aes(y = yval,
                x = logFC,
                label = ifelse(is.na(label), "", label)),
            min.segment.length = 0,
            size = 1.5,
            show.legend=FALSE,
            #nudge_x=1.25*sign(nudges$logFC),
            #seed=10403, 
            max.iter=100000,
            max.overlaps=10000,
            max.time=3,
            segment.size=0.25)+
    guides(alpha="none",color=guide_legend(override.aes=list(size=1.25,alpha=1)))+
    theme(axis.text.x=element_text(size=8),
          axis.text.y=element_text(size=8),
          axis.title.x=element_text(size=9),
          axis.title.y = element_text(size=9),
          legend.title=element_text(size=8,hjust=0.5),
          legend.text=element_text(size=7,margin=margin(0,0,0,-0.025,unit="in")),
          axis.ticks = element_line(linewidth=0.5),
          panel.grid.major = element_line(linewidth=0.5),
          panel.grid.minor = element_line(linewidth=0.25),
          legend.margin = margin(-0.25,0,0,-0.25,unit="in"),
          legend.background = element_blank(),
          legend.box.background = element_blank(),
          legend.key.height = unit(0.083,"in"),
          legend.key.width = unit(0.156,"in"),
          legend.key=element_blank())

p <- rasterize(p,layers="Points",dpi=300,dev="cairo_png")


## point labels are wildly screwed up and will need fixing in illustrator.
pdf("plots/manuscript/Fig2/3C-LCnmDE_volcano.pdf",height=2.5,width=3)
p
dev.off()

rm(p,nmde)
gc(full=T)
```

Fig 3D: NM DEG GSEA
```{r}
msignm <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/05-LCNMposVsNeg_msigdb_and_tftarg_GSEAs.RDS")
## ^ only grab the msigdb ones here; TF-targ enrichments were filtered beyond this point to those where TF is at least marginally expressed later
msignm <- msignm$msigdb_signed

tfnm <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/08-LCNM_APOESexDE_PredomAncesDE_LCnmf_LCnmfLct18Filt_TFtarg_GSEA_filterToClusAndDEgroupExpressed.RDS")
tfnm <- tfnm$LCNM_tftarg_signed

## also, the filtering allows for the possibility of transcriptional repressors (i.e., "targets" of the TF expressed enriched in cell group 1, TF expressed in cell group 2), which is going to be confusing for a figure. so make sure LCNM_neg_up TFs are expressed in LC_NMneg and vice versa for this
tfnm <- tfnm[(TF_LC_NMneg_Expr=="yes"&dir=="LCNM_neg_up")|(TF_LC_NMpos_Expr=="yes"&dir=="LCNM_pos_up")]
# that's the vast majority of them anyhow

#### NMpos TFs: 
# STAT3, 
# ATF4-upregulated (PMC8444148-NIHMS1529754-supplement-1529754_Tab_S2.xlsx-upregulated_In_ATF4OE_txt-RNA-seq_Of_C3_ATF4_OE_And_C3_Neurons_ATF4_Rummagene_transcription_factors), 
# MEF2D-down
# EBF3-up (hum prot atlas shows this as being most highly expressed in human LC)

## NMpos-enriched TFtarg sets to keep:
# PMC8444148-NIHMS1529754-supplement-1529754_Tab_S2.xlsx-downregulated_In_ATF4OE-RNA-seq_Of_C3_ATF4_OE_And_C3_Neurons_ATF4_Rummagene_transcription_factors
# PMC8444148-NIHMS1529754-supplement-1529754_Tab_S2.xlsx-upregulated_In_ATF4OE_txt-RNA-seq_Of_C3_ATF4_OE_And_C3_Neurons_ATF4_Rummagene_transcription_factors
# EBF3_OE_SKNSH_HUMAN_GSE90682_RNASEQ_UP_TF_Perturbations_Followed_by_Expression
# MEF2D_KD_MOUSE_GSE47150_CREEDSID_GENE_956_DOWN_TF_Perturbations_Followed_by_Expression
# PMC7801680-41598_2020_79740_MOESM9_ESM.xlsx-STAT3_Target_Genes_In_AD1-STAT3_Target_Genes_Predicted_By_CHIP-seq_Overlap_With_Genes_In_AD_M1_STAT3_Rummagene_transcription_factors

#### NMneg TFs: ATF6-down, EBF3-down, MEF2D-UP, STAT3-UP, AR, ATF4-downregulated
## NMneg TFtarg sets to keep:
# PMC6345883-41598_2018_37397_MOESM10_ESM.xlsx-AR_D0_VS_D7_WO_B29_TRANS-Gene_Name_AR_Rummagene_transcription_factors
# PMC8444148-NIHMS1529754-supplement-1529754_Tab_S2.xlsx-downregulated_In_ATF4OE-RNA-seq_Of_C3_ATF4_OE_And_C3_Neurons_ATF4_Rummagene_transcription_factors
# ATF6_ACTIVATION_HESC_HUMAN_GSE106847_RNASEQ_DOWN_TF_Perturbations_Followed_by_Expression
# EBF3_OE_SKNSH_HUMAN_GSE90682_RNASEQ_DOWN_TF_Perturbations_Followed_by_Expression
# MEF2D_KD_MOUSE_GSE47150_CREEDSID_GENE_956_UP_TF_Perturbations_Followed_by_Expression
# STAT3_KO_MOUSE_GSE38375_CREEDSID_GENE_2479_UP_TF_Perturbations_Followed_by_Expression

## this is very interesting--many bidirectional relationships here!

tfnm <- tfnm[pathway %in% c(
  "PMC8444148-NIHMS1529754-supplement-1529754_Tab_S2.xlsx-downregulated_In_ATF4OE-RNA-seq_Of_C3_ATF4_OE_And_C3_Neurons_ATF4_Rummagene_transcription_factors",
  "PMC8444148-NIHMS1529754-supplement-1529754_Tab_S2.xlsx-upregulated_In_ATF4OE_txt-RNA-seq_Of_C3_ATF4_OE_And_C3_Neurons_ATF4_Rummagene_transcription_factors",
  "EBF3_OE_SKNSH_HUMAN_GSE90682_RNASEQ_UP_TF_Perturbations_Followed_by_Expression",
  "MEF2D_KD_MOUSE_GSE47150_CREEDSID_GENE_956_DOWN_TF_Perturbations_Followed_by_Expression",
  "PMC7801680-41598_2020_79740_MOESM9_ESM.xlsx-STAT3_Target_Genes_In_AD1-STAT3_Target_Genes_Predicted_By_CHIP-seq_Overlap_With_Genes_In_AD_M1_STAT3_Rummagene_transcription_factors",
  "PMC6345883-41598_2018_37397_MOESM10_ESM.xlsx-AR_D0_VS_D7_WO_B29_TRANS-Gene_Name_AR_Rummagene_transcription_factors",
  "PMC8444148-NIHMS1529754-supplement-1529754_Tab_S2.xlsx-downregulated_In_ATF4OE-RNA-seq_Of_C3_ATF4_OE_And_C3_Neurons_ATF4_Rummagene_transcription_factors",
  "ATF6_ACTIVATION_HESC_HUMAN_GSE106847_RNASEQ_DOWN_TF_Perturbations_Followed_by_Expression",
  "EBF3_OE_SKNSH_HUMAN_GSE90682_RNASEQ_DOWN_TF_Perturbations_Followed_by_Expression",
  "MEF2D_KD_MOUSE_GSE47150_CREEDSID_GENE_956_UP_TF_Perturbations_Followed_by_Expression",
  "STAT3_KO_MOUSE_GSE38375_CREEDSID_GENE_2479_UP_TF_Perturbations_Followed_by_Expression")]

### annotate TF with target regulation direction
tfnm[pathway=="PMC8444148-NIHMS1529754-supplement-1529754_Tab_S2.xlsx-downregulated_In_ATF4OE-RNA-seq_Of_C3_ATF4_OE_And_C3_Neurons_ATF4_Rummagene_transcription_factors",tf:=paste0(tf," ","Downreg.")]
tfnm[pathway=="PMC8444148-NIHMS1529754-supplement-1529754_Tab_S2.xlsx-upregulated_In_ATF4OE_txt-RNA-seq_Of_C3_ATF4_OE_And_C3_Neurons_ATF4_Rummagene_transcription_factors",tf:=paste0(tf," ","Upreg.")]
tfnm[pathway=="EBF3_OE_SKNSH_HUMAN_GSE90682_RNASEQ_UP_TF_Perturbations_Followed_by_Expression",tf:=paste0(tf," ","Upreg.")]
tfnm[pathway=="MEF2D_KD_MOUSE_GSE47150_CREEDSID_GENE_956_DOWN_TF_Perturbations_Followed_by_Expression",tf:=paste0(tf," ","Upreg.")]
tfnm[pathway=="PMC7801680-41598_2020_79740_MOESM9_ESM.xlsx-STAT3_Target_Genes_In_AD1-STAT3_Target_Genes_Predicted_By_CHIP-seq_Overlap_With_Genes_In_AD_M1_STAT3_Rummagene_transcription_factors",tf:=paste0(tf," ","ChIP")]
tfnm[pathway=="PMC6345883-41598_2018_37397_MOESM10_ESM.xlsx-AR_D0_VS_D7_WO_B29_TRANS-Gene_Name_AR_Rummagene_transcription_factors",tf:=paste0(tf," ","Bidir.")]
tfnm[pathway=="ATF6_ACTIVATION_HESC_HUMAN_GSE106847_RNASEQ_DOWN_TF_Perturbations_Followed_by_Expression",tf:=paste0(tf," ","Downreg.")]
tfnm[pathway=="EBF3_OE_SKNSH_HUMAN_GSE90682_RNASEQ_DOWN_TF_Perturbations_Followed_by_Expression",tf:=paste0(tf," ","Downreg.")]
tfnm[pathway=="MEF2D_KD_MOUSE_GSE47150_CREEDSID_GENE_956_UP_TF_Perturbations_Followed_by_Expression",tf:=paste0(tf," ","Downreg.")]
tfnm[pathway=="STAT3_KO_MOUSE_GSE38375_CREEDSID_GENE_2479_UP_TF_Perturbations_Followed_by_Expression",tf:=paste0(tf," ","Downreg.")]

## how the shit to plot this..I think this is not getting the key point across, which is that NM+ neurons are both enriched in TF-upregulated genes AND depleted of TF-repressed genes


ggplot(tfnm,aes(x=-log10(pval),y=tf,color=NES,shape=`LC Spot Type\nEnriched for Set`,size=abs(NES)))+
  geom_point()+
  ylab("TF and effect on genes in geneset")+
  scale_color_gradient2(breaks=c(-2,0,2,4),low="#F15FFF",mid="gray",high = "darkorange4")+labs(color="Norm.\nEnrichment\nScore")
```

repinf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.2

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] BiocParallel_1.40.0         parallelly_1.43.0          
 [3] colorout_1.3-0.2            ggrepel_0.9.6              
 [5] viridis_0.6.5               viridisLite_0.4.2          
 [7] circlize_0.4.16             Cairo_1.6-2                
 [9] ComplexHeatmap_2.22.0       ggrastr_1.0.2              
[11] spatialLIBD_1.18.0          SpatialExperiment_1.16.0   
[13] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0
[15] Biobase_2.66.0              GenomicRanges_1.58.0       
[17] GenomeInfoDb_1.42.3         IRanges_2.40.1             
[19] S4Vectors_0.44.0            BiocGenerics_0.52.0        
[21] MatrixGenerics_1.18.1       matrixStats_1.5.0          
[23] gridExtra_2.3               ggtext_0.1.2               
[25] ggplot2_3.5.2               data.table_1.17.0          
[27] rlang_1.1.5                

loaded via a namespace (and not attached):
  [1] later_1.4.2              BiocIO_1.16.0            bitops_1.0-9            
  [4] filelock_1.0.3           fields_16.3.1            tibble_3.2.1            
  [7] XML_3.99-0.18            lifecycle_1.0.4          edgeR_4.4.2             
 [10] doParallel_1.0.17        rprojroot_2.0.4          lattice_0.22-6          
 [13] magrittr_2.0.3           limma_3.62.2             plotly_4.10.4           
 [16] sass_0.4.10              rmarkdown_2.29           jquerylib_0.1.4         
 [19] yaml_2.3.10              httpuv_1.6.16            spam_2.11-1             
 [22] sessioninfo_1.2.3        cowplot_1.1.3            DBI_1.2.3               
 [25] RColorBrewer_1.1-3       golem_0.5.1              maps_3.4.2.1            
 [28] abind_1.4-8              zlibbioc_1.52.0          purrr_1.0.4             
 [31] RCurl_1.98-1.17          rappdirs_0.3.3           GenomeInfoDbData_1.2.13 
 [34] irlba_2.3.5.1            commonmark_1.9.5         codetools_0.2-20        
 [37] DelayedArray_0.32.0      DT_0.33                  scuttle_1.16.0          
 [40] xml2_1.3.8               tidyselect_1.2.1         shape_1.4.6.1           
 [43] UCSC.utils_1.2.0         farver_2.1.2             ScaledMatrix_1.14.0     
 [46] shinyWidgets_0.9.0       BiocFileCache_2.14.0     GenomicAlignments_1.42.0
 [49] jsonlite_2.0.0           GetoptLong_1.0.5         BiocNeighbors_2.0.1     
 [52] scater_1.34.1            iterators_1.0.14         foreach_1.5.2           
 [55] tools_4.4.3              Rcpp_1.0.14              glue_1.8.0              
 [58] SparseArray_1.6.2        xfun_0.52                here_1.0.1              
 [61] dplyr_1.1.4              withr_3.0.2              BiocManager_1.30.25     
 [64] fastmap_1.2.0            litedown_0.6             digest_0.6.37           
 [67] rsvd_1.0.5               R6_2.6.1                 mime_0.13               
 [70] colorspace_2.1-1         markdown_2.0             dichromat_2.0-0.1       
 [73] RSQLite_2.3.9            config_0.3.2             tidyr_1.3.1             
 [76] generics_0.1.3           rtracklayer_1.66.0       httr_1.4.7              
 [79] htmlwidgets_1.6.4        S4Arrays_1.6.0           pkgconfig_2.0.3         
 [82] gtable_0.3.6             blob_1.2.4               XVector_0.46.0          
 [85] htmltools_0.5.8.1        dotCall64_1.2            clue_0.3-66             
 [88] scales_1.4.0             png_0.1-8                attempt_0.3.1           
 [91] knitr_1.50               rstudioapi_0.17.1        rjson_0.2.23            
 [94] curl_6.2.2               cachem_1.1.0             GlobalOptions_0.1.2     
 [97] stringr_1.5.1            BiocVersion_3.20.0       parallel_4.4.3          
[100] vipor_0.4.7              AnnotationDbi_1.68.0     restfulr_0.0.15         
[103] pillar_1.10.2            vctrs_0.6.5              promises_1.3.2          
[106] BiocSingular_1.22.0      dbplyr_2.5.0             beachmat_2.22.0         
[109] cluster_2.1.8.1          xtable_1.8-4             beeswarm_0.4.0          
[112] paletteer_1.6.0          evaluate_1.0.3           magick_2.8.6            
[115] cli_3.6.4                locfit_1.5-9.12          compiler_4.4.3          
[118] Rsamtools_2.22.0         crayon_1.5.3             labeling_0.4.3          
[121] rematch2_2.1.2           ggbeeswarm_0.7.2         stringi_1.8.7           
[124] Biostrings_2.74.1        lazyeval_0.2.2           Matrix_1.7-3            
[127] ExperimentHub_2.14.0     benchmarkme_1.0.8        bit64_4.6.0-1           
[130] KEGGREST_1.46.0          statmod_1.5.0            shiny_1.10.0            
[133] AnnotationHub_3.14.0     gridtext_0.1.5           memoise_2.0.1           
[136] bslib_0.9.0              benchmarkmeData_1.0.4    bit_4.6.0               
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sequoia 15.3.2
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-05-24
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)
 quarto   1.4.555 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/quarto

─ Packages ────────────────────────────────────────────────────────────────────────
 !  package              * version   date (UTC) lib source
    abind                  1.4-8     2024-09-12 [1] CRAN (R 4.4.1)
    AnnotationDbi          1.68.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    AnnotationHub          3.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    attempt                0.3.1     2020-05-03 [1] CRAN (R 4.4.0)
    beachmat               2.22.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    beeswarm               0.4.0     2021-06-01 [1] CRAN (R 4.4.0)
    benchmarkme            1.0.8     2022-06-12 [1] CRAN (R 4.4.0)
    benchmarkmeData        1.0.4     2020-04-23 [1] CRAN (R 4.4.0)
    Biobase              * 2.66.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocFileCache          2.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocIO                 1.16.0    2024-12-12 [1] Github (Bioconductor/BiocIO@8e80624)
    BiocManager            1.30.25   2024-08-28 [1] CRAN (R 4.4.1)
    BiocNeighbors          2.0.1     2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
    BiocParallel         * 1.40.0    2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    BiocSingular           1.22.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocVersion            3.20.0    2024-05-04 [1] Bioconductor 3.20 (R 4.4.0)
    Biostrings             2.74.1    2024-12-16 [1] Bioconductor 3.20 (R 4.4.2)
    bit                    4.6.0     2025-03-06 [1] CRAN (R 4.4.1)
    bit64                  4.6.0-1   2025-01-16 [1] CRAN (R 4.4.1)
    bitops                 1.0-9     2024-10-03 [1] CRAN (R 4.4.1)
    blob                   1.2.4     2023-03-17 [1] CRAN (R 4.4.0)
    bslib                  0.9.0     2025-01-30 [1] CRAN (R 4.4.1)
    cachem                 1.1.0     2024-05-16 [1] CRAN (R 4.4.0)
    Cairo                * 1.6-2     2023-11-28 [1] CRAN (R 4.4.0)
    circlize             * 0.4.16    2024-02-20 [1] CRAN (R 4.4.0)
 VP cli                    3.6.4     2025-04-23 [2] CRAN (R 4.4.1) (on disk 3.6.5)
    clue                   0.3-66    2024-11-13 [1] CRAN (R 4.4.1)
    cluster                2.1.8.1   2025-03-12 [1] CRAN (R 4.4.1)
    codetools              0.2-20    2024-03-31 [1] CRAN (R 4.4.3)
    colorout             * 1.3-0.2   2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    colorspace             2.1-1     2024-07-26 [1] CRAN (R 4.4.0)
    commonmark             1.9.5     2025-03-17 [1] CRAN (R 4.4.1)
    ComplexHeatmap       * 2.22.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    config                 0.3.2     2023-08-30 [1] CRAN (R 4.4.0)
    cowplot                1.1.3     2024-01-22 [1] CRAN (R 4.4.0)
    crayon                 1.5.3     2024-06-20 [1] CRAN (R 4.4.0)
    curl                   6.2.2     2025-03-24 [1] CRAN (R 4.4.1)
    data.table           * 1.17.0    2025-02-22 [1] CRAN (R 4.4.1)
    DBI                    1.2.3     2024-06-02 [1] CRAN (R 4.4.0)
    dbplyr                 2.5.0     2024-03-19 [1] CRAN (R 4.4.0)
    DelayedArray           0.32.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    dichromat              2.0-0.1   2022-05-02 [1] CRAN (R 4.4.0)
    digest                 0.6.37    2024-08-19 [1] CRAN (R 4.4.1)
    doParallel             1.0.17    2022-02-07 [1] CRAN (R 4.4.1)
    dotCall64              1.2       2024-10-04 [1] CRAN (R 4.4.1)
    dplyr                  1.1.4     2023-11-17 [1] CRAN (R 4.4.0)
    DT                     0.33      2024-04-04 [1] CRAN (R 4.4.0)
    edgeR                  4.4.2     2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    evaluate               1.0.3     2025-01-10 [1] CRAN (R 4.4.1)
    ExperimentHub          2.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    farver                 2.1.2     2024-05-13 [1] CRAN (R 4.4.0)
    fastmap                1.2.0     2024-05-15 [1] CRAN (R 4.4.0)
    fields                 16.3.1    2025-03-08 [1] CRAN (R 4.4.1)
    filelock               1.0.3     2023-12-11 [1] CRAN (R 4.4.0)
    foreach                1.5.2     2022-02-02 [1] CRAN (R 4.4.0)
    generics               0.1.3     2022-07-05 [1] CRAN (R 4.4.0)
    GenomeInfoDb         * 1.42.3    2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13    2024-12-12 [1] Bioconductor
    GenomicAlignments      1.42.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    GenomicRanges        * 1.58.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    GetoptLong             1.0.5     2020-12-15 [1] CRAN (R 4.4.0)
    ggbeeswarm             0.7.2     2023-04-29 [1] CRAN (R 4.4.0)
    ggplot2              * 3.5.2     2025-04-09 [1] CRAN (R 4.4.1)
    ggrastr              * 1.0.2     2023-06-01 [1] CRAN (R 4.4.0)
    ggrepel              * 0.9.6     2024-09-07 [1] CRAN (R 4.4.1)
    ggtext               * 0.1.2     2022-09-16 [1] CRAN (R 4.4.0)
    GlobalOptions          0.1.2     2020-06-10 [1] CRAN (R 4.4.0)
    glue                   1.8.0     2024-09-30 [1] CRAN (R 4.4.1)
    golem                  0.5.1     2024-08-27 [1] CRAN (R 4.4.1)
    gridExtra            * 2.3       2017-09-09 [1] CRAN (R 4.4.0)
    gridtext               0.1.5     2022-09-16 [1] CRAN (R 4.4.0)
    gtable                 0.3.6     2024-10-25 [1] CRAN (R 4.4.1)
    here                   1.0.1     2020-12-13 [1] CRAN (R 4.4.0)
    htmltools              0.5.8.1   2024-04-04 [1] CRAN (R 4.4.0)
    htmlwidgets            1.6.4     2023-12-06 [1] CRAN (R 4.4.0)
    httpuv                 1.6.16    2025-04-16 [1] CRAN (R 4.4.1)
    httr                   1.4.7     2023-08-15 [1] CRAN (R 4.4.0)
    IRanges              * 2.40.1    2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    irlba                  2.3.5.1   2022-10-03 [1] CRAN (R 4.4.0)
    iterators              1.0.14    2022-02-05 [1] CRAN (R 4.4.0)
    jquerylib              0.1.4     2021-04-26 [1] CRAN (R 4.4.0)
    jsonlite               2.0.0     2025-03-27 [1] CRAN (R 4.4.1)
    KEGGREST               1.46.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    knitr                  1.50      2025-03-16 [1] CRAN (R 4.4.2)
    labeling               0.4.3     2023-08-29 [1] CRAN (R 4.4.0)
    later                  1.4.2     2025-04-08 [1] CRAN (R 4.4.1)
    lattice                0.22-6    2024-03-20 [1] CRAN (R 4.4.3)
    lazyeval               0.2.2     2019-03-15 [1] CRAN (R 4.4.0)
    lifecycle              1.0.4     2023-11-07 [1] CRAN (R 4.4.0)
    limma                  3.62.2    2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    litedown               0.6       2025-02-27 [1] CRAN (R 4.4.3)
    locfit                 1.5-9.12  2025-03-05 [1] CRAN (R 4.4.1)
    magick                 2.8.6     2025-03-23 [1] CRAN (R 4.4.1)
    magrittr               2.0.3     2022-03-30 [1] CRAN (R 4.4.0)
    maps                   3.4.2.1   2024-11-10 [1] CRAN (R 4.4.1)
    markdown               2.0       2025-03-23 [1] CRAN (R 4.4.3)
    Matrix                 1.7-3     2025-03-11 [1] CRAN (R 4.4.1)
    MatrixGenerics       * 1.18.1    2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    matrixStats          * 1.5.0     2025-01-07 [1] CRAN (R 4.4.1)
    memoise                2.0.1     2021-11-26 [1] CRAN (R 4.4.0)
    mime                   0.13      2025-03-17 [1] CRAN (R 4.4.1)
    paletteer              1.6.0     2024-01-21 [1] CRAN (R 4.4.0)
    parallelly           * 1.43.0    2025-03-24 [1] CRAN (R 4.4.1)
    pillar                 1.10.2    2025-04-05 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3     2019-09-22 [1] CRAN (R 4.4.0)
    plotly                 4.10.4    2024-01-13 [1] CRAN (R 4.4.0)
    png                    0.1-8     2022-11-29 [1] CRAN (R 4.4.0)
    promises               1.3.2     2024-11-28 [1] CRAN (R 4.4.1)
    purrr                  1.0.4     2025-02-05 [1] CRAN (R 4.4.1)
    R6                     2.6.1     2025-02-15 [1] CRAN (R 4.4.1)
    rappdirs               0.3.3     2021-01-31 [1] CRAN (R 4.4.0)
    RColorBrewer           1.1-3     2022-04-03 [1] CRAN (R 4.4.0)
    Rcpp                   1.0.14    2025-01-12 [1] CRAN (R 4.4.1)
    RCurl                  1.98-1.17 2025-03-22 [1] CRAN (R 4.4.1)
    rematch2               2.1.2     2020-05-01 [1] CRAN (R 4.4.0)
    restfulr               0.0.15    2022-06-16 [1] CRAN (R 4.4.0)
    rjson                  0.2.23    2024-09-16 [1] CRAN (R 4.4.1)
 VP rlang                * 1.1.5     2025-04-11 [2] CRAN (R 4.4.1) (on disk 1.1.6)
    rmarkdown              2.29      2024-11-04 [1] CRAN (R 4.4.1)
    rprojroot              2.0.4     2023-11-05 [1] CRAN (R 4.4.0)
    Rsamtools              2.22.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    RSQLite                2.3.9     2024-12-03 [1] CRAN (R 4.4.1)
    rstudioapi             0.17.1    2024-10-22 [1] CRAN (R 4.4.1)
    rsvd                   1.0.5     2021-04-16 [1] CRAN (R 4.4.0)
    rtracklayer            1.66.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Arrays               1.6.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    sass                   0.4.10    2025-04-11 [1] CRAN (R 4.4.1)
    ScaledMatrix           1.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.4.0     2025-04-24 [1] CRAN (R 4.4.1)
    scater                 1.34.1    2025-03-03 [1] Bioconductor 3.20 (R 4.4.2)
    scuttle                1.16.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    sessioninfo            1.2.3     2025-02-05 [1] CRAN (R 4.4.1)
    shape                  1.4.6.1   2024-02-23 [1] CRAN (R 4.4.0)
    shiny                  1.10.0    2024-12-14 [1] CRAN (R 4.4.1)
    shinyWidgets           0.9.0     2025-02-21 [1] CRAN (R 4.4.1)
    SingleCellExperiment * 1.28.1    2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    spam                   2.11-1    2025-01-20 [1] CRAN (R 4.4.1)
    SparseArray            1.6.2     2025-02-20 [1] Bioconductor 3.20 (R 4.4.2)
    SpatialExperiment    * 1.16.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    spatialLIBD          * 1.18.0    2024-11-07 [1] Bioconductor 3.20 (R 4.4.3)
    statmod                1.5.0     2023-01-06 [1] CRAN (R 4.4.0)
    stringi                1.8.7     2025-03-27 [1] CRAN (R 4.4.1)
    stringr                1.5.1     2023-11-14 [1] CRAN (R 4.4.0)
    SummarizedExperiment * 1.36.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    tibble                 3.2.1     2023-03-20 [1] CRAN (R 4.4.0)
    tidyr                  1.3.1     2024-01-24 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1     2024-03-11 [1] CRAN (R 4.4.0)
    UCSC.utils             1.2.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    vctrs                  0.6.5     2023-12-01 [1] CRAN (R 4.4.0)
    vipor                  0.4.7     2023-12-18 [1] CRAN (R 4.4.0)
    viridis              * 0.6.5     2024-01-29 [1] CRAN (R 4.4.0)
    viridisLite          * 0.4.2     2023-05-02 [1] CRAN (R 4.4.0)
    withr                  3.0.2     2024-10-28 [1] CRAN (R 4.4.1)
    xfun                   0.52      2025-04-02 [1] CRAN (R 4.4.1)
    XML                    3.99-0.18 2025-01-01 [1] CRAN (R 4.4.1)
    xml2                   1.3.8     2025-03-14 [1] CRAN (R 4.4.1)
    xtable                 1.8-4     2019-04-21 [1] CRAN (R 4.4.0)
    XVector                0.46.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    yaml                   2.3.10    2024-07-26 [1] CRAN (R 4.4.0)
    zlibbioc               1.52.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * ── Packages attached to the search path.
 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

───────────────────────────────────────────────────────────────────────────────────
