---
title: "Fig5-APOE DE and GSEA"
author: "Bernie Mulvey"
date: "2025-07-20"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(ggplot2)
library(ggtext)
library(gridExtra)
library(ggrastr)
library(ggrepel)
library(SpatialExperiment)
library(escheR)
library(scater)
library(jaffelab) # cleaningY
library(SpatialExperiment)
library(ggvenn)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")


## ggplot defaults
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 8,color = "#000000"), axis.title.x = element_text(size = 9,color = "#000000"), axis.text.y = element_text(size = 8,color = "#000000"), axis.title.y = element_text(size =9,color = "#000000"), plot.title = element_markdown(size = 10,hjust=0.5,color = "#000000"), strip.text = element_text(size=11), legend.text = element_text(size=7,color = "#000000"), legend.title = element_text(size=8,hjust=0.5,color = "#000000"),plot.title.position="plot",axis.ticks = element_line(linewidth=0.5),panel.grid.major = element_line(linewidth=0.5),panel.grid.minor=element_line(linewidth=0.25)))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```


### helper function to extract a legend and pipe back to a diff plot: from https://statisticsglobe.com/add-common-legend-to-combined-ggplot2-plots-in-r/#example-2-add-shared-legend-to-ggplot2-plots-using-gridextra-package 
```{r}
extract_legend <- function(my_ggp) {
    step1 <- ggplot_gtable(ggplot_build(my_ggp))
    step2 <- which(sapply(step1$grobs, function(x) x$name) == "guide-box")
    step3 <- step1$grobs[[step2]]
    return(step3)
}
```


Fig 5A/B Volcanos (LC, Astro E4vE2)
```{r}
## load cluster palette
pal <- readRDS("plots/finalpal_somuchfornorainbows.RDS")
pal <- pal[c("LC","Astro")]

detabs <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/02b-voomLmFit_objs_and_DE_tables_LabSex_LabE2vE4_LabSingleGenos_LabAnces_LabAncesE4E2_noMixClusts_YRIcontAndCat_JUL2025.RDS")

## get the results from ancestry as continuous
ab4 <- detabs$cont_ances$E2E4label$detabs
ab4 <- ab4[c(3,4)] # astro and LC

names(ab4) <- gsub(names(ab4),pattern="_E4vE2",replacement="")
names(ab4) <- gsub(names(ab4),pattern="_",replacement=" ")

## identify genes to be labeled in plots
# LC SLC7A11, PLPP3, SLC1A3, NTSR2, GJA1, APOC1
# astro CP, SLC7A11, HSPA1B, HSPA1A, CD74, SPARC

ab4$LC[,label:=""]
ab4$LC[gene_name %in% c("SLC7A11","PLPP3","SLC1A3","NTSR2","GJA1","APOC1"),label:=gene_name]

ab4$Astro[,label:=""]
ab4$Astro[gene_name %in% c("HSPA1B", "HSPA1A", "CD74","SPARC","CP","SLC7A11"),label:=gene_name]

## make a giant table of the results for all 3 clusters for legend purposes

pltdetabs <- rbindlist(ab4,idcol = "clust")

### make plot
p <- ggplot(pltdetabs[!(chr %in% c("chrMT"))],aes(x=logFC,y=-log10(P.Value),col=clust)) +
    geom_point(size=0.9,stroke=0,alpha=0.7) +
    facet_wrap(~clust,scales="free_x")+
    labs(x="log2FC (+: APOE4-up)",y="-log10 P-value") +
    geom_hline(yintercept = -log10(0.05), linetype="dashed", color="red",linewidth=0.5) +
    geom_text_repel(seed = 23942,aes(y = -log10(P.Value),
            x = logFC,
            label = ifelse(is.na(label), "", label)),
        min.segment.length = 0,
        size = 2,
        show.legend=FALSE,
        max.iter=100000,
        max.overlaps=10000,
        max.time=3,
        segment.size=0.4,
        force_pull = 0.65,
        fontface = "italic")+
    scale_color_manual(values=pal)+
    labs(color="Domain")+
    theme(legend.text=element_markdown(size=7,color="#000000"),
        legend.box.background=element_blank(),
        legend.title=element_markdown(size=8,color = "#000000"),
        legend.key.width=unit(0.083,"in"),
        legend.key.height=unit(0.083,"in"),
        strip.background = element_blank(),
        strip.text = element_blank(),
        legend.margin = margin(0,-0.08,0,-0.07,unit="in"))


p<-rasterize(p,layers="points",dpi=450,dev="cairo_png")

pdf(paste0("plots/manuscript/Fig5/5AB-AllE4vsAllE2_Volcano_LCandAstro.pdf"),width=5,height=2,onefile=FALSE)
p
dev.off()

rm(p,pltdetabs,legplt,leg.plt,ab4,detabs)
```

5C: GSEA: msigdb
```{r}
msigs <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/03bi-MulticlusterDE_OneInterxnPerModel_dropMixClusts_GSEAmsigdb.RDS")

msiglc <- msigs$LC_E4vE2[pathway %in% c("GOBP_ANTEROGRADE_AXONAL_TRANSPORT",
"GOCC_ENDOPEPTIDASE_COMPLEX",
"REACTOME_CROSS_PRESENTATION_OF_SOLUBLE_EXOGENOUS_ANTIGENS_ENDOSOMES",
"GOCC_PROTEASOME_COMPLEX",
"WP_PROTEASOME_DEGRADATION",
"REACTOME_UCH_PROTEINASES",
"REACTOME_THE_ROLE_OF_GTSE1_IN_G2_M_PROGRESSION_AFTER_G2_CHECKPOINT",
"REACTOME_G1_S_DNA_DAMAGE_CHECKPOINTS",
"LEIN_ASTROCYTE_MARKERS",
"GOCC_ASTROCYTE_PROJECTION",
"GOBP_ACIDIC_AMINO_ACID_TRANSPORT",
"REACTOME_BINDING_AND_UPTAKE_OF_LIGANDS_BY_SCAVENGER_RECEPTORS",
"GOBP_NEURON_FATE_COMMITMENT")]

## there's an entry per direction here so get the significant ones
msiglc <- msiglc[padj<0.05]

msiglc[pathway=="GOBP_ANTEROGRADE_AXONAL_TRANSPORT",pathway:="Anterograde axonal\nTransport"]
msiglc[pathway=="GOCC_ENDOPEPTIDASE_COMPLEX",pathway:="Endopeptidase Complex"]
msiglc[pathway=="REACTOME_CROSS_PRESENTATION_OF_SOLUBLE_EXOGENOUS_ANTIGENS_ENDOSOMES",pathway:="Endosomal Transport of Exogenous Antigens\nto Proteasome for MHC-I Presentation"]
msiglc[pathway=="GOCC_PROTEASOME_COMPLEX",pathway:="Proteasome Complex"]
msiglc[pathway=="WP_PROTEASOME_DEGRADATION",pathway:="Proteasome Degradation"]
msiglc[pathway=="REACTOME_UCH_PROTEINASES",pathway:="UCH Proteinases"]
msiglc[pathway=="REACTOME_THE_ROLE_OF_GTSE1_IN_G2_M_PROGRESSION_AFTER_G2_CHECKPOINT",pathway:="GTSE1 in G2-M\nProgression after G2\nCheckpoint"]
msiglc[pathway=="REACTOME_G1_S_DNA_DAMAGE_CHECKPOINTS",pathway:="G1/S DNA Damage\nCheckpoints"]
msiglc[pathway=="LEIN_ASTROCYTE_MARKERS",pathway:="Lein Astrocyte\nMarkers"]
msiglc[pathway=="GOCC_ASTROCYTE_PROJECTION",pathway:="Astrocyte Projection"]
msiglc[pathway=="GOBP_ACIDIC_AMINO_ACID_TRANSPORT",pathway:="Acidic Amino Acid\nTransport"]
msiglc[pathway=="REACTOME_BINDING_AND_UPTAKE_OF_LIGANDS_BY_SCAVENGER_RECEPTORS",pathway:="Ligand Uptake by\nScavenger Receptors"]
msiglc[pathway=="GOBP_NEURON_FATE_COMMITMENT",pathway:="Neuron Fate\nCommitment"]

pdf("plots/manuscript/Fig5/5C-LC_allE4vsAllE2_msigdbGSEA.pdf",height=4,width=4.25)
ggplot(msiglc,aes(y=pathway,x=-log10(pval),color=NES,size=abs(NES)))+
    geom_point()+
    scale_color_gradient2(low="red",high="blue")+
    geom_vline(xintercept = -log10(0.05), linetype="dashed", color="black",linewidth=0.5) +
    scale_x_continuous(breaks=seq(0,15,3),expand=c(0,3)) +
    #scale_y_discrete(expand = c(0,0)) +
    labs(x="-log10(p-value)",y="Msigdb Term",color="NES (+: E4-Up\nEnriched)")+
    theme(axis.text.x=element_text(hjust=1,size=7,colour = "#000000"),
          axis.text.y=element_text(size=7,colour = "#000000"),
          axis.title.x=element_text(size=8,colour = "#000000"),
          axis.title.y=element_text(size=8,colour = "#000000", margin=margin(0,0,0,-0.02,unit="in")),
          legend.title=element_text(size=8,hjust=0.5,colour = "#000000"),
          legend.text=element_text(size=7,colour = "#000000"),
          legend.background=element_blank(),
          legend.margin=margin(0,-0.08,0,-0.1,unit="in"))+
    guides(size="none")
dev.off()

rm(fn,i,w,p,x,msiglc,msigs)
```

### 5DE: Ancestry-E4E2 comparisons (logFCs colored by signif)
```{r}
detabs <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/02b-voomLmFit_objs_and_DE_tables_LabSex_LabE2vE4_LabSingleGenos_LabAnces_LabAncesE4E2_noMixClusts_YRIcontAndCat_JUL2025.RDS")
detabs <- detabs$cat_ances$YRIAPOlabel$detabs

lcancapo <- detabs[c("LC_E4CEU_vs_E2CEU","LC_E4YRI_vs_E2YRI")]
lcancapo <- rbindlist(lcancapo,idcol="clust_contrast")
lcancapo[,domain:="LC"]
lcancapo[,contrast:=gsub(clust_contrast,pattern="^.*_E4(...).*$",replacement="\\1")]
lcancapo[,clust_contrast:=NULL]

astancapo <- detabs[c("Astro_E4CEU_vs_E2CEU","Astro_E4YRI_vs_E2YRI")]
astancapo <- rbindlist(astancapo,idcol="clust_contrast")
astancapo[,domain:="Astro"]
astancapo[,contrast:=gsub(clust_contrast,pattern="^.*_E4(...).*$",replacement="\\1")]
astancapo[,clust_contrast:=NULL]

ancapo <- as.data.table(rbind(lcancapo,astancapo))
rm(astancapo,lcancapo)

ancapo2 <- dcast(ancapo,gene_id+gene_name+domain~contrast,value.var = c("logFC","adj.P.Val","t"))

ancapo2[adj.P.Val_CEU<0.05&adj.P.Val_YRI>0.05,sig:="CEU"]
ancapo2[adj.P.Val_CEU>0.05&adj.P.Val_YRI<0.05,sig:="YRI"]
ancapo2[adj.P.Val_CEU<0.05&adj.P.Val_YRI<0.05&logFC_CEU&logFC_YRI>0,sig:="Both"] # only label concordant points here where we can confidently speak to the direction of both tests
ancapo2[is.na(sig),sig:="None"]

ancapo2[domain=="Astro"&gene_name %in% c("HSPA1B", "HSPA1A", "CD74","SPARC","CP","SLC7A11"),genelab:=gene_name]

ancapo2[domain=="LC"&gene_name %in% c("SLC7A11","PLPP3","SLC1A3","NTSR2","GJA1","APOC1"),genelab:=gene_name]

ancapo2[,sig:=factor(sig,levels=c("CEU","YRI","Both","None"))]

lapply(unique(ancapo2$domain),function(d){
    tmppal <- c("#999999","red","green",pal[d])
    names(tmppal) <- c("None","CEU","YRI","Both")
    
    p <-  ggplot(ancapo2[domain==d],aes(x=logFC_YRI,y=logFC_CEU))+
        geom_point(stroke=0,size=ifelse(ancapo2[domain==d]$sig=="None",0.5,0.75),
                   alpha=ifelse(ancapo2[domain==d]$sig=="None",0.15,0.7),
                   aes(color=sig))+
        geom_text_repel(seed = 23942,aes(y = logFC_CEU,
                x = logFC_YRI,
                label = ifelse(is.na(genelab), "", genelab)),
            min.segment.length = 0,
            size = 2,
            show.legend=FALSE,
            max.iter=100000,
            max.overlaps=10000,
            max.time=3,
            segment.size=0.4,
            force_pull = 0.65,
            fontface = "italic")+
        geom_abline(slope=1,intercept=0,linetype="dashed",linewidth = 0.25)+
        xlim(min(c(ancapo2$logFC_CEU,ancapo2$logFC_YRI)),
             max(c(ancapo2$logFC_CEU,ancapo2$logFC_YRI)))+
        ylim(min(c(ancapo2$logFC_CEU,ancapo2$logFC_YRI)),
             max(c(ancapo2$logFC_CEU,ancapo2$logFC_YRI)))+
        xlab("logFC E4 vs. E2 (YRI)")+
        ylab("logFC E4 vs. E2 (CEU)")+
        labs(color="FDR<0.05 in:")+
        scale_color_manual(values=tmppal)+
        guides(alpha="none",size="none")+
        ggtitle(d)+
        theme(plot.title.position = "panel",
              title = element_text(size=10,colour="#000000",hjust=0.5),
              axis.text.x=element_text(hjust=1,size=8,colour = "#000000"),
              axis.text.y=element_text(size=8,colour = "#000000"),
              axis.title.x=element_text(size=9,colour = "#000000"),
              axis.title.y=element_text(size=9,colour = "#000000", margin=margin(0,0,0,-0.02,unit="in")),
              legend.title=element_text(size=9,hjust=0.5,colour = "#000000"),
              legend.text=element_text(size=8,colour = "#000000"),
              legend.background=element_blank(),
              legend.key.height = unit(0.15,"in"),
              legend.key.width = unit(0,"in"),
              legend.margin=margin(0,-0.08,0,-0.1,unit="in"))
    
    pdf(paste0("plots/manuscript/Fig5/5DE-AncestryE4E2_",d,".pdf"),height=2.5,width=3)
    print(p)
    dev.off()
})
```

5FG: Venns of signif genes
```{r}
yriastup <- ancapo2[adj.P.Val_YRI<0.05&logFC_YRI>0&domain=="Astro",gene_name]
yriastdn <- ancapo2[adj.P.Val_YRI<0.05&logFC_YRI<0&domain=="Astro",gene_name]
ceuastup <- ancapo2[adj.P.Val_CEU<0.05&logFC_CEU>0&domain=="Astro",gene_name]
ceuastdn <- ancapo2[adj.P.Val_CEU<0.05&logFC_CEU<0&domain=="Astro",gene_name]

yrilcup <- ancapo2[adj.P.Val_YRI<0.05&logFC_YRI>0&domain=="LC",gene_name]
yrilcdn <- ancapo2[adj.P.Val_YRI<0.05&logFC_YRI<0&domain=="LC",gene_name]
ceulcup <- ancapo2[adj.P.Val_CEU<0.05&logFC_CEU>0&domain=="LC",gene_name]
ceulcdn <- ancapo2[adj.P.Val_CEU<0.05&logFC_CEU<0&domain=="LC",gene_name]

astde <- list("Astro YRI E4 Up\nvs. YRI E2"=yriastup,
              "Astro YRI E4 Down\nvs. YRI E2"=yriastdn,
              "Astro CEU E4 Down\nvs. CEU E2"=ceuastdn,
              "Astro CEU E4 Up\nvs. CEU E2"=ceuastup
)

lcde <- list("LC YRI E4 Up\nvs. YRI E2"=yrilcup,
              "LC YRI E4 Down\nvs. YRI E2"=yrilcdn,
              "LC CEU E4 Down\nvs. CEU E2"=ceulcdn,
              "LC CEU E4 Up\nvs. CEU E2"=ceulcup
)

pdf("plots/manuscript/Fig5/5FG-APOAnc_Astro_Overlaps.pdf",height=3,width=3.75)
ggvenn(astde,names(astde),set_name_size = 3,text_size = 2,stroke_size = 0.5)
dev.off()

pdf("plots/manuscript/Fig5/5FG-APOAnc_LC_Overlaps.pdf",height=3,width=3.75)
ggvenn(lcde,names(lcde),set_name_size = 3,text_size = 2,stroke_size = 0.5)
dev.off()
```

seshinf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.2

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] BiocParallel_1.40.0         parallelly_1.43.0           colorout_1.3-0.2           
 [4] ggvenn_0.1.10               dplyr_1.1.4                 jaffelab_0.99.34           
 [7] rafalib_1.0.4               scater_1.34.1               scuttle_1.16.0             
[10] escheR_1.6.0                SpatialExperiment_1.16.0    SingleCellExperiment_1.28.1
[13] SummarizedExperiment_1.36.0 Biobase_2.66.0              GenomicRanges_1.58.0       
[16] GenomeInfoDb_1.42.3         IRanges_2.40.1              S4Vectors_0.44.0           
[19] BiocGenerics_0.52.0         MatrixGenerics_1.18.1       matrixStats_1.5.0          
[22] ggrepel_0.9.6               ggrastr_1.0.2               gridExtra_2.3              
[25] ggtext_0.1.2                ggplot2_3.5.2               data.table_1.17.4          
[28] rlang_1.1.5                

loaded via a namespace (and not attached):
 [1] magrittr_2.0.3          compiler_4.4.3          vctrs_0.6.5            
 [4] stringr_1.5.1           pkgconfig_2.0.3         crayon_1.5.3           
 [7] fastmap_1.2.0           magick_2.8.6            XVector_0.46.0         
[10] labeling_0.4.3          rmarkdown_2.29          markdown_2.0           
[13] UCSC.utils_1.2.0        ggbeeswarm_0.7.2        purrr_1.0.4            
[16] xfun_0.52               zlibbioc_1.52.0         beachmat_2.22.0        
[19] litedown_0.6            jsonlite_2.0.0          DelayedArray_0.32.0    
[22] irlba_2.3.5.1           parallel_4.4.3          R6_2.6.1               
[25] stringi_1.8.7           RColorBrewer_1.1-3      limma_3.62.2           
[28] Rcpp_1.0.14             knitr_1.50              Matrix_1.7-3           
[31] splines_4.4.3           tidyselect_1.2.1        rstudioapi_0.17.1      
[34] dichromat_2.0-0.1       abind_1.4-8             yaml_2.3.10            
[37] viridis_0.6.5           codetools_0.2-20        lattice_0.22-6         
[40] tibble_3.2.1            withr_3.0.2             evaluate_1.0.3         
[43] xml2_1.3.8              pillar_1.10.2           generics_0.1.4         
[46] rprojroot_2.0.4         commonmark_1.9.5        scales_1.4.0           
[49] glue_1.8.0              tools_4.4.3             BiocNeighbors_2.0.1    
[52] ScaledMatrix_1.14.0     fs_1.6.6                nlme_3.1-167           
[55] GenomeInfoDbData_1.2.13 beeswarm_0.4.0          BiocSingular_1.22.0    
[58] googledrive_2.1.1       vipor_0.4.7             cli_3.6.4              
[61] rsvd_1.0.5              gargle_1.5.2            segmented_2.1-4        
[64] S4Arrays_1.6.0          viridisLite_0.4.2       gtable_0.3.6           
[67] digest_0.6.37           SparseArray_1.6.2       rjson_0.2.23           
[70] farver_2.1.2            htmltools_0.5.8.1       lifecycle_1.0.4        
[73] httr_1.4.7              here_1.0.1              statmod_1.5.0          
[76] gridtext_0.1.5          MASS_7.3-65            
> sessioninfo::session_info()
─ Session info ──────────────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sequoia 15.3.2
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-07-23
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)
 quarto   1.4.555 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/quarto

─ Packages ──────────────────────────────────────────────────────────────────────────────────
 !  package              * version date (UTC) lib source
    abind                  1.4-8   2024-09-12 [1] CRAN (R 4.4.1)
    beachmat               2.22.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    beeswarm               0.4.0   2021-06-01 [1] CRAN (R 4.4.0)
    Biobase              * 2.66.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocNeighbors          2.0.1   2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
    BiocParallel         * 1.40.0  2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    BiocSingular           1.22.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
 VP cli                    3.6.4   2025-04-23 [2] CRAN (R 4.4.1) (on disk 3.6.5)
    codetools              0.2-20  2024-03-31 [1] CRAN (R 4.4.3)
    colorout             * 1.3-0.2 2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    commonmark             1.9.5   2025-03-17 [1] CRAN (R 4.4.1)
    crayon                 1.5.3   2024-06-20 [1] CRAN (R 4.4.0)
    data.table           * 1.17.4  2025-05-26 [1] CRAN (R 4.4.3)
    DelayedArray           0.32.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    dichromat              2.0-0.1 2022-05-02 [1] CRAN (R 4.4.0)
    digest                 0.6.37  2024-08-19 [1] CRAN (R 4.4.1)
    dplyr                * 1.1.4   2023-11-17 [1] CRAN (R 4.4.0)
    escheR               * 1.6.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    evaluate               1.0.3   2025-01-10 [1] CRAN (R 4.4.1)
    farver                 2.1.2   2024-05-13 [1] CRAN (R 4.4.0)
    fastmap                1.2.0   2024-05-15 [1] CRAN (R 4.4.0)
    fs                     1.6.6   2025-04-12 [1] CRAN (R 4.4.1)
    gargle                 1.5.2   2023-07-20 [1] CRAN (R 4.4.0)
    generics               0.1.4   2025-05-09 [1] CRAN (R 4.4.1)
    GenomeInfoDb         * 1.42.3  2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13  2024-12-12 [1] Bioconductor
    GenomicRanges        * 1.58.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ggbeeswarm             0.7.2   2023-04-29 [1] CRAN (R 4.4.0)
    ggplot2              * 3.5.2   2025-04-09 [1] CRAN (R 4.4.1)
    ggrastr              * 1.0.2   2023-06-01 [1] CRAN (R 4.4.0)
    ggrepel              * 0.9.6   2024-09-07 [1] CRAN (R 4.4.1)
    ggtext               * 0.1.2   2022-09-16 [1] CRAN (R 4.4.0)
    ggvenn               * 0.1.10  2023-03-31 [1] CRAN (R 4.4.0)
    glue                   1.8.0   2024-09-30 [1] CRAN (R 4.4.1)
    googledrive            2.1.1   2023-06-11 [1] CRAN (R 4.4.0)
    gridExtra            * 2.3     2017-09-09 [1] CRAN (R 4.4.0)
    gridtext               0.1.5   2022-09-16 [1] CRAN (R 4.4.0)
    gtable                 0.3.6   2024-10-25 [1] CRAN (R 4.4.1)
    here                   1.0.1   2020-12-13 [1] CRAN (R 4.4.0)
    htmltools              0.5.8.1 2024-04-04 [1] CRAN (R 4.4.0)
    httr                   1.4.7   2023-08-15 [1] CRAN (R 4.4.0)
    IRanges              * 2.40.1  2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    irlba                  2.3.5.1 2022-10-03 [1] CRAN (R 4.4.0)
    jaffelab             * 0.99.34 2025-05-27 [1] Github (LieberInstitute/jaffelab@e7a26ca)
    jsonlite               2.0.0   2025-03-27 [1] CRAN (R 4.4.1)
    knitr                  1.50    2025-03-16 [1] CRAN (R 4.4.2)
    labeling               0.4.3   2023-08-29 [1] CRAN (R 4.4.0)
    lattice                0.22-6  2024-03-20 [1] CRAN (R 4.4.3)
    lifecycle              1.0.4   2023-11-07 [1] CRAN (R 4.4.0)
    limma                  3.62.2  2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    litedown               0.6     2025-02-27 [1] CRAN (R 4.4.3)
    magick                 2.8.6   2025-03-23 [1] CRAN (R 4.4.1)
    magrittr               2.0.3   2022-03-30 [1] CRAN (R 4.4.0)
    markdown               2.0     2025-03-23 [1] CRAN (R 4.4.3)
    MASS                   7.3-65  2025-02-28 [1] CRAN (R 4.4.1)
    Matrix                 1.7-3   2025-03-11 [1] CRAN (R 4.4.1)
    MatrixGenerics       * 1.18.1  2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    matrixStats          * 1.5.0   2025-01-07 [1] CRAN (R 4.4.1)
    nlme                   3.1-167 2025-01-27 [1] CRAN (R 4.4.3)
    parallelly           * 1.43.0  2025-03-24 [1] CRAN (R 4.4.1)
    pillar                 1.10.2  2025-04-05 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3   2019-09-22 [1] CRAN (R 4.4.0)
    purrr                  1.0.4   2025-02-05 [1] CRAN (R 4.4.1)
    R6                     2.6.1   2025-02-15 [1] CRAN (R 4.4.1)
    rafalib              * 1.0.4   2025-04-08 [1] CRAN (R 4.4.1)
    RColorBrewer           1.1-3   2022-04-03 [1] CRAN (R 4.4.0)
    Rcpp                   1.0.14  2025-01-12 [1] CRAN (R 4.4.1)
    rjson                  0.2.23  2024-09-16 [1] CRAN (R 4.4.1)
 VP rlang                * 1.1.5   2025-04-11 [2] CRAN (R 4.4.1) (on disk 1.1.6)
    rmarkdown              2.29    2024-11-04 [1] CRAN (R 4.4.1)
    rprojroot              2.0.4   2023-11-05 [1] CRAN (R 4.4.0)
    rstudioapi             0.17.1  2024-10-22 [1] CRAN (R 4.4.1)
    rsvd                   1.0.5   2021-04-16 [1] CRAN (R 4.4.0)
    S4Arrays               1.6.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ScaledMatrix           1.14.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.4.0   2025-04-24 [1] CRAN (R 4.4.1)
    scater               * 1.34.1  2025-03-03 [1] Bioconductor 3.20 (R 4.4.2)
    scuttle              * 1.16.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    segmented              2.1-4   2025-02-28 [1] CRAN (R 4.4.1)
    sessioninfo            1.2.3   2025-02-05 [1] CRAN (R 4.4.1)
    SingleCellExperiment * 1.28.1  2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    SparseArray            1.6.2   2025-02-20 [1] Bioconductor 3.20 (R 4.4.2)
    SpatialExperiment    * 1.16.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    statmod                1.5.0   2023-01-06 [1] CRAN (R 4.4.0)
    stringi                1.8.7   2025-03-27 [1] CRAN (R 4.4.1)
    stringr                1.5.1   2023-11-14 [1] CRAN (R 4.4.0)
    SummarizedExperiment * 1.36.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    tibble                 3.2.1   2023-03-20 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1   2024-03-11 [1] CRAN (R 4.4.0)
    UCSC.utils             1.2.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    vctrs                  0.6.5   2023-12-01 [1] CRAN (R 4.4.0)
    vipor                  0.4.7   2023-12-18 [1] CRAN (R 4.4.0)
    viridis                0.6.5   2024-01-29 [1] CRAN (R 4.4.0)
    viridisLite            0.4.2   2023-05-02 [1] CRAN (R 4.4.0)
    withr                  3.0.2   2024-10-28 [1] CRAN (R 4.4.1)
    xfun                   0.52    2025-04-02 [1] CRAN (R 4.4.1)
    xml2                   1.3.8   2025-03-14 [1] CRAN (R 4.4.1)
    XVector                0.46.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    yaml                   2.3.10  2024-07-26 [1] CRAN (R 4.4.0)
    zlibbioc               1.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * ── Packages attached to the search path.
 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

─────────────────────────────────────────────────────────────────────────────────────────────
