---
title: "Fig6-SpotNM_LMMs"
format: html
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(ggplot2)
library(ggtext)
library(ggtext)
library(gridExtra)
# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
# require(colorout)
# ColorOut()
options('styler.addins_style_transformer' = 'biocthis::bioc_style()')
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn='')
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost='localhost')


## ggplot defaults
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 8,color = '#000000'), axis.title.x = element_text(size = 9,color = '#000000'), axis.text.y = element_text(size = 8,color = '#000000'), axis.title.y = element_text(size =9,color = '#000000'), plot.title = element_markdown(size = 10,hjust=0.5,color = '#000000'), strip.text = element_text(size=11), legend.text = element_text(size=7,color = '#000000'), legend.title = element_text(size=8,hjust=0.5,color = '#000000'),axis.ticks = element_line(linewidth=0.5),panel.grid.major = element_line(linewidth=0.5),panel.grid.minor=element_line(linewidth=0.25),plot.title.position='plot'))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace('datasets')
```

### using an E4 carrier yes/no (i.e., E4 or E2) model
```{r}
nmlm <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/11-LCNMspots_NMmetric_LMMed_by_geneLogcounts_3apoeFactorizations.RDS")

nmi <- nmlm$E4carrierYN$Intensity_NM

nmi[variable==gene_name|variable=="APOE.x",bonfP:=p*(length(unique(nmi$gene_name)))]
nmi[bonfP>1,bonfP:=1]

## check if any non-gene features were significantly associated
nmi[gene_name!=variable &variable!="LRTvs_no_gene_coef"& p<0.05]
#     gene_name variable          coef
#        <char>   <char>         <num>
#  1:       DBH      Age -0.0009718147
#  2:    SLC6A2      Age -0.0009717981
#  3:      APOE   APOE.x  0.0101268892
#  4:      CHGB      Age -0.0009920979
#  5:    MT.ND1      Age -0.0010982648
#  6:    MT.CO1      Age -0.0011546129
#  7:    MT.CO1      Sex  0.0236634038
#  8:    MT.ND4      Age -0.0010313770
#  9:    MT.CYB      Age -0.0010245239
# 10:    MT.CYB      Sex  0.0200976755
#var_group     Chisq    Df            p
#        <char>     <num> <num>        <num>
#  1:      <NA>  4.189558     1 4.067370e-02
#  2:      <NA>  4.003680     1 4.540102e-02
#  3:      <NA> 67.058580     1 2.635580e-16
#  4:      <NA>  3.927644     1 4.749843e-02
#  5:      <NA>  4.554034     1 3.284128e-02
#  6:      <NA>  4.498759     1 3.391947e-02
#  7:         M  4.631333     1 3.139316e-02
#  8:      <NA>  4.180122     1 4.090076e-02
#  9:      <NA>  4.155482     1 4.149996e-02
# 10:         M  3.929867     1 4.743570e-02

## weak decreases in LC markers and mitochondrial genes with age; increased mitochondrial genes in males (which we knew)
## also, APOE expression significantly assoc with lower NM intensity


## check if any non-gene features were significantly associated w nm proportion
nmp[gene_name!=variable &variable!="LRTvs_no_gene_coef"& p<0.05]
#    gene_name  variable        coef var_group      Chisq    Df            p logLik   NMvar
#       <char>    <char>       <num>    <char>      <num> <num>        <num>  <num>  <char>
# 1: MTRNR2L12 E4carrier  0.01033318       Yes   4.408608     1 3.575801e-02     NA Prop_NM
# 2:   C5orf63       Sex -0.01123681         M   5.389835     1 2.025438e-02     NA Prop_NM
# 3:      APOE    APOE.x -0.01068722      <NA> 133.703074     1 6.344919e-31     NA Prop_NM

## mtrnr2l12 is a sample-specifically variable gene that will be discarded from feature selection in the future and should be disregarded generally
## again, APOE is associated
```

### panel A: NM intensity and proportion associations with APOE expression
found the best way to plot this for intensity was make plot of tissue section-level NMscore by section mean expression of gene for sections with 20+ NM+ LC spots. proportion is harder to interpret given the deadspace on visium, so we won't go there.

# so we'll need the SPE with LC NM+ cells, log counts, and metadata with tissue NM scores
```{r}
lc3 <- readRDS("processed-data/06-countsOnly_noImgData_QCed_SPE_split_to_tissSections.RDS")

# calc spot log counts
lc3 <- scater::computeLibraryFactors(lc3)
lc3 <- scater::logNormCounts(lc3)

# load clusters + annots
louvs <- readRDS("processed-data/07_featureSelection_dimred_harmony_clustering/04-Initial_louvLeid_res0.5-1-2_clusterings.RDS")
louvs <- louvs[["HARMONYlmbna_HDG_SVG_2575"]][,.(rn,snnHARMONYlmbna_HDG_SVG_2575_louv_res1)]
setnames(louvs,2,"clusid")

finalann <- fread("processed-data/08_validitycheck_25hdg75svg_louv1/10-25hdg75svg_louv1_annots.txt")

louvs <- merge.data.table(louvs,finalann,by="clusid")
louvs <- DataFrame(louvs,row.names=louvs$rn)[colnames(lc3),]

colLabels(lc3) <- louvs$anno
lc3 <- lc3[,!(lc3$brnum %in% paste0("Br",c(5517,5276,5712)))]

## read in nm subclusterings
lcnm <- fread("processed-data/10_subclustering/04-LCsubclus-by-NM_otherclusts-by-LCorLCNMadjacency.txt")
lcnm <- lcnm[,.(rn,NMseg_subclus)]
lcnm <- DataFrame(lcnm,row.names=lcnm$rn)[colnames(lc3),]

## assign
lc3$lcnm <- lcnm$NMseg_subclus

## get LC NM+ spot spe and its coldata w/ nm metrics; calculate mean and median NM intensity per sample
lc3 <- lc3[,lc3$lcnm=="LC_NMpos"]
lcnmcd <- as.data.table(colData(lc3[,lc3$lcnm=="LC_NMpos"]))
lcnmcd[,med_NMI:=median(Intensity_NM,na.rm=T),by="sample_id"]
lcnmcd[,mn_NMI:=mean(Intensity_NM,na.rm=T),by="sample_id"]

# rownames to gene symbs
rownames(lc3) <- rowData(lc3)$gene_name
```

## we can wrap this all in a function and make these plots for multiple genes of interest (GPX3, APOE)
```{r}
saveplt <- function(g){
    pltdat <- as.data.frame(t(as.matrix(logcounts(lc3[g,]))))
    pltdat <- as.data.table(pltdat,keep.rownames=T)
    setnames(pltdat,g,"g_xpr") # since the value of g will be a variable in coldata
    pltdat <- merge.data.table(lcnmcd,pltdat,by.x="rowname",by.y="rn")
    pltdat[,nspot:=.N,by="sample_id"] # determine the number of spots from each sample INCLUDING those with logcounts=0 (though we won't plot those)
    
    pltdat <- pltdat[g_xpr!=0]
    quartile_breaks <- quantile(pltdat$g_xpr, probs = seq(0, 1, by = 0.25), na.rm = TRUE)
    
    pltdat[, x_quartile := cut(g_xpr, 
                     breaks = quartile_breaks, 
                     labels = 1:4,         # Label bins accoring to nth decile
                     include.lowest = TRUE)]
    
    pltdat[,calcfill:=mean(Intensity_NM),by="x_quartile"]

    p <- ggplot(unique(pltdat[nspot>19,.(x_quartile,Intensity_NM,calcfill)]),aes(x=x_quartile,group=x_quartile,y=Intensity_NM,fill=calcfill))+
        scale_fill_gradient(low="brown",high="beige")+
        guides(fill="none")+
        geom_boxplot()+
        labs(y="Tissue NM Intensity Score",x=paste0("Quartile logcounts *",g,"*"))+
        theme(axis.title.x=element_markdown(size=9),
              axis.title.y=element_markdown(size=9),
              axis.text.x=element_text(size=8),
              axis.text.y=element_text(size=8))

    pdf(paste0("plots/manuscript/Fig6/6x-Spotlogcountquartiles_",g,"_spotNMintensity_nozeros.pdf"),height=2.5,width=2.5)
    print(p)
    dev.off()
}

saveplt("APOE")
saveplt("GPX3")

rm(saveplt,flares)
gc(full=T)
```

then we have GSEA for the NM enriched genes
bonf-sig genes, bidirectional, (calculated using E4YN model)

GO_Cellular_Component_2025	Neuron Projection (GO:0043005)
Reactome_Pathways_2024	Transmission Across Chemical Synapses
Reactome_Pathways_2024	Vesicle-mediated Transport
KEGG_2021_Human	Synaptic vesicle cycle
PPI_Hub_Proteins	APP
ChEA_2022	CREB1 23762244 ChIP-Seq HIPPOCAMPUS Rat
HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression	Inh L1 Pax6 Chrfam7A (id:271)
HuBMAP_ASCTplusB_augmented_2022	Inh L5-6 PVALB FAM150B - Brain
LINCS_L1000_Ligand_Perturbations_up	IL17-MCF7
GO_Cellular_Component_2025	Axon (GO:0030424)
HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression	Inh L2-5 Vip Bspry (id:307)
HuBMAP_ASCTplusB_augmented_2022	Inh L5-6 PVALB GAPDHP60 - Brain
Reactome_Pathways_2024	Kinesins
MGI_Mammalian_Phenotype_Level_4_2024	Axon Degeneration MP:0005405
HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression	Melanocyte (id:406)
```{r}
nmgsea <- readRDS("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/13-enrichr_fdrsig_and_bonfP_genes_NMintensityAndProp-to-geneXpr_lmms.RDS")
nmgsea <- nmgsea$nondir$bonfP_gene$Intensity_NM

plt <- nmgsea[Enrichr_database=="GO_Cellular_Component_2025" & 
    Term %in% c("Neuron Projection (GO:0043005)","Axon (GO:0030424)"),.(Enrichr_database,Term,Adjusted.P.value,Genes,Odds.Ratio)] 

plt[Term=="Neuron Projection (GO:0043005)",Term:="Neuron Projection"]
plt[Term=="Axon (GO:0030424)",Term:="Axon"]

plt <- rbind(plt,
             nmgsea[Enrichr_database=="Reactome_Pathways_2024" & Term %in% c("Transmission Across Chemical Synapses","Vesicle-mediated Transport","Kinesins"),.(Enrichr_database,Term,Adjusted.P.value,Genes,Odds.Ratio)])

plt <- rbind(plt, 
             nmgsea[Enrichr_database=="KEGG_2021_Human" & Term %in% c("Synaptic vesicle cycle","Mineral absorption"),.(Enrichr_database,Term,Adjusted.P.value,Genes,Odds.Ratio)])

plt <- rbind(plt, nmgsea[Term=="APP" & Enrichr_database=="PPI_Hub_Proteins",.(Enrichr_database,Term,Adjusted.P.value,Genes,Odds.Ratio)])

plt <- rbind(plt, nmgsea[Term=="CREB1 23762244 ChIP-Seq HIPPOCAMPUS Rat" & Enrichr_database=="ChEA_2022",.(Enrichr_database,Term,Adjusted.P.value,Genes,Odds.Ratio)])

plt <- rbind(plt, nmgsea[Term=="Inh L1 Pax6 Chrfam7A (id:271)" & Enrichr_database=="HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression",.(Enrichr_database,Term,Adjusted.P.value,Genes,Odds.Ratio)])

plt <- rbind(plt, nmgsea[Term=="Inh L2-5 Vip Bspry (id:307)" & Enrichr_database=="HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression",.(Enrichr_database,Term,Adjusted.P.value,Genes,Odds.Ratio)])

plt <- rbind(plt, nmgsea[Term=="Inh L5-6 PVALB FAM150B - Brain" & Enrichr_database=="HuBMAP_ASCTplusB_augmented_2022",.(Enrichr_database,Term,Adjusted.P.value,Genes,Odds.Ratio)])

plt <- rbind(plt, nmgsea[Term=="Axon Degeneration MP:0005405" & Enrichr_database=="MGI_Mammalian_Phenotype_Level_4_2024",.(Enrichr_database,Term,Adjusted.P.value,Genes,Odds.Ratio)])

plt <- rbind(plt, nmgsea[Term=="Melanocyte (id:406)" & Enrichr_database=="HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression",.(Enrichr_database,Term,Adjusted.P.value,Genes,Odds.Ratio)])

plt <- rbind(plt, nmgsea[Term=="IL17-MCF7" & Enrichr_database=="LINCS_L1000_Ligand_Perturbations_up",.(Enrichr_database,Term,Adjusted.P.value,Genes,Odds.Ratio)])

## condense terms
plt[Term=="Inh L1 Pax6 Chrfam7A (id:271)",Term:="Inh L1 *Pax6* *Chrfam7A*"]
plt[Term=="CREB1 23762244 ChIP-Seq HIPPOCAMPUS Rat",Term:="Creb1 Targets (Rat Hpc)"]
plt[Term=="Inh L2-5 Vip Bspry (id:307)",Term:="Inh L2-5 *Vip* *Bspry*"]
plt[Term=="Inh L5-6 PVALB FAM150B - Brain",Term:="Inh L5-6 *PVALB* *FAM150B*"]
plt[Term=="Axon Degeneration MP:0005405",Term:="Axon Degeneration"]
plt[Term=="Melanocyte (id:406)",Term:="Melanocyte"]
plt[Term=="Transmission Across Chemical Synapses",Term:="Chem. Synapse Transmission"]

unique(plt$Enrichr_database)
plt[Enrichr_database %in% c("GO_Cellular_Component_2025","KEGG_2021_Human","Reactome_Pathways_2024"),Databases:="Ontology and\nPathways"]
plt[Enrichr_database %in% c("PPI_Hub_Proteins","ChEA_2022"),Databases:="TF Targets\nand PPIs"]
plt[Enrichr_database %in% c("HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression","HuBMAP_ASCTplusB_augmented_2022"),Databases:="HuBMAP Cell\nMarkers"]
plt[Enrichr_database %in% c("MGI_Mammalian_Phenotype_Level_4_2024","LINCS_L1000_Ligand_Perturbations_up"),Databases:="Phenotypes\nand Perturbations"]

setorderv(plt,"Databases")
plt[,Term:=factor(Term,levels=plt$Term)]

pdf("plots/manuscript/Fig6/6C-Enrichr_NMintensity_bonfsig_bidir.pdf",width=5,height=4)
ggplot(plt,aes(y=-log10(Adjusted.P.value),x=Term,color=Databases))+
    geom_point(aes(size=Odds.Ratio))+
    geom_hline(yintercept = -log10(0.05), linetype="dashed", color="black",linewidth=0.5) +
    #scale_x_continuous(expand = c(0,2)) +
    #scale_y_discrete(expand = c(0,0)) +
    labs(y="-log10 (FDR)",x="Geneset Term",color="Gene Set\nLibrary Type",size="Odds Ratio")+
    theme(axis.text.x=element_markdown(hjust=0.05,size=7,angle=-45,vjust=0.1,colour = "#000000"),
          axis.text.y=element_text(size=7,colour = "#000000"),
          axis.title.x=element_text(size=8,colour = "#000000"),
          axis.title.y=element_text(size=8,colour = "#000000", margin=margin(0,0,0,-0.02,unit="in")),
          legend.title=element_text(size=8,hjust=0.5,colour = "#000000"),
          legend.text=element_text(size=7,colour = "#000000"),
          legend.background=element_blank(),
          legend.margin=margin(0,-0.08,0,-0.1,unit="in"))
dev.off()
```

seshinf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.2

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     methods   base     

other attached packages:
[1] BiocParallel_1.40.0 parallelly_1.43.0   gridExtra_2.3       ggtext_0.1.2       
[5] ggplot2_3.5.2       data.table_1.17.4   rlang_1.1.5        

loaded via a namespace (and not attached):
 [1] gtable_0.3.6       dplyr_1.1.4        compiler_4.4.3     tidyselect_1.2.1  
 [5] Rcpp_1.0.14        stringr_1.5.1      xml2_1.3.8         parallel_4.4.3    
 [9] dichromat_2.0-0.1  scales_1.4.0       yaml_2.3.10        fastmap_1.2.0     
[13] here_1.0.1         R6_2.6.1           commonmark_1.9.5   labeling_0.4.3    
[17] generics_0.1.4     knitr_1.50         tibble_3.2.1       rprojroot_2.0.4   
[21] pillar_1.10.2      RColorBrewer_1.1-3 stringi_1.8.7      litedown_0.6      
[25] xfun_0.52          cli_3.6.4          withr_3.0.2        magrittr_2.0.3    
[29] digest_0.6.37      grid_4.4.3         gridtext_0.1.5     rstudioapi_0.17.1 
[33] markdown_2.0       lifecycle_1.0.4    vctrs_0.6.5        evaluate_1.0.3    
[37] glue_1.8.0         farver_2.1.2       sessioninfo_1.2.3  codetools_0.2-20  
[41] rmarkdown_2.29     tools_4.4.3        pkgconfig_2.0.3    htmltools_0.5.8.1 
> sessioninfo::session_info()
─ Session info ─────────────────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sequoia 15.3.2
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-08-15
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)
 quarto   1.4.555 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/quarto

─ Packages ─────────────────────────────────────────────────────────────────────────────────────
 !  package      * version date (UTC) lib source
    BiocParallel * 1.40.0  2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
 VP cli            3.6.4   2025-04-23 [2] CRAN (R 4.4.1) (on disk 3.6.5)
    codetools      0.2-20  2024-03-31 [1] CRAN (R 4.4.3)
    commonmark     1.9.5   2025-03-17 [1] CRAN (R 4.4.1)
    data.table   * 1.17.4  2025-05-26 [1] CRAN (R 4.4.3)
    dichromat      2.0-0.1 2022-05-02 [1] CRAN (R 4.4.0)
    digest         0.6.37  2024-08-19 [1] CRAN (R 4.4.1)
    dplyr          1.1.4   2023-11-17 [1] CRAN (R 4.4.0)
    evaluate       1.0.3   2025-01-10 [1] CRAN (R 4.4.1)
    farver         2.1.2   2024-05-13 [1] CRAN (R 4.4.0)
    fastmap        1.2.0   2024-05-15 [1] CRAN (R 4.4.0)
    generics       0.1.4   2025-05-09 [1] CRAN (R 4.4.1)
    ggplot2      * 3.5.2   2025-04-09 [1] CRAN (R 4.4.1)
    ggtext       * 0.1.2   2022-09-16 [1] CRAN (R 4.4.0)
    glue           1.8.0   2024-09-30 [1] CRAN (R 4.4.1)
    gridExtra    * 2.3     2017-09-09 [1] CRAN (R 4.4.0)
    gridtext       0.1.5   2022-09-16 [1] CRAN (R 4.4.0)
    gtable         0.3.6   2024-10-25 [1] CRAN (R 4.4.1)
    here           1.0.1   2020-12-13 [1] CRAN (R 4.4.0)
    htmltools      0.5.8.1 2024-04-04 [1] CRAN (R 4.4.0)
    knitr          1.50    2025-03-16 [1] CRAN (R 4.4.2)
    labeling       0.4.3   2023-08-29 [1] CRAN (R 4.4.0)
    lifecycle      1.0.4   2023-11-07 [1] CRAN (R 4.4.0)
    litedown       0.6     2025-02-27 [1] CRAN (R 4.4.3)
    magrittr       2.0.3   2022-03-30 [1] CRAN (R 4.4.0)
    markdown       2.0     2025-03-23 [1] CRAN (R 4.4.3)
    parallelly   * 1.43.0  2025-03-24 [1] CRAN (R 4.4.1)
    pillar         1.10.2  2025-04-05 [1] CRAN (R 4.4.1)
    pkgconfig      2.0.3   2019-09-22 [1] CRAN (R 4.4.0)
    R6             2.6.1   2025-02-15 [1] CRAN (R 4.4.1)
    RColorBrewer   1.1-3   2022-04-03 [1] CRAN (R 4.4.0)
    Rcpp           1.0.14  2025-01-12 [1] CRAN (R 4.4.1)
 VP rlang        * 1.1.5   2025-04-11 [2] CRAN (R 4.4.1) (on disk 1.1.6)
    rmarkdown      2.29    2024-11-04 [1] CRAN (R 4.4.1)
    rprojroot      2.0.4   2023-11-05 [1] CRAN (R 4.4.0)
    rstudioapi     0.17.1  2024-10-22 [1] CRAN (R 4.4.1)
    scales         1.4.0   2025-04-24 [1] CRAN (R 4.4.1)
    sessioninfo    1.2.3   2025-02-05 [1] CRAN (R 4.4.1)
    stringi        1.8.7   2025-03-27 [1] CRAN (R 4.4.1)
    stringr        1.5.1   2023-11-14 [1] CRAN (R 4.4.0)
    tibble         3.2.1   2023-03-20 [1] CRAN (R 4.4.0)
    tidyselect     1.2.1   2024-03-11 [1] CRAN (R 4.4.0)
    vctrs          0.6.5   2023-12-01 [1] CRAN (R 4.4.0)
    withr          3.0.2   2024-10-28 [1] CRAN (R 4.4.1)
    xfun           0.52    2025-04-02 [1] CRAN (R 4.4.1)
    xml2           1.3.8   2025-03-14 [1] CRAN (R 4.4.1)
    yaml           2.3.10  2024-07-26 [1] CRAN (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * ── Packages attached to the search path.
 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

────────────────────────────────────────────────────────────────────────────────────────────────
