---
title: "01-TopDEG_boxplots_10spotmin"
author: "Bernie Mulvey"
date: "2025-02-26"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(ggplot2)
library(ggtext)
library(gridExtra)
library(edgeR)
# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")


## ggplot defaults
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 9), axis.title.x = element_text(size = 8), axis.text.y = element_text(size = 8), axis.title.y = element_text(size =9), plot.title = element_markdown(size = 11,hjust=0.5), strip.text = element_text(size=11), legend.text = element_text(size=8), legend.title = element_text(size=9,hjust=0.5)))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

load candidates table and DGElists
```{r}
gl <- fread("processed-data/Xenium_prioritization/01a-DEGs_byCluster_andNModels_FDR0.2_noXY_maxAbsLFC0.5_10spotmin.txt")

dgels <- readRDS("processed-data/11_DEanalyses/01a-DGElist_per_25HDG75SVGlouv1_cluster_normFilt_10spotmin.RDS")
```

## append addnl metadata used in some DE tests: global ancestry estimates, e4 copies
```{r}
dgels <- lapply(dgels,FUN=function(d){
    d$samples$brnum[d$samples$brnum=="Br6119(re-dis)"] <- "Br6119"

    ## convert APOE genotype to E4 copy number
    
    d$samples$e4copies <- 0
    d$samples$e4copies[d$samples$APOE=="E4/ E4"] <- 2
    d$samples$e4copies[d$samples$APOE=="E3/ E4"] <- 1 # all hets are E4/E3
    
    return(d)
})

## global ancest
flares <- fread("raw-data/ancestryAndHaplotypes/FLARE/global_ancestry_estimates.txt")

dgels <- lapply(dgels,FUN=function(d){
    tmpsamp <- as.data.table(d$samples,keep.rownames=T)
    tmpsamp <- merge.data.table(tmpsamp,flares,by.x="brnum",by.y="ID")
    tmpsamp <- as.data.frame(tmpsamp,row.names=tmpsamp$rn)[rownames(d$samples),]
    tmpsamp$rn <- NULL
    d$samples <- tmpsamp
    return(d)
})

rm(flares)
```


# convert to cpm, melt to long form for plotting, and append metadata vars of interest
```{r}
dgels.logcpm <- lapply(dgels,FUN = function(x){
    # get logcpm
    y <- as.data.table(cpm(x,log=T),keep.rownames=T)
    setnames(y,"rn","gene_id")
    
    # append gene symbols and chromosome (so we can e.g. plot sex DE without skewing by X/Y genes)
    y <- merge.data.table(y,as.data.table(x$genes)[,.(gene_id,gene_name,chr)],by="gene_id")
    
    # long format
    z <- melt(y,id.vars=c("gene_id","gene_name","chr"))
    setnames(z,c("variable","value"),c("sample_id","logCPM"))
    
    # grab metadata vars of interest
    s <- as.data.table(x$samples,keep.rownames=T)[,.(rn,Sex,Age,CEU,Rin,e4,e4copies)]
    setnames(s,"rn","sample_id")
    
    # merge
    z <- merge.data.table(z,s,by="sample_id")
    
    # spit back
    return(z)
})
```

### APOE4 carrier DE
```{r}
gl.e4pos <- gl[Nmodel_E4carrier_DE>0,.(gene_name,louvCluster)]

e4pos.plt <- mapply(g=gl.e4pos$gene_name,d=gl.e4pos$louvCluster,SIMPLIFY=FALSE,FUN=function(g,d){
    pltdat <- dgels.logcpm[[d]][gene_name==g]
    
    p <- ggplot(pltdat,aes(x=e4,y=logCPM,col=e4))+
        geom_boxplot(outliers=FALSE)+
        geom_point()+
        guides(color="none",fill="none")+
        ggtitle(paste0(g," in ",d))
    
    return(p)
    
})

pdf("plots/Xenium_prioritization/01_boxplots_10spotmin/APOE4carrier-non_candidates_10spotmin.pdf",height=(1+(2.5*length(e4pos.plt))),width=8)
do.call("grid.arrange",c(e4pos.plt,ncol=2))
dev.off()

rm(e4pos.plt,gl.e4pos)
```

e4copies
```{r}
gl.e4copies <- gl[Nmodel_E4copies_DE>0,.(gene_name,louvCluster)]

e4copies.plt <- mapply(g=gl.e4copies$gene_name,d=gl.e4copies$louvCluster,SIMPLIFY=FALSE,FUN=function(g,d){
    pltdat <- dgels.logcpm[[d]][gene_name==g]
    
    p <- ggplot(pltdat,aes(x=as.factor(e4copies),y=logCPM,col=as.factor(e4copies)))+
        geom_boxplot(outliers=FALSE)+
        geom_point()+
        guides(color="none")+
        ggtitle(paste0(g," in ",d))

    
    return(p)
    
})

pdf("plots/Xenium_prioritization/01_boxplots_10spotmin/APOE4copies_candidates_10spotmin.pdf",height=(1+(2.5*length(e4copies.plt))),width=8)
do.call("grid.arrange",c(e4copies.plt,ncol=2))
dev.off()

rm(e4copies.plt,gl.e4copies)
```

sex DE in e4 carrier model or e4copies model

```{r}
gl.sex <- gl[maxAbsLFC>0.75&(Nmodel_SexDE_E4carrierCov>0|Nmodel_SexDE_E4copiesCov>0),.(gene_name,louvCluster)]


sex.plt <- mapply(g=gl.sex$gene_name,d=gl.sex$louvCluster,SIMPLIFY=FALSE,FUN=function(g,d){
    pltdat <- dgels.logcpm[[d]][gene_name==g]
    
    p <- ggplot(pltdat,aes(x=Sex,y=logCPM,col=Sex))+
        geom_boxplot(outliers=FALSE)+
        geom_point()+
        guides(color="none")+
        ggtitle(paste0(g," in ",d))
    
    return(p)
    
})

pdf("plots/Xenium_prioritization/01_boxplots_10spotmin/sexDE_candidates_10spotmin.pdf",height=(1+(2.5*length(sex.plt))),width=12)
do.call("grid.arrange",c(sex.plt,ncol=3))
dev.off()

rm(gl.sex,sex.plt)
```

anc scatters
```{r}
gl.anc <- gl[maxAbsLFC>0.75&Nmodel_GlobAncDE_E4copiescov>0,.(gene_name,louvCluster)]

anc.plt <- mapply(g=gl.anc$gene_name,d=gl.anc$louvCluster,SIMPLIFY=FALSE,FUN=function(g,d){
    pltdat <- dgels.logcpm[[d]][gene_name==g]
    
    p <- ggplot(pltdat,aes(x=CEU,y=logCPM,col=CEU))+
        geom_point()+
        guides(color="none")+
        ggtitle(paste0(g," in ",d))
    
    return(p)
    
})

pdf("plots/Xenium_prioritization/01_boxplots_10spotmin/ancDE_candidates_10spotmin.pdf",height=(1+(2.5*length(anc.plt))),width=12)
do.call("grid.arrange",c(anc.plt,ncol=3))
dev.off()

rm(gl.anc,anc.plt)
``` 

repro inf
```{r}
sessionInfo()
sessioninfo::session_info()
```

R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] BiocParallel_1.40.0 parallelly_1.42.0   colorout_1.3-0.2    edgeR_4.4.1        
 [5] limma_3.62.1        gridExtra_2.3       ggtext_0.1.2        ggplot2_3.5.1      
 [9] data.table_1.16.4   rlang_1.1.4        

loaded via a namespace (and not attached):
 [1] generics_0.1.3    xml2_1.3.6        stringi_1.8.4     lattice_0.22-6   
 [5] digest_0.6.37     magrittr_2.0.3    evaluate_1.0.3    grid_4.4.2       
 [9] fastmap_1.2.0     rprojroot_2.0.4   sessioninfo_1.2.2 scales_1.3.0     
[13] codetools_0.2-20  cli_3.6.3         munsell_0.5.1     commonmark_1.9.2 
[17] withr_3.0.2       yaml_2.3.10       tools_4.4.2       parallel_4.4.2   
[21] dplyr_1.1.4       colorspace_2.1-1  locfit_1.5-9.10   here_1.0.1       
[25] vctrs_0.6.5       R6_2.5.1          lifecycle_1.0.4   stringr_1.5.1    
[29] pkgconfig_2.0.3   pillar_1.10.1     gtable_0.3.6      glue_1.8.0       
[33] Rcpp_1.0.14       statmod_1.5.0     xfun_0.50         tibble_3.2.1     
[37] tidyselect_1.2.1  rstudioapi_0.17.1 knitr_1.49        farver_2.1.2     
[41] htmltools_0.5.8.1 rmarkdown_2.29    labeling_0.4.3    compiler_4.4.2   
[45] markdown_1.13     gridtext_0.1.5   
> sessioninfo::session_info()
─ Session info ─────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.2 (2024-10-31)
 os       macOS Sonoma 14.7.1
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-02-24
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ─────────────────────────────────────────────────────────────────────────
 !  package      * version  date (UTC) lib source
    BiocParallel * 1.40.0   2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
 P  cli            3.6.3    2024-06-21 [2] CRAN (R 4.4.0)
    codetools      0.2-20   2024-03-31 [1] CRAN (R 4.4.2)
    colorout     * 1.3-0.2  2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    colorspace     2.1-1    2024-07-26 [1] CRAN (R 4.4.0)
    commonmark     1.9.2    2024-10-04 [1] CRAN (R 4.4.1)
    data.table   * 1.16.4   2024-12-06 [1] CRAN (R 4.4.1)
    digest         0.6.37   2024-08-19 [1] CRAN (R 4.4.1)
    dplyr          1.1.4    2023-11-17 [1] CRAN (R 4.4.0)
    edgeR        * 4.4.1    2024-12-02 [1] Bioconductor 3.20 (R 4.4.2)
    evaluate       1.0.3    2025-01-10 [1] CRAN (R 4.4.1)
    farver         2.1.2    2024-05-13 [1] CRAN (R 4.4.0)
    fastmap        1.2.0    2024-05-15 [1] CRAN (R 4.4.0)
    generics       0.1.3    2022-07-05 [1] CRAN (R 4.4.0)
    ggplot2      * 3.5.1    2024-04-23 [1] CRAN (R 4.4.0)
    ggtext       * 0.1.2    2022-09-16 [1] CRAN (R 4.4.0)
    glue           1.8.0    2024-09-30 [1] CRAN (R 4.4.1)
    gridExtra    * 2.3      2017-09-09 [1] CRAN (R 4.4.0)
    gridtext       0.1.5    2022-09-16 [1] CRAN (R 4.4.0)
    gtable         0.3.6    2024-10-25 [1] CRAN (R 4.4.1)
    here           1.0.1    2020-12-13 [1] CRAN (R 4.4.0)
    htmltools      0.5.8.1  2024-04-04 [1] CRAN (R 4.4.0)
    knitr          1.49     2024-11-08 [1] CRAN (R 4.4.1)
    labeling       0.4.3    2023-08-29 [1] CRAN (R 4.4.0)
    lattice        0.22-6   2024-03-20 [1] CRAN (R 4.4.2)
    lifecycle      1.0.4    2023-11-07 [1] CRAN (R 4.4.0)
    limma        * 3.62.1   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    locfit         1.5-9.10 2024-06-24 [1] CRAN (R 4.4.0)
    magrittr       2.0.3    2022-03-30 [1] CRAN (R 4.4.0)
    markdown       1.13     2024-06-04 [1] CRAN (R 4.4.0)
    munsell        0.5.1    2024-04-01 [1] CRAN (R 4.4.0)
    parallelly   * 1.42.0   2025-01-30 [1] CRAN (R 4.4.1)
    pillar         1.10.1   2025-01-07 [1] CRAN (R 4.4.1)
    pkgconfig      2.0.3    2019-09-22 [1] CRAN (R 4.4.0)
    R6             2.5.1    2021-08-19 [1] CRAN (R 4.4.0)
    Rcpp           1.0.14   2025-01-12 [1] CRAN (R 4.4.1)
 VP rlang        * 1.1.4    2025-01-17 [2] CRAN (R 4.4.1) (on disk 1.1.5)
    rmarkdown      2.29     2024-11-04 [1] CRAN (R 4.4.1)
    rprojroot      2.0.4    2023-11-05 [1] CRAN (R 4.4.0)
    rstudioapi     0.17.1   2024-10-22 [1] CRAN (R 4.4.1)
    scales         1.3.0    2023-11-28 [1] CRAN (R 4.4.0)
    sessioninfo    1.2.2    2021-12-06 [1] CRAN (R 4.4.0)
    statmod        1.5.0    2023-01-06 [1] CRAN (R 4.4.0)
    stringi        1.8.4    2024-05-06 [1] CRAN (R 4.4.0)
    stringr        1.5.1    2023-11-14 [1] CRAN (R 4.4.0)
    tibble         3.2.1    2023-03-20 [1] CRAN (R 4.4.0)
    tidyselect     1.2.1    2024-03-11 [1] CRAN (R 4.4.0)
    vctrs          0.6.5    2023-12-01 [1] CRAN (R 4.4.0)
    withr          3.0.2    2024-10-28 [1] CRAN (R 4.4.1)
    xfun           0.50     2025-01-07 [1] CRAN (R 4.4.1)
    xml2           1.3.6    2023-12-04 [1] CRAN (R 4.4.0)
    yaml           2.3.10   2024-07-26 [1] CRAN (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

────────────────────────────────────────────────────────────────────────────────────
