---
title: "02-escheR plot Xen candidates in clusts of interest"
author: "Bernie Mulvey"
date: "2025-02-25"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(ggplot2)
library(ggtext)
library(gridExtra)
library(SpatialExperiment)
library(escheR)

# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")


## ggplot defaults
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 9), axis.title.x = element_text(size = 8), axis.text.y = element_text(size = 8), axis.title.y = element_text(size =9), plot.title = element_markdown(size = 11,hjust=0.5), strip.text = element_text(size=11), legend.text = element_text(size=8), legend.title = element_text(size=9,hjust=0.5)))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

## load SPE and assign hdg_svg_2575_louv_res1 clusterings as autoannotated
```{r}
lc3 <- readRDS("processed-data/06-countsOnly_noImgData_QCed_SPE_split_to_tissSections.RDS")

louvs <- readRDS("processed-data/07_featureSelection_dimred_harmony_clustering/04-Initial_louvLeid_res0.5-1-2_clusterings.RDS")
louvs <- louvs[["HARMONYlmbna_HDG_SVG_2575"]][,.(rn,snnHARMONYlmbna_HDG_SVG_2575_louv_res1)]
setnames(louvs,2,"clusid")


anno <- readRDS("processed-data/07_featureSelection_dimred_harmony_clustering/05b-init_clusterings_autoannoLCrapheAndGABAergic.RDS")
anno <- anno$snnHARMONYlmbna_HDG_SVG_2575_louv_res1

louvs <- merge(louvs,anno,by="clusid",all.x=T)
louvs <- DataFrame(louvs,row.names=louvs$rn)[colnames(lc3),]
colLabels(lc3) <- louvs$autoanno

## calculate log counts
lc3 <- scater::computeLibraryFactors(lc3)
lc3 <- scater::logNormCounts(lc3)
```

### append FLARE CEU estimate, APOE carrier status, and e4 allele count columns so we can order the plotted samples on these DE variables
```{r}
tmpcd <- as.data.table(colData(lc3),keep.rownames=T)

tmpcd[brnum=="Br6119(re-dis)",brnum:="Br6119"]

tmpcd[,E4carrier:="E4-"]
tmpcd[APOE %in% c("E4/ E4","E3/ E4"),E4carrier:="E4+"]

tmpcd[,E4copies:=0]
tmpcd[APOE %in% c("E4/ E4"),E4copies:=2]
tmpcd[APOE %in% c("E3/ E4"),E4copies:=1]

## get flare ests
flares <- fread("raw-data/ancestryAndHaplotypes/FLARE/global_ancestry_estimates.txt")
flares <- flares[ID %in% tmpcd$brnum,.(ID,CEU)]

tmpcd <- merge.data.table(tmpcd,flares,by.x="brnum",by.y="ID",all.x=T)

tmpcd <- DataFrame(tmpcd,row.names=tmpcd$rn)[colnames(lc3),]
tmpcd$rn <- NULL

colData(lc3) <- tmpcd
```

### get candidate genes and their cluster / DE variable
```{r}
gn <- fread("processed-data/Xenium_prioritization/02-Boxplot_candidates_andnotes_cleaned.txt")
```

### systematically plot for non-LCNM subsets

```{r}
gn <- gn[clust %in% c("LC.1","5HT.1","5HT.2",paste0("X",c(1:9)))]

## make spe rownames gene symbols
rownames(lc3) <- rowData(lc3)$gene_name

# spotplot with fill scaled to max expression IN the cluster of interest
mapply(g=gn$gene,cl=gn$clust,t=gn$test,SIMPLIFY=FALSE,FUN=function(g,cl,t){
    tmpspe <- lc3[g,]
    tmpspe$label <- ifelse(tmpspe$label==cl,yes=cl,no="other")
    
    if (t=="sex"){
        sampord <- c(unique(colData(tmpspe[,tmpspe$Sex=="M"])$sample_id),
                     unique(colData(tmpspe[,tmpspe$Sex=="F"])$sample_id))
    } else if (t=="apo_carrier") {
        sampord <- c(unique(colData(tmpspe[,tmpspe$E4carrier=="E4+"])$sample_id),
                     unique(colData(tmpspe[,tmpspe$E4carrier=="E4-"])$sample_id))
    } else if (t=="apo_copies") {
        sampord <- c(unique(colData(tmpspe[,tmpspe$E4copies==0])$sample_id),
                     unique(colData(tmpspe[,tmpspe$E4copies==1])$sample_id),
                     unique(colData(tmpspe[,tmpspe$E4copies==2])$sample_id))
    } else if (t=="CEU") {
        sampord <- unique(colData(tmpspe)[order(tmpspe$CEU),"sample_id"])
    } else {
        stop("unrecognized test")
    }
    
    tmpsamp <- lapply(sampord,FUN=function(x){tmpspe[,tmpspe$sample_id==x]})
    pal <- c("red","darkblue")
    names(pal) <- c(cl,"other")
    
    plt <- lapply(tmpsamp,FUN=function(x){
        colData(x)$`log counts` <- logcounts(x)[g,]
        p <- make_escheR(x)
    
        p <- p |> add_fill("log counts",size=0.7,point_size = 0.95)
        p <- p |> add_ground(var = "label",stroke=0.25,point_size = 0.95)

    p <- p+
        scale_fill_gradient(low="white",high="black",limits=c(0,max(logcounts(tmpspe)[g,tmpspe$label==cl])))+
        scale_color_manual(values=pal)+
        guides(color="none",fill="none")+
        ggtitle(paste0(unique(x$sample_id)," ",unique(x$Sex),"/",unique(x$E4carrier),"/",unique(x$E4copies),"/",round(unique(x$CEU),2)))+
        theme(title = element_text(size=9,hjust=0.5))
    
        return(p)
    })
    
    # save
    png(paste0("plots/Xenium_prioritization/02_spotplots/",g,"_",cl,"_",t,".png"),width=3000,height=700*22,res=300)
    do.call("grid.arrange",c(plt,ncol=4))
    dev.off()
})


```

rep inf
```{r}
sessionInfo()
sessioninfo::session_info()
```

R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] BiocParallel_1.40.0         parallelly_1.42.0          
 [3] colorout_1.3-0.2            escheR_1.6.0               
 [5] SpatialExperiment_1.16.0    SingleCellExperiment_1.28.1
 [7] SummarizedExperiment_1.36.0 Biobase_2.66.0             
 [9] GenomicRanges_1.58.0        GenomeInfoDb_1.42.1        
[11] IRanges_2.40.1              S4Vectors_0.44.0           
[13] BiocGenerics_0.52.0         MatrixGenerics_1.18.0      
[15] matrixStats_1.5.0           gridExtra_2.3              
[17] ggtext_0.1.2                ggplot2_3.5.1              
[19] data.table_1.16.4           rlang_1.1.4                

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.1        viridisLite_0.4.2       farver_2.1.2           
 [4] dplyr_1.1.4             vipor_0.4.7             viridis_0.6.5          
 [7] fastmap_1.2.0           digest_0.6.37           rsvd_1.0.5             
[10] lifecycle_1.0.4         magrittr_2.0.3          compiler_4.4.2         
[13] tools_4.4.2             yaml_2.3.10             knitr_1.49             
[16] labeling_0.4.3          S4Arrays_1.6.0          here_1.0.1             
[19] DelayedArray_0.32.0     xml2_1.3.6              abind_1.4-8            
[22] withr_3.0.2             grid_4.4.2              beachmat_2.22.0        
[25] colorspace_2.1-1        scales_1.3.0            cli_3.6.3              
[28] rmarkdown_2.29          crayon_1.5.3            generics_0.1.3         
[31] rstudioapi_0.17.1       httr_1.4.7              rjson_0.2.23           
[34] scuttle_1.16.0          ggbeeswarm_0.7.2        zlibbioc_1.52.0        
[37] parallel_4.4.2          XVector_0.46.0          vctrs_0.6.5            
[40] Matrix_1.7-1            jsonlite_1.8.9          BiocSingular_1.22.0    
[43] BiocNeighbors_2.0.1     ggrepel_0.9.6           irlba_2.3.5.1          
[46] beeswarm_0.4.0          scater_1.34.0           magick_2.8.5           
[49] glue_1.8.0              codetools_0.2-20        gtable_0.3.6           
[52] UCSC.utils_1.2.0        ScaledMatrix_1.14.0     munsell_0.5.1          
[55] tibble_3.2.1            pillar_1.10.1           htmltools_0.5.8.1      
[58] GenomeInfoDbData_1.2.13 R6_2.5.1                rprojroot_2.0.4        
[61] evaluate_1.0.3          lattice_0.22-6          gridtext_0.1.5         
[64] Rcpp_1.0.14             SparseArray_1.6.0       xfun_0.50              
[67] pkgconfig_2.0.3        
> sessioninfo::session_info()
─ Session info ─────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.2 (2024-10-31)
 os       macOS Sonoma 14.7.1
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-02-25
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ─────────────────────────────────────────────────────────────────────────
 !  package              * version date (UTC) lib source
    abind                  1.4-8   2024-09-12 [1] CRAN (R 4.4.1)
    beachmat               2.22.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    beeswarm               0.4.0   2021-06-01 [1] CRAN (R 4.4.0)
    Biobase              * 2.66.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocNeighbors          2.0.1   2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
    BiocParallel         * 1.40.0  2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    BiocSingular           1.22.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
 P  cli                    3.6.3   2024-06-21 [2] CRAN (R 4.4.0)
    codetools              0.2-20  2024-03-31 [1] CRAN (R 4.4.2)
    colorout             * 1.3-0.2 2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    colorspace             2.1-1   2024-07-26 [1] CRAN (R 4.4.0)
    crayon                 1.5.3   2024-06-20 [1] CRAN (R 4.4.0)
    data.table           * 1.16.4  2024-12-06 [1] CRAN (R 4.4.1)
    DelayedArray           0.32.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    digest                 0.6.37  2024-08-19 [1] CRAN (R 4.4.1)
    dplyr                  1.1.4   2023-11-17 [1] CRAN (R 4.4.0)
    escheR               * 1.6.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    evaluate               1.0.3   2025-01-10 [1] CRAN (R 4.4.1)
    farver                 2.1.2   2024-05-13 [1] CRAN (R 4.4.0)
    fastmap                1.2.0   2024-05-15 [1] CRAN (R 4.4.0)
    generics               0.1.3   2022-07-05 [1] CRAN (R 4.4.0)
    GenomeInfoDb         * 1.42.1  2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13  2024-12-12 [1] Bioconductor
    GenomicRanges        * 1.58.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ggbeeswarm             0.7.2   2023-04-29 [1] CRAN (R 4.4.0)
    ggplot2              * 3.5.1   2024-04-23 [1] CRAN (R 4.4.0)
    ggrepel                0.9.6   2024-09-07 [1] CRAN (R 4.4.1)
    ggtext               * 0.1.2   2022-09-16 [1] CRAN (R 4.4.0)
    glue                   1.8.0   2024-09-30 [1] CRAN (R 4.4.1)
    gridExtra            * 2.3     2017-09-09 [1] CRAN (R 4.4.0)
    gridtext               0.1.5   2022-09-16 [1] CRAN (R 4.4.0)
    gtable                 0.3.6   2024-10-25 [1] CRAN (R 4.4.1)
    here                   1.0.1   2020-12-13 [1] CRAN (R 4.4.0)
    htmltools              0.5.8.1 2024-04-04 [1] CRAN (R 4.4.0)
    httr                   1.4.7   2023-08-15 [1] CRAN (R 4.4.0)
    IRanges              * 2.40.1  2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    irlba                  2.3.5.1 2022-10-03 [1] CRAN (R 4.4.0)
    jsonlite               1.8.9   2024-09-20 [1] CRAN (R 4.4.1)
    knitr                  1.49    2024-11-08 [1] CRAN (R 4.4.1)
    labeling               0.4.3   2023-08-29 [1] CRAN (R 4.4.0)
    lattice                0.22-6  2024-03-20 [1] CRAN (R 4.4.2)
    lifecycle              1.0.4   2023-11-07 [1] CRAN (R 4.4.0)
    magick                 2.8.5   2024-09-20 [1] CRAN (R 4.4.1)
    magrittr               2.0.3   2022-03-30 [1] CRAN (R 4.4.0)
    Matrix                 1.7-1   2024-10-18 [1] CRAN (R 4.4.2)
    MatrixGenerics       * 1.18.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    matrixStats          * 1.5.0   2025-01-07 [1] CRAN (R 4.4.1)
    munsell                0.5.1   2024-04-01 [1] CRAN (R 4.4.0)
    parallelly           * 1.42.0  2025-01-30 [1] CRAN (R 4.4.1)
    pillar                 1.10.1  2025-01-07 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3   2019-09-22 [1] CRAN (R 4.4.0)
    R6                     2.5.1   2021-08-19 [1] CRAN (R 4.4.0)
    Rcpp                   1.0.14  2025-01-12 [1] CRAN (R 4.4.1)
    rjson                  0.2.23  2024-09-16 [1] CRAN (R 4.4.1)
 VP rlang                * 1.1.4   2025-01-17 [2] CRAN (R 4.4.1) (on disk 1.1.5)
    rmarkdown              2.29    2024-11-04 [1] CRAN (R 4.4.1)
    rprojroot              2.0.4   2023-11-05 [1] CRAN (R 4.4.0)
    rstudioapi             0.17.1  2024-10-22 [1] CRAN (R 4.4.1)
    rsvd                   1.0.5   2021-04-16 [1] CRAN (R 4.4.0)
    S4Arrays               1.6.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ScaledMatrix           1.14.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.3.0   2023-11-28 [1] CRAN (R 4.4.0)
    scater                 1.34.0  2024-10-29 [1] Bioconductor 3.20 (R 4.4.1)
    scuttle                1.16.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    sessioninfo            1.2.2   2021-12-06 [1] CRAN (R 4.4.0)
    SingleCellExperiment * 1.28.1  2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    SparseArray            1.6.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SpatialExperiment    * 1.16.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SummarizedExperiment * 1.36.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    tibble                 3.2.1   2023-03-20 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1   2024-03-11 [1] CRAN (R 4.4.0)
    UCSC.utils             1.2.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    vctrs                  0.6.5   2023-12-01 [1] CRAN (R 4.4.0)
    vipor                  0.4.7   2023-12-18 [1] CRAN (R 4.4.0)
    viridis                0.6.5   2024-01-29 [1] CRAN (R 4.4.0)
    viridisLite            0.4.2   2023-05-02 [1] CRAN (R 4.4.0)
    withr                  3.0.2   2024-10-28 [1] CRAN (R 4.4.1)
    xfun                   0.50    2025-01-07 [1] CRAN (R 4.4.1)
    xml2                   1.3.6   2023-12-04 [1] CRAN (R 4.4.0)
    XVector                0.46.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    yaml                   2.3.10  2024-07-26 [1] CRAN (R 4.4.0)
    zlibbioc               1.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

────────────────────────────────────────────────────────────────────────────────────
