---
title: "02-calc_APOE_haplolocus_ancestry"
author: "Bernie Mulvey"
date: "2025-02-21"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
setDTthreads(8)
```

## load chr19 local ancestry estimates from /dcs04/lieber/shared/statsgen/local_ancestry/flare_array_data/merged-R.9_local_ancestry_chr19.anc.vcf.gz

vcf header:
##fileformat=VCFv4.2
##filedate=20250113
##source=flare.15Jan24.5f5.jar
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=AN1,Number=1,Type=Integer,Description="Ancestry of first haplotype">
##FORMAT=<ID=AN2,Number=1,Type=Integer,Description="Ancestry of second haplotype">
##ANCESTRY=<CEU=0,YRI=1,CHB=2>
```{r}
# get donor ids
lc3cd <- fread("processed-data/06-QCed_SPE_splitToTissSect_colData.txt")
brs <- unique(lc3cd$brnum)
brs[brs=="Br6119(re-dis)"] <- "Br6119"

anc <- fread("~/Desktop/merged-R.9_local_ancestry_chr19.anc.vcf")
setnames(anc,1,"CHROM")
keepn <- c("POS","REF","ALT",brs)
anc <- anc[,..keepn]
gc(full=T)

## coords in the vcf are in hg38. for comparability to prior work, UCSC liftover the hg19 window, 40416921-51727962 as described in https://www.nature.com/articles/s41380-022-01729-x .
## liftover: chr19	39911014	51224706

anc <- anc[POS>=39911014&POS<=51224706]
## extract haplotype 1 and haplotype 2 into separate columns for each donor
posn <- c("POS",brs)
anc1 <- copy(anc[,..posn])
anc1 <- melt.data.table(anc1,id.vars="POS")
anc1[,value:=as.numeric(gsub(value,pattern="^.\\|.:(.):.$",replacement="\\1"))]
## drop the occassional CHB genotypes (treat as 0)
anc1[value==2,value:=0]
hap1 <- anc1[,mean(value),by="variable"]
setnames(hap1,"V1","apoe_hap1_estPropYRI")

anc2 <- copy(anc[,..posn])
anc2 <- melt.data.table(anc2,id.vars="POS")
anc2[,value:=as.numeric(gsub(value,pattern="^.\\|.:.:(.)$",replacement="\\1"))]
## drop the occassional CHB genotypes (treat as 0)
anc2[value==2,value:=0]
hap2 <- anc2[,mean(value),by="variable"]
setnames(hap2,"V1","apoe_hap2_estPropYRI")

## get haplotype 1 APOE isoform and hap2 APOE isoform
# 44908684:rs429358
# 44908822:rs7412

# rs42 rs74 apoe
# T 	T 	ε2
# T 	C 	ε3 
# C 	C	ε4

hapg1 <- copy(anc)
hapg1.42 <- hapg1[POS==44908684]
hapg1.42 <- melt.data.table(hapg1.42,id.vars=c("POS","REF","ALT"))
hapg1.42[,hap1gt42:=gsub(value,pattern="^(.)\\|.*$",replacement="\\1")]
hapg1.42[hap1gt42=="0",gt42:=REF]
hapg1.42[hap1gt42=="1",gt42:=ALT]

hapg1.74 <- hapg1[POS==44908822]
hapg1.74 <- melt.data.table(hapg1.74,id.vars=c("POS","REF","ALT"))
hapg1.74[,hap1gt74:=gsub(value,pattern="^(.)\\|.*$",replacement="\\1")]
hapg1.74[hap1gt74=="0",gt74:=REF]
hapg1.74[hap1gt74=="1",gt74:=ALT]

hapg1 <- merge(hapg1.42[,.(variable,gt42)],hapg1.74[,.(variable,gt74)],by="variable")
hapg1[,hap1gt:=paste0(gt42,gt74)]

## haplotype 2 APOE isoform
hapg2 <- copy(anc)
hapg2.42 <- hapg2[POS==44908684]
hapg2.42 <- melt.data.table(hapg2.42,id.vars=c("POS","REF","ALT"))
hapg2.42[,hap2gt42:=gsub(value,pattern="^.\\|(.):.*$",replacement="\\1")]
hapg2.42[hap2gt42=="0",gt42:=REF]
hapg2.42[hap2gt42=="1",gt42:=ALT]

hapg2.74 <- hapg2[POS==44908822]
hapg2.74 <- melt.data.table(hapg2.74,id.vars=c("POS","REF","ALT"))
hapg2.74[,hap2gt74:=gsub(value,pattern="^.\\|(.):.*$",replacement="\\1")]
hapg2.74[hap2gt74=="0",gt74:=REF]
hapg2.74[hap2gt74=="1",gt74:=ALT]

hapg2 <- merge(hapg2.42[,.(variable,gt42)],hapg2.74[,.(variable,gt74)],by="variable")
hapg2[,hap2gt:=paste0(gt42,gt74)]

## merge haplotype 1 and haplotype 2 APOE isoform
hapg <- merge(hapg1[,.(variable,hap1gt)],hapg2[,.(variable,hap2gt)],by="variable")
## append local ancestry estimates per haplotype
hapg <- merge(hapg,hap1[,.(variable,apoe_hap1_estPropYRI)],by="variable")
hapg <- merge(hapg,hap2[,.(variable,apoe_hap2_estPropYRI)],by="variable")

hapg[hap1gt=="TT",hap1gt:="E2"]
hapg[hap1gt=="TC",hap1gt:="E3"]
hapg[hap1gt=="CC",hap1gt:="E4"]

hapg[hap2gt=="TT",hap2gt:="E2"]
hapg[hap2gt=="TC",hap2gt:="E3"]
hapg[hap2gt=="CC",hap2gt:="E4"]

## make sure the genotypes we're getting here agree with the source metadata
check <- merge.data.table(unique(lc3cd[,.(brnum,APOE)]),hapg,by.x="brnum",by.y="variable")
## looks good.
```

## save
```{r}
fwrite(hapg,"raw-data/ancestryAndHaplotypes/FLARE/apoe_haplotypes_w_localYRI_est.txt",sep='\t',quote=F,row.names=F,col.names=T)
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils    
[5] datasets  methods   base     

other attached packages:
[1] data.table_1.16.4 rlang_1.1.4      

loaded via a namespace (and not attached):
[1] compiler_4.4.2  here_1.0.1     
[3] rprojroot_2.0.4 cli_3.6.3      
[5] tools_4.4.2     knitr_1.49     
[7] xfun_0.50       evaluate_1.0.3 
> sessioninfo::session_info()
─ Session info ────────────────────────────────
 setting  value
 version  R version 4.4.2 (2024-10-31)
 os       macOS Sonoma 14.7.1
 system   aarch64, darwin20
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-02-21
 pandoc   NA

─ Packages ────────────────────────────────────
 !  package     * version date (UTC) lib source
 P  cli           3.6.3   2024-06-21 [2] CRAN (R 4.4.0)
    data.table  * 1.16.4  2024-12-06 [1] CRAN (R 4.4.1)
    evaluate      1.0.3   2025-01-10 [1] CRAN (R 4.4.1)
    here          1.0.1   2020-12-13 [1] CRAN (R 4.4.0)
    knitr         1.49    2024-11-08 [1] CRAN (R 4.4.1)
 VP rlang       * 1.1.4   2025-01-17 [2] CRAN (R 4.4.1) (on disk 1.1.5)
    rprojroot     2.0.4   2023-11-05 [1] CRAN (R 4.4.0)
    sessioninfo   1.2.2   2021-12-06 [1] CRAN (R 4.4.0)
    xfun          0.50    2025-01-07 [1] CRAN (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

───────────────────────────────────────────────