---
title: "SupFig-Registration_to_LuskinMouseLCsnseq"
author: "Bernie Mulvey"
date: "2025-09-10"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(ggplot2)
library(ggtext)
library(gridExtra)
# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")


## ggplot defaults
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 8,color = "#000000"), axis.title.x = element_text(size = 9,color = "#000000"), axis.text.y = element_text(size = 8,color = "#000000"), axis.title.y = element_text(size =9,color = "#000000"), plot.title = element_markdown(size = 10,hjust=0.5,color = "#000000"), strip.text = element_text(size=11), legend.text = element_text(size=7,color = "#000000"), legend.title = element_text(size=8,hjust=0.5,color = "#000000"),axis.ticks = element_line(linewidth=0.5),panel.grid.major = element_line(linewidth=0.5),panel.grid.minor=element_line(linewidth=0.25),plot.title.position="plot"))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

## plot the primary clusters against this data using top 700 mouse genes with an ortholog--does the job across glia and neurons
```{r}
lff.luskin.reg <- readRDS("processed-data/08_validitycheck_25hdg75svg_louv1/05-Registrations_to_LuskinMs2024.RDS")
anno <- fread("processed-data/08_validitycheck_25hdg75svg_louv1/10-25hdg75svg_louv1_annots.txt")

lff.luskin.reg <- lff.luskin.reg$allRetainedClusts
## map to plotting domain names
anno[,anno_prelim_alt:=gsub(anno_prelim_alt,pattern="xX",replacement="X")]
anno[,anno_prelim_alt:=gsub(anno_prelim_alt,pattern="xLC",replacement="LC")]

lff.luskin.reg <- merge.data.table(lff.luskin.reg,anno[,.(anno_prelim_alt,plot_anno)],by.x="LFFclus",by.y="anno_prelim_alt")

## top 700 does a good job for both glial and (to the extent possible, i.e. lighting up DBH-to-LC domain at all) neurons
plotn=700

lff.luskin.reg[,msAnnot:=factor(msAnnot,levels=rev(c("NeuronDbh","NeuronChat",paste0("Neuron",c(1,3:6,8:14)),paste0("Ambiguous",c(1:2)),"Oligo","OPC1","Astrocyte","Ependymal","Mural","Epithelial","Microglia")))]

lff.luskin.reg[,plot_anno:=factor(plot_anno,levels=c("Vascular","Astro","Astro-Oligo Mixed","Oligo","Oligo-Astro Mixed","Oligo and Other Types Mixed","*ACHE*, *SERT*, Neuropeptide Mixed","*ACHE*, *SERT*, LC, Oligo Mixed","LC"))]

pdf("plots/manuscript/SuppFigs/SupFig-Registration_to_LuskinMouseLCsnseq.pdf",height=7,width=5.5)
ggplot(lff.luskin.reg[topNmsmk==plotn],aes(y=msAnnot,x=plot_anno,fill=value))+
    geom_tile()+
    scale_fill_continuous(type="viridis",
        limits=c(
            min(lff.luskin.reg[topNmsmk==plotn,value]),
            max(lff.luskin.reg[topNmsmk==plotn,value])))+
    theme(axis.text.x=element_markdown(angle=45,hjust=1,vjust=1))+
    labs(x="Domain",
         y="Mouse snRNAseq Cluster (Luskin 2025)",
         fill="Spearman Cor.\n(Top 700 Mouse\nMarkers w/ Human\nOrtholog)")+
    scale_x_discrete(expand = c(0,0))+
    scale_y_discrete(expand = c(0,0))
dev.off()
```

