---
title: "01-Make_cluster_DGEs_withMinSpots_reqmnt"
author: "Bernie Mulvey"
date: "2025-02-26"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(SpatialExperiment)
library(dreamlet)
library(limma)
library(edgeR)
library(gridExtra)
library(ggplot2)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")
library(parallel) # use PSOCK clusters for best performance here

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

## load SPE and assign hdg_svg_2575_louv_res1 clusterings as autoannotated
```{r}
lc3 <- readRDS("processed-data/06-countsOnly_noImgData_QCed_SPE_split_to_tissSections.RDS")

louvs <- readRDS("processed-data/07_featureSelection_dimred_harmony_clustering/04-Initial_louvLeid_res0.5-1-2_clusterings.RDS")
louvs <- louvs[["HARMONYlmbna_HDG_SVG_2575"]][,.(rn,snnHARMONYlmbna_HDG_SVG_2575_louv_res1)]
setnames(louvs,2,"clusid")

autoan <- readRDS("processed-data/07_featureSelection_dimred_harmony_clustering/05b-init_clusterings_autoannoLCrapheAndGABAergic.RDS")
autoan <- autoan[["snnHARMONYlmbna_HDG_SVG_2575_louv_res1"]]

louvs <- merge.data.table(louvs,autoan,by="clusid")
louvs <- DataFrame(louvs,row.names=louvs$rn)[colnames(lc3),]

colLabels(lc3) <- louvs$autoanno

## for plotting by Visium slide, we need to edit the colData(spe)$Visium_slide to JUST be the slide (it still has the slide capture area ID appended to it)
colData(lc3)$Visium_slide <- gsub(colData(lc3)$Visium_slide,pattern="^(V.....-...)_..$",replacement="\\1")

rm(louvs,autoan)
```

#### TOP LEVEL CLUSTER DGELists ####

# make pseudobulk object (dreamlet is v. fast at this)
```{r}
ag <- dreamlet::aggregateToPseudoBulk(lc3,sample_id="sample_id",cluster_id = "label",BPPARAM = MulticoreParam(6))

# extract number of spots in a pseudobulk sample to filter by. this is in the inverse of the format we want, list entries are sample and subentries are cluster, so rearrange to cluster listed by sample
agn <- ag@int_colData$n_cells
agn <- lapply(agn,function(x){as.data.frame(t(x))})
agn <- rbindlist(agn,idcol="samp")

# extract colData to feed into DGEList as the "samples" data
agcd <- as.data.frame(colData(ag))

## only retain clusters for which at least 43 samples (of 85 total, i.e over half) of â‰¥10 spots each are available:  
ag <- as.list(assays(ag))
agc <- mapply(ag,names(ag),SIMPLIFY=FALSE,FUN=function(a,n){
  tenspot <- agn[get(n)>=10,samp]
  ncol(a[,colSums(a)>0][,tenspot])
})

ag <- ag[!(agc<43)]

## drop clusters where there are 20+ samples, but under 18 samples for one of the sexes (i.e., < half of the samples from the smaller group--females w 35 sections total):
agsc <-  mapply(ag,names(ag),SIMPLIFY=FALSE,FUN=function(a,n){
    cn <- colnames(a[,colSums(a)>0]) # any counts
    cn <- cn[cn %in% agn[get(n)>=10,samp]] # 10+ spots included
    tmp <- as.data.table(agcd[cn,])
    if(nrow(tmp[,.N,by="Sex"][N<18])>0){
      return(0)
    } else {return(1)}
})
ag <- ag[agsc==1]

# stratify by APOE4 positive / negative  
agcd$e4 <- ifelse(agcd$APOE %in% c("E4/ E4","E3/ E4"),yes="carrier",no="negative")
```

## make the DGELists per remaining cluster:
## ADDED 02/14/25: store plots of the entire count distribution with the low-expression cutoff marked to check that this is working as expected.
## ADDED 02/26/25: filter out samples with under 10 spots in the current cluster
```{r}
## get rowData as a data.table from the SPE to append to the DGElist--namely so we can make MDS plots of the pseudobulks while excluding mitochondrial and sex chromosomal genes
tmprd <- as.data.table(rowData(lc3)) # rownames are the same as $gene_id

dgels <- mapply(a=ag,n=names(ag),SIMPLIFY=FALSE,FUN=function(a,n){
    cn <- colnames(a[,colSums(a)>0]) # has counts
    
    tenspot <- cn[cn %in% agn[get(n)>=10,samp]] # and has 10+spots contributed
    X <- a[,tenspot]
    
    s <- copy(agcd)[colnames(X),]
    if(length(unique(s$Sex))<2){return()} else 
    {

        dge <- DGEList(counts=as.matrix(X),samples=s,genes=as.data.frame(tmprd,row.names = tmprd$rn)[rownames(X),])
        dge <- normLibSizes(dge,method = "TMM")
        
        ## find the smaller group for this cluster
        smolgrp <- table(dge$samples$Sex)[which.min(table(dge$samples$Sex))]
        
        ## for the smallest group, determine low expression cutoff
        rowmn.m <- as.numeric(rowMeans(cpm(dge[,(dge$samples$Sex==names(smolgrp)[1])],log=TRUE)))
        
        ## get histogram distribution with 10 bins at the low end of the logcpm space
        h <- hist(rowmn.m[which(rowmn.m>min(rowmn.m)&rowmn.m<(4+min(rowmn.m)))],breaks=10,plot=FALSE)
        ## get the minimum in this low logcpm range and its corresponding bin (x axis value, i.e. log cpm)
        minc <- h$breaks[which(h$counts==min(h$counts))]
        
        ## get the corresponding smallest-library count value for this minimum^
        smollib <- min(dge$samples[(dge$samples$Sex==names(smolgrp)[1]),"lib.size"])
        countthresh <- (minc^2)*(as.numeric(smollib)/10^6)
        
        ## and convert that smallest count to the logcpm value in said smallest library to use as a low-expression threshold
        cpmthresh <- log2(countthresh/(as.numeric(smollib)*10^-6))
        
        ## 2/14/25: make a plot to also return
        mnplt <- as.data.table(rowmn.m)
        setnames(mnplt,1,"V1")
        
        ## exclude genes with a mean logcpm under 0.1
        mnplt <- mnplt[V1>0.1]
        
        plt <- ggplot(mnplt,aes(x=V1))+
          geom_histogram(bins = 100)+
          geom_vline(xintercept=cpmthresh,col="red")+
          guides(color="none")+
          ggtitle(paste0(n," genewise mean logCPM in\nlower n sex (",smolgrp,")"))
        
        ## filter for genes with at least thresh counts in at least (smaller sex group size) samples
        keep <- rowSums(cpm(dge,log=T)>cpmthresh) >= as.numeric(smolgrp)
        
        ## recalc lib factors after dropping low-expressed genes
        dge <- dge[keep,,keep.lib.sizes=FALSE]
        dge <- calcNormFactors(dge,method = "TMM")
        
        retlist <- list(dge,plt)
        names(retlist) <- c(n,"plt")
        return(retlist)
    }
})

## get and arrange plots
plts <- lapply(dgels,function(x){x$plt})

pdf("plots/11_DEanalyses/01-logCPM_cutoff_checks_10spotfilt.pdf",height=10,width=10)
do.call("grid.arrange",c(plts,ncol=3))
dev.off()

dgels <- lapply(dgels,function(x){x[[1]]})
```

## save -- with different name for 10spot filtering on 022625
```{r}
saveRDS(dgels,"processed-data/11_DEanalyses/01a-DGElist_per_25HDG75SVGlouv1_cluster_normFilt_10spotmin.RDS")
```

## make MDS plots per pseudobulk cluster per potential variable of interest
```{r}
clus <- makeCluster(9,"PSOCK")
clusterMap(d=dgels,n=names(dgels),SIMPLIFY=FALSE,cl=clus,fun=function(d,n){
  covcols <- c("Sex","e4","Visium_slide","brnum")
  lapply(covcols,FUN=function(cov){
    pal <- rainbow(n=length(unique(d$samples[,cov])))
    d$samples[,cov] <- as.factor(d$samples[,cov])
    
    png(paste0("plots/11_DEanalyses/01_pseudobulk_cluster_MDSes/01a-",n,"_",cov,"_1_2_10spotfilt.png"),width=1000,height=800)
    plotMDS(d,col=pal[factor(d$samples[,cov])], dim.plot = c(1,2),pch = 18)
    legend('bottomright', legend=levels(d$samples[,cov]), col=pal, cex=1, pch=18)
    dev.off()
    
    png(paste0("plots/11_DEanalyses/01_pseudobulk_cluster_MDSes/01a-",n,"_",cov,"_1_3_10spotfilt.png"),width=1000,height=800)
    plotMDS(d,col=pal[factor(d$samples[,cov])], dim.plot = c(1,3),pch = 18)
    legend('bottomright', legend=levels(d$samples[,cov]), col=pal, cex=1, pch=18)
    dev.off()
    
    png(paste0("plots/11_DEanalyses/01_pseudobulk_cluster_MDSes/01a-",n,"_",cov,"_2_3_10spotfilt.png"),width=1000,height=800)
    plotMDS(d,col=pal[factor(d$samples[,cov])], dim.plot = c(2,3),pch = 18)
    legend('bottomright', legend=levels(d$samples[,cov]), col=pal, cex=1, pch=18)
    dev.off()
  })
})
```

# clean up
```{r}
rm(ag,agc,agcd,agsc,clus,dgels,plts)
```

#### LCNM-SEGMENTATION-SUBCLUSTER DGELists -- only LC NM+/NM- frxns. ####
#### not done yet ####

## load and appened the subcluster labels to spe
```{r}
nmseg <- fread("processed-data/10_subclustering/04-LCsubclus-by-NM_otherclusts-by-LCorLCNMadjacency.txt")

nmseg <- DataFrame(nmseg,row.names=nmseg$rn)[colnames(lc3),c(3,4)]
colData(lc3) <- cbind(colData(lc3),nmseg)

rm(nmseg)
```

## # make pseudobulk object (dreamlet is v. fast at this)
```{r}
LCNMadj.ag <- dreamlet::aggregateToPseudoBulk(lc3[,lc3$NMseg_subclus!=""],sample_id="sample_id",cluster_id = "NMseg_subclus",BPPARAM = MulticoreParam(6))

# extract colData to feed into DGEList as the "samples" data
nm.agcd <- as.data.frame(colData(LCNMadj.ag))

# extract number of spots in a pseudobulk sample to filter by. this is in the inverse of the format we want, list entries are sample and subentries are cluster, so rearrange to cluster listed by sample
nmagn <- LCNMadj.ag@int_colData$n_cells
nmagn <- lapply(nmagn,function(x){as.data.frame(t(x))})
nmagn <- rbindlist(nmagn,idcol="samp")

## only retain LCNM+- clusters, and then only if at least 43 samples with 10+ spots of a type (of 85 total, i.e over half) are available after pseudobulking through dreamlet:  

LCNMadj.ag <- as.list(assays(LCNMadj.ag))
LCNMadj.ag <- LCNMadj.ag[c("LC.1_NMpos","LC.1_NMneg")]
LCNMadj.agc <- mapply(LCNMadj.ag,names(LCNMadj.ag),SIMPLIFY=FALSE,FUN=function(a,n){
  tenspot <- nmagn[get(n)>=10,samp]
  ncol(a[,colSums(a)>0][,tenspot])
})

LCNMadj.ag <- LCNMadj.ag[!(LCNMadj.agc<43)]

## drop clusters where there are 20+ samples, but under 18 samples for one of the sexes (i.e., < half of the samples from the smaller group--females w 35 sections total):
agsc <- mapply(LCNMadj.ag,n=names(LCNMadj.ag),SIMPLIFY=FALSE,FUN=function(a,n){
    cn <- colnames(a[,colSums(a)>0])
    cn <- cn[cn %in% nmagn[get(n)>=10,samp]] # 10+ spots included

    tmp <- as.data.table(nm.agcd[cn,])
    if(nrow(tmp[,.N,by="Sex"][N<18])>0){
      return(0)
    } else {return(1)}
})

LCNMadj.ag <- LCNMadj.ag[agsc==1]

rm(agscs,LCadj.agc,LCNMadj.agc)

# stratify by APOE4 positive / negative  
nm.agcd$e4 <- ifelse(nm.agcd$APOE %in% c("E4/ E4","E3/ E4"),yes="carrier",no="negative")

```

## make the DGELists per retained subcluster:
```{r}
## get rowData as a data.table from the SPE to append to the DGElist--namely so we can make MDS plots of the pseudobulks while excluding mitochondrial and sex chromosomal genes
tmprd <- as.data.table(rowData(lc3)) # rownames are the same as $gene_id

retdgels <- mapply(a=LCNMadj.ag,n=names(LCNMadj.ag),SIMPLIFY=FALSE,FUN=function(a,n){
    cn <- colnames(a[,colSums(a)>0]) # has counts
    
    tenspot <- cn[cn %in% nmagn[get(n)>=10,samp]] # and has 10+spots contributed
    X <- a[,tenspot]
    
    s <- copy(nm.agcd)[colnames(X),]
    if(length(unique(s$Sex))<2){return()} else {

        dge <- DGEList(counts=as.matrix(X),samples=s,genes=as.data.frame(tmprd,row.names = tmprd$rn)[rownames(X),])
        dge <- normLibSizes(dge,method = "TMM")
        
        ## find the smaller group for this cluster
        smolgrp <- table(dge$samples$Sex)[which.min(table(dge$samples$Sex))]
        
        ## for the smallest group, determine low expression cutoff
        rowmn.m <- as.numeric(rowMeans(cpm(dge[,(dge$samples$Sex==names(smolgrp)[1])],log=TRUE)))
        
        ## get histogram distribution with 10 bins at the low end of the logcpm space
        h <- hist(rowmn.m[which(rowmn.m>min(rowmn.m)&rowmn.m<(4+min(rowmn.m)))],breaks=10,plot=FALSE)
        ## get the minimum in this low logcpm range and its corresponding bin (x axis value, i.e. log cpm)
        minc <- h$breaks[which(h$counts==min(h$counts))]
        
        # if there's multiple points meeting this^ minimum, take the larger
        if(length(minc)>1){minc <- max(minc)}
        
        ## get the corresponding smallest-library count value for this minimum^
        smollib <- min(dge$samples[(dge$samples$Sex==names(smolgrp)[1]),"lib.size"])
        countthresh <- (minc^2)*(as.numeric(smollib)/10^6)
        
        ## and convert that smallest count to the logcpm value in said smallest library to use as a low-expression threshold
        cpmthresh <- log2(countthresh/(as.numeric(smollib)*10^-6))
        
        ## 2/14/25: make a plot to also return
        mnplt <- as.data.table(rowmn.m)
        setnames(mnplt,1,"V1")
        
        ## exclude genes with a mean logcpm under 0.1
        mnplt <- mnplt[V1>0.1]
        
        plt <- ggplot(mnplt,aes(x=V1))+
          geom_histogram(bins = 100)+
          guides(color="none")+
          ggtitle(paste0(n," genewise mean logCPM in\nlower n sex (",smolgrp,")"))
        
        ## only draw the threshold line if it's within the rowmn.m range
        if(cpmthresh>min(rowmn.m)&cpmthresh<max(rowmn.m)){
          plt <- plt + geom_vline(xintercept=cpmthresh,col="red")
        }
        
        ## filter for genes with at least thresh counts in at least (smaller sex group size) samples
        keep <- rowSums(cpm(dge,log=T)>cpmthresh) >= as.numeric(smolgrp)
        
        ## recalc lib factors after dropping low-expressed genes
        dge <- dge[keep,,keep.lib.sizes=FALSE]
        dge <- calcNormFactors(dge,method = "TMM")
        
        retlist <- list(dge,plt)
        names(retlist) <- c(n,"plt")
        return(retlist)
    }
    return(retdgels)
})
```

## get and arrange low-expression-cutoff plots
```{r}
plts <- lapply(retdgels,FUN=function(y){
    y$plt
})

pdf(paste0("plots/11_DEanalyses/01b-logCPM_cutoff_checks_LC-NM-subsets.pdf"),height=3.33*ceiling(length(plts)/3),width=10)
do.call("grid.arrange",c(plts,ncol=3))
dev.off()
```

## get the dgelists themselves to save
```{r}
retdgels <- lapply(retdgels,function(x){x[[1]]}) 

saveRDS(retdgels,"processed-data/11_DEanalyses/01b-DGElist_per_LCNM_subcluster_normFilt_10spotmin.RDS")
```

reprod info
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] parallelly_1.42.0           colorout_1.3-0.2           
 [3] gridExtra_2.3               edgeR_4.4.1                
 [5] dreamlet_1.4.1              variancePartition_1.36.2   
 [7] BiocParallel_1.40.0         limma_3.62.1               
 [9] ggplot2_3.5.1               SpatialExperiment_1.16.0   
[11] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0
[13] Biobase_2.66.0              GenomicRanges_1.58.0       
[15] GenomeInfoDb_1.42.1         IRanges_2.40.1             
[17] S4Vectors_0.44.0            BiocGenerics_0.52.0        
[19] MatrixGenerics_1.18.0       matrixStats_1.5.0          
[21] data.table_1.16.4           rlang_1.1.4                

loaded via a namespace (and not attached):
  [1] splines_4.4.2             bitops_1.0-9              tibble_3.2.1             
  [4] graph_1.84.0              XML_3.99-0.17             lifecycle_1.0.4          
  [7] Rdpack_2.6.2              mixsqp_0.3-54             rprojroot_2.0.4          
 [10] lattice_0.22-6            MASS_7.3-61               backports_1.5.0          
 [13] magrittr_2.0.3            metafor_4.6-0             rmarkdown_2.29           
 [16] yaml_2.3.10               DBI_1.2.3                 minqa_1.2.8              
 [19] abind_1.4-8               zlibbioc_1.52.0           EnvStats_3.0.0           
 [22] purrr_1.0.2               rmeta_3.0                 msigdbr_7.5.1            
 [25] RCurl_1.98-1.16           GenomeInfoDbData_1.2.13   ggrepel_0.9.6            
 [28] pbkrtest_0.5.3            irlba_2.3.5.1             annotate_1.84.0          
 [31] DelayedMatrixStats_1.28.0 codetools_0.2-20          DelayedArray_0.32.0      
 [34] tidyselect_1.2.1          farver_2.1.2              UCSC.utils_1.2.0         
 [37] lme4_1.1-35.5             mathjaxr_1.6-0            jsonlite_1.8.9           
 [40] iterators_1.0.14          tools_4.4.2               progress_1.2.3           
 [43] Rcpp_1.0.14               zenith_1.8.0              glue_1.8.0               
 [46] SparseArray_1.6.0         xfun_0.50                 here_1.0.1               
 [49] dplyr_1.1.4               withr_3.0.2               numDeriv_2016.8-1.1      
 [52] fastmap_1.2.0             boot_1.3-31               caTools_1.18.3           
 [55] digest_0.6.37             truncnorm_1.0-9           R6_2.5.1                 
 [58] colorspace_2.1-1          scattermore_1.2           gtools_3.9.5             
 [61] RSQLite_2.3.9             RhpcBLASctl_0.23-42       tidyr_1.3.1              
 [64] generics_0.1.3            corpcor_1.6.10            prettyunits_1.2.0        
 [67] httr_1.4.7                S4Arrays_1.6.0            pkgconfig_2.0.3          
 [70] gtable_0.3.6              blob_1.2.4                XVector_0.46.0           
 [73] remaCor_0.0.18            htmltools_0.5.8.1         GSEABase_1.68.0          
 [76] scales_1.3.0              png_0.1-8                 fANCOVA_0.6-1            
 [79] ashr_2.2-63               knitr_1.49                rstudioapi_0.17.1        
 [82] reshape2_1.4.4            rjson_0.2.23              nlme_3.1-166             
 [85] nloptr_2.1.1              cachem_1.1.0              stringr_1.5.1            
 [88] KernSmooth_2.23-24        vipor_0.4.7               metadat_1.2-0            
 [91] RcppZiggurat_0.1.6        AnnotationDbi_1.68.0      pillar_1.10.1            
 [94] grid_4.4.2                vctrs_0.6.5               gplots_3.2.0             
 [97] mashr_0.2.79              xtable_1.8-4              beeswarm_0.4.0           
[100] Rgraphviz_2.50.0          evaluate_1.0.3            KEGGgraph_1.66.0         
[103] invgamma_1.1              magick_2.8.5              mvtnorm_1.3-2            
[106] cli_3.6.3                 locfit_1.5-9.10           compiler_4.4.2           
[109] crayon_1.5.3              SQUAREM_2021.1            labeling_0.4.3           
[112] plyr_1.8.9                ggbeeswarm_0.7.2          stringi_1.8.4            
[115] assertthat_0.2.1          babelgene_22.9            lmerTest_3.1-3           
[118] munsell_0.5.1             Biostrings_2.74.0         aod_1.3.3                
[121] Matrix_1.7-1              hms_1.1.3                 sparseMatrixStats_1.18.0 
[124] bit64_4.5.2               KEGGREST_1.46.0           statmod_1.5.0            
[127] rbibutils_2.3             Rfast_2.1.0               broom_1.0.7              
[130] memoise_2.0.1             RcppParallel_5.1.9        bit_4.5.0.1              
[133] EnrichmentBrowser_2.36.0 
> sessioninfo::session_info()
â”€ Session info â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
 setting  value
 version  R version 4.4.2 (2024-10-31)
 os       macOS Sonoma 14.7.1
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-02-26
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

â”€ Packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
 !  package              * version    date (UTC) lib source
    abind                  1.4-8      2024-09-12 [1] CRAN (R 4.4.1)
    annotate               1.84.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    AnnotationDbi          1.68.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    aod                    1.3.3      2023-12-13 [1] CRAN (R 4.4.0)
    ashr                   2.2-63     2023-08-21 [1] CRAN (R 4.4.0)
    assertthat             0.2.1      2019-03-21 [1] CRAN (R 4.4.0)
    babelgene              22.9       2022-09-29 [1] CRAN (R 4.4.0)
    backports              1.5.0      2024-05-23 [1] CRAN (R 4.4.0)
    beeswarm               0.4.0      2021-06-01 [1] CRAN (R 4.4.0)
    Biobase              * 2.66.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocParallel         * 1.40.0     2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    Biostrings             2.74.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    bit                    4.5.0.1    2024-12-03 [1] CRAN (R 4.4.1)
    bit64                  4.5.2      2024-09-22 [1] CRAN (R 4.4.1)
    bitops                 1.0-9      2024-10-03 [1] CRAN (R 4.4.1)
    blob                   1.2.4      2023-03-17 [1] CRAN (R 4.4.0)
    boot                   1.3-31     2024-08-28 [1] CRAN (R 4.4.2)
    broom                  1.0.7      2024-09-26 [1] CRAN (R 4.4.1)
    cachem                 1.1.0      2024-05-16 [1] CRAN (R 4.4.0)
    caTools                1.18.3     2024-09-04 [1] CRAN (R 4.4.1)
 P  cli                    3.6.3      2024-06-21 [2] CRAN (R 4.4.0)
    codetools              0.2-20     2024-03-31 [1] CRAN (R 4.4.2)
    colorout             * 1.3-0.2    2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    colorspace             2.1-1      2024-07-26 [1] CRAN (R 4.4.0)
    corpcor                1.6.10     2021-09-16 [1] CRAN (R 4.4.0)
    crayon                 1.5.3      2024-06-20 [1] CRAN (R 4.4.0)
    data.table           * 1.16.4     2024-12-06 [1] CRAN (R 4.4.1)
    DBI                    1.2.3      2024-06-02 [1] CRAN (R 4.4.0)
    DelayedArray           0.32.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    DelayedMatrixStats     1.28.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    digest                 0.6.37     2024-08-19 [1] CRAN (R 4.4.1)
    dplyr                  1.1.4      2023-11-17 [1] CRAN (R 4.4.0)
    dreamlet             * 1.4.1      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    edgeR                * 4.4.1      2024-12-02 [1] Bioconductor 3.20 (R 4.4.2)
    EnrichmentBrowser      2.36.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    EnvStats               3.0.0      2024-08-24 [1] CRAN (R 4.4.1)
    evaluate               1.0.3      2025-01-10 [1] CRAN (R 4.4.1)
    fANCOVA                0.6-1      2020-11-13 [1] CRAN (R 4.4.0)
    farver                 2.1.2      2024-05-13 [1] CRAN (R 4.4.0)
    fastmap                1.2.0      2024-05-15 [1] CRAN (R 4.4.0)
    generics               0.1.3      2022-07-05 [1] CRAN (R 4.4.0)
    GenomeInfoDb         * 1.42.1     2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13     2024-12-12 [1] Bioconductor
    GenomicRanges        * 1.58.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ggbeeswarm             0.7.2      2023-04-29 [1] CRAN (R 4.4.0)
    ggplot2              * 3.5.1      2024-04-23 [1] CRAN (R 4.4.0)
    ggrepel                0.9.6      2024-09-07 [1] CRAN (R 4.4.1)
    glue                   1.8.0      2024-09-30 [1] CRAN (R 4.4.1)
    gplots                 3.2.0      2024-10-05 [1] CRAN (R 4.4.1)
    graph                  1.84.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    gridExtra            * 2.3        2017-09-09 [1] CRAN (R 4.4.0)
    GSEABase               1.68.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    gtable                 0.3.6      2024-10-25 [1] CRAN (R 4.4.1)
    gtools                 3.9.5      2023-11-20 [1] CRAN (R 4.4.0)
    here                   1.0.1      2020-12-13 [1] CRAN (R 4.4.0)
    hms                    1.1.3      2023-03-21 [1] CRAN (R 4.4.0)
    htmltools              0.5.8.1    2024-04-04 [1] CRAN (R 4.4.0)
    httr                   1.4.7      2023-08-15 [1] CRAN (R 4.4.0)
    invgamma               1.1        2017-05-07 [1] CRAN (R 4.4.0)
    IRanges              * 2.40.1     2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    irlba                  2.3.5.1    2022-10-03 [1] CRAN (R 4.4.0)
    iterators              1.0.14     2022-02-05 [1] CRAN (R 4.4.0)
    jsonlite               1.8.9      2024-09-20 [1] CRAN (R 4.4.1)
    KEGGgraph              1.66.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    KEGGREST               1.46.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    KernSmooth             2.23-24    2024-05-17 [1] CRAN (R 4.4.2)
    knitr                  1.49       2024-11-08 [1] CRAN (R 4.4.1)
    labeling               0.4.3      2023-08-29 [1] CRAN (R 4.4.0)
    lattice                0.22-6     2024-03-20 [1] CRAN (R 4.4.2)
    lifecycle              1.0.4      2023-11-07 [1] CRAN (R 4.4.0)
    limma                * 3.62.1     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    lme4                   1.1-35.5   2024-07-03 [1] CRAN (R 4.4.1)
    lmerTest               3.1-3      2020-10-23 [1] CRAN (R 4.4.0)
    locfit                 1.5-9.10   2024-06-24 [1] CRAN (R 4.4.0)
    magick                 2.8.5      2024-09-20 [1] CRAN (R 4.4.1)
    magrittr               2.0.3      2022-03-30 [1] CRAN (R 4.4.0)
    mashr                  0.2.79     2023-10-18 [1] CRAN (R 4.4.0)
    MASS                   7.3-61     2024-06-13 [1] CRAN (R 4.4.2)
    mathjaxr               1.6-0      2022-02-28 [1] CRAN (R 4.4.0)
    Matrix                 1.7-1      2024-10-18 [1] CRAN (R 4.4.2)
    MatrixGenerics       * 1.18.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    matrixStats          * 1.5.0      2025-01-07 [1] CRAN (R 4.4.1)
    memoise                2.0.1      2021-11-26 [1] CRAN (R 4.4.0)
    metadat                1.2-0      2022-04-06 [1] CRAN (R 4.4.0)
    metafor                4.6-0      2024-03-28 [1] CRAN (R 4.4.0)
    minqa                  1.2.8      2024-08-17 [1] CRAN (R 4.4.0)
    mixsqp                 0.3-54     2023-12-20 [1] CRAN (R 4.4.0)
    msigdbr                7.5.1      2022-03-30 [1] CRAN (R 4.4.0)
    munsell                0.5.1      2024-04-01 [1] CRAN (R 4.4.0)
    mvtnorm                1.3-2      2024-11-04 [1] CRAN (R 4.4.1)
    nlme                   3.1-166    2024-08-14 [1] CRAN (R 4.4.2)
    nloptr                 2.1.1      2024-06-25 [1] CRAN (R 4.4.0)
    numDeriv               2016.8-1.1 2019-06-06 [1] CRAN (R 4.4.0)
    parallelly           * 1.42.0     2025-01-30 [1] CRAN (R 4.4.1)
    pbkrtest               0.5.3      2024-06-26 [1] CRAN (R 4.4.0)
    pillar                 1.10.1     2025-01-07 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3      2019-09-22 [1] CRAN (R 4.4.0)
    plyr                   1.8.9      2023-10-02 [1] CRAN (R 4.4.0)
    png                    0.1-8      2022-11-29 [1] CRAN (R 4.4.0)
    prettyunits            1.2.0      2023-09-24 [1] CRAN (R 4.4.0)
    progress               1.2.3      2023-12-06 [1] CRAN (R 4.4.0)
    purrr                  1.0.2      2023-08-10 [1] CRAN (R 4.4.0)
    R6                     2.5.1      2021-08-19 [1] CRAN (R 4.4.0)
    rbibutils              2.3        2024-10-04 [1] CRAN (R 4.4.1)
    Rcpp                   1.0.14     2025-01-12 [1] CRAN (R 4.4.1)
    RcppParallel           5.1.9      2024-08-19 [1] CRAN (R 4.4.1)
    RcppZiggurat           0.1.6      2020-10-20 [1] CRAN (R 4.4.0)
    RCurl                  1.98-1.16  2024-07-11 [1] CRAN (R 4.4.0)
    Rdpack                 2.6.2      2024-11-15 [1] CRAN (R 4.4.1)
    remaCor                0.0.18     2024-02-08 [1] CRAN (R 4.4.0)
    reshape2               1.4.4      2020-04-09 [1] CRAN (R 4.4.0)
    Rfast                  2.1.0      2023-11-09 [1] CRAN (R 4.4.0)
    Rgraphviz              2.50.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    RhpcBLASctl            0.23-42    2023-02-11 [1] CRAN (R 4.4.0)
    rjson                  0.2.23     2024-09-16 [1] CRAN (R 4.4.1)
 VP rlang                * 1.1.4      2025-01-17 [2] CRAN (R 4.4.1) (on disk 1.1.5)
    rmarkdown              2.29       2024-11-04 [1] CRAN (R 4.4.1)
    rmeta                  3.0        2018-03-20 [1] CRAN (R 4.4.0)
    rprojroot              2.0.4      2023-11-05 [1] CRAN (R 4.4.0)
    RSQLite                2.3.9      2024-12-03 [1] CRAN (R 4.4.1)
    rstudioapi             0.17.1     2024-10-22 [1] CRAN (R 4.4.1)
    S4Arrays               1.6.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.3.0      2023-11-28 [1] CRAN (R 4.4.0)
    scattermore            1.2        2023-06-12 [1] CRAN (R 4.4.0)
    sessioninfo            1.2.2      2021-12-06 [1] CRAN (R 4.4.0)
    SingleCellExperiment * 1.28.1     2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    SparseArray            1.6.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    sparseMatrixStats      1.18.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SpatialExperiment    * 1.16.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SQUAREM                2021.1     2021-01-13 [1] CRAN (R 4.4.0)
    statmod                1.5.0      2023-01-06 [1] CRAN (R 4.4.0)
    stringi                1.8.4      2024-05-06 [1] CRAN (R 4.4.0)
    stringr                1.5.1      2023-11-14 [1] CRAN (R 4.4.0)
    SummarizedExperiment * 1.36.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    tibble                 3.2.1      2023-03-20 [1] CRAN (R 4.4.0)
    tidyr                  1.3.1      2024-01-24 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1      2024-03-11 [1] CRAN (R 4.4.0)
    truncnorm              1.0-9      2023-03-20 [1] CRAN (R 4.4.0)
    UCSC.utils             1.2.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    variancePartition    * 1.36.2     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    vctrs                  0.6.5      2023-12-01 [1] CRAN (R 4.4.0)
    vipor                  0.4.7      2023-12-18 [1] CRAN (R 4.4.0)
    withr                  3.0.2      2024-10-28 [1] CRAN (R 4.4.1)
    xfun                   0.50       2025-01-07 [1] CRAN (R 4.4.1)
    XML                    3.99-0.17  2024-06-25 [1] CRAN (R 4.3.3)
    xtable                 1.8-4      2019-04-21 [1] CRAN (R 4.4.0)
    XVector                0.46.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    yaml                   2.3.10     2024-07-26 [1] CRAN (R 4.4.0)
    zenith                 1.8.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    zlibbioc               1.52.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 V â”€â”€ Loaded and on-disk version mismatch.
 P â”€â”€ Loaded and on-disk path mismatch.

â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
