---
title: "0203a-Boxplot_top_APOE4_allelecount_DEGs"
author: "Bernie Mulvey"
date: "2025-02-11"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(ggplot2)
library(ggtext)
library(gridExtra)
library(limma)
library(edgeR)
# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")
library(parallel)


## ggplot defaults
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 9), axis.title.x = element_text(size = 8), axis.text.y = element_text(size = 8), axis.title.y = element_text(size =9), plot.title = element_markdown(size = 10,hjust=0.5), strip.text = element_text(size=11), legend.text = element_text(size=8), legend.title = element_text(size=9,hjust=0.5)))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

load the tabulations of number of DE models wherein a gene hits as DE for APOE4 allele count # only
```{r}
# number of signif models tabulations
dgehittabs <- list()
dgehittabs[[1]] <- fread("processed-data/11_DEanalyses/03b-E4alleleCount_DE_FDRunder0.2_byGene_byNmodelsOf4max_byCluster.txt")
dgehittabs[[2]] <- fread("processed-data/11_DEanalyses/03c-E4alleleCount_DE_unadjP.05_byGene_byNmodelsOf4max_byCluster.txt")

names(dgehittabs) <- c("E4count_FDRpt2","E4count_nom")

### DGEList objects as used for the DE testing ###

dgels <- readRDS("processed-data/11_DEanalyses/01-DGElist_per_25HDG75SVGlouv1_cluster_normFilt.RDS")

# append global genetic ancestry proportion estimates, which we used as a covar in some models (FLARE, 3-component CEU ancestry--the other two being YRI and CHB). this source file is intentionally not included in in github but is on JHPCE. just in case we want to make plots by ancestry prop

flares <- fread("raw-data/ancestryAndHaplotypes/FLARE/global_ancestry_estimates.txt")

dgels <- lapply(dgels,FUN=function(d){
    tmpsamp <- as.data.table(d$samples,keep.rownames=T)
    # fix Br6119(re-dis) to be Br6119
    tmpsamp[brnum=="Br6119(re-dis)",brnum:="Br6119"]
    
    ## convert APOE genotype to E4 copy number
    d$samples$e4copies <- 0
    d$samples$e4copies[d$samples$APOE=="E4/ E4"] <- 2
    d$samples$e4copies[d$samples$APOE=="E3/ E4"] <- 1 # all hets are E4/E3
    
    # append ancestry estimates
    tmpsamp <- merge.data.table(tmpsamp,flares,by.x="brnum",by.y="ID")
    tmpsamp <- as.data.frame(tmpsamp,row.names=tmpsamp$rn)[rownames(d$samples),]
    tmpsamp$rn <- NULL
    d$samples <- tmpsamp
    return(d)
})
rm(flares)
# convert to cpm, melt to long form for plotting, and append metadata vars of interest

dgels.logcpm <- lapply(dgels,FUN = function(x){
    # get logcpm
    y <- as.data.table(cpm(x,log=T),keep.rownames=T)
    setnames(y,"rn","gene_id")
    
    # append gene symbols and chromosome (so we can e.g. plot sex DE without skewing by X/Y genes)
    y <- merge.data.table(y,as.data.table(x$genes)[,.(gene_id,gene_name,chr)],by="gene_id")
    
    # long format
    z <- melt(y,id.vars=c("gene_id","gene_name","chr"))
    setnames(z,c("variable","value"),c("sample_id","logCPM"))
    
    # grab metadata vars of interest
    s <- as.data.table(x$samples,keep.rownames=T)[,.(rn,Sex,Age,CEU,Rin,e4)]
    setnames(s,"rn","sample_id")
    
    # merge
    z <- merge.data.table(z,s,by="sample_id")
    
    # spit back
    return(z)
})

# clean up
rm(dgels)
gc(full=T)
```

### now, get genes that we actually want to plot from this table; recurrent FDR 0.2 hits for APOE4 DE first. (max possible for a cluster is all 4 DE models used).
### if we have too many genes, drop certain clusters: X9, X7
X9 is best marked by hemoglobins so that's probably blood etc. vascular could be interesting; blood less so
X7 has oligo markers -- WM is glial type of least interest here for now
```{r}
# nrow(unique(dgehittabs$E4_FDRpt2[Nsigmodels_forCluster>=3,.(gene_name,louvCluster)]))
# nrow(unique(dgehittabs$E4_FDRpt2[Nsigmodels_forCluster==4,.(gene_name,louvCluster)]))
# 52 gene-cluster pairs that were FDR<0.2 in 3+ models for e4 de; 26 that were FDR<0.2 in all 4 models
apofg <- unique(dgehittabs$E4count_FDRpt2[Nsigmodels_forCluster>=3,.(gene_name,louvCluster,Nsigmodels_forCluster)])

setDTthreads(1,restore_after_fork=FALSE)

apoplot <- mcmapply(ct=apofg$louvCluster,g=apofg$gene_name,nsig=apofg$Nsigmodels_forCluster,SIMPLIFY=FALSE,mc.cores=8,FUN=function(ct,g,nsig){
    # get the logcpm data for this gene
    tmp <- copy(dgels.logcpm[[ct]][gene_name==g])
    
    # make the boxplot
    p <- ggplot(tmp,aes(x=e4,y=logCPM,col=e4))+
        geom_boxplot(outliers = FALSE)+
        geom_point()+
        guides(color="none")+
        ggtitle(paste0(g," in ", ct,"<br>FDR<0.2 in ",nsig,"/4 models"))
    return(p)
})

pdf("plots/11_DEanalyses/02a_gene_boxplots/E4_DE_3plusmodels_FDR0.2_boxplots.pdf",height=length(apoplot)*4,width=9)
do.call("grid.arrange",c(apoplot,list(ncol=2)))
dev.off()


```

### next, do the same for sex DE tests.
```{r}
# nrow(unique(dgehittabs$sex_FDRpt2[Nsigmodels_forCluster>=3,.(gene_name,louvCluster)]))
# 2563 gene cluster pairs-yowza
# nrow(unique(dgehittabs$sex_FDRpt2[Nsigmodels_forCluster==4,.(gene_name,louvCluster)]))
# 1446 gene cluster pairs-still yowza
# what about excluding X7, X9
# nrow(unique(dgehittabs$sex_FDRpt2[Nsigmodels_forCluster==4 & !(louvCluster %in% c("X7","X9")),.(gene_name,louvCluster)]))
# 1389. still a lot. we'll start with boxplots here

sexfg <- unique(dgehittabs$sex_FDRpt2[Nsigmodels_forCluster==4 & !(louvCluster %in% c("X7","X9")),.(gene_name,louvCluster,Nsigmodels_forCluster)])

setDTthreads(1,restore_after_fork=FALSE)

## to make more manageable plot grids, split out by cluster for this one
sexplots <- lapply(unique(sexfg$louvCluster),function(ct){
    clustplots <- mcmapply(ct=ct,g=sexfg[louvCluster==ct,gene_name],SIMPLIFY=FALSE,mc.cores=8,FUN=function(ct,g){
    # get the logcpm data for this gene
        tmp <- copy(dgels.logcpm[[ct]][gene_name==g])
        
        # make the boxplot
        p <- ggplot(tmp,aes(x=Sex,y=logCPM,col=Sex))+
            geom_boxplot(outliers = FALSE)+
            geom_point()+
            guides(color="none")+
            ggtitle(paste0(g," in ", ct,"<br>FDR<0.2 in 4/4 models"))
        return(p)
    })
    return(clustplots)
})
names(sexplots) <- unique(sexfg$louvCluster)

# save a plot grid per louvclust
i<-1
for (i in c(1:length(sexplots))){
    pdf(paste0("plots/11_DEanalyses/02a_gene_boxplots/",names(sexplots)[i],"_Sex_DE_FDR0.2_boxplots.pdf"),height=length(sexplots[[i]])*4,width=9)
    do.call("grid.arrange",c(sexplots[[i]],list(ncol=2)))
    dev.off()
}
```

Now, inspect these boxplots offline to prioritize genes to further examine by spatial expression plots. back after these messages.
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] BiocParallel_1.40.0 parallelly_1.42.0   colorout_1.3-0.2    edgeR_4.4.1        
 [5] limma_3.62.1        gridExtra_2.3       ggtext_0.1.2        ggplot2_3.5.1      
 [9] data.table_1.16.4   rlang_1.1.4        

loaded via a namespace (and not attached):
 [1] generics_0.1.3    xml2_1.3.6        stringi_1.8.4     lattice_0.22-6   
 [5] digest_0.6.37     magrittr_2.0.3    evaluate_1.0.3    grid_4.4.2       
 [9] fastmap_1.2.0     rprojroot_2.0.4   sessioninfo_1.2.2 scales_1.3.0     
[13] codetools_0.2-20  cli_3.6.3         munsell_0.5.1     commonmark_1.9.2 
[17] withr_3.0.2       yaml_2.3.10       tools_4.4.2       dplyr_1.1.4      
[21] colorspace_2.1-1  locfit_1.5-9.10   here_1.0.1        vctrs_0.6.5      
[25] R6_2.5.1          lifecycle_1.0.4   stringr_1.5.1     pkgconfig_2.0.3  
[29] pillar_1.10.1     gtable_0.3.6      glue_1.8.0        Rcpp_1.0.14      
[33] statmod_1.5.0     xfun_0.50         tibble_3.2.1      tidyselect_1.2.1 
[37] rstudioapi_0.17.1 knitr_1.49        farver_2.1.2      htmltools_0.5.8.1
[41] rmarkdown_2.29    labeling_0.4.3    compiler_4.4.2    markdown_1.13    
[45] gridtext_0.1.5   
> sessioninfo::session_info()
─ Session info ─────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.2 (2024-10-31)
 os       macOS Sonoma 14.7.1
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-02-11
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ─────────────────────────────────────────────────────────────────────────
 !  package      * version  date (UTC) lib source
    BiocParallel * 1.40.0   2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
 P  cli            3.6.3    2024-06-21 [2] CRAN (R 4.4.0)
    codetools      0.2-20   2024-03-31 [1] CRAN (R 4.4.2)
    colorout     * 1.3-0.2  2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    colorspace     2.1-1    2024-07-26 [1] CRAN (R 4.4.0)
    commonmark     1.9.2    2024-10-04 [1] CRAN (R 4.4.1)
    data.table   * 1.16.4   2024-12-06 [1] CRAN (R 4.4.1)
    digest         0.6.37   2024-08-19 [1] CRAN (R 4.4.1)
    dplyr          1.1.4    2023-11-17 [1] CRAN (R 4.4.0)
    edgeR        * 4.4.1    2024-12-02 [1] Bioconductor 3.20 (R 4.4.2)
    evaluate       1.0.3    2025-01-10 [1] CRAN (R 4.4.1)
    farver         2.1.2    2024-05-13 [1] CRAN (R 4.4.0)
    fastmap        1.2.0    2024-05-15 [1] CRAN (R 4.4.0)
    generics       0.1.3    2022-07-05 [1] CRAN (R 4.4.0)
    ggplot2      * 3.5.1    2024-04-23 [1] CRAN (R 4.4.0)
    ggtext       * 0.1.2    2022-09-16 [1] CRAN (R 4.4.0)
    glue           1.8.0    2024-09-30 [1] CRAN (R 4.4.1)
    gridExtra    * 2.3      2017-09-09 [1] CRAN (R 4.4.0)
    gridtext       0.1.5    2022-09-16 [1] CRAN (R 4.4.0)
    gtable         0.3.6    2024-10-25 [1] CRAN (R 4.4.1)
    here           1.0.1    2020-12-13 [1] CRAN (R 4.4.0)
    htmltools      0.5.8.1  2024-04-04 [1] CRAN (R 4.4.0)
    knitr          1.49     2024-11-08 [1] CRAN (R 4.4.1)
    labeling       0.4.3    2023-08-29 [1] CRAN (R 4.4.0)
    lattice        0.22-6   2024-03-20 [1] CRAN (R 4.4.2)
    lifecycle      1.0.4    2023-11-07 [1] CRAN (R 4.4.0)
    limma        * 3.62.1   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    locfit         1.5-9.10 2024-06-24 [1] CRAN (R 4.4.0)
    magrittr       2.0.3    2022-03-30 [1] CRAN (R 4.4.0)
    markdown       1.13     2024-06-04 [1] CRAN (R 4.4.0)
    munsell        0.5.1    2024-04-01 [1] CRAN (R 4.4.0)
    parallelly   * 1.42.0   2025-01-30 [1] CRAN (R 4.4.1)
    pillar         1.10.1   2025-01-07 [1] CRAN (R 4.4.1)
    pkgconfig      2.0.3    2019-09-22 [1] CRAN (R 4.4.0)
    R6             2.5.1    2021-08-19 [1] CRAN (R 4.4.0)
    Rcpp           1.0.14   2025-01-12 [1] CRAN (R 4.4.1)
 VP rlang        * 1.1.4    2025-01-17 [2] CRAN (R 4.4.1) (on disk 1.1.5)
    rmarkdown      2.29     2024-11-04 [1] CRAN (R 4.4.1)
    rprojroot      2.0.4    2023-11-05 [1] CRAN (R 4.4.0)
    rstudioapi     0.17.1   2024-10-22 [1] CRAN (R 4.4.1)
    scales         1.3.0    2023-11-28 [1] CRAN (R 4.4.0)
    sessioninfo    1.2.2    2021-12-06 [1] CRAN (R 4.4.0)
    statmod        1.5.0    2023-01-06 [1] CRAN (R 4.4.0)
    stringi        1.8.4    2024-05-06 [1] CRAN (R 4.4.0)
    stringr        1.5.1    2023-11-14 [1] CRAN (R 4.4.0)
    tibble         3.2.1    2023-03-20 [1] CRAN (R 4.4.0)
    tidyselect     1.2.1    2024-03-11 [1] CRAN (R 4.4.0)
    vctrs          0.6.5    2023-12-01 [1] CRAN (R 4.4.0)
    withr          3.0.2    2024-10-28 [1] CRAN (R 4.4.1)
    xfun           0.50     2025-01-07 [1] CRAN (R 4.4.1)
    xml2           1.3.6    2023-12-04 [1] CRAN (R 4.4.0)
    yaml           2.3.10   2024-07-26 [1] CRAN (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

────────────────────────────────────────────────────────────────────────────────────
