---
title: "06c-Examine_LuskinBruchas-pixelseq_for_LCsubclusts"
author: "Bernie Mulvey"
date: "2025-03-03"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(ggplot2)
library(ggtext)
library(gridExtra)
library(Seurat)
library(dplyr)
library(SpatialExperiment)
library(spatialLIBD)
library(orthogene)
# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")


## ggplot defaults
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 9), axis.title.x = element_text(size = 8), axis.text.y = element_text(size = 8), axis.title.y = element_text(size =9), plot.title = element_markdown(size = 11,hjust=0.5), strip.text = element_text(size=11), legend.text = element_text(size=8), legend.title = element_text(size=9,hjust=0.5)))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

### Data stored on GEO, GSE244659; download the two RDS files and untar, ungzip.

```{r}
pxm <- readRDS("~/Desktop/GSE244659_RAW/GSM7823492_male.combined.sct.rds")
pxf <- readRDS("~/Desktop/GSE244659_RAW/GSM7823491_female.combined.sct.rds")
```


define function to convert Seurat spatial object to SpatialExperiment object: see https://github.com/drighelli/SpatialExperiment/issues/115
```{r}
seurat_to_spe <- function(seu, sample_id, img_id) {
    ## Convert to SCE with the spatial assay slot carried over (there's also integrated and single cell data that we aren't doing anything with, but only in the male dataset which fucks everything up downstream)
    sce <- Seurat::as.SingleCellExperiment(seu,assay="Spatial")
    
    ## Extract spatial coordinates
    spatialCoords <- as.matrix(
        seu@images[[img_id]]@coordinates[, c("x", "y")])
    
    ## Extract and process image data

    # imgData <- DataFrame(
    #     sample_id = sample_id,
    #     image_id = img_id,
    #     data = I(list(img)),
    #     scaleFactor = seu@images[[img_id]]@scale.factors$lowres)
    
    # Convert to SpatialExperiment
    spe <- SpatialExperiment(
        assays = assays(sce),
        rowData = rowData(sce),
        colData = colData(sce),
        metadata = metadata(sce),
        reducedDims = reducedDims(sce),
        altExps = altExps(sce),
        sample_id = sample_id,
        spatialCoords = spatialCoords,
        # imgData = imgData
    )
    # indicate all spots are on the tissue
    spe$in_tissue <- 1
    spe$sample_id <- sample_id
    # Return Spatial Experiment object
    spe
}
```

### male seu to spe
```{r}
sampid <- unique(pxm@meta.data$orig.ident)
imgn <- unique(names(pxm@images)) ## assuming these are in the same order as sample names occur

mspes <- mapply(s=sampid,i=imgn,SIMPLIFY=FALSE,FUN=function(s,i){
  sampspe <- seurat_to_spe(pxm[,pxm$orig.ident==s],sample_id = s,img_id = i)
  return(sampspe)
})

mspe <- Reduce(cbind,mspes)

rm(mspes,imgn,sampid,pxm)
gc(full=T)

### same now for fem
sampid <- unique(pxf$orig.ident)
imgn <- unique(names(pxf@images)) ## assuming these are in the same order as sample names occur

fspes <- mapply(s=sampid,i=imgn,SIMPLIFY=FALSE,FUN=function(s,i){
  sampspe <- seurat_to_spe(pxf[,pxf$orig.ident==s],sample_id = s,img_id = i)
  return(sampspe)
})

fspe <- Reduce(cbind,fspes)

rm(fspes,imgn,sampid,pxf)
gc(full=T)
```

combining will be tricky -- rownames aren't matched between the male and female data. let's just get the union of genes and subset to those for joining
```{r}
keepg <- rownames(fspe)[rownames(fspe) %in% rownames(mspe)]
stopifnot(sum(rownames(fspe) %in% keepg)==sum(rownames(mspe) %in% keepg))

mspe <- mspe[keepg,]
fspe <- fspe[keepg,]

### something else doesn't jive now -- something about unmatched colnames. and indeed coldata for male has addnl columns
keepcdn <- colnames(colData(mspe))[colnames(colData(mspe)) %in% colnames(colData(fspe))]

colData(mspe) <- colData(mspe)[,keepcdn]
colData(fspe) <- colData(fspe)[,keepcdn]

## ugh FFS now binding gives error "objects to bind must have the same number of assays"
assayNames(mspe)
assayNames(fspe)

## MOTHER FUCKER failed to rbind column 'reducedDims' across DataFrame objects: the DFrame objects to combine must have the same column names

reducedDims(mspe) <- list()
reducedDims(fspe) <- list()

allspe <- cbind(mspe,fspe)

## THERE we go.
rm(mspe,fspe,keepcdn,keepg)
gc(full=T)

## save
saveRDS(allspe,"processed-data/Xenium_prioritization/06c-Luskin_pixelseq_spe.RDS")
```


```{r}
View(as.data.table(colData(allspe)))

allspe$ident <- paste0("X",allspe$ident)

## add some rowdata
rowData(allspe)[,1] <- rownames(allspe)
rowData(allspe)[,2] <- rownames(allspe)

colnames(rowData(allspe)) <- c("gene_name","fakeens")

## there are just numerical clusters, so we'll need to figure out which are probable NE neurons based on markers.
mks <- registration_wrapper(allspe,var_registration = "ident",var_sample_id = "sample_id",gene_ensembl = "fakeens",gene_name = "gene_name",min_ncells = 25)

mks2 <- as.data.table(mks$enrichment)
mks2[,ensembl:=NULL]
mks2 <- melt.data.table(mks2,id.vars="gene")

mks2[,clus:=gsub(variable,pattern="^.*_(X.*)$",replacement="\\1")]
mks2[,stat:=gsub(variable,pattern="(.*)_X.*$",replacement="\\1")]

mks2[,variable:=NA]

mks3 <- dcast(mks2,gene+clus~stat,value.var="value")
mks3 <- mks3[fdr<0.05&logFC>0]
```

X3 and X21 are both signif DBH enriched (4+ logFC); get their pairwise mks
```{r}
mks.sublc <- as.data.table(mks$pairwise)
keepn <- c("gene",grep(names(mks.sublc),pattern="X3-X21|X21-X3",value=T))
mks.sublc <- mks.sublc[,..keepn]

## comparison is X21 vs X3
mks.sublc[,comparison:="X21.v.X3"]
setnames(mks.sublc,gsub(names(mks.sublc),pattern="_X21-X3",replacement=""))

mks.sublc <- mks.sublc[fdr<0.05]
## are any of those genes markers of one but not the other?
mks.sublc <- mks.sublc[!(gene %in% mks3[clus %in% c("X3","X21"),gene])]
## yes, 9 of them, all X3 enriched over X21. X3 also had top comparative markers here like ALDH1A1, so some concern these could be mixed w astrocyte, but:
write.table(mks.sublc$gene,"processed-data/Xenium_prioritization/06c-Luskin_pixelseq_DBH-X3_butnot_DBH-X21_mks.txt",quote=F,row.names=F,col.names=F)

rm(allspe)
gc(full=T)
```

## now load in LFF LC and tabulate prop expressing per cluster as before

```{r}
## load SPE
lc3 <- readRDS("processed-data/06-countsOnly_noImgData_QCed_SPE_split_to_tissSections.RDS")

## calculate log counts
lc3 <- scater::computeLibraryFactors(lc3)
lc3 <- scater::logNormCounts(lc3)

## read in clustering and annots
louvs <- readRDS("processed-data/07_featureSelection_dimred_harmony_clustering/04-Initial_louvLeid_res0.5-1-2_clusterings.RDS")
louvs <- louvs$HARMONYlmbna_HDG_SVG_2575[,.(rn,snnHARMONYlmbna_HDG_SVG_2575_louv_res1)]
setnames(louvs,2,"clusid")

anno <- fread("processed-data/08_validitycheck_25hdg75svg_louv1/10-25hdg75svg_louv1_annots.txt")
louvs <- merge.data.table(louvs,anno,by="clusid")
louvs <- DataFrame(louvs,row.names=louvs$rn)[colnames(lc3),]

## assign
colLabels(lc3) <- louvs$anno

## rownames to symbols
rownames(lc3) <- rowData(lc3)$gene_name

## subset to human equivalents of these genes. since we pulled down an orthogene table manually already for this data, let's just keep using that for consistency
orthos <- fread("raw-data/ref_data/gProfiler_mmusculus_hsapiens_1-28-2025_4-42-13 PM.csv")

orthos <- orthos[initial_alias %in% toupper(mks.sublc$gene)]
## well, three of the 8 left are mitochondrial genes..
lc3 <- lc3[orthos[!(ortholog_name %in% c("COX17","COX7C","NDUFS8")),ortholog_name],]
```

## tabulate prop of each cluster with counts > 0
```{r}
xprprops <- lapply(rownames(lc3),FUN=function(g){
    ans <- lapply(unique(lc3$label),FUN=function(a){
        subans <- sum(counts(lc3[g,lc3$label==a])>0)/ncol(lc3[g,lc3$label==a])
        return(subans)
    })
    names(ans) <- unique(lc3$label)
    return(ans)
})
names(xprprops) <- rownames(lc3)

anstab <- lapply(xprprops,FUN=function(x){
    y <- cbind(as.numeric(x),names(x))
    y <- as.data.table(y)
    setnames(y,c("prop_nonzero","cluster"))
    return(y)
})
anstab <- rbindlist(anstab,idcol="gene")

# save
fwrite(anstab,"processed-data/Xenium_prioritization/06c-ProportionSpots_per_clus_w_LuskinPixelseq_LCclusPairDiffMks.txt",sep='\t',quote=F,row.names=F)
```

repinf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] spatialLIBD_1.18.0          orthogene_1.12.0           
 [3] BiocParallel_1.40.0         parallelly_1.42.0          
 [5] colorout_1.3-0.2            SpatialExperiment_1.16.0   
 [7] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0
 [9] Biobase_2.66.0              GenomicRanges_1.58.0       
[11] GenomeInfoDb_1.42.1         IRanges_2.40.1             
[13] S4Vectors_0.44.0            BiocGenerics_0.52.0        
[15] MatrixGenerics_1.18.0       matrixStats_1.5.0          
[17] dplyr_1.1.4                 Seurat_5.2.1               
[19] SeuratObject_5.0.2          sp_2.1-4                   
[21] gridExtra_2.3               ggtext_0.1.2               
[23] ggplot2_3.5.1               data.table_1.16.4          
[25] rlang_1.1.4                

loaded via a namespace (and not attached):
  [1] bitops_1.0-9              fs_1.6.5                  spatstat.sparse_3.1-0    
  [4] doParallel_1.0.17         httr_1.4.7                RColorBrewer_1.1-3       
  [7] tools_4.4.2               sctransform_0.4.1         backports_1.5.0          
 [10] DT_0.33                   R6_2.5.1                  lazyeval_0.2.2           
 [13] uwot_0.2.2                withr_3.0.2               progressr_0.15.1         
 [16] cli_3.6.3                 spatstat.explore_3.3-4    fastDummies_1.7.5        
 [19] sass_0.4.9                spatstat.data_3.1-4       ggridges_0.5.6           
 [22] pbapply_1.7-2             Rsamtools_2.22.0          yulab.utils_0.1.8        
 [25] scater_1.34.0             sessioninfo_1.2.2         attempt_0.3.1            
 [28] maps_3.4.2.1              limma_3.62.1              rstudioapi_0.17.1        
 [31] RSQLite_2.3.9             BiocIO_1.16.0             generics_0.1.3           
 [34] gridGraphics_0.5-1        ica_1.0-3                 spatstat.random_3.3-2    
 [37] car_3.1-3                 homologene_1.4.68.19.3.27 Matrix_1.7-1             
 [40] ggbeeswarm_0.7.2          abind_1.4-8               lifecycle_1.0.4          
 [43] edgeR_4.4.1               yaml_2.3.10               carData_3.0-5            
 [46] SparseArray_1.6.0         BiocFileCache_2.14.0      Rtsne_0.17               
 [49] paletteer_1.6.0           grid_4.4.2                blob_1.2.4               
 [52] promises_1.3.2            ExperimentHub_2.14.0      crayon_1.5.3             
 [55] miniUI_0.1.1.1            lattice_0.22-6            beachmat_2.22.0          
 [58] cowplot_1.1.3             KEGGREST_1.46.0           magick_2.8.5             
 [61] pillar_1.10.1             knitr_1.49                rjson_0.2.23             
 [64] future.apply_1.11.3       codetools_0.2-20          glue_1.8.0               
 [67] ggfun_0.1.8               spatstat.univar_3.1-1     vctrs_0.6.5              
 [70] png_0.1-8                 treeio_1.30.0             spam_2.11-1              
 [73] gtable_0.3.6              rematch2_2.1.2            cachem_1.1.0             
 [76] xfun_0.50                 S4Arrays_1.6.0            mime_0.12                
 [79] survival_3.7-0            iterators_1.0.14          fields_16.3              
 [82] statmod_1.5.0             fitdistrplus_1.2-2        ROCR_1.0-11              
 [85] nlme_3.1-166              ggtree_3.14.0             bit64_4.5.2              
 [88] filelock_1.0.3            RcppAnnoy_0.0.22          rprojroot_2.0.4          
 [91] bslib_0.9.0               irlba_2.3.5.1             vipor_0.4.7              
 [94] KernSmooth_2.23-24        colorspace_2.1-1          DBI_1.2.3                
 [97] tidyselect_1.2.1          curl_6.2.0                bit_4.5.0.1              
[100] compiler_4.4.2            BiocNeighbors_2.0.1       xml2_1.3.6               
[103] DelayedArray_0.32.0       plotly_4.10.4             rtracklayer_1.66.0       
[106] scales_1.3.0              lmtest_0.9-40             rappdirs_0.3.3           
[109] stringr_1.5.1             digest_0.6.37             goftest_1.2-3            
[112] spatstat.utils_3.1-2      rmarkdown_2.29            benchmarkmeData_1.0.4    
[115] XVector_0.46.0            htmltools_0.5.8.1         pkgconfig_2.0.3          
[118] dbplyr_2.5.0              fastmap_1.2.0             htmlwidgets_1.6.4        
[121] UCSC.utils_1.2.0          shiny_1.10.0              jquerylib_0.1.4          
[124] farver_2.1.2              zoo_1.8-12                jsonlite_1.8.9           
[127] config_0.3.2              RCurl_1.98-1.16           BiocSingular_1.22.0      
[130] magrittr_2.0.3            Formula_1.2-5             scuttle_1.16.0           
[133] GenomeInfoDbData_1.2.13   ggplotify_0.1.2           dotCall64_1.2            
[136] patchwork_1.3.0           munsell_0.5.1             Rcpp_1.0.14              
[139] ape_5.8                   babelgene_22.9            viridis_0.6.5            
[142] reticulate_1.40.0         stringi_1.8.4             zlibbioc_1.52.0          
[145] MASS_7.3-61               AnnotationHub_3.14.0      plyr_1.8.9               
[148] parallel_4.4.2            listenv_0.9.1             ggrepel_0.9.6            
[151] deldir_2.0-4              Biostrings_2.74.0         splines_4.4.2            
[154] gridtext_0.1.5            tensor_1.5                locfit_1.5-9.10          
[157] igraph_2.1.4              ggpubr_0.6.0              spatstat.geom_3.3-5      
[160] ggsignif_0.6.4            RcppHNSW_0.6.0            reshape2_1.4.4           
[163] ScaledMatrix_1.14.0       XML_3.99-0.17             gprofiler2_0.2.3         
[166] BiocVersion_3.20.0        evaluate_1.0.3            golem_0.5.1              
[169] BiocManager_1.30.25       foreach_1.5.2             httpuv_1.6.15            
[172] grr_0.9.5                 RANN_2.6.2                tidyr_1.3.1              
[175] purrr_1.0.2               polyclip_1.10-7           benchmarkme_1.0.8        
[178] future_1.34.0             scattermore_1.2           rsvd_1.0.5               
[181] broom_1.0.7               xtable_1.8-4              restfulr_0.0.15          
[184] RSpectra_0.16-2           tidytree_0.4.6            rstatix_0.7.2            
[187] later_1.4.1               viridisLite_0.4.2         tibble_3.2.1             
[190] aplot_0.2.3               GenomicAlignments_1.42.0  AnnotationDbi_1.68.0     
[193] memoise_2.0.1             beeswarm_0.4.0            cluster_2.1.6            
[196] shinyWidgets_0.8.7        globals_0.16.3            here_1.0.1               
> sessioninfo::session_info()
─ Session info ─────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.2 (2024-10-31)
 os       macOS Sonoma 14.7.1
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-03-02
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ─────────────────────────────────────────────────────────────────────────
 !  package              * version        date (UTC) lib source
    abind                  1.4-8          2024-09-12 [1] CRAN (R 4.4.1)
    AnnotationDbi          1.68.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    AnnotationHub          3.14.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ape                    5.8            2024-04-11 [1] CRAN (R 4.4.0)
    aplot                  0.2.3          2024-06-17 [1] CRAN (R 4.4.0)
    attempt                0.3.1          2020-05-03 [1] CRAN (R 4.4.0)
    babelgene              22.9           2022-09-29 [1] CRAN (R 4.4.0)
    backports              1.5.0          2024-05-23 [1] CRAN (R 4.4.0)
    beachmat               2.22.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    beeswarm               0.4.0          2021-06-01 [1] CRAN (R 4.4.0)
    benchmarkme            1.0.8          2022-06-12 [1] CRAN (R 4.4.0)
    benchmarkmeData        1.0.4          2020-04-23 [1] CRAN (R 4.4.0)
    Biobase              * 2.66.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocFileCache          2.14.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocIO                 1.16.0         2024-12-12 [1] Github (Bioconductor/BiocIO@8e80624)
    BiocManager            1.30.25        2024-08-28 [1] CRAN (R 4.4.1)
    BiocNeighbors          2.0.1          2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
    BiocParallel         * 1.40.0         2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    BiocSingular           1.22.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocVersion            3.20.0         2024-05-04 [1] Bioconductor 3.20 (R 4.4.0)
    Biostrings             2.74.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    bit                    4.5.0.1        2024-12-03 [1] CRAN (R 4.4.1)
    bit64                  4.5.2          2024-09-22 [1] CRAN (R 4.4.1)
    bitops                 1.0-9          2024-10-03 [1] CRAN (R 4.4.1)
    blob                   1.2.4          2023-03-17 [1] CRAN (R 4.4.0)
    broom                  1.0.7          2024-09-26 [1] CRAN (R 4.4.1)
    bslib                  0.9.0          2025-01-30 [1] CRAN (R 4.4.1)
    cachem                 1.1.0          2024-05-16 [1] CRAN (R 4.4.0)
    car                    3.1-3          2024-09-27 [1] CRAN (R 4.4.1)
    carData                3.0-5          2022-01-06 [1] CRAN (R 4.4.0)
 P  cli                    3.6.3          2024-06-21 [2] CRAN (R 4.4.0)
    cluster                2.1.6          2023-12-01 [1] CRAN (R 4.4.2)
    codetools              0.2-20         2024-03-31 [1] CRAN (R 4.4.2)
    colorout             * 1.3-0.2        2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    colorspace             2.1-1          2024-07-26 [1] CRAN (R 4.4.0)
    config                 0.3.2          2023-08-30 [1] CRAN (R 4.4.0)
    cowplot                1.1.3          2024-01-22 [1] CRAN (R 4.4.0)
    crayon                 1.5.3          2024-06-20 [1] CRAN (R 4.4.0)
    curl                   6.2.0          2025-01-23 [1] CRAN (R 4.4.1)
    data.table           * 1.16.4         2024-12-06 [1] CRAN (R 4.4.1)
    DBI                    1.2.3          2024-06-02 [1] CRAN (R 4.4.0)
    dbplyr                 2.5.0          2024-03-19 [1] CRAN (R 4.4.0)
    DelayedArray           0.32.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    deldir                 2.0-4          2024-02-28 [1] CRAN (R 4.4.0)
    digest                 0.6.37         2024-08-19 [1] CRAN (R 4.4.1)
    doParallel             1.0.17         2022-02-07 [1] CRAN (R 4.4.1)
    dotCall64              1.2            2024-10-04 [1] CRAN (R 4.4.1)
    dplyr                * 1.1.4          2023-11-17 [1] CRAN (R 4.4.0)
    DT                     0.33           2024-04-04 [1] CRAN (R 4.4.0)
    edgeR                  4.4.1          2024-12-02 [1] Bioconductor 3.20 (R 4.4.2)
    evaluate               1.0.3          2025-01-10 [1] CRAN (R 4.4.1)
    ExperimentHub          2.14.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    farver                 2.1.2          2024-05-13 [1] CRAN (R 4.4.0)
    fastDummies            1.7.5          2025-01-20 [1] CRAN (R 4.4.1)
    fastmap                1.2.0          2024-05-15 [1] CRAN (R 4.4.0)
    fields                 16.3           2024-09-30 [1] CRAN (R 4.4.1)
    filelock               1.0.3          2023-12-11 [1] CRAN (R 4.4.0)
    fitdistrplus           1.2-2          2025-01-07 [1] CRAN (R 4.4.1)
    foreach                1.5.2          2022-02-02 [1] CRAN (R 4.4.0)
    Formula                1.2-5          2023-02-24 [1] CRAN (R 4.4.0)
    fs                     1.6.5          2024-10-30 [1] CRAN (R 4.4.1)
    future                 1.34.0         2024-07-29 [1] CRAN (R 4.4.0)
    future.apply           1.11.3         2024-10-27 [1] CRAN (R 4.4.1)
    generics               0.1.3          2022-07-05 [1] CRAN (R 4.4.0)
    GenomeInfoDb         * 1.42.1         2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13         2024-12-12 [1] Bioconductor
    GenomicAlignments      1.42.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    GenomicRanges        * 1.58.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ggbeeswarm             0.7.2          2023-04-29 [1] CRAN (R 4.4.0)
    ggfun                  0.1.8          2024-12-03 [1] CRAN (R 4.4.1)
    ggplot2              * 3.5.1          2024-04-23 [1] CRAN (R 4.4.0)
    ggplotify              0.1.2          2023-08-09 [1] CRAN (R 4.4.0)
    ggpubr                 0.6.0          2023-02-10 [1] CRAN (R 4.4.0)
    ggrepel                0.9.6          2024-09-07 [1] CRAN (R 4.4.1)
    ggridges               0.5.6          2024-01-23 [1] CRAN (R 4.4.0)
    ggsignif               0.6.4          2022-10-13 [1] CRAN (R 4.4.0)
    ggtext               * 0.1.2          2022-09-16 [1] CRAN (R 4.4.0)
    ggtree                 3.14.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    globals                0.16.3         2024-03-08 [1] CRAN (R 4.4.0)
    glue                   1.8.0          2024-09-30 [1] CRAN (R 4.4.1)
    goftest                1.2-3          2021-10-07 [1] CRAN (R 4.4.0)
    golem                  0.5.1          2024-08-27 [1] CRAN (R 4.4.1)
    gprofiler2             0.2.3          2024-02-23 [1] CRAN (R 4.4.0)
    gridExtra            * 2.3            2017-09-09 [1] CRAN (R 4.4.0)
    gridGraphics           0.5-1          2020-12-13 [1] CRAN (R 4.4.0)
    gridtext               0.1.5          2022-09-16 [1] CRAN (R 4.4.0)
    grr                    0.9.5          2016-08-26 [1] CRAN (R 4.4.0)
    gtable                 0.3.6          2024-10-25 [1] CRAN (R 4.4.1)
    here                   1.0.1          2020-12-13 [1] CRAN (R 4.4.0)
    homologene             1.4.68.19.3.27 2019-03-28 [1] CRAN (R 4.4.0)
    htmltools              0.5.8.1        2024-04-04 [1] CRAN (R 4.4.0)
    htmlwidgets            1.6.4          2023-12-06 [1] CRAN (R 4.4.0)
    httpuv                 1.6.15         2024-03-26 [1] CRAN (R 4.4.0)
    httr                   1.4.7          2023-08-15 [1] CRAN (R 4.4.0)
    ica                    1.0-3          2022-07-08 [1] CRAN (R 4.4.0)
    igraph                 2.1.4          2025-01-23 [1] CRAN (R 4.4.1)
    IRanges              * 2.40.1         2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    irlba                  2.3.5.1        2022-10-03 [1] CRAN (R 4.4.0)
    iterators              1.0.14         2022-02-05 [1] CRAN (R 4.4.0)
    jquerylib              0.1.4          2021-04-26 [1] CRAN (R 4.4.0)
    jsonlite               1.8.9          2024-09-20 [1] CRAN (R 4.4.1)
    KEGGREST               1.46.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    KernSmooth             2.23-24        2024-05-17 [1] CRAN (R 4.4.2)
    knitr                  1.49           2024-11-08 [1] CRAN (R 4.4.1)
    later                  1.4.1          2024-11-27 [1] CRAN (R 4.4.1)
    lattice                0.22-6         2024-03-20 [1] CRAN (R 4.4.2)
    lazyeval               0.2.2          2019-03-15 [1] CRAN (R 4.4.0)
    lifecycle              1.0.4          2023-11-07 [1] CRAN (R 4.4.0)
    limma                  3.62.1         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    listenv                0.9.1          2024-01-29 [1] CRAN (R 4.4.0)
    lmtest                 0.9-40         2022-03-21 [1] CRAN (R 4.4.0)
    locfit                 1.5-9.10       2024-06-24 [1] CRAN (R 4.4.0)
    magick                 2.8.5          2024-09-20 [1] CRAN (R 4.4.1)
    magrittr               2.0.3          2022-03-30 [1] CRAN (R 4.4.0)
    maps                   3.4.2.1        2024-11-10 [1] CRAN (R 4.4.1)
    MASS                   7.3-61         2024-06-13 [1] CRAN (R 4.4.2)
    Matrix                 1.7-1          2024-10-18 [1] CRAN (R 4.4.2)
    MatrixGenerics       * 1.18.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    matrixStats          * 1.5.0          2025-01-07 [1] CRAN (R 4.4.1)
    memoise                2.0.1          2021-11-26 [1] CRAN (R 4.4.0)
    mime                   0.12           2021-09-28 [1] CRAN (R 4.4.0)
    miniUI                 0.1.1.1        2018-05-18 [1] CRAN (R 4.4.0)
    munsell                0.5.1          2024-04-01 [1] CRAN (R 4.4.0)
    nlme                   3.1-166        2024-08-14 [1] CRAN (R 4.4.2)
    orthogene            * 1.12.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    paletteer              1.6.0          2024-01-21 [1] CRAN (R 4.4.0)
    parallelly           * 1.42.0         2025-01-30 [1] CRAN (R 4.4.1)
    patchwork              1.3.0          2024-09-16 [1] CRAN (R 4.4.1)
    pbapply                1.7-2          2023-06-27 [1] CRAN (R 4.4.0)
    pillar                 1.10.1         2025-01-07 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3          2019-09-22 [1] CRAN (R 4.4.0)
    plotly                 4.10.4         2024-01-13 [1] CRAN (R 4.4.0)
    plyr                   1.8.9          2023-10-02 [1] CRAN (R 4.4.0)
    png                    0.1-8          2022-11-29 [1] CRAN (R 4.4.0)
    polyclip               1.10-7         2024-07-23 [1] CRAN (R 4.4.0)
    progressr              0.15.1         2024-11-22 [1] CRAN (R 4.4.1)
    promises               1.3.2          2024-11-28 [1] CRAN (R 4.4.1)
    purrr                  1.0.2          2023-08-10 [1] CRAN (R 4.4.0)
    R6                     2.5.1          2021-08-19 [1] CRAN (R 4.4.0)
    RANN                   2.6.2          2024-08-25 [1] CRAN (R 4.4.1)
    rappdirs               0.3.3          2021-01-31 [1] CRAN (R 4.4.0)
    RColorBrewer           1.1-3          2022-04-03 [1] CRAN (R 4.4.0)
    Rcpp                   1.0.14         2025-01-12 [1] CRAN (R 4.4.1)
    RcppAnnoy              0.0.22         2024-01-23 [1] CRAN (R 4.4.0)
    RcppHNSW               0.6.0          2024-02-04 [1] CRAN (R 4.4.0)
    RCurl                  1.98-1.16      2024-07-11 [1] CRAN (R 4.4.0)
    rematch2               2.1.2          2020-05-01 [1] CRAN (R 4.4.0)
    reshape2               1.4.4          2020-04-09 [1] CRAN (R 4.4.0)
    restfulr               0.0.15         2022-06-16 [1] CRAN (R 4.4.0)
    reticulate             1.40.0         2024-11-15 [1] CRAN (R 4.4.1)
    rjson                  0.2.23         2024-09-16 [1] CRAN (R 4.4.1)
 VP rlang                * 1.1.4          2025-01-17 [2] CRAN (R 4.4.1) (on disk 1.1.5)
    rmarkdown              2.29           2024-11-04 [1] CRAN (R 4.4.1)
    ROCR                   1.0-11         2020-05-02 [1] CRAN (R 4.4.0)
    rprojroot              2.0.4          2023-11-05 [1] CRAN (R 4.4.0)
    Rsamtools              2.22.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    RSpectra               0.16-2         2024-07-18 [1] CRAN (R 4.4.0)
    RSQLite                2.3.9          2024-12-03 [1] CRAN (R 4.4.1)
    rstatix                0.7.2          2023-02-01 [1] CRAN (R 4.4.0)
    rstudioapi             0.17.1         2024-10-22 [1] CRAN (R 4.4.1)
    rsvd                   1.0.5          2021-04-16 [1] CRAN (R 4.4.0)
    rtracklayer            1.66.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    Rtsne                  0.17           2023-12-07 [1] CRAN (R 4.4.0)
    S4Arrays               1.6.0          2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    sass                   0.4.9          2024-03-15 [1] CRAN (R 4.4.0)
    ScaledMatrix           1.14.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.3.0          2023-11-28 [1] CRAN (R 4.4.0)
    scater                 1.34.0         2024-10-29 [1] Bioconductor 3.20 (R 4.4.1)
    scattermore            1.2            2023-06-12 [1] CRAN (R 4.4.0)
    sctransform            0.4.1          2023-10-19 [1] CRAN (R 4.4.0)
    scuttle                1.16.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    sessioninfo            1.2.2          2021-12-06 [1] CRAN (R 4.4.0)
    Seurat               * 5.2.1          2025-01-24 [1] CRAN (R 4.4.1)
    SeuratObject         * 5.0.2          2024-05-08 [1] CRAN (R 4.4.0)
    shiny                  1.10.0         2024-12-14 [1] CRAN (R 4.4.1)
    shinyWidgets           0.8.7          2024-09-23 [1] CRAN (R 4.4.1)
    SingleCellExperiment * 1.28.1         2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    sp                   * 2.1-4          2024-04-30 [1] CRAN (R 4.4.0)
    spam                   2.11-1         2025-01-20 [1] CRAN (R 4.4.1)
    SparseArray            1.6.0          2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SpatialExperiment    * 1.16.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    spatialLIBD          * 1.18.0         2024-11-07 [1] Bioconductor 3.20 (R 4.4.1)
    spatstat.data          3.1-4          2024-11-15 [1] CRAN (R 4.4.1)
    spatstat.explore       3.3-4          2025-01-08 [1] CRAN (R 4.4.1)
    spatstat.geom          3.3-5          2025-01-18 [1] CRAN (R 4.4.1)
    spatstat.random        3.3-2          2024-09-18 [1] CRAN (R 4.4.1)
    spatstat.sparse        3.1-0          2024-06-21 [1] CRAN (R 4.4.0)
    spatstat.univar        3.1-1          2024-11-05 [1] CRAN (R 4.4.1)
    spatstat.utils         3.1-2          2025-01-08 [1] CRAN (R 4.4.1)
    statmod                1.5.0          2023-01-06 [1] CRAN (R 4.4.0)
    stringi                1.8.4          2024-05-06 [1] CRAN (R 4.4.0)
    stringr                1.5.1          2023-11-14 [1] CRAN (R 4.4.0)
    SummarizedExperiment * 1.36.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    survival               3.7-0          2024-06-05 [1] CRAN (R 4.4.2)
    tensor                 1.5            2012-05-05 [1] CRAN (R 4.4.0)
    tibble                 3.2.1          2023-03-20 [1] CRAN (R 4.4.0)
    tidyr                  1.3.1          2024-01-24 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1          2024-03-11 [1] CRAN (R 4.4.0)
    tidytree               0.4.6          2023-12-12 [1] CRAN (R 4.4.0)
    treeio                 1.30.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    UCSC.utils             1.2.0          2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    uwot                   0.2.2          2024-04-21 [1] CRAN (R 4.4.0)
    vctrs                  0.6.5          2023-12-01 [1] CRAN (R 4.4.0)
    vipor                  0.4.7          2023-12-18 [1] CRAN (R 4.4.0)
    viridis                0.6.5          2024-01-29 [1] CRAN (R 4.4.0)
    viridisLite            0.4.2          2023-05-02 [1] CRAN (R 4.4.0)
    withr                  3.0.2          2024-10-28 [1] CRAN (R 4.4.1)
    xfun                   0.50           2025-01-07 [1] CRAN (R 4.4.1)
    XML                    3.99-0.17      2024-06-25 [1] CRAN (R 4.3.3)
    xml2                   1.3.6          2023-12-04 [1] CRAN (R 4.4.0)
    xtable                 1.8-4          2019-04-21 [1] CRAN (R 4.4.0)
    XVector                0.46.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    yaml                   2.3.10         2024-07-26 [1] CRAN (R 4.4.0)
    yulab.utils            0.1.8          2024-11-07 [1] CRAN (R 4.4.1)
    zlibbioc               1.52.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    zoo                    1.8-12         2023-04-13 [1] CRAN (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

────────────────────────────────────────────────────────────────────────────────────
