---
title: "07b-Tabulate_xpr_mouse_hTyrLC_DEGs"
author: "Bernie Mulvey"
date: "2025-03-08"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(SpatialExperiment)
library(orthogene)
# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##
```

read in spe
```{r}
lc3 <- readRDS("processed-data/06-countsOnly_noImgData_QCed_SPE_split_to_tissSections.RDS")

## calculate log counts
lc3 <- scater::computeLibraryFactors(lc3)
lc3 <- scater::logNormCounts(lc3)

## read in clustering and annots
louvs <- readRDS("processed-data/07_featureSelection_dimred_harmony_clustering/04-Initial_louvLeid_res0.5-1-2_clusterings.RDS")
louvs <- louvs$HARMONYlmbna_HDG_SVG_2575[,.(rn,snnHARMONYlmbna_HDG_SVG_2575_louv_res1)]
setnames(louvs,2,"clusid")

anno <- fread("processed-data/08_validitycheck_25hdg75svg_louv1/10-25hdg75svg_louv1_annots.txt")
louvs <- merge.data.table(louvs,anno,by="clusid")
louvs <- DataFrame(louvs,row.names=louvs$rn)[colnames(lc3),]

## assign
colLabels(lc3) <- louvs$anno

## rownames to symbols
rownames(lc3) <- rowData(lc3)$gene_name
```

fetch table of htyr-assoc DEGs from LC TRAP analysis for htyr mar2025 resub
```{r}
htyr <- fread("raw-data/ref_data/ianitelli_weinshenker_htyrLCtrapDEGs_mar2025.txt")


## get orthologs
orths <- convert_orthologs(gene_df = htyr,gene_input="mouse_gene",gene_output = "columns",input_species = "mouse",output_species = "human",non121_strategy = "drop_both_species")

## subset DE table and spe to these genes
htyr <- htyr[mouse_gene %in% orths$input_gene]

rownames(lc3) <- rowData(lc3)$gene_name
lc3sub <- lc3[orths$ortholog_gene,]
```

## tabulate prop of each cluster with counts > 0
```{r}
xprprops <- lapply(rownames(lc3sub),FUN=function(g){
    ans <- lapply(unique(lc3sub$label),FUN=function(a){
        subans <- sum(counts(lc3sub[g,lc3sub$label==a])>0)/ncol(lc3sub[g,lc3sub$label==a])
        return(subans)
    })
    names(ans) <- unique(lc3sub$label)
    return(ans)
})
names(xprprops) <- rownames(lc3sub)

anstab <- lapply(xprprops,FUN=function(x){
    y <- cbind(as.numeric(x),names(x))
    y <- as.data.table(y)
    setnames(y,c("prop_nonzero","cluster"))
    return(y)
})
anstab <- rbindlist(anstab,idcol="gene")

# save
fwrite(anstab,"processed-data/Xenium_prioritization/07b-ProportionSpots_per_clus_w_msHtyrLC_fdr20DEGs.txt",sep='\t',quote=F,row.names=F)
```

## rep inf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] orthogene_1.12.0            colorout_1.3-0.2           
 [3] SpatialExperiment_1.16.0    SingleCellExperiment_1.28.1
 [5] SummarizedExperiment_1.36.0 Biobase_2.66.0             
 [7] GenomicRanges_1.58.0        GenomeInfoDb_1.42.1        
 [9] IRanges_2.40.1              S4Vectors_0.44.0           
[11] BiocGenerics_0.52.0         MatrixGenerics_1.18.0      
[13] matrixStats_1.5.0           data.table_1.16.4          
[15] rlang_1.1.4                

loaded via a namespace (and not attached):
 [1] bitops_1.0-9              grr_0.9.5                 gridExtra_2.3            
 [4] magrittr_2.0.3            scater_1.34.0             compiler_4.4.2           
 [7] gprofiler2_0.2.3          vctrs_0.6.5               pkgconfig_2.0.3          
[10] crayon_1.5.3              fastmap_1.2.0             backports_1.5.0          
[13] magick_2.8.5              XVector_0.46.0            scuttle_1.16.0           
[16] rmarkdown_2.29            UCSC.utils_1.2.0          ggbeeswarm_0.7.2         
[19] purrr_1.0.2               xfun_0.50                 zlibbioc_1.52.0          
[22] beachmat_2.22.0           aplot_0.2.3               jsonlite_1.8.9           
[25] DelayedArray_0.32.0       BiocParallel_1.40.0       broom_1.0.7              
[28] irlba_2.3.5.1             parallel_4.4.2            R6_2.5.1                 
[31] car_3.1-3                 Rcpp_1.0.14               knitr_1.49               
[34] Matrix_1.7-1              tidyselect_1.2.1          rstudioapi_0.17.1        
[37] abind_1.4-8               yaml_2.3.10               viridis_0.6.5            
[40] codetools_0.2-20          lattice_0.22-6            tibble_3.2.1             
[43] withr_3.0.2               treeio_1.30.0             evaluate_1.0.3           
[46] gridGraphics_0.5-1        pillar_1.10.1             ggtree_3.14.0            
[49] ggpubr_0.6.0              carData_3.0-5             rsconnect_1.3.3          
[52] plotly_4.10.4             ggfun_0.1.8               generics_0.1.3           
[55] RCurl_1.98-1.16           rprojroot_2.0.4           ggplot2_3.5.1            
[58] munsell_0.5.1             scales_1.3.0              tidytree_0.4.6           
[61] glue_1.8.0                lazyeval_0.2.2            tools_4.4.2              
[64] BiocNeighbors_2.0.1       ScaledMatrix_1.14.0       ggsignif_0.6.4           
[67] babelgene_22.9            fs_1.6.5                  grid_4.4.2               
[70] tidyr_1.3.1               ape_5.8                   colorspace_2.1-1         
[73] patchwork_1.3.0           nlme_3.1-166              GenomeInfoDbData_1.2.13  
[76] homologene_1.4.68.19.3.27 beeswarm_0.4.0            BiocSingular_1.22.0      
[79] vipor_0.4.7               Formula_1.2-5             cli_3.6.3                
[82] rsvd_1.0.5                S4Arrays_1.6.0            viridisLite_0.4.2        
[85] dplyr_1.1.4               gtable_0.3.6              rstatix_0.7.2            
[88] yulab.utils_0.1.8         digest_0.6.37             ggplotify_0.1.2          
[91] SparseArray_1.6.0         ggrepel_0.9.6             htmlwidgets_1.6.4        
[94] farver_2.1.2              rjson_0.2.23              htmltools_0.5.8.1        
[97] lifecycle_1.0.4           httr_1.4.7                here_1.0.1               
> sessioninfo::session_info()
─ Session info ─────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.2 (2024-10-31)
 os       macOS Sonoma 14.7.1
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-03-08
 rstudio  2024.12.1+563 Kousa Dogwood (desktop)
 pandoc   3.2 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ─────────────────────────────────────────────────────────────────────────
 !  package              * version        date (UTC) lib source
    abind                  1.4-8          2024-09-12 [1] CRAN (R 4.4.1)
    ape                    5.8            2024-04-11 [1] CRAN (R 4.4.0)
    aplot                  0.2.3          2024-06-17 [1] CRAN (R 4.4.0)
    babelgene              22.9           2022-09-29 [1] CRAN (R 4.4.0)
    backports              1.5.0          2024-05-23 [1] CRAN (R 4.4.0)
    beachmat               2.22.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    beeswarm               0.4.0          2021-06-01 [1] CRAN (R 4.4.0)
    Biobase              * 2.66.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocNeighbors          2.0.1          2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
    BiocParallel           1.40.0         2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    BiocSingular           1.22.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    bitops                 1.0-9          2024-10-03 [1] CRAN (R 4.4.1)
    broom                  1.0.7          2024-09-26 [1] CRAN (R 4.4.1)
    car                    3.1-3          2024-09-27 [1] CRAN (R 4.4.1)
    carData                3.0-5          2022-01-06 [1] CRAN (R 4.4.0)
 P  cli                    3.6.3          2024-06-21 [2] CRAN (R 4.4.0)
    codetools              0.2-20         2024-03-31 [1] CRAN (R 4.4.2)
    colorout             * 1.3-0.2        2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    colorspace             2.1-1          2024-07-26 [1] CRAN (R 4.4.0)
    crayon                 1.5.3          2024-06-20 [1] CRAN (R 4.4.0)
    data.table           * 1.16.4         2024-12-06 [1] CRAN (R 4.4.1)
    DelayedArray           0.32.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    digest                 0.6.37         2024-08-19 [1] CRAN (R 4.4.1)
    dplyr                  1.1.4          2023-11-17 [1] CRAN (R 4.4.0)
    evaluate               1.0.3          2025-01-10 [1] CRAN (R 4.4.1)
    farver                 2.1.2          2024-05-13 [1] CRAN (R 4.4.0)
    fastmap                1.2.0          2024-05-15 [1] CRAN (R 4.4.0)
    Formula                1.2-5          2023-02-24 [1] CRAN (R 4.4.0)
    fs                     1.6.5          2024-10-30 [1] CRAN (R 4.4.1)
    generics               0.1.3          2022-07-05 [1] CRAN (R 4.4.0)
    GenomeInfoDb         * 1.42.1         2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13         2024-12-12 [1] Bioconductor
    GenomicRanges        * 1.58.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ggbeeswarm             0.7.2          2023-04-29 [1] CRAN (R 4.4.0)
    ggfun                  0.1.8          2024-12-03 [1] CRAN (R 4.4.1)
    ggplot2                3.5.1          2024-04-23 [1] CRAN (R 4.4.0)
    ggplotify              0.1.2          2023-08-09 [1] CRAN (R 4.4.0)
    ggpubr                 0.6.0          2023-02-10 [1] CRAN (R 4.4.0)
    ggrepel                0.9.6          2024-09-07 [1] CRAN (R 4.4.1)
    ggsignif               0.6.4          2022-10-13 [1] CRAN (R 4.4.0)
    ggtree                 3.14.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    glue                   1.8.0          2024-09-30 [1] CRAN (R 4.4.1)
    gprofiler2             0.2.3          2024-02-23 [1] CRAN (R 4.4.0)
    gridExtra              2.3            2017-09-09 [1] CRAN (R 4.4.0)
    gridGraphics           0.5-1          2020-12-13 [1] CRAN (R 4.4.0)
    grr                    0.9.5          2016-08-26 [1] CRAN (R 4.4.0)
    gtable                 0.3.6          2024-10-25 [1] CRAN (R 4.4.1)
    here                   1.0.1          2020-12-13 [1] CRAN (R 4.4.0)
    homologene             1.4.68.19.3.27 2019-03-28 [1] CRAN (R 4.4.0)
    htmltools              0.5.8.1        2024-04-04 [1] CRAN (R 4.4.0)
    htmlwidgets            1.6.4          2023-12-06 [1] CRAN (R 4.4.0)
    httr                   1.4.7          2023-08-15 [1] CRAN (R 4.4.0)
    IRanges              * 2.40.1         2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    irlba                  2.3.5.1        2022-10-03 [1] CRAN (R 4.4.0)
    jsonlite               1.8.9          2024-09-20 [1] CRAN (R 4.4.1)
    knitr                  1.49           2024-11-08 [1] CRAN (R 4.4.1)
    lattice                0.22-6         2024-03-20 [1] CRAN (R 4.4.2)
    lazyeval               0.2.2          2019-03-15 [1] CRAN (R 4.4.0)
    lifecycle              1.0.4          2023-11-07 [1] CRAN (R 4.4.0)
    magick                 2.8.5          2024-09-20 [1] CRAN (R 4.4.1)
    magrittr               2.0.3          2022-03-30 [1] CRAN (R 4.4.0)
    Matrix                 1.7-1          2024-10-18 [1] CRAN (R 4.4.2)
    MatrixGenerics       * 1.18.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    matrixStats          * 1.5.0          2025-01-07 [1] CRAN (R 4.4.1)
    munsell                0.5.1          2024-04-01 [1] CRAN (R 4.4.0)
    nlme                   3.1-166        2024-08-14 [1] CRAN (R 4.4.2)
    orthogene            * 1.12.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    patchwork              1.3.0          2024-09-16 [1] CRAN (R 4.4.1)
    pillar                 1.10.1         2025-01-07 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3          2019-09-22 [1] CRAN (R 4.4.0)
    plotly                 4.10.4         2024-01-13 [1] CRAN (R 4.4.0)
    purrr                  1.0.2          2023-08-10 [1] CRAN (R 4.4.0)
    R6                     2.5.1          2021-08-19 [1] CRAN (R 4.4.0)
    Rcpp                   1.0.14         2025-01-12 [1] CRAN (R 4.4.2)
    RCurl                  1.98-1.16      2024-07-11 [1] CRAN (R 4.4.0)
    rjson                  0.2.23         2024-09-16 [1] CRAN (R 4.4.1)
 VP rlang                * 1.1.4          2025-01-17 [2] CRAN (R 4.4.1) (on disk 1.1.5)
    rmarkdown              2.29           2024-11-04 [1] CRAN (R 4.4.1)
    rprojroot              2.0.4          2023-11-05 [1] CRAN (R 4.4.0)
    rsconnect              1.3.3          2024-11-19 [1] CRAN (R 4.4.1)
    rstatix                0.7.2          2023-02-01 [1] CRAN (R 4.4.0)
    rstudioapi             0.17.1         2024-10-22 [1] CRAN (R 4.4.1)
    rsvd                   1.0.5          2021-04-16 [1] CRAN (R 4.4.0)
    S4Arrays               1.6.0          2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ScaledMatrix           1.14.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.3.0          2023-11-28 [1] CRAN (R 4.4.0)
    scater                 1.34.0         2024-10-29 [1] Bioconductor 3.20 (R 4.4.1)
    scuttle                1.16.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    sessioninfo            1.2.2          2021-12-06 [1] CRAN (R 4.4.0)
    SingleCellExperiment * 1.28.1         2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    SparseArray            1.6.0          2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SpatialExperiment    * 1.16.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SummarizedExperiment * 1.36.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    tibble                 3.2.1          2023-03-20 [1] CRAN (R 4.4.0)
    tidyr                  1.3.1          2024-01-24 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1          2024-03-11 [1] CRAN (R 4.4.0)
    tidytree               0.4.6          2023-12-12 [1] CRAN (R 4.4.0)
    treeio                 1.30.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    UCSC.utils             1.2.0          2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    vctrs                  0.6.5          2023-12-01 [1] CRAN (R 4.4.0)
    vipor                  0.4.7          2023-12-18 [1] CRAN (R 4.4.0)
    viridis                0.6.5          2024-01-29 [1] CRAN (R 4.4.0)
    viridisLite            0.4.2          2023-05-02 [1] CRAN (R 4.4.0)
    withr                  3.0.2          2024-10-28 [1] CRAN (R 4.4.1)
    xfun                   0.50           2025-01-07 [1] CRAN (R 4.4.1)
    XVector                0.46.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    yaml                   2.3.10         2024-07-26 [1] CRAN (R 4.4.0)
    yulab.utils            0.1.8          2024-11-07 [1] CRAN (R 4.4.1)
    zlibbioc               1.52.0         2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

────────────────────────────────────────────────────────────────────────────────────
