---
title: "03-Oligo_Astro_1vothers_markers"
format: html
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(spatialLIBD)
library(SpatialExperiment)
# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```


## load SPE and assign hdg_svg_2575_louv_res1 clusterings as autoannotated (to start with), weber divecha lc datasets, 
```{r}
lc3 <- readRDS("processed-data/06-countsOnly_noImgData_QCed_SPE_split_to_tissSections.RDS")

louvs <- readRDS("processed-data/07_featureSelection_dimred_harmony_clustering/04-Initial_louvLeid_res0.5-1-2_clusterings.RDS")
louvs <- louvs[["HARMONYlmbna_HDG_SVG_2575"]][,.(rn,snnHARMONYlmbna_HDG_SVG_2575_louv_res1)]
setnames(louvs,2,"clusid")

anno <- fread("processed-data/08_validitycheck_25hdg75svg_louv1/10-25hdg75svg_louv1_annots.txt")
anno <- anno[,.(clusid,anno)]

louvs <- merge.data.table(louvs,anno,by="clusid")
louvs <- DataFrame(louvs,row.names=louvs$rn)[colnames(lc3),]

colLabels(lc3) <- louvs$anno

rm(louvs,anno)

## drop questionable samples
lc3 <- lc3[,!(lc3$brnum %in% paste0("Br",c(5517,5276,5712)))]
```

## make APOE columns tolerable as model variables; append genomwide YRI estimates
```{r}
tmpcd <- as.data.table(colData(lc3),keep.rownames=T)
tmpcd[APOE=="E4/ E4",APOE:="E4hom"]
tmpcd[APOE=="E3/ E4",APOE:="E4het"]
tmpcd[APOE=="E2/ E3",APOE:="E2het"]
tmpcd[APOE=="E2/ E2",APOE:="E2hom"]

flares<-fread("raw-data/ancestryAndHaplotypes/FLARE/global_ancestry_estimates.txt")
flares[ID=="Br6119",ID:="Br6119(re-dis)"]
stopifnot(sum(flares$ID %in% unique(tmpcd$brnum))==30)

tmpcd <- merge.data.table(tmpcd,flares[,.(ID,YRI)],by.x="brnum",by.y="ID")
stopifnot(nrow(tmpcd)==ncol(lc3))

tmpcd <- DataFrame(tmpcd,row.names=tmpcd$rn)[colnames(lc3),]
tmpcd$rn <- NULL
colData(lc3) <- tmpcd

## make an oligo subobject
oligs <- grep(unique(lc3$label),pattern="Olig",value=T)
oligs <- oligs[oligs!="Astro_Oligo"]
lc3 <- lc3[,lc3$label %in% oligs]

rm(oligs,tmpcd,flares)
```


## run registration wrapper
```{r}
regwrap <- registration_wrapper(lc3,var_registration="label",var_sample_id = "sample_id",covars=c("Sex","APOE","Age","YRI"),gene_ensembl = "gene_id",gene_name="gene_name",min_ncells=10)

## save a table of one-vs-all markers
regwraptab <- as.data.table(regwrap$enrichment)
regwraptab <- melt.data.table(regwraptab,id.vars=c("ensembl","gene"))
regwraptab[,stat:=gsub(variable,pattern="^(logFC|fdr|p_value|t_stat)_.*$",replacement="\\1")]
regwraptab[,cluster:=gsub(variable,pattern="^(logFC|fdr|p_value|t_stat)_(.*)$",replacement="\\2")]

regwraptab[,variable:=NULL]
regwraptab2 <- dcast(regwraptab,cluster+gene+ensembl~stat)
fwrite(regwraptab2,"processed-data/13_markers_and_registration_final/03-HDGSVG2575louv1_regWrapOut_Oligo_1vOthers_markers.txt",sep="\t",quote=F,row.names=F,col.names=T)
```

repinf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.2

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices
[5] utils     methods   base     

other attached packages:
 [1] BiocParallel_1.40.0        
 [2] parallelly_1.43.0          
 [3] spatialLIBD_1.18.0         
 [4] SpatialExperiment_1.16.0   
 [5] SingleCellExperiment_1.28.1
 [6] SummarizedExperiment_1.36.0
 [7] Biobase_2.66.0             
 [8] GenomicRanges_1.58.0       
 [9] GenomeInfoDb_1.42.3        
[10] IRanges_2.40.1             
[11] S4Vectors_0.44.0           
[12] BiocGenerics_0.52.0        
[13] MatrixGenerics_1.18.1      
[14] matrixStats_1.5.0          
[15] data.table_1.17.4          
[16] rlang_1.1.5                

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3      
  [2] jsonlite_2.0.0          
  [3] magrittr_2.0.3          
  [4] ggbeeswarm_0.7.2        
  [5] magick_2.8.6            
  [6] farver_2.1.2            
  [7] BiocIO_1.16.0           
  [8] fields_16.3.1           
  [9] zlibbioc_1.52.0         
 [10] vctrs_0.6.5             
 [11] memoise_2.0.1           
 [12] config_0.3.2            
 [13] Rsamtools_2.22.0        
 [14] paletteer_1.6.0         
 [15] RCurl_1.98-1.17         
 [16] benchmarkme_1.0.8       
 [17] htmltools_0.5.8.1       
 [18] S4Arrays_1.6.0          
 [19] AnnotationHub_3.14.0    
 [20] curl_6.2.2              
 [21] BiocNeighbors_2.0.1     
 [22] SparseArray_1.6.2       
 [23] sass_0.4.10             
 [24] bslib_0.9.0             
 [25] htmlwidgets_1.6.4       
 [26] plotly_4.10.4           
 [27] cachem_1.1.0            
 [28] GenomicAlignments_1.42.0
 [29] mime_0.13               
 [30] lifecycle_1.0.4         
 [31] iterators_1.0.14        
 [32] pkgconfig_2.0.3         
 [33] rsvd_1.0.5              
 [34] Matrix_1.7-3            
 [35] R6_2.6.1                
 [36] fastmap_1.2.0           
 [37] GenomeInfoDbData_1.2.13 
 [38] shiny_1.10.0            
 [39] digest_0.6.37           
 [40] rematch2_2.1.2          
 [41] AnnotationDbi_1.68.0    
 [42] rprojroot_2.0.4         
 [43] scater_1.34.1           
 [44] irlba_2.3.5.1           
 [45] ExperimentHub_2.14.0    
 [46] RSQLite_2.3.9           
 [47] beachmat_2.22.0         
 [48] filelock_1.0.3          
 [49] httr_1.4.7              
 [50] abind_1.4-8             
 [51] compiler_4.4.3          
 [52] here_1.0.1              
 [53] bit64_4.6.0-1           
 [54] doParallel_1.0.17       
 [55] attempt_0.3.1           
 [56] viridis_0.6.5           
 [57] DBI_1.2.3               
 [58] maps_3.4.3              
 [59] sessioninfo_1.2.3       
 [60] rappdirs_0.3.3          
 [61] DelayedArray_0.32.0     
 [62] rjson_0.2.23            
 [63] tools_4.4.3             
 [64] vipor_0.4.7             
 [65] beeswarm_0.4.0          
 [66] httpuv_1.6.16           
 [67] glue_1.8.0              
 [68] restfulr_0.0.15         
 [69] promises_1.3.2          
 [70] grid_4.4.3              
 [71] generics_0.1.4          
 [72] gtable_0.3.6            
 [73] tidyr_1.3.1             
 [74] ScaledMatrix_1.14.0     
 [75] BiocSingular_1.22.0     
 [76] XVector_0.46.0          
 [77] ggrepel_0.9.6           
 [78] BiocVersion_3.20.0      
 [79] foreach_1.5.2           
 [80] pillar_1.10.2           
 [81] spam_2.11-1             
 [82] limma_3.62.2            
 [83] later_1.4.2             
 [84] benchmarkmeData_1.0.4   
 [85] dplyr_1.1.4             
 [86] BiocFileCache_2.14.0    
 [87] lattice_0.22-6          
 [88] rtracklayer_1.66.0      
 [89] bit_4.6.0               
 [90] tidyselect_1.2.1        
 [91] locfit_1.5-9.12         
 [92] Biostrings_2.74.1       
 [93] scuttle_1.16.0          
 [94] knitr_1.50              
 [95] gridExtra_2.3           
 [96] edgeR_4.4.2             
 [97] xfun_0.52               
 [98] statmod_1.5.0           
 [99] DT_0.33                 
[100] UCSC.utils_1.2.0        
[101] lazyeval_0.2.2          
[102] yaml_2.3.10             
[103] shinyWidgets_0.9.0      
[104] evaluate_1.0.3          
[105] codetools_0.2-20        
[106] tibble_3.2.1            
[107] BiocManager_1.30.25     
[108] cli_3.6.4               
[109] xtable_1.8-4            
[110] jquerylib_0.1.4         
[111] golem_0.5.1             
[112] dichromat_2.0-0.1       
[113] Rcpp_1.0.14             
[114] dbplyr_2.5.0            
[115] png_0.1-8               
[116] XML_3.99-0.18           
[117] parallel_4.4.3          
[118] ggplot2_3.5.2           
[119] blob_1.2.4              
[120] dotCall64_1.2           
[121] bitops_1.0-9            
[122] viridisLite_0.4.2       
[123] scales_1.4.0            
[124] purrr_1.0.4             
[125] crayon_1.5.3            
[126] cowplot_1.1.3           
[127] KEGGREST_1.46.0         
> sessioninfo::session_info()
─ Session info ──────────────────────────
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sequoia 15.3.2
 system   aarch64, darwin20
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-07-28
 pandoc   NA
 quarto   NA

─ Packages ──────────────────────────────
 !  package              * version   date (UTC) lib source
    abind                  1.4-8     2024-09-12 [1] CRAN (R 4.4.1)
    AnnotationDbi          1.68.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    AnnotationHub          3.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    attempt                0.3.1     2020-05-03 [1] CRAN (R 4.4.0)
    beachmat               2.22.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    beeswarm               0.4.0     2021-06-01 [1] CRAN (R 4.4.0)
    benchmarkme            1.0.8     2022-06-12 [1] CRAN (R 4.4.0)
    benchmarkmeData        1.0.4     2020-04-23 [1] CRAN (R 4.4.0)
    Biobase              * 2.66.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocFileCache          2.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocIO                 1.16.0    2024-12-12 [1] Github (Bioconductor/BiocIO@8e80624)
    BiocManager            1.30.25   2024-08-28 [1] CRAN (R 4.4.1)
    BiocNeighbors          2.0.1     2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
    BiocParallel         * 1.40.0    2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    BiocSingular           1.22.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocVersion            3.20.0    2024-05-04 [1] Bioconductor 3.20 (R 4.4.0)
    Biostrings             2.74.1    2024-12-16 [1] Bioconductor 3.20 (R 4.4.2)
    bit                    4.6.0     2025-03-06 [1] CRAN (R 4.4.1)
    bit64                  4.6.0-1   2025-01-16 [1] CRAN (R 4.4.1)
    bitops                 1.0-9     2024-10-03 [1] CRAN (R 4.4.1)
    blob                   1.2.4     2023-03-17 [1] CRAN (R 4.4.0)
    bslib                  0.9.0     2025-01-30 [1] CRAN (R 4.4.1)
    cachem                 1.1.0     2024-05-16 [1] CRAN (R 4.4.0)
 VP cli                    3.6.4     2025-04-23 [2] CRAN (R 4.4.1) (on disk 3.6.5)
    codetools              0.2-20    2024-03-31 [1] CRAN (R 4.4.3)
    config                 0.3.2     2023-08-30 [1] CRAN (R 4.4.0)
    cowplot                1.1.3     2024-01-22 [1] CRAN (R 4.4.0)
    crayon                 1.5.3     2024-06-20 [1] CRAN (R 4.4.0)
    curl                   6.2.2     2025-03-24 [1] CRAN (R 4.4.1)
    data.table           * 1.17.4    2025-05-26 [1] CRAN (R 4.4.3)
    DBI                    1.2.3     2024-06-02 [1] CRAN (R 4.4.0)
    dbplyr                 2.5.0     2024-03-19 [1] CRAN (R 4.4.0)
    DelayedArray           0.32.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    dichromat              2.0-0.1   2022-05-02 [1] CRAN (R 4.4.0)
    digest                 0.6.37    2024-08-19 [1] CRAN (R 4.4.1)
    doParallel             1.0.17    2022-02-07 [1] CRAN (R 4.4.1)
    dotCall64              1.2       2024-10-04 [1] CRAN (R 4.4.1)
    dplyr                  1.1.4     2023-11-17 [1] CRAN (R 4.4.0)
    DT                     0.33      2024-04-04 [1] CRAN (R 4.4.0)
    edgeR                  4.4.2     2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    evaluate               1.0.3     2025-01-10 [1] CRAN (R 4.4.1)
    ExperimentHub          2.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    farver                 2.1.2     2024-05-13 [1] CRAN (R 4.4.0)
    fastmap                1.2.0     2024-05-15 [1] CRAN (R 4.4.0)
    fields                 16.3.1    2025-03-08 [1] CRAN (R 4.4.1)
    filelock               1.0.3     2023-12-11 [1] CRAN (R 4.4.0)
    foreach                1.5.2     2022-02-02 [1] CRAN (R 4.4.0)
    generics               0.1.4     2025-05-09 [1] CRAN (R 4.4.1)
    GenomeInfoDb         * 1.42.3    2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13    2024-12-12 [1] Bioconductor
    GenomicAlignments      1.42.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    GenomicRanges        * 1.58.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ggbeeswarm             0.7.2     2023-04-29 [1] CRAN (R 4.4.0)
    ggplot2                3.5.2     2025-04-09 [1] CRAN (R 4.4.1)
    ggrepel                0.9.6     2024-09-07 [1] CRAN (R 4.4.1)
    glue                   1.8.0     2024-09-30 [1] CRAN (R 4.4.1)
    golem                  0.5.1     2024-08-27 [1] CRAN (R 4.4.1)
    gridExtra              2.3       2017-09-09 [1] CRAN (R 4.4.0)
    gtable                 0.3.6     2024-10-25 [1] CRAN (R 4.4.1)
    here                   1.0.1     2020-12-13 [1] CRAN (R 4.4.0)
    htmltools              0.5.8.1   2024-04-04 [1] CRAN (R 4.4.0)
    htmlwidgets            1.6.4     2023-12-06 [1] CRAN (R 4.4.0)
    httpuv                 1.6.16    2025-04-16 [1] CRAN (R 4.4.1)
    httr                   1.4.7     2023-08-15 [1] CRAN (R 4.4.0)
    IRanges              * 2.40.1    2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    irlba                  2.3.5.1   2022-10-03 [1] CRAN (R 4.4.0)
    iterators              1.0.14    2022-02-05 [1] CRAN (R 4.4.0)
    jquerylib              0.1.4     2021-04-26 [1] CRAN (R 4.4.0)
    jsonlite               2.0.0     2025-03-27 [1] CRAN (R 4.4.1)
    KEGGREST               1.46.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    knitr                  1.50      2025-03-16 [1] CRAN (R 4.4.2)
    later                  1.4.2     2025-04-08 [1] CRAN (R 4.4.1)
    lattice                0.22-6    2024-03-20 [1] CRAN (R 4.4.3)
    lazyeval               0.2.2     2019-03-15 [1] CRAN (R 4.4.0)
    lifecycle              1.0.4     2023-11-07 [1] CRAN (R 4.4.0)
    limma                  3.62.2    2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    locfit                 1.5-9.12  2025-03-05 [1] CRAN (R 4.4.1)
    magick                 2.8.6     2025-03-23 [1] CRAN (R 4.4.1)
    magrittr               2.0.3     2022-03-30 [1] CRAN (R 4.4.0)
    maps                   3.4.3     2025-05-26 [1] CRAN (R 4.4.1)
    Matrix                 1.7-3     2025-03-11 [1] CRAN (R 4.4.1)
    MatrixGenerics       * 1.18.1    2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    matrixStats          * 1.5.0     2025-01-07 [1] CRAN (R 4.4.1)
    memoise                2.0.1     2021-11-26 [1] CRAN (R 4.4.0)
    mime                   0.13      2025-03-17 [1] CRAN (R 4.4.1)
    paletteer              1.6.0     2024-01-21 [1] CRAN (R 4.4.0)
    parallelly           * 1.43.0    2025-03-24 [1] CRAN (R 4.4.1)
    pillar                 1.10.2    2025-04-05 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3     2019-09-22 [1] CRAN (R 4.4.0)
    plotly                 4.10.4    2024-01-13 [1] CRAN (R 4.4.0)
    png                    0.1-8     2022-11-29 [1] CRAN (R 4.4.0)
    promises               1.3.2     2024-11-28 [1] CRAN (R 4.4.1)
    purrr                  1.0.4     2025-02-05 [1] CRAN (R 4.4.1)
    R6                     2.6.1     2025-02-15 [1] CRAN (R 4.4.1)
    rappdirs               0.3.3     2021-01-31 [1] CRAN (R 4.4.0)
    RColorBrewer           1.1-3     2022-04-03 [1] CRAN (R 4.4.0)
    Rcpp                   1.0.14    2025-01-12 [1] CRAN (R 4.4.1)
    RCurl                  1.98-1.17 2025-03-22 [1] CRAN (R 4.4.1)
    rematch2               2.1.2     2020-05-01 [1] CRAN (R 4.4.0)
    restfulr               0.0.15    2022-06-16 [1] CRAN (R 4.4.0)
    rjson                  0.2.23    2024-09-16 [1] CRAN (R 4.4.1)
 VP rlang                * 1.1.5     2025-04-11 [2] CRAN (R 4.4.1) (on disk 1.1.6)
    rprojroot              2.0.4     2023-11-05 [1] CRAN (R 4.4.0)
    Rsamtools              2.22.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    RSQLite                2.3.9     2024-12-03 [1] CRAN (R 4.4.1)
    rsvd                   1.0.5     2021-04-16 [1] CRAN (R 4.4.0)
    rtracklayer            1.66.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Arrays               1.6.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    sass                   0.4.10    2025-04-11 [1] CRAN (R 4.4.1)
    ScaledMatrix           1.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.4.0     2025-04-24 [1] CRAN (R 4.4.1)
    scater                 1.34.1    2025-03-03 [1] Bioconductor 3.20 (R 4.4.2)
    scuttle                1.16.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    sessioninfo            1.2.3     2025-02-05 [1] CRAN (R 4.4.1)
    shiny                  1.10.0    2024-12-14 [1] CRAN (R 4.4.1)
    shinyWidgets           0.9.0     2025-02-21 [1] CRAN (R 4.4.1)
    SingleCellExperiment * 1.28.1    2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    spam                   2.11-1    2025-01-20 [1] CRAN (R 4.4.1)
    SparseArray            1.6.2     2025-02-20 [1] Bioconductor 3.20 (R 4.4.2)
    SpatialExperiment    * 1.16.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    spatialLIBD          * 1.18.0    2024-11-07 [1] Bioconductor 3.20 (R 4.4.3)
    statmod                1.5.0     2023-01-06 [1] CRAN (R 4.4.0)
    SummarizedExperiment * 1.36.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    tibble                 3.2.1     2023-03-20 [1] CRAN (R 4.4.0)
    tidyr                  1.3.1     2024-01-24 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1     2024-03-11 [1] CRAN (R 4.4.0)
    UCSC.utils             1.2.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    vctrs                  0.6.5     2023-12-01 [1] CRAN (R 4.4.0)
    vipor                  0.4.7     2023-12-18 [1] CRAN (R 4.4.0)
    viridis                0.6.5     2024-01-29 [1] CRAN (R 4.4.0)
    viridisLite            0.4.2     2023-05-02 [1] CRAN (R 4.4.0)
    xfun                   0.52      2025-04-02 [1] CRAN (R 4.4.1)
    XML                    3.99-0.18 2025-01-01 [1] CRAN (R 4.4.1)
    xtable                 1.8-4     2019-04-21 [1] CRAN (R 4.4.0)
    XVector                0.46.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    yaml                   2.3.10    2024-07-26 [1] CRAN (R 4.4.0)
    zlibbioc               1.52.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * ── Packages attached to the search path.
 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

─────────────────────────────────────────