---
title: "10-LuskinNMF_LCsubclusts_projectToSPE"
author: "Bernie Mulvey"
date: "2025-02-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(RcppML)
library(Seurat)
library(SingleCellExperiment)
library(SpatialExperiment)
library(Matrix) # otherwise rcppml gives error 'no item called "package:Matrix" on the search list'
library(dplyr)
library(ggplot2)
# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

## load luskin preprint Seurat object with LC neurons and subclusster labels (obtained by request from authors), convert to SCE
```{r}
mslc <- readRDS("/Users/bmulvey/Desktop/KM Lab/Other Groups' Published Data/Luskin Bruchas mouse periLC snseq publication RDSes/All_DBH_or_TH_subclustering.rds")

mslc <- as.SingleCellExperiment(mslc)

# here, the clusters are just NE.1 to NE.11 , no further specific labeling
mslc$ident <- paste0("NE_",mslc$ident)
mslc$ident <- factor(mslc$ident,levels=paste0("NE_",c(1:11)))
```

# cleanup
```{r}
gc(full=T)

setDTthreads(1,restore_after_fork = FALSE)
```

## run NMF using Erik Nelson's final parameter set (100 factors, tol 1e-6, 1k max iters, L1 penalty 0.1, diag = T) https://github.com/LieberInstitute/spatial_hpc/blob/main/snRNAseq_hpc/code/nmf/nmf.R
```{r}
mslc.nmf <-RcppML::nmf(assay(mslc,'logcounts'),
    k=100,
    tol = 1e-06,
    maxit = 1000,
    verbose = T,
    L1 = 0.1,
    seed = 42,
    mask_zeros = FALSE,
    diag = TRUE,
    nonneg = TRUE
)
```

## Verbose output: ##
iter |      tol 
---------------
   1 | 9.32e-01
   2 | 1.32e-01
   3 | 4.84e-02
   4 | 2.69e-02
   5 | 1.71e-02
   6 | 1.22e-02
   7 | 9.53e-03
   8 | 7.45e-03
   9 | 5.22e-03
  10 | 3.27e-03
  11 | 2.33e-03
  12 | 1.93e-03
  13 | 1.73e-03
  14 | 1.37e-03
  15 | 8.86e-04
  16 | 6.71e-04
  17 | 5.58e-04
  18 | 4.73e-04
  19 | 4.15e-04
  20 | 3.82e-04
  21 | 3.41e-04
  22 | 2.85e-04
  23 | 2.42e-04
  24 | 2.11e-04
  25 | 1.95e-04
  26 | 1.91e-04
  27 | 1.92e-04
  28 | 1.81e-04
  29 | 1.55e-04
  30 | 1.18e-04
  31 | 9.67e-05
  32 | 7.57e-05
  33 | 6.18e-05
  34 | 5.38e-05
  35 | 4.73e-05
  36 | 4.18e-05
  37 | 3.75e-05
  38 | 3.38e-05
  39 | 3.19e-05
  40 | 2.99e-05
  41 | 2.90e-05
  42 | 3.07e-05
  43 | 3.37e-05
  44 | 3.70e-05
  45 | 3.93e-05
  46 | 4.35e-05
  47 | 5.05e-05
  48 | 5.70e-05
  49 | 6.21e-05
  50 | 5.91e-05
  51 | 5.39e-05
  52 | 4.38e-05
  53 | 1.98e-05
  54 | 2.21e-05
  55 | 2.83e-05
  56 | 3.72e-05
  57 | 4.89e-05
  58 | 6.24e-05
  59 | 7.41e-05
  60 | 8.13e-05
  61 | 7.62e-05
  62 | 3.97e-05
  63 | 1.37e-05
  64 | 6.71e-06
  65 | 4.51e-06
  66 | 3.79e-06
  67 | 3.32e-06
  68 | 3.00e-06
  69 | 2.77e-06
  70 | 2.62e-06
  71 | 2.56e-06
  72 | 2.52e-06
  73 | 2.56e-06
  74 | 2.66e-06
  75 | 2.89e-06
  76 | 3.17e-06
  77 | 3.65e-06
  78 | 4.37e-06
  79 | 5.43e-06
  80 | 6.92e-06
  81 | 9.00e-06
  82 | 1.18e-05
  83 | 1.58e-05
  84 | 2.11e-05
  85 | 2.79e-05
  86 | 3.60e-05
  87 | 4.42e-05
  88 | 4.99e-05
  89 | 4.32e-05
  90 | 2.49e-05
  91 | 9.65e-06
  92 | 2.71e-06
  93 | 8.84e-07
 
# append cell patterns (h, 100 row x 1324 col matrix) to sce 
```{r}
th <- t(mslc.nmf$h)
colData(mslc)<-cbind(colData(mslc),th)

saveRDS(mslc.nmf,"processed-data/10_subclustering/06a-LuskinMs2024NEcells_nmf100_out.RDS")
saveRDS(mslc,"processed-data/10_subclustering/06b-LuskinMs2024NEcells_sce_w_nmf100.RDS")
```

## now to figure out what factor is what in the source data. see jacqui's code at https://github.com/LieberInstitute/spatial_hpc/blob/main/code/revision/plots_nmf-general-specific.R
```{r}
seed1 = as.matrix(colData(mslc)[,paste0("V",1:100)])
seed1 = seed1>0
d1 = cbind.data.frame(authlabel=as.data.frame(colData(mslc))[,"ident"],
                      seed1) %>% 
  group_by(authlabel) %>% add_tally(name="total") %>%
    group_by(authlabel, total) %>%
  summarise_at(paste0("V",1:100), sum) %>%
  tidyr::pivot_longer(paste0("V",1:100), values_to="n", names_to="nmf") %>%
  mutate(prop=n/total)

seed2 = as.matrix(colData(mslc)[,paste0("V",1:100)])
seed2 = apply(seed2, 2, scale)
d2 = cbind.data.frame(authlabel=as.data.frame(colData(mslc))[,"ident"],
                      seed2) %>% 
  group_by(authlabel) %>%
  summarise_at(paste0("V",1:100), mean) %>% 
  tidyr::pivot_longer(paste0("V",1:100), values_to="scaled.avg", names_to="nmf")

pltdat <- left_join(d1[,c("authlabel","nmf","prop")], 
                   d2[,c("authlabel","nmf","scaled.avg")])

nmford <- c(paste0("V",c(1:4,11,12)),
            paste0("V",c(5:7)),
            paste0("V",c(8:9,13)),
            "V10",
            "V14",
            "V15","V26","V32",
            "V17","V21",
            "V23",
            "V25",
            "V31",
            "V33",
            "V35")

authlabelord <- rev(c("NE_1",
                      "NE_3",
                      "NE_2",
                      "NE_7",
                      "NE_6",
                      "NE_4",
                      "NE_10",
                      "NE_8",
                      "NE_9",
                      "NE_5",
                      "NE_11"))

## make sure no repeats
stopifnot(length(nmford)==length(unique(nmford)))

nmfmisc <- paste0("V",c(1:100))
nmford <- c(nmford,nmfmisc[!(nmfmisc %in% nmford)])

# data table-ize for ease of use -- dplyr code was jacqui's almost verbatim
pltdat <- as.data.table(pltdat)

stopifnot(all(unique(pltdat$authlabel) %in% authlabelord))
pltdat[,authlabel:=factor(authlabel,levels=authlabelord)]
pltdat[,nmf:=factor(nmf,levels=nmford)]


## save plot
png("plots/10_subclustering/06a-msLC_NMF100_scaledAvgs_and_props_by_msCluster.png",height=1000,width=1600)
ggplot(pltdat, aes(x=nmf, y=authlabel, size=prop, color=scaled.avg))+
  geom_count()+
  theme_bw()+
  scale_size(range=c(0,3))+
  scale_color_viridis_c(option="F", direction=-1)+
  labs(x="NMF Factor",y="Luskin Bruchas scSeq Label")+
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=.5,size=10),
        axis.text.y=element_text(size=10),
        axis.title.x=element_text(size=12),
        axis.title.y=element_text(size=12))
dev.off()
```

## clean up
```{r}
rm(d1,d2,seed1,seed2,authlabelord,nmfmisc,pltdat)
gc(full=T)
```


#######################################################################
###### PART 2: Project mouse single cell into LFF SPE LC SPOTS ########
#######################################################################

## load SPE and assign unannotated SUBCLUSTER labels we made from hdg_svg_2575_louv_res1 LC (hvg20, louv res 1 being the highest res we went to)
```{r}
lc3 <- readRDS("processed-data/06-countsOnly_noImgData_QCed_SPE_split_to_tissSections.RDS")

louvs <- readRDS("processed-data/10_subclustering/02b-subclustering_HVG5-10-20_HARMONY_louv_res0.2-1.RDS")
louvs <- louvs$HARMONYlmbna_LC.1_hvg20[,.(rn,HARMONYlmbna_LC.1_hvg20_louv1)]
setnames(louvs,2,"clusid")

## subset SPE to LC spots and assign them subcluster labels
lc3 <- lc3[,colnames(lc3) %in% louvs$rn]

louvs <- DataFrame(louvs,row.names=louvs$rn)[colnames(lc3),]
colLabels(lc3) <- louvs$clusid

## calculate log counts (needed for nmf projection)
lc3 <- scater::computeLibraryFactors(lc3)
lc3 <- scater::logNormCounts(lc3)

rm(louvs)
```


## get everything matched down to 1:1 orthologs between mouse and human. note that, contrary to erik's code cited below, there are NOT rownames on the NMF output w matrix, so we need to keep the pre-filtered rownames of the sce to know what the rows of the w matrix correspond to.
```{r}
nmf.w.rn <- rownames(mslc)

# library(orthogene)
# 
# msg <- as.data.frame(rownames(mslc))
# rownames(msg) <- msg$V1
# gprofiler web service was down--again-- error 301 on queries both from my machine and using JHPCE. manually queried via https://biit.cs.ut.ee/gprofiler_beta/orth
mhorth <- fread("raw-data/ref_data/gProfiler_mmusculus_hsapiens_1-28-2025_4-42-13 PM.csv",header=T)
## drop multimatches
dropg <- mhorth[,.N,by="initial_alias"][N>1,initial_alias]
mhorth <- mhorth[!(initial_alias %in% dropg)]

dropg <- mhorth[,.N,by="ortholog_ensg"][N>1,ortholog_ensg]
mhorth <- mhorth[!(ortholog_ensg %in% dropg)]

## mouse genes come back all caps there so caps the mslc rownames (and w mat rn vector)
rownames(mslc) <- toupper(rownames(mslc))
nmf.w.rn <- toupper(nmf.w.rn)

## subset to genes whose all caps mouse symbol is in mslc and ensg is in lc3
mhorth <- mhorth[initial_alias %in% rownames(mslc) & ortholog_ensg %in% rownames(lc3)]

## subset SCE, SPE, w rowname vector, and w accordingly
nmf.w.keeprows <- which(nmf.w.rn %in% mhorth$initial_alias)
orthw <- mslc.nmf$w[nmf.w.keeprows,]

mslc <- mslc[mhorth$initial_alias,]
lc3 <- lc3[mhorth$ortholog_ensg,]

stopifnot(nrow(orthw)==nrow(mslc))
## swap mouse LC rownames to be the HUMAN ensg ortholog. since we just reordered both objects by paired mouse-human names, we can pass ortholog_ensg directly and have the appropriate ortholog mappings.
# rownames(mslc) <- mhorth$ortholog_ensg

## clean up
rm(tmprn,dropg,nmf.w.keeprows,nmf.w.rn)
gc(full=T)
```

## use rcppml::project (pretty straightforward but erik's script for posterity and the L2 parameter used). Actually, there is no L2 parameter argument here so idk wtf. also, the argument order has changed for project relative to this code, i think. nonetheless:
https://github.com/LieberInstitute/spatial_hpc/blob/main/code/NMF/project_nmf_patterns.R

Note that Erik's code implies there are rownames on the nmf w matrix, which there aren't. so we also need to re-load the full mouse SPE to determine which genes were there to start with (i.e. are the rows of the w matrix) and which remained after ortholog filtering above.
```{r}
set.seed(42)
msprojinto.lff <- RcppML::project(A=logcounts(lc3),w=orthw)#,L2=0.00005)
msprojinto.lff <- t(msprojinto.lff)

stopifnot(nrow(msprojinto.lff)==ncol(lc3))
rownames(msprojinto.lff) <- colnames(lc3)
colnames(msprojinto.lff) <- paste0("V",c(1:100))
saveRDS(msprojinto.lff,"processed-data/10_subclustering/06c-LuskinMs2024allCells_nmf100_projected_into_LFF_LCsubclustsHVG20louv1_mat.RDS")

## append loadings to SPE, only for those factors that were cell type specific in the mouse data, and in the same order for cross-data labeling convenience
nmford <- c(paste0("V",c(1:4,11,12)),
            paste0("V",c(5:7)),
            paste0("V",c(8:9,13)),
            "V10",
            "V14",
            "V15","V26","V32",
            "V17","V21",
            "V23",
            "V25",
            "V31",
            "V33",
            "V35")

colData(lc3) <- cbind(colData(lc3),msprojinto.lff[,nmford])

## make a ms factor- ms cluster name for readability

nmfspeclab <- c(rep("NE1_NE3",6),
                      rep("NE1_NE2_NE3",3),
                      rep("NE_2",3),
                      "NE2_NE6_NE10",
                      "NE_7",
                      rep("NE_6",3),
                      rep("NE_4",2),
                      "NE_10",
                      "NE_8",
                      "NE_9",
                      "NE_5",
                      "NE_11")


stopifnot(length(nmfspeclab)==length(nmford))
usefulname <- paste0(nmford,"_",nmfspeclab)

colnames(colData(lc3))[c(40:ncol(colData(lc3)))] <- usefulname
```

## back to jacqui's plotting code linked above: https://github.com/LieberInstitute/spatial_hpc/blob/main/code/revision/plots_nmf-general-specific.R 
```{r}
seed3 = as.matrix(colData(lc3)[,usefulname])
seed3 = seed3>0
d3 = cbind.data.frame(LFFclust=as.data.frame(colData(lc3))[,"label"],
                      seed3) %>% 
  group_by(LFFclust) %>% add_tally(name="total") %>%
  group_by(LFFclust, total) %>%
  summarise_at(usefulname, sum) %>%
  tidyr::pivot_longer(usefulname, values_to="n", names_to="nmf") %>%
  mutate(prop=n/total)

seed4 = as.matrix(colData(lc3)[,usefulname])
seed4 = apply(seed4, 2, scale)
d4 = cbind.data.frame(LFFclust=as.data.frame(colData(lc3))[,"label"],
                      seed4) %>% 
  group_by(LFFclust) %>%
  summarise_at(usefulname, mean) %>% 
  tidyr::pivot_longer(usefulname, values_to="scaled.avg", names_to="nmf")


pltdat <- left_join(d3[,c("LFFclust","nmf","prop")], 
                    d4[,c("LFFclust","nmf","scaled.avg")])

# to d.t.
pltdat <- as.data.table(pltdat)
pltdat[,nmf:=factor(nmf,usefulname)]

pltdat[,LFFclust:=factor(LFFclust,rev(paste0("X",c(1:9))))]

png("plots/10_subclustering/06b-LuskinMs2024NEcells_cellspecNMFs_into_LFFSPE_scaledAvgs_and_props_by_HVG20louv1-LCsubclust.png",height=750,width=1000)
ggplot(pltdat, aes(x=nmf, y=LFFclust, size=prop, color=scaled.avg))+
  geom_count()+
  theme_bw()+
  scale_size(range=c(0,3))+
  scale_color_viridis_c(option="F", direction=-1)+
  labs(x="Mouse LC NMF Factor and Corresp Cluster",y="LFF LC HVG20-Louv SUBcluster")+
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=.5,size=10),
        axis.text.y=element_text(size=10),
        axis.title.x=element_text(size=12),
        axis.title.y=element_text(size=12))
dev.off()
```

## well THAT's disappointing.

reprod inf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] dplyr_1.1.4                 Matrix_1.7-2               
 [3] Seurat_5.2.1                SeuratObject_5.0.2         
 [5] sp_2.2-0                    RcppML_0.3.7               
 [7] BiocParallel_1.40.0         parallelly_1.42.0          
 [9] colorout_1.3-0.2            escheR_1.6.0               
[11] SpatialExperiment_1.16.0    SingleCellExperiment_1.28.1
[13] SummarizedExperiment_1.36.0 Biobase_2.66.0             
[15] GenomicRanges_1.58.0        GenomeInfoDb_1.42.3        
[17] IRanges_2.40.1              S4Vectors_0.44.0           
[19] BiocGenerics_0.52.0         MatrixGenerics_1.18.1      
[21] matrixStats_1.5.0           gridExtra_2.3              
[23] ggtext_0.1.2                ggplot2_3.5.1              
[25] data.table_1.17.0           rlang_1.1.5                

loaded via a namespace (and not attached):
  [1] RcppAnnoy_0.0.22        splines_4.4.3           later_1.4.1            
  [4] tibble_3.2.1            polyclip_1.10-7         fastDummies_1.7.5      
  [7] lifecycle_1.0.4         rprojroot_2.0.4         globals_0.16.3         
 [10] lattice_0.22-6          MASS_7.3-64             magrittr_2.0.3         
 [13] plotly_4.10.4           rmarkdown_2.29          yaml_2.3.10            
 [16] httpuv_1.6.15           sctransform_0.4.1       spam_2.11-1            
 [19] spatstat.sparse_3.1-0   reticulate_1.41.0.1     cowplot_1.1.3          
 [22] pbapply_1.7-2           RColorBrewer_1.1-3      abind_1.4-8            
 [25] zlibbioc_1.52.0         Rtsne_0.17              purrr_1.0.4            
 [28] GenomeInfoDbData_1.2.13 ggrepel_0.9.6           irlba_2.3.5.1          
 [31] listenv_0.9.1           spatstat.utils_3.1-2    goftest_1.2-3          
 [34] RSpectra_0.16-2         spatstat.random_3.3-2   fitdistrplus_1.2-2     
 [37] codetools_0.2-20        DelayedArray_0.32.0     scuttle_1.16.0         
 [40] xml2_1.3.6              tidyselect_1.2.1        UCSC.utils_1.2.0       
 [43] farver_2.1.2            ScaledMatrix_1.14.0     viridis_0.6.5          
 [46] spatstat.explore_3.3-4  jsonlite_1.9.1          BiocNeighbors_2.0.1    
 [49] progressr_0.15.1        ggridges_0.5.6          survival_3.8-3         
 [52] scater_1.34.1           tools_4.4.3             ica_1.0-3              
 [55] Rcpp_1.0.14             glue_1.8.0              SparseArray_1.6.2      
 [58] xfun_0.51               here_1.0.1              withr_3.0.2            
 [61] fastmap_1.2.0           digest_0.6.37           rsvd_1.0.5             
 [64] R6_2.6.1                mime_0.12               colorspace_2.1-1       
 [67] scattermore_1.2         tensor_1.5              spatstat.data_3.1-6    
 [70] tidyr_1.3.1             generics_0.1.3          httr_1.4.7             
 [73] htmlwidgets_1.6.4       S4Arrays_1.6.0          uwot_0.2.3             
 [76] pkgconfig_2.0.3         gtable_0.3.6            lmtest_0.9-40          
 [79] XVector_0.46.0          htmltools_0.5.8.1       dotCall64_1.2          
 [82] scales_1.3.0            png_0.1-8               spatstat.univar_3.1-2  
 [85] knitr_1.50              rstudioapi_0.17.1       reshape2_1.4.4         
 [88] rjson_0.2.23            nlme_3.1-167            zoo_1.8-13             
 [91] stringr_1.5.1           KernSmooth_2.23-26      parallel_4.4.3         
 [94] miniUI_0.1.1.1          vipor_0.4.7             pillar_1.10.1          
 [97] grid_4.4.3              vctrs_0.6.5             RANN_2.6.2             
[100] promises_1.3.2          BiocSingular_1.22.0     beachmat_2.22.0        
[103] xtable_1.8-4            cluster_2.1.8           beeswarm_0.4.0         
[106] evaluate_1.0.3          magick_2.8.5            cli_3.6.3              
[109] compiler_4.4.3          crayon_1.5.3            future.apply_1.11.3    
[112] labeling_0.4.3          plyr_1.8.9              ggbeeswarm_0.7.2       
[115] stringi_1.8.4           viridisLite_0.4.2       deldir_2.0-4           
[118] munsell_0.5.1           lazyeval_0.2.2          spatstat.geom_3.3-5    
[121] RcppHNSW_0.6.0          patchwork_1.3.0         future_1.34.0          
[124] shiny_1.10.0            ROCR_1.0-11             gridtext_0.1.5         
[127] igraph_2.1.4           
> sessioninfo::session_info()
─ Session info ─────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sonoma 14.7.1
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/New_York
 date     2025-03-18
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)
 quarto   1.4.555 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/quarto

─ Packages ─────────────────────────────────────────────────────────────────────────
 !  package              * version  date (UTC) lib source
    abind                  1.4-8    2024-09-12 [1] CRAN (R 4.4.1)
    beachmat               2.22.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    beeswarm               0.4.0    2021-06-01 [1] CRAN (R 4.4.0)
    Biobase              * 2.66.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocNeighbors          2.0.1    2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
    BiocParallel         * 1.40.0   2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    BiocSingular           1.22.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
 VP cli                    3.6.3    2025-02-13 [2] CRAN (R 4.4.1) (on disk 3.6.4)
    cluster                2.1.8    2024-12-11 [1] CRAN (R 4.4.3)
    codetools              0.2-20   2024-03-31 [1] CRAN (R 4.4.3)
    colorout             * 1.3-0.2  2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    colorspace             2.1-1    2024-07-26 [1] CRAN (R 4.4.0)
    cowplot                1.1.3    2024-01-22 [1] CRAN (R 4.4.0)
    crayon                 1.5.3    2024-06-20 [1] CRAN (R 4.4.0)
    data.table           * 1.17.0   2025-02-22 [1] CRAN (R 4.4.1)
    DelayedArray           0.32.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    deldir                 2.0-4    2024-02-28 [1] CRAN (R 4.4.0)
    digest                 0.6.37   2024-08-19 [1] CRAN (R 4.4.1)
    dotCall64              1.2      2024-10-04 [1] CRAN (R 4.4.1)
    dplyr                * 1.1.4    2023-11-17 [1] CRAN (R 4.4.0)
    escheR               * 1.6.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    evaluate               1.0.3    2025-01-10 [1] CRAN (R 4.4.1)
    farver                 2.1.2    2024-05-13 [1] CRAN (R 4.4.0)
    fastDummies            1.7.5    2025-01-20 [1] CRAN (R 4.4.1)
    fastmap                1.2.0    2024-05-15 [1] CRAN (R 4.4.0)
    fitdistrplus           1.2-2    2025-01-07 [1] CRAN (R 4.4.1)
    future                 1.34.0   2024-07-29 [1] CRAN (R 4.4.0)
    future.apply           1.11.3   2024-10-27 [1] CRAN (R 4.4.1)
    generics               0.1.3    2022-07-05 [1] CRAN (R 4.4.0)
    GenomeInfoDb         * 1.42.3   2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13   2024-12-12 [1] Bioconductor
    GenomicRanges        * 1.58.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ggbeeswarm             0.7.2    2023-04-29 [1] CRAN (R 4.4.0)
    ggplot2              * 3.5.1    2024-04-23 [1] CRAN (R 4.4.0)
    ggrepel                0.9.6    2024-09-07 [1] CRAN (R 4.4.1)
    ggridges               0.5.6    2024-01-23 [1] CRAN (R 4.4.0)
    ggtext               * 0.1.2    2022-09-16 [1] CRAN (R 4.4.0)
    globals                0.16.3   2024-03-08 [1] CRAN (R 4.4.0)
    glue                   1.8.0    2024-09-30 [1] CRAN (R 4.4.1)
    goftest                1.2-3    2021-10-07 [1] CRAN (R 4.4.0)
    gridExtra            * 2.3      2017-09-09 [1] CRAN (R 4.4.0)
    gridtext               0.1.5    2022-09-16 [1] CRAN (R 4.4.0)
    gtable                 0.3.6    2024-10-25 [1] CRAN (R 4.4.1)
    here                   1.0.1    2020-12-13 [1] CRAN (R 4.4.0)
    htmltools              0.5.8.1  2024-04-04 [1] CRAN (R 4.4.0)
    htmlwidgets            1.6.4    2023-12-06 [1] CRAN (R 4.4.0)
    httpuv                 1.6.15   2024-03-26 [1] CRAN (R 4.4.0)
    httr                   1.4.7    2023-08-15 [1] CRAN (R 4.4.0)
    ica                    1.0-3    2022-07-08 [1] CRAN (R 4.4.0)
    igraph                 2.1.4    2025-01-23 [1] CRAN (R 4.4.1)
    IRanges              * 2.40.1   2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    irlba                  2.3.5.1  2022-10-03 [1] CRAN (R 4.4.0)
    jsonlite               1.9.1    2025-03-03 [1] CRAN (R 4.4.1)
    KernSmooth             2.23-26  2025-01-01 [1] CRAN (R 4.4.3)
    knitr                  1.50     2025-03-16 [1] CRAN (R 4.4.2)
    labeling               0.4.3    2023-08-29 [1] CRAN (R 4.4.0)
    later                  1.4.1    2024-11-27 [1] CRAN (R 4.4.1)
    lattice                0.22-6   2024-03-20 [1] CRAN (R 4.4.3)
    lazyeval               0.2.2    2019-03-15 [1] CRAN (R 4.4.0)
    lifecycle              1.0.4    2023-11-07 [1] CRAN (R 4.4.0)
    listenv                0.9.1    2024-01-29 [1] CRAN (R 4.4.0)
    lmtest                 0.9-40   2022-03-21 [1] CRAN (R 4.4.0)
    magick                 2.8.5    2024-09-20 [1] CRAN (R 4.4.1)
    magrittr               2.0.3    2022-03-30 [1] CRAN (R 4.4.0)
    MASS                   7.3-64   2025-01-04 [1] CRAN (R 4.4.3)
    Matrix               * 1.7-2    2025-01-23 [1] CRAN (R 4.4.3)
    MatrixGenerics       * 1.18.1   2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    matrixStats          * 1.5.0    2025-01-07 [1] CRAN (R 4.4.1)
    mime                   0.12     2021-09-28 [1] CRAN (R 4.4.0)
    miniUI                 0.1.1.1  2018-05-18 [1] CRAN (R 4.4.0)
    munsell                0.5.1    2024-04-01 [1] CRAN (R 4.4.0)
    nlme                   3.1-167  2025-01-27 [1] CRAN (R 4.4.3)
    parallelly           * 1.42.0   2025-01-30 [1] CRAN (R 4.4.1)
    patchwork              1.3.0    2024-09-16 [1] CRAN (R 4.4.1)
    pbapply                1.7-2    2023-06-27 [1] CRAN (R 4.4.0)
    pillar                 1.10.1   2025-01-07 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3    2019-09-22 [1] CRAN (R 4.4.0)
    plotly                 4.10.4   2024-01-13 [1] CRAN (R 4.4.0)
    plyr                   1.8.9    2023-10-02 [1] CRAN (R 4.4.0)
    png                    0.1-8    2022-11-29 [1] CRAN (R 4.4.0)
    polyclip               1.10-7   2024-07-23 [1] CRAN (R 4.4.0)
    progressr              0.15.1   2024-11-22 [1] CRAN (R 4.4.1)
    promises               1.3.2    2024-11-28 [1] CRAN (R 4.4.1)
    purrr                  1.0.4    2025-02-05 [1] CRAN (R 4.4.1)
    R6                     2.6.1    2025-02-15 [1] CRAN (R 4.4.1)
    RANN                   2.6.2    2024-08-25 [1] CRAN (R 4.4.1)
    RColorBrewer           1.1-3    2022-04-03 [1] CRAN (R 4.4.0)
    Rcpp                   1.0.14   2025-01-12 [1] CRAN (R 4.4.1)
    RcppAnnoy              0.0.22   2024-01-23 [1] CRAN (R 4.4.0)
    RcppHNSW               0.6.0    2024-02-04 [1] CRAN (R 4.4.0)
    RcppML               * 0.3.7    2021-09-21 [1] CRAN (R 4.4.2)
    reshape2               1.4.4    2020-04-09 [1] CRAN (R 4.4.0)
    reticulate             1.41.0.1 2025-03-09 [1] CRAN (R 4.4.1)
    rjson                  0.2.23   2024-09-16 [1] CRAN (R 4.4.1)
 P  rlang                * 1.1.5    2025-01-17 [2] CRAN (R 4.4.1)
    rmarkdown              2.29     2024-11-04 [1] CRAN (R 4.4.1)
    ROCR                   1.0-11   2020-05-02 [1] CRAN (R 4.4.0)
    rprojroot              2.0.4    2023-11-05 [1] CRAN (R 4.4.0)
    RSpectra               0.16-2   2024-07-18 [1] CRAN (R 4.4.0)
    rstudioapi             0.17.1   2024-10-22 [1] CRAN (R 4.4.1)
    rsvd                   1.0.5    2021-04-16 [1] CRAN (R 4.4.0)
    Rtsne                  0.17     2023-12-07 [1] CRAN (R 4.4.0)
    S4Arrays               1.6.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ScaledMatrix           1.14.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.3.0    2023-11-28 [1] CRAN (R 4.4.0)
    scater                 1.34.1   2025-03-03 [1] Bioconductor 3.20 (R 4.4.2)
    scattermore            1.2      2023-06-12 [1] CRAN (R 4.4.0)
    sctransform            0.4.1    2023-10-19 [1] CRAN (R 4.4.0)
    scuttle                1.16.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    sessioninfo            1.2.3    2025-02-05 [1] CRAN (R 4.4.1)
    Seurat               * 5.2.1    2025-01-24 [1] CRAN (R 4.4.1)
    SeuratObject         * 5.0.2    2024-05-08 [1] CRAN (R 4.4.0)
    shiny                  1.10.0   2024-12-14 [1] CRAN (R 4.4.1)
    SingleCellExperiment * 1.28.1   2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    sp                   * 2.2-0    2025-02-01 [1] CRAN (R 4.4.1)
    spam                   2.11-1   2025-01-20 [1] CRAN (R 4.4.1)
    SparseArray            1.6.2    2025-02-20 [1] Bioconductor 3.20 (R 4.4.2)
    SpatialExperiment    * 1.16.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    spatstat.data          3.1-6    2025-03-17 [1] CRAN (R 4.4.2)
    spatstat.explore       3.3-4    2025-01-08 [1] CRAN (R 4.4.1)
    spatstat.geom          3.3-5    2025-01-18 [1] CRAN (R 4.4.1)
    spatstat.random        3.3-2    2024-09-18 [1] CRAN (R 4.4.1)
    spatstat.sparse        3.1-0    2024-06-21 [1] CRAN (R 4.4.0)
    spatstat.univar        3.1-2    2025-03-05 [1] CRAN (R 4.4.1)
    spatstat.utils         3.1-2    2025-01-08 [1] CRAN (R 4.4.1)
    stringi                1.8.4    2024-05-06 [1] CRAN (R 4.4.0)
    stringr                1.5.1    2023-11-14 [1] CRAN (R 4.4.0)
    SummarizedExperiment * 1.36.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    survival               3.8-3    2024-12-17 [1] CRAN (R 4.4.3)
    tensor                 1.5      2012-05-05 [1] CRAN (R 4.4.0)
    tibble                 3.2.1    2023-03-20 [1] CRAN (R 4.4.0)
    tidyr                  1.3.1    2024-01-24 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1    2024-03-11 [1] CRAN (R 4.4.0)
    UCSC.utils             1.2.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    uwot                   0.2.3    2025-02-24 [1] CRAN (R 4.4.1)
    vctrs                  0.6.5    2023-12-01 [1] CRAN (R 4.4.0)
    vipor                  0.4.7    2023-12-18 [1] CRAN (R 4.4.0)
    viridis                0.6.5    2024-01-29 [1] CRAN (R 4.4.0)
    viridisLite            0.4.2    2023-05-02 [1] CRAN (R 4.4.0)
    withr                  3.0.2    2024-10-28 [1] CRAN (R 4.4.1)
    xfun                   0.51     2025-02-19 [1] CRAN (R 4.4.1)
    xml2                   1.3.6    2023-12-04 [1] CRAN (R 4.4.0)
    xtable                 1.8-4    2019-04-21 [1] CRAN (R 4.4.0)
    XVector                0.46.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    yaml                   2.3.10   2024-07-26 [1] CRAN (R 4.4.0)
    zlibbioc               1.52.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    zoo                    1.8-13   2025-02-22 [1] CRAN (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * ── Packages attached to the search path.
 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

────────────────────────────────────────────────────────────────────────────────────
