---
title: "07-NMF decomp LC for GSEA weights"
author: "Bernie Mulvey"
date: "2025-05-16"
output: html_document
---

```{r}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(SpatialExperiment)
library(RcppML)
library(data.table)
library(Matrix)

```

setup
```{r}
lc3 <- readRDS("processed-data/06-countsOnly_noImgData_QCed_SPE_split_to_tissSections.RDS")
lc3 <- scater::computeLibraryFactors(lc3)
lc3 <- scater::logNormCounts(lc3)
```

## Drop Br5517 (unknown anatomy/orientation), Br5276 (LC has been depleted by QC), Br5712 (unknown anatomy/orientation). Keep 6297; no LC but anatomy is right in vicinity.
```{r}
lc3 <- lc3[,!(lc3$brnum %in% paste0("Br",c(5517,5276,5712)))]
```

### annotate and subset to LC spots
```{r}
clus <- readRDS("processed-data/07_featureSelection_dimred_harmony_clustering/04-Initial_louvLeid_res0.5-1-2_clusterings.RDS")
clus <- clus$HARMONYlmbna_HDG_SVG_2575[,.(rn,snnHARMONYlmbna_HDG_SVG_2575_louv_res1)]
setnames(clus,2,"clusid")
anno <- fread("processed-data/08_validitycheck_25hdg75svg_louv1/10-25hdg75svg_louv1_annots.txt")

clus <- merge.data.table(clus,anno,by="clusid",all.x=T)
clus <- DataFrame(clus,row.names=clus$rn)[colnames(lc3),]

lc3$label <- clus$anno

lc3 <- lc3[,lc3$label=="LC"]

lclct <- logcounts(lc3)
lcrn <- rownames(lclct)
lccn <- colnames(lclct)

# get gene metadata to add symbol info later
rddt <- as.data.table(rowData(lc3)) 
```


## run NMF using Erik Nelson's final parameter set for some of these args (1k max iters, L1 penalty 0.1, diag = T) https://github.com/LieberInstitute/spatial_hpc/blob/main/snRNAseq_hpc/code/nmf/nmf.R

but let's set tol lower for more precision, which we can afford since we only have a few thousands spots here, and contempl8 a granular k
```{r}
setDTthreads(1,restore_after_fork = FALSE)

# let's suppose 10 factors per group that are distinct in each group. 2 sex * 4 apoe * 2 predominant ancestries = 16 groupings * 10 = 160 factors. stringency-ize tol from 10-6 to 10-9.
setRcppMLthreads(11)
lc.nmf <-nmf(lclct,
    k=160,
    tol = 1e-09, 
    maxit = 5000,
    verbose = T,
    L1 = 0.1,
    seed = 42,
    mask_zeros = FALSE,
    diag = TRUE,
    nonneg = TRUE
)

w <- as.data.frame(lc.nmf$w)
rownames(w) <- lcrn
colnames(w) <- paste0("LCnmf",c(1:160))

h <- as.data.frame(lc.nmf$h)
rownames(h) <- paste0("LCnmf",c(1:160))
colnames(h) <- colnames(lclct)

lc.nmf2 <- list(w=w,h=h)
saveRDS(lc.nmf2,"processed-data/10_subclustering/07-lcnmf160_tol1e-9.RDS")
```

### 5/20/25 second run: same parameters, but dropping low/no detection genes
```{r}
check <- hist(log10(rowSums(logcounts(lc3))),breaks=seq(-2,8,0.25))
plot(check)
## looks like the minimum we want is somewhere around log10 1 to log10 2
View(cbind(check$breaks[2:length(check$breaks)],check$counts))
# 0.75-1: 1392; 1-1.25: 1254 
# <10^1.25 is the last major drop-off point, so let's set total gene logcounts 10^1.25 as the cutoff
dev.off()

## drop genes w less than 56 counts
keepg <- rownames(lc3)[rowSums(logcounts(lc3))>=10^1.25]
length(keepg) ## 15170
lc3 <- lc3[keepg,]

## clean up for second run
rm(h,lc.nmf2,w,lc.nmf,lclct,lcrn,lccn,check,keepg,rddt)
gc(full=T)

## get new logcount matrix, rownames colnames and gene info
lclct <- logcounts(lc3)
lcrn <- rownames(lclct)
lccn <- colnames(lclct)

# get gene metadata to add symbol info later
rddt <- as.data.table(rowData(lc3)) 
```

run again
# let's suppose 10 factors per group that are distinct in each group. 2 sex * 4 apoe * 2 predominant ancestries = 16 groupings * 10 = 160 factors. stringency-ize tol from 10-6 to 10-9.
```{r}
setDTthreads(1,restore_after_fork = FALSE)
setRcppMLthreads(11)

lc.nmf <-nmf(lclct,
    k=160,
    tol = 1e-09, 
    maxit = 5000,
    verbose = T,
    L1 = 0.1,
    seed = 42,
    mask_zeros = FALSE,
    diag = TRUE,
    nonneg = TRUE
)

w <- as.data.frame(lc.nmf$w)
rownames(w) <- lcrn
colnames(w) <- paste0("LCnmf",c(1:160))

h <- as.data.frame(lc.nmf$h)
rownames(h) <- paste0("LCnmf",c(1:160))
colnames(h) <- colnames(lclct)

lc.nmf2 <- list(w=w,h=h)
saveRDS(lc.nmf2,"processed-data/10_subclustering/07b-lcnmf160_tol1e-9_rowSumLogctFilt18.RDS")
```

repinfo
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.2

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] Matrix_1.7-3                data.table_1.17.0          
 [3] RcppML_0.3.7                SpatialExperiment_1.16.0   
 [5] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0
 [7] Biobase_2.66.0              GenomicRanges_1.58.0       
 [9] GenomeInfoDb_1.42.3         IRanges_2.40.1             
[11] S4Vectors_0.44.0            BiocGenerics_0.52.0        
[13] MatrixGenerics_1.18.1       matrixStats_1.5.0          
[15] rlang_1.1.5                

loaded via a namespace (and not attached):
 [1] beeswarm_0.4.0          gtable_0.3.6            rjson_0.2.23           
 [4] xfun_0.52               ggplot2_3.5.2           ggrepel_0.9.6          
 [7] lattice_0.22-6          generics_0.1.3          vctrs_0.6.5            
[10] tools_4.4.3             parallel_4.4.3          tibble_3.2.1           
[13] pkgconfig_2.0.3         BiocNeighbors_2.0.1     RColorBrewer_1.1-3     
[16] lifecycle_1.0.4         GenomeInfoDbData_1.2.13 compiler_4.4.3         
[19] farver_2.1.2            codetools_0.2-20        vipor_0.4.7            
[22] htmltools_0.5.8.1       yaml_2.3.10             pillar_1.10.2          
[25] crayon_1.5.3            BiocParallel_1.40.0     DelayedArray_0.32.0    
[28] viridis_0.6.5           magick_2.8.6            abind_1.4-8            
[31] tidyselect_1.2.1        rsvd_1.0.5              digest_0.6.37          
[34] dplyr_1.1.4             BiocSingular_1.22.0     rprojroot_2.0.4        
[37] fastmap_1.2.0           grid_4.4.3              here_1.0.1             
[40] cli_3.6.4               SparseArray_1.6.2       scater_1.34.1          
[43] magrittr_2.0.3          S4Arrays_1.6.0          dichromat_2.0-0.1      
[46] UCSC.utils_1.2.0        scales_1.4.0            ggbeeswarm_0.7.2       
[49] rmarkdown_2.29          XVector_0.46.0          httr_1.4.7             
[52] gridExtra_2.3           ScaledMatrix_1.14.0     beachmat_2.22.0        
[55] evaluate_1.0.3          knitr_1.50              viridisLite_0.4.2      
[58] irlba_2.3.5.1           Rcpp_1.0.14             glue_1.8.0             
[61] scuttle_1.16.0          rstudioapi_0.17.1       jsonlite_2.0.0         
[64] R6_2.6.1                zlibbioc_1.52.0        
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sequoia 15.3.2
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-05-16
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)
 quarto   1.4.555 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/quarto

─ Packages ────────────────────────────────────────────────────────────────────────
 !  package              * version date (UTC) lib source
    abind                  1.4-8   2024-09-12 [1] CRAN (R 4.4.1)
    beachmat               2.22.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    beeswarm               0.4.0   2021-06-01 [1] CRAN (R 4.4.0)
    Biobase              * 2.66.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocNeighbors          2.0.1   2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
    BiocParallel           1.40.0  2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    BiocSingular           1.22.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
 VP cli                    3.6.4   2025-04-23 [2] CRAN (R 4.4.1) (on disk 3.6.5)
    codetools              0.2-20  2024-03-31 [1] CRAN (R 4.4.3)
    crayon                 1.5.3   2024-06-20 [1] CRAN (R 4.4.0)
    data.table           * 1.17.0  2025-02-22 [1] CRAN (R 4.4.1)
    DelayedArray           0.32.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    dichromat              2.0-0.1 2022-05-02 [1] CRAN (R 4.4.0)
    digest                 0.6.37  2024-08-19 [1] CRAN (R 4.4.1)
    dplyr                  1.1.4   2023-11-17 [1] CRAN (R 4.4.0)
    evaluate               1.0.3   2025-01-10 [1] CRAN (R 4.4.1)
    farver                 2.1.2   2024-05-13 [1] CRAN (R 4.4.0)
    fastmap                1.2.0   2024-05-15 [1] CRAN (R 4.4.0)
    generics               0.1.3   2022-07-05 [1] CRAN (R 4.4.0)
    GenomeInfoDb         * 1.42.3  2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13  2024-12-12 [1] Bioconductor
    GenomicRanges        * 1.58.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ggbeeswarm             0.7.2   2023-04-29 [1] CRAN (R 4.4.0)
    ggplot2                3.5.2   2025-04-09 [1] CRAN (R 4.4.1)
    ggrepel                0.9.6   2024-09-07 [1] CRAN (R 4.4.1)
    glue                   1.8.0   2024-09-30 [1] CRAN (R 4.4.1)
    gridExtra              2.3     2017-09-09 [1] CRAN (R 4.4.0)
    gtable                 0.3.6   2024-10-25 [1] CRAN (R 4.4.1)
    here                   1.0.1   2020-12-13 [1] CRAN (R 4.4.0)
    htmltools              0.5.8.1 2024-04-04 [1] CRAN (R 4.4.0)
    httr                   1.4.7   2023-08-15 [1] CRAN (R 4.4.0)
    IRanges              * 2.40.1  2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    irlba                  2.3.5.1 2022-10-03 [1] CRAN (R 4.4.0)
    jsonlite               2.0.0   2025-03-27 [1] CRAN (R 4.4.1)
    knitr                  1.50    2025-03-16 [1] CRAN (R 4.4.2)
    lattice                0.22-6  2024-03-20 [1] CRAN (R 4.4.3)
    lifecycle              1.0.4   2023-11-07 [1] CRAN (R 4.4.0)
    magick                 2.8.6   2025-03-23 [1] CRAN (R 4.4.1)
    magrittr               2.0.3   2022-03-30 [1] CRAN (R 4.4.0)
    Matrix               * 1.7-3   2025-03-11 [1] CRAN (R 4.4.1)
    MatrixGenerics       * 1.18.1  2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    matrixStats          * 1.5.0   2025-01-07 [1] CRAN (R 4.4.1)
    pillar                 1.10.2  2025-04-05 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3   2019-09-22 [1] CRAN (R 4.4.0)
    R6                     2.6.1   2025-02-15 [1] CRAN (R 4.4.1)
    RColorBrewer           1.1-3   2022-04-03 [1] CRAN (R 4.4.0)
    Rcpp                   1.0.14  2025-01-12 [1] CRAN (R 4.4.1)
    RcppML               * 0.3.7   2021-09-21 [1] CRAN (R 4.4.2)
    rjson                  0.2.23  2024-09-16 [1] CRAN (R 4.4.1)
 VP rlang                * 1.1.5   2025-04-11 [2] CRAN (R 4.4.1) (on disk 1.1.6)
    rmarkdown              2.29    2024-11-04 [1] CRAN (R 4.4.1)
    rprojroot              2.0.4   2023-11-05 [1] CRAN (R 4.4.0)
    rstudioapi             0.17.1  2024-10-22 [1] CRAN (R 4.4.1)
    rsvd                   1.0.5   2021-04-16 [1] CRAN (R 4.4.0)
    S4Arrays               1.6.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ScaledMatrix           1.14.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.4.0   2025-04-24 [1] CRAN (R 4.4.1)
    scater                 1.34.1  2025-03-03 [1] Bioconductor 3.20 (R 4.4.2)
    scuttle                1.16.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    sessioninfo            1.2.3   2025-02-05 [1] CRAN (R 4.4.1)
    SingleCellExperiment * 1.28.1  2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    SparseArray            1.6.2   2025-02-20 [1] Bioconductor 3.20 (R 4.4.2)
    SpatialExperiment    * 1.16.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SummarizedExperiment * 1.36.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    tibble                 3.2.1   2023-03-20 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1   2024-03-11 [1] CRAN (R 4.4.0)
    UCSC.utils             1.2.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    vctrs                  0.6.5   2023-12-01 [1] CRAN (R 4.4.0)
    vipor                  0.4.7   2023-12-18 [1] CRAN (R 4.4.0)
    viridis                0.6.5   2024-01-29 [1] CRAN (R 4.4.0)
    viridisLite            0.4.2   2023-05-02 [1] CRAN (R 4.4.0)
    xfun                   0.52    2025-04-02 [1] CRAN (R 4.4.1)
    XVector                0.46.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    yaml                   2.3.10  2024-07-26 [1] CRAN (R 4.4.0)
    zlibbioc               1.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * ── Packages attached to the search path.
 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

───────────────────────────────────────────────────────────────────────────────────
