---
title: "10d-speLCnmfNOfilt_projectToLuskinFullDataset"
author: "Bernie Mulvey"
date: "2025-05-26"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(RcppML)
library(Seurat)
library(SingleCellExperiment)
library(SpatialExperiment)
library(Matrix) # otherwise rcppml gives error 'no item called "package:Matrix" on the search list'
library(dplyr)
library(ggplot2)
# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

load the all cells seurat object, which has the problematic female sample already removed as it was in the manuscript. the object has has unannotated numerical cluster labels, but andrew luskin provided the look-up table between these values and the annotations used in the paper (along with the number of cells that are in each cluster for sanity checking).
```{r}
lca <- readRDS("~/Desktop/KM Lab/Other Groups\' Published Data/Luskin Bruchas mouse periLC snseq publication RDSes/LC_integrated_allcells.rds")

## convert to singlecellexperiment
lca <- as.SingleCellExperiment(lca)

## read lookup table to add real cluster annots. for neuron subtype labels, there is a blank for nonneuron clusters
cluslut <- fread("~/Desktop/KM Lab/Other Groups\' Published Data/Luskin Bruchas mouse periLC snseq publication RDSes/luskin_allcellsRDS_snnres03_cluster_lookuptable_013025.txt")

## the cluster numbers (Index) here start at 1, while in the SCE they start at 0, so subtracted 1 and make character
cluslut[,Index:=paste0("x",as.character(Index-1))]

## there's actually two oligo clusters listed here so add identifying info
cluslut[Index=="x0",cellType_label:="Oligo_1"]
cluslut[Index=="x3",cellType_label:="Oligo_2"]

## the neuronSubtypelabel is actually a 1-1 mapping between the neuron clusters, just with more informative names, so swap those in for cases where cellType_label is not ""
cluslut[neuronSubtype_label!="",cellType_label:=neuronSubtype_label]
## doublecheck
stopifnot(length(unique(cluslut$Index))==length(unique(cluslut$cellType_label)))


## add real cluster annots
tmpcd <- as.data.table(colData(lca),keep.rownames=T)

## mscript annots (corresponding to Index) are in integrated_snn_res.0.3
tmpcd[,integrated_snn_res.0.3:=as.character(integrated_snn_res.0.3)]
tmpcd[,integrated_snn_res.0.3:=paste0("x",integrated_snn_res.0.3)]

stopifnot(all(cluslut$Index %in% tmpcd$integrated_snn_res.0.3))

tmpcd <- merge(tmpcd,cluslut,by.x="integrated_snn_res.0.3",by.y="Index",all.x=T)

## there also isn't an actual sample_id column here (though the sample ids are appended to the cell barcodes), so we can split a sample id column off
tmpcd[,sample_id:=gsub(rn,pattern="^.*(LC.*$)",replacement="\\1")]
# there should be three samples (2 male one female)
stopifnot(length(unique(tmpcd$sample_id))==3)

tmpcd <- DataFrame(tmpcd,row.names=tmpcd$rn)[colnames(lca),]

colData(lca) <- tmpcd
```

## load filtered LC NMF output
```{r}
lcnmf <- readRDS("processed-data/10_subclustering/07-lcnmf160_tol1e-9.RDS")
```

```{r}
# gprofiler web service was down--again-- error 301 on queries both from my machine and using JHPCE. manually queried via https://biit.cs.ut.ee/gprofiler_beta/orth
mhorth <- fread("raw-data/ref_data/gProfiler_mmusculus_hsapiens_1-28-2025_4-42-13 PM.csv",header=T)
## drop multimatches
dropg <- mhorth[,.N,by="initial_alias"][N>1,initial_alias]
mhorth <- mhorth[!(initial_alias %in% dropg)]

dropg <- mhorth[,.N,by="ortholog_ensg"][N>1,ortholog_ensg]
mhorth <- mhorth[!(ortholog_ensg %in% dropg)]

## mouse genes come back all caps there so caps the mslc rownames (and w mat rn vector)
rownames(lca) <- toupper(rownames(lca))

## subset to genes whose all caps mouse symbol is in mslc and ensg is in lc3
mhorth <- mhorth[initial_alias %in% rownames(lca) & ortholog_ensg %in% rownames(lcnmf$w)]

## subset SCE, SPE, w rowname vector, and w accordingly
nmf.w.keeprows <- which(rownames(lcnmf$w) %in% mhorth$ortholog_ensg)
orthw <- lcnmf$w[nmf.w.keeprows,]

lca <- lca[mhorth$initial_alias,]
stopifnot(nrow(orthw)==nrow(lca))
```

## use rcppml::project (pretty straightforward but erik's script for posterity and the L2 parameter used). Actually, there is no L2 parameter argument here so idk wtf. also, the argument order has changed for project relative to this code, i think. nonetheless:
https://github.com/LieberInstitute/spatial_hpc/blob/main/code/NMF/project_nmf_patterns.R

Note that Erik's code implies there are rownames on the nmf w matrix, which there aren't. so we also need to re-load the full mouse SPE to determine which genes were there to start with (i.e. are the rows of the w matrix) and which remained after ortholog filtering above.
```{r}
setDTthreads(1,restore_after_fork = FALSE)

set.seed(42)
lffprojinto.ms <- RcppML::project(A=logcounts(lca),w=orthw)#,L2=0.00005)
lffprojinto.ms <- t(lffprojinto.ms)


stopifnot(nrow(lffprojinto.ms)==ncol(lca))
rownames(lffprojinto.ms) <- colnames(lca)
colnames(lffprojinto.ms) <- paste0("V",c(1:160))
## append to sce
colData(lca) <- cbind(colData(lca),lffprojinto.ms)
colnames(colData(lca))[c(14:ncol(colData(lca)))] <- paste0("LCspeNOfiltNMF_",c(1:160))

usefulname <- colnames(colData(lca))[14:ncol(colData(lca))]

seed3 = as.matrix(colData(lca)[,usefulname])
seed3 = seed3>0
d3 = cbind.data.frame(msclust=as.data.frame(colData(lca))[,"cellType_label"],
                      seed3) %>% 
  group_by(msclust) %>% add_tally(name="total") %>%
  group_by(msclust, total) %>%
  summarise_at(usefulname, sum) %>%
  tidyr::pivot_longer(usefulname, values_to="n", names_to="nmf") %>%
  mutate(prop=n/total)

seed4 = as.matrix(colData(lca)[,usefulname])
seed4 = apply(seed4, 2, scale)
d4 = cbind.data.frame(msclust=as.data.frame(colData(lca))[,"cellType_label"],
                      seed4) %>% 
  group_by(msclust) %>%
  summarise_at(usefulname, mean) %>% 
  tidyr::pivot_longer(usefulname, values_to="scaled.avg", names_to="nmf")


pltdat <- left_join(d3[,c("msclust","nmf","prop")], 
                    d4[,c("msclust","nmf","scaled.avg")])

# to d.t.
pltdat <- as.data.table(pltdat)
pltdat[,nmf:=factor(nmf,usefulname)]

#reorder clusters to put LC neurons first
lcfirst <- c("DBH_neurons","Chat_neurons",paste0("vGAT_neurons_",c(1:6)),"Astrocyte","Microglia")
lcfirst <- c(lcfirst,unique(pltdat[!(msclust %in% lcfirst),msclust]))
pltdat[,msclust:=factor(msclust,rev(lcfirst))]

png("plots/10_subclustering/10d-speLCnmfNOfilt_allfactors_projectToLuskinMainClusts.png",height=1000,width=8000)
ggplot(pltdat, aes(x=nmf, y=msclust, size=prop, color=scaled.avg))+
  geom_count()+
  theme_bw()+
  scale_size(range=c(0,3))+
  scale_color_viridis_c(option="F", direction=-1)+
  labs(x="Spatial LC NMF Factor",y="Luskin Ms Cluster")+
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=.5,size=10),
        axis.text.y=element_text(size=10),
        axis.title.x=element_text(size=12),
        axis.title.y=element_text(size=12))
dev.off()

## also plot for the factors with factor-specific genes that we analyzed in GSEA; this will give us a better idea as to whether those factors are LC neuron-driven or something else.
factspec <- fread("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/06-LCNMF_factors_and_factorSpecGenes_min25specGenes.txt")
gsfacts <- unique(factspec$factor)
gsfacts <- gsub(gsfacts,pattern="LCnmf(.*)$",replacement="\\1")
gsfacts <- paste0("LCspeNOfiltNMF_",gsfacts)

png("plots/10_subclustering/10d-speLCnmfNOfilt_GSEAdFactors_projectToLuskinMainClusts.png",height=1000,width=1500)
ggplot(pltdat[nmf %in% gsfacts], aes(x=nmf, y=msclust, size=prop, color=scaled.avg))+
  geom_count()+
  theme_bw()+
  scale_size(range=c(0,6))+
  scale_color_viridis_c(option="F", direction=-1)+
  labs(x="Spatial LC NMF Factor",y="Luskin Ms Cluster")+
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=.5,size=10),
        axis.text.y=element_text(size=10),
        axis.title.x=element_text(size=12),
        axis.title.y=element_text(size=12))
dev.off()

## well basically all of them have at least a non-negative projection value and are represented in virtually all cells of the LC subclusters. good sign!
```


rep inf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.2

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] colorout_1.3-0.2            ggplot2_3.5.2              
 [3] dplyr_1.1.4                 Matrix_1.7-3               
 [5] SpatialExperiment_1.16.0    SingleCellExperiment_1.28.1
 [7] SummarizedExperiment_1.36.0 Biobase_2.66.0             
 [9] GenomicRanges_1.58.0        GenomeInfoDb_1.42.3        
[11] IRanges_2.40.1              S4Vectors_0.44.0           
[13] BiocGenerics_0.52.0         MatrixGenerics_1.18.1      
[15] matrixStats_1.5.0           Seurat_5.3.0               
[17] SeuratObject_5.1.0          sp_2.2-0                   
[19] RcppML_0.3.7                data.table_1.17.0          
[21] rlang_1.1.5                

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3      rstudioapi_0.17.1       jsonlite_2.0.0         
  [4] magrittr_2.0.3          magick_2.8.6            spatstat.utils_3.1-3   
  [7] farver_2.1.2            rmarkdown_2.29          zlibbioc_1.52.0        
 [10] vctrs_0.6.5             ROCR_1.0-11             spatstat.explore_3.4-2 
 [13] S4Arrays_1.6.0          htmltools_0.5.8.1       SparseArray_1.6.2      
 [16] sctransform_0.4.2       parallelly_1.43.0       KernSmooth_2.23-26     
 [19] htmlwidgets_1.6.4       ica_1.0-3               plyr_1.8.9             
 [22] plotly_4.10.4           zoo_1.8-14              igraph_2.1.4           
 [25] mime_0.13               lifecycle_1.0.4         pkgconfig_2.0.3        
 [28] R6_2.6.1                fastmap_1.2.0           GenomeInfoDbData_1.2.13
 [31] fitdistrplus_1.2-2      future_1.40.0           shiny_1.10.0           
 [34] digest_0.6.37           colorspace_2.1-1        patchwork_1.3.0        
 [37] rprojroot_2.0.4         tensor_1.5              RSpectra_0.16-2        
 [40] irlba_2.3.5.1           labeling_0.4.3          progressr_0.15.1       
 [43] spatstat.sparse_3.1-0   httr_1.4.7              polyclip_1.10-7        
 [46] abind_1.4-8             compiler_4.4.3          here_1.0.1             
 [49] withr_3.0.2             fastDummies_1.7.5       MASS_7.3-65            
 [52] sessioninfo_1.2.3       DelayedArray_0.32.0     rjson_0.2.23           
 [55] tools_4.4.3             lmtest_0.9-40           httpuv_1.6.16          
 [58] future.apply_1.11.3     goftest_1.2-3           glue_1.8.0             
 [61] nlme_3.1-167            promises_1.3.2          grid_4.4.3             
 [64] Rtsne_0.17              cluster_2.1.8.1         reshape2_1.4.4         
 [67] generics_0.1.3          gtable_0.3.6            spatstat.data_3.1-6    
 [70] tidyr_1.3.1             XVector_0.46.0          spatstat.geom_3.3-6    
 [73] RcppAnnoy_0.0.22        ggrepel_0.9.6           RANN_2.6.2             
 [76] pillar_1.10.2           stringr_1.5.1           spam_2.11-1            
 [79] RcppHNSW_0.6.0          later_1.4.2             splines_4.4.3          
 [82] lattice_0.22-6          survival_3.8-3          deldir_2.0-4           
 [85] tidyselect_1.2.1        miniUI_0.1.2            pbapply_1.7-2          
 [88] knitr_1.50              gridExtra_2.3           scattermore_1.2        
 [91] xfun_0.52               stringi_1.8.7           UCSC.utils_1.2.0       
 [94] lazyeval_0.2.2          yaml_2.3.10             evaluate_1.0.3         
 [97] codetools_0.2-20        tibble_3.2.1            cli_3.6.4              
[100] uwot_0.2.3              xtable_1.8-4            reticulate_1.42.0      
[103] dichromat_2.0-0.1       Rcpp_1.0.14             globals_0.17.0         
[106] spatstat.random_3.3-3   png_0.1-8               spatstat.univar_3.1-2  
[109] parallel_4.4.3          dotCall64_1.2           listenv_0.9.1          
[112] viridisLite_0.4.2       scales_1.4.0            ggridges_0.5.6         
[115] crayon_1.5.3            purrr_1.0.4             cowplot_1.1.3          
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sequoia 15.3.2
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-05-25
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)
 quarto   1.4.555 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/quarto

─ Packages ────────────────────────────────────────────────────────────────────────
 !  package              * version date (UTC) lib source
    abind                  1.4-8   2024-09-12 [1] CRAN (R 4.4.1)
    Biobase              * 2.66.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
 VP cli                    3.6.4   2025-04-23 [2] CRAN (R 4.4.1) (on disk 3.6.5)
    cluster                2.1.8.1 2025-03-12 [1] CRAN (R 4.4.1)
    codetools              0.2-20  2024-03-31 [1] CRAN (R 4.4.3)
    colorout             * 1.3-0.2 2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    colorspace             2.1-1   2024-07-26 [1] CRAN (R 4.4.0)
    cowplot                1.1.3   2024-01-22 [1] CRAN (R 4.4.0)
    crayon                 1.5.3   2024-06-20 [1] CRAN (R 4.4.0)
    data.table           * 1.17.0  2025-02-22 [1] CRAN (R 4.4.1)
    DelayedArray           0.32.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    deldir                 2.0-4   2024-02-28 [1] CRAN (R 4.4.0)
    dichromat              2.0-0.1 2022-05-02 [1] CRAN (R 4.4.0)
    digest                 0.6.37  2024-08-19 [1] CRAN (R 4.4.1)
    dotCall64              1.2     2024-10-04 [1] CRAN (R 4.4.1)
    dplyr                * 1.1.4   2023-11-17 [1] CRAN (R 4.4.0)
    evaluate               1.0.3   2025-01-10 [1] CRAN (R 4.4.1)
    farver                 2.1.2   2024-05-13 [1] CRAN (R 4.4.0)
    fastDummies            1.7.5   2025-01-20 [1] CRAN (R 4.4.1)
    fastmap                1.2.0   2024-05-15 [1] CRAN (R 4.4.0)
    fitdistrplus           1.2-2   2025-01-07 [1] CRAN (R 4.4.1)
    future                 1.40.0  2025-04-10 [1] CRAN (R 4.4.1)
    future.apply           1.11.3  2024-10-27 [1] CRAN (R 4.4.1)
    generics               0.1.3   2022-07-05 [1] CRAN (R 4.4.0)
    GenomeInfoDb         * 1.42.3  2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13  2024-12-12 [1] Bioconductor
    GenomicRanges        * 1.58.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ggplot2              * 3.5.2   2025-04-09 [1] CRAN (R 4.4.1)
    ggrepel                0.9.6   2024-09-07 [1] CRAN (R 4.4.1)
    ggridges               0.5.6   2024-01-23 [1] CRAN (R 4.4.0)
    globals                0.17.0  2025-04-16 [1] CRAN (R 4.4.1)
    glue                   1.8.0   2024-09-30 [1] CRAN (R 4.4.1)
    goftest                1.2-3   2021-10-07 [1] CRAN (R 4.4.0)
    gridExtra              2.3     2017-09-09 [1] CRAN (R 4.4.0)
    gtable                 0.3.6   2024-10-25 [1] CRAN (R 4.4.1)
    here                   1.0.1   2020-12-13 [1] CRAN (R 4.4.0)
    htmltools              0.5.8.1 2024-04-04 [1] CRAN (R 4.4.0)
    htmlwidgets            1.6.4   2023-12-06 [1] CRAN (R 4.4.0)
    httpuv                 1.6.16  2025-04-16 [1] CRAN (R 4.4.1)
    httr                   1.4.7   2023-08-15 [1] CRAN (R 4.4.0)
    ica                    1.0-3   2022-07-08 [1] CRAN (R 4.4.0)
    igraph                 2.1.4   2025-01-23 [1] CRAN (R 4.4.1)
    IRanges              * 2.40.1  2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    irlba                  2.3.5.1 2022-10-03 [1] CRAN (R 4.4.0)
    jsonlite               2.0.0   2025-03-27 [1] CRAN (R 4.4.1)
    KernSmooth             2.23-26 2025-01-01 [1] CRAN (R 4.4.3)
    knitr                  1.50    2025-03-16 [1] CRAN (R 4.4.2)
    labeling               0.4.3   2023-08-29 [1] CRAN (R 4.4.0)
    later                  1.4.2   2025-04-08 [1] CRAN (R 4.4.1)
    lattice                0.22-6  2024-03-20 [1] CRAN (R 4.4.3)
    lazyeval               0.2.2   2019-03-15 [1] CRAN (R 4.4.0)
    lifecycle              1.0.4   2023-11-07 [1] CRAN (R 4.4.0)
    listenv                0.9.1   2024-01-29 [1] CRAN (R 4.4.0)
    lmtest                 0.9-40  2022-03-21 [1] CRAN (R 4.4.0)
    magick                 2.8.6   2025-03-23 [1] CRAN (R 4.4.1)
    magrittr               2.0.3   2022-03-30 [1] CRAN (R 4.4.0)
    MASS                   7.3-65  2025-02-28 [1] CRAN (R 4.4.1)
    Matrix               * 1.7-3   2025-03-11 [1] CRAN (R 4.4.1)
    MatrixGenerics       * 1.18.1  2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    matrixStats          * 1.5.0   2025-01-07 [1] CRAN (R 4.4.1)
    mime                   0.13    2025-03-17 [1] CRAN (R 4.4.1)
    miniUI                 0.1.2   2025-04-17 [1] CRAN (R 4.4.1)
    nlme                   3.1-167 2025-01-27 [1] CRAN (R 4.4.3)
    parallelly             1.43.0  2025-03-24 [1] CRAN (R 4.4.1)
    patchwork              1.3.0   2024-09-16 [1] CRAN (R 4.4.1)
    pbapply                1.7-2   2023-06-27 [1] CRAN (R 4.4.0)
    pillar                 1.10.2  2025-04-05 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3   2019-09-22 [1] CRAN (R 4.4.0)
    plotly                 4.10.4  2024-01-13 [1] CRAN (R 4.4.0)
    plyr                   1.8.9   2023-10-02 [1] CRAN (R 4.4.0)
    png                    0.1-8   2022-11-29 [1] CRAN (R 4.4.0)
    polyclip               1.10-7  2024-07-23 [1] CRAN (R 4.4.0)
    progressr              0.15.1  2024-11-22 [1] CRAN (R 4.4.1)
    promises               1.3.2   2024-11-28 [1] CRAN (R 4.4.1)
    purrr                  1.0.4   2025-02-05 [1] CRAN (R 4.4.1)
    R6                     2.6.1   2025-02-15 [1] CRAN (R 4.4.1)
    RANN                   2.6.2   2024-08-25 [1] CRAN (R 4.4.1)
    RColorBrewer           1.1-3   2022-04-03 [1] CRAN (R 4.4.0)
    Rcpp                   1.0.14  2025-01-12 [1] CRAN (R 4.4.1)
    RcppAnnoy              0.0.22  2024-01-23 [1] CRAN (R 4.4.0)
    RcppHNSW               0.6.0   2024-02-04 [1] CRAN (R 4.4.0)
    RcppML               * 0.3.7   2021-09-21 [1] CRAN (R 4.4.2)
    reshape2               1.4.4   2020-04-09 [1] CRAN (R 4.4.0)
    reticulate             1.42.0  2025-03-25 [1] CRAN (R 4.4.1)
    rjson                  0.2.23  2024-09-16 [1] CRAN (R 4.4.1)
 VP rlang                * 1.1.5   2025-04-11 [2] CRAN (R 4.4.1) (on disk 1.1.6)
    rmarkdown              2.29    2024-11-04 [1] CRAN (R 4.4.1)
    ROCR                   1.0-11  2020-05-02 [1] CRAN (R 4.4.0)
    rprojroot              2.0.4   2023-11-05 [1] CRAN (R 4.4.0)
    RSpectra               0.16-2  2024-07-18 [1] CRAN (R 4.4.0)
    rstudioapi             0.17.1  2024-10-22 [1] CRAN (R 4.4.1)
    Rtsne                  0.17    2023-12-07 [1] CRAN (R 4.4.0)
    S4Arrays               1.6.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.4.0   2025-04-24 [1] CRAN (R 4.4.1)
    scattermore            1.2     2023-06-12 [1] CRAN (R 4.4.0)
    sctransform            0.4.2   2025-04-30 [1] CRAN (R 4.4.1)
    sessioninfo            1.2.3   2025-02-05 [1] CRAN (R 4.4.1)
    Seurat               * 5.3.0   2025-04-23 [1] CRAN (R 4.4.1)
    SeuratObject         * 5.1.0   2025-04-22 [1] CRAN (R 4.4.1)
    shiny                  1.10.0  2024-12-14 [1] CRAN (R 4.4.1)
    SingleCellExperiment * 1.28.1  2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    sp                   * 2.2-0   2025-02-01 [1] CRAN (R 4.4.1)
    spam                   2.11-1  2025-01-20 [1] CRAN (R 4.4.1)
    SparseArray            1.6.2   2025-02-20 [1] Bioconductor 3.20 (R 4.4.2)
    SpatialExperiment    * 1.16.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    spatstat.data          3.1-6   2025-03-17 [1] CRAN (R 4.4.2)
    spatstat.explore       3.4-2   2025-03-21 [1] CRAN (R 4.4.1)
    spatstat.geom          3.3-6   2025-03-18 [1] CRAN (R 4.4.1)
    spatstat.random        3.3-3   2025-03-19 [1] CRAN (R 4.4.1)
    spatstat.sparse        3.1-0   2024-06-21 [1] CRAN (R 4.4.0)
    spatstat.univar        3.1-2   2025-03-05 [1] CRAN (R 4.4.1)
    spatstat.utils         3.1-3   2025-03-15 [1] CRAN (R 4.4.1)
    stringi                1.8.7   2025-03-27 [1] CRAN (R 4.4.1)
    stringr                1.5.1   2023-11-14 [1] CRAN (R 4.4.0)
    SummarizedExperiment * 1.36.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    survival               3.8-3   2024-12-17 [1] CRAN (R 4.4.3)
    tensor                 1.5     2012-05-05 [1] CRAN (R 4.4.0)
    tibble                 3.2.1   2023-03-20 [1] CRAN (R 4.4.0)
    tidyr                  1.3.1   2024-01-24 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1   2024-03-11 [1] CRAN (R 4.4.0)
    UCSC.utils             1.2.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    uwot                   0.2.3   2025-02-24 [1] CRAN (R 4.4.1)
    vctrs                  0.6.5   2023-12-01 [1] CRAN (R 4.4.0)
    viridisLite            0.4.2   2023-05-02 [1] CRAN (R 4.4.0)
    withr                  3.0.2   2024-10-28 [1] CRAN (R 4.4.1)
    xfun                   0.52    2025-04-02 [1] CRAN (R 4.4.1)
    xtable                 1.8-4   2019-04-21 [1] CRAN (R 4.4.0)
    XVector                0.46.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    yaml                   2.3.10  2024-07-26 [1] CRAN (R 4.4.0)
    zlibbioc               1.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    zoo                    1.8-14  2025-04-10 [1] CRAN (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * ── Packages attached to the search path.
 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

───────────────────────────────────────────────────────────────────────────────────
