---
title: "10e-speLCnmfPrefilt_projectToWeberDivecha_snSeq"
author: "Bernie Mulvey"
date: "2025-06-03"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(RcppML)
library(SpatialExperiment)
library(Matrix) # otherwise rcppml gives error 'no item called "package:Matrix" on the search list'
library(dplyr)
library(ggplot2)
library(WeberDivechaLCdata)
# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

```{r}
lcsn <- WeberDivechaLCdata_singleNucleus()

colnames(colData(lcsn))
## several clusterings
# label-30, numerical
# label_inhibitory (14 subclusters)
# label_merged (10 clusters annotated)
# supervised_NE,unsupervised_NE,unsupervised_5HT (all T/F)
# we'll use label_merged here. label_inhibitory could be interesting for NMF on the other neuronal clusters some other time though.
```

## load filtered LC NMF output
```{r}
lcnmf <- readRDS("processed-data/10_subclustering/07b-lcnmf160_tol1e-9_rowSumLogctFilt18.RDS")
```

```{r}
## drop multimatches
dropg <- as.data.table(table(rownames(lcsn)))[N>1] # none
keepg <- rownames(lcnmf$w)[rownames(lcnmf$w) %in% rownames(lcsn)] 

## subset SCE and w accordingly
lcsn <- lcsn[keepg,]
lcnmf$w <- lcnmf$w[keepg,]

stopifnot(nrow(lcnmf$w)==nrow(lcsn))
```

## use rcppml::project (pretty straightforward but erik's script for posterity and the L2 parameter used). Actually, there is no L2 parameter argument here so idk wtf. also, the argument order has changed for project relative to this code, i think. nonetheless:
https://github.com/LieberInstitute/spatial_hpc/blob/main/code/NMF/project_nmf_patterns.R

Note that Erik's code implies there are rownames on the nmf w matrix, which there aren't. so we also need to re-load the full mouse SPE to determine which genes were there to start with (i.e. are the rows of the w matrix) and which remained after ortholog filtering above.
```{r}
setDTthreads(1,restore_after_fork = FALSE)

set.seed(42)
lffprojinto.webdiv <- RcppML::project(A=logcounts(lcsn),w=lcnmf$w)#,L2=0.00005)
lffprojinto.webdiv <- t(lffprojinto.webdiv)


stopifnot(nrow(lffprojinto.webdiv)==ncol(lcsn))
rownames(lffprojinto.webdiv) <- colnames(lcsn)
colnames(lffprojinto.webdiv) <- paste0("V",c(1:160))
## append to sce
colData(lcsn) <- cbind(colData(lcsn),lffprojinto.webdiv)
colnames(colData(lcsn))[c(17:ncol(colData(lcsn)))] <- paste0("LCspeFiltNMF_",c(1:160))

usefulname <- colnames(colData(lcsn))[17:ncol(colData(lcsn))]

seed3 = as.matrix(colData(lcsn)[,usefulname])
seed3 = seed3>0
d3 = cbind.data.frame(webdivClus=as.data.frame(colData(lcsn))[,"label_merged"],
                      seed3) %>% 
  group_by(webdivClus) %>% add_tally(name="total") %>%
  group_by(webdivClus, total) %>%
  summarise_at(usefulname, sum) %>%
  tidyr::pivot_longer(usefulname, values_to="n", names_to="nmf") %>%
  mutate(prop=n/total)

seed4 = as.matrix(colData(lcsn)[,usefulname])
seed4 = apply(seed4, 2, scale)
d4 = cbind.data.frame(webdivClus=as.data.frame(colData(lcsn))[,"label_merged"],
                      seed4) %>% 
  group_by(webdivClus) %>%
  summarise_at(usefulname, mean) %>% 
  tidyr::pivot_longer(usefulname, values_to="scaled.avg", names_to="nmf")


pltdat <- left_join(d3[,c("webdivClus","nmf","prop")], 
                    d4[,c("webdivClus","nmf","scaled.avg")])

# to d.t.
pltdat <- as.data.table(pltdat)
pltdat[,nmf:=factor(nmf,usefulname)]

#reorder clusters to put LC neurons first
lcfirst <- c("NE","5HT","excitatory","inhibitory","neurons_ambiguous","astrocytes","macrophages_microglia","OPCs","oligodendrocytes","endothelial_mural")
pltdat[,webdivClus:=factor(webdivClus,rev(lcfirst))]

png("plots/10_subclustering/10e-speLCnmfPrefilt_allfactors_projectToWeberDivechaSnseq.png",height=1000,width=8000)
ggplot(pltdat, aes(x=nmf, y=webdivClus, size=prop, color=scaled.avg))+
  geom_count()+
  theme_bw()+
  scale_size(range=c(0,3))+
  scale_color_viridis_c(option="F", direction=-1)+
  labs(x="Spatial LC NMF Factor",y="Luskin Ms Cluster")+
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=.5,size=10),
        axis.text.y=element_text(size=10),
        axis.title.x=element_text(size=12),
        axis.title.y=element_text(size=12))
dev.off()

## also plot for the factors with factor-specific genes that we analyzed in GSEA; this will give us a better idea as to whether those factors are LC neuron-driven or something else.
factspec <- fread("processed-data/12_DEanalyses_removedsampsAndFinalNMseg/06b-LCNMFLogCt18Filt_factors_and_factorSpecGenes_min25specGenes.txt")
gsfacts <- unique(factspec$factor)
gsfacts <- gsub(gsfacts,pattern="LCnmf(.*)$",replacement="\\1")
gsfacts <- paste0("LCspeFiltNMF_",gsfacts)

png("plots/10_subclustering/10e-speLCnmfPrefilt_GSEAdFactors_projectToWeberDivechaSnseq.png",height=1000,width=1500)
ggplot(pltdat[nmf %in% gsfacts], aes(x=nmf, y=webdivClus, size=prop, color=scaled.avg))+
  geom_count()+
  theme_bw()+
  scale_size(range=c(0,6))+
  scale_color_viridis_c(option="F", direction=-1)+
  labs(x="Spatial LC NMF Factor",y="Luskin Ms Cluster")+
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=.5,size=10),
        axis.text.y=element_text(size=10),
        axis.title.x=element_text(size=12),
        axis.title.y=element_text(size=12))
dev.off()

## nice, that clears up a couple that are astrocytic, vascular and/or microglial
```


rep inf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.2

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] WeberDivechaLCdata_1.8.0    ExperimentHub_2.14.0        AnnotationHub_3.14.0       
 [4] BiocFileCache_2.14.0        dbplyr_2.5.0                colorout_1.3-0.2           
 [7] ggplot2_3.5.2               dplyr_1.1.4                 Matrix_1.7-3               
[10] SpatialExperiment_1.16.0    SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0
[13] Biobase_2.66.0              GenomicRanges_1.58.0        GenomeInfoDb_1.42.3        
[16] IRanges_2.40.1              S4Vectors_0.44.0            BiocGenerics_0.52.0        
[19] MatrixGenerics_1.18.1       matrixStats_1.5.0           RcppML_0.3.7               
[22] data.table_1.17.0           rlang_1.1.5                

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.1        viridisLite_0.4.2       farver_2.1.2           
 [4] blob_1.2.4              filelock_1.0.3          Biostrings_2.74.1      
 [7] fastmap_1.2.0           digest_0.6.37           mime_0.13              
[10] lifecycle_1.0.4         KEGGREST_1.46.0         RSQLite_2.3.9          
[13] magrittr_2.0.3          compiler_4.4.3          tools_4.4.3            
[16] yaml_2.3.10             knitr_1.50              labeling_0.4.3         
[19] S4Arrays_1.6.0          bit_4.6.0               curl_6.2.2             
[22] here_1.0.1              DelayedArray_0.32.0     RColorBrewer_1.1-3     
[25] abind_1.4-8             withr_3.0.2             purrr_1.0.4            
[28] grid_4.4.3              scales_1.4.0            dichromat_2.0-0.1      
[31] cli_3.6.4               rmarkdown_2.29          crayon_1.5.3           
[34] generics_0.1.3          rstudioapi_0.17.1       httr_1.4.7             
[37] rjson_0.2.23            DBI_1.2.3               cachem_1.1.0           
[40] zlibbioc_1.52.0         AnnotationDbi_1.68.0    BiocManager_1.30.25    
[43] XVector_0.46.0          vctrs_0.6.5             jsonlite_2.0.0         
[46] bit64_4.6.0-1           magick_2.8.6            tidyr_1.3.1            
[49] glue_1.8.0              gtable_0.3.6            BiocVersion_3.20.0     
[52] UCSC.utils_1.2.0        tibble_3.2.1            pillar_1.10.2          
[55] rappdirs_0.3.3          htmltools_0.5.8.1       GenomeInfoDbData_1.2.13
[58] R6_2.6.1                rprojroot_2.0.4         evaluate_1.0.3         
[61] lattice_0.22-6          png_0.1-8               memoise_2.0.1          
[64] Rcpp_1.0.14             SparseArray_1.6.2       xfun_0.52              
[67] pkgconfig_2.0.3        
> sessioninfo::session_info()
─ Session info ───────────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sequoia 15.3.2
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-06-03
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)
 quarto   1.4.555 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/quarto

─ Packages ───────────────────────────────────────────────────────────────────────────────
 !  package              * version date (UTC) lib source
    abind                  1.4-8   2024-09-12 [1] CRAN (R 4.4.1)
    AnnotationDbi          1.68.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    AnnotationHub        * 3.14.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    Biobase              * 2.66.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocFileCache        * 2.14.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocManager            1.30.25 2024-08-28 [1] CRAN (R 4.4.1)
    BiocVersion            3.20.0  2024-05-04 [1] Bioconductor 3.20 (R 4.4.0)
    Biostrings             2.74.1  2024-12-16 [1] Bioconductor 3.20 (R 4.4.2)
    bit                    4.6.0   2025-03-06 [1] CRAN (R 4.4.1)
    bit64                  4.6.0-1 2025-01-16 [1] CRAN (R 4.4.1)
    blob                   1.2.4   2023-03-17 [1] CRAN (R 4.4.0)
    cachem                 1.1.0   2024-05-16 [1] CRAN (R 4.4.0)
 VP cli                    3.6.4   2025-04-23 [2] CRAN (R 4.4.1) (on disk 3.6.5)
    colorout             * 1.3-0.2 2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    crayon                 1.5.3   2024-06-20 [1] CRAN (R 4.4.0)
    curl                   6.2.2   2025-03-24 [1] CRAN (R 4.4.1)
    data.table           * 1.17.0  2025-02-22 [1] CRAN (R 4.4.1)
    DBI                    1.2.3   2024-06-02 [1] CRAN (R 4.4.0)
    dbplyr               * 2.5.0   2024-03-19 [1] CRAN (R 4.4.0)
    DelayedArray           0.32.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    dichromat              2.0-0.1 2022-05-02 [1] CRAN (R 4.4.0)
    digest                 0.6.37  2024-08-19 [1] CRAN (R 4.4.1)
    dplyr                * 1.1.4   2023-11-17 [1] CRAN (R 4.4.0)
    evaluate               1.0.3   2025-01-10 [1] CRAN (R 4.4.1)
    ExperimentHub        * 2.14.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    farver                 2.1.2   2024-05-13 [1] CRAN (R 4.4.0)
    fastmap                1.2.0   2024-05-15 [1] CRAN (R 4.4.0)
    filelock               1.0.3   2023-12-11 [1] CRAN (R 4.4.0)
    generics               0.1.3   2022-07-05 [1] CRAN (R 4.4.0)
    GenomeInfoDb         * 1.42.3  2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13  2024-12-12 [1] Bioconductor
    GenomicRanges        * 1.58.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ggplot2              * 3.5.2   2025-04-09 [1] CRAN (R 4.4.1)
    glue                   1.8.0   2024-09-30 [1] CRAN (R 4.4.1)
    gtable                 0.3.6   2024-10-25 [1] CRAN (R 4.4.1)
    here                   1.0.1   2020-12-13 [1] CRAN (R 4.4.0)
    htmltools              0.5.8.1 2024-04-04 [1] CRAN (R 4.4.0)
    httr                   1.4.7   2023-08-15 [1] CRAN (R 4.4.0)
    IRanges              * 2.40.1  2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    jsonlite               2.0.0   2025-03-27 [1] CRAN (R 4.4.1)
    KEGGREST               1.46.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    knitr                  1.50    2025-03-16 [1] CRAN (R 4.4.2)
    labeling               0.4.3   2023-08-29 [1] CRAN (R 4.4.0)
    lattice                0.22-6  2024-03-20 [1] CRAN (R 4.4.3)
    lifecycle              1.0.4   2023-11-07 [1] CRAN (R 4.4.0)
    magick                 2.8.6   2025-03-23 [1] CRAN (R 4.4.1)
    magrittr               2.0.3   2022-03-30 [1] CRAN (R 4.4.0)
    Matrix               * 1.7-3   2025-03-11 [1] CRAN (R 4.4.1)
    MatrixGenerics       * 1.18.1  2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    matrixStats          * 1.5.0   2025-01-07 [1] CRAN (R 4.4.1)
    memoise                2.0.1   2021-11-26 [1] CRAN (R 4.4.0)
    mime                   0.13    2025-03-17 [1] CRAN (R 4.4.1)
    pillar                 1.10.2  2025-04-05 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3   2019-09-22 [1] CRAN (R 4.4.0)
    png                    0.1-8   2022-11-29 [1] CRAN (R 4.4.0)
    purrr                  1.0.4   2025-02-05 [1] CRAN (R 4.4.1)
    R6                     2.6.1   2025-02-15 [1] CRAN (R 4.4.1)
    rappdirs               0.3.3   2021-01-31 [1] CRAN (R 4.4.0)
    RColorBrewer           1.1-3   2022-04-03 [1] CRAN (R 4.4.0)
    Rcpp                   1.0.14  2025-01-12 [1] CRAN (R 4.4.1)
    RcppML               * 0.3.7   2021-09-21 [1] CRAN (R 4.4.2)
    rjson                  0.2.23  2024-09-16 [1] CRAN (R 4.4.1)
 VP rlang                * 1.1.5   2025-04-11 [2] CRAN (R 4.4.1) (on disk 1.1.6)
    rmarkdown              2.29    2024-11-04 [1] CRAN (R 4.4.1)
    rprojroot              2.0.4   2023-11-05 [1] CRAN (R 4.4.0)
    RSQLite                2.3.9   2024-12-03 [1] CRAN (R 4.4.1)
    rstudioapi             0.17.1  2024-10-22 [1] CRAN (R 4.4.1)
    S4Arrays               1.6.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.4.0   2025-04-24 [1] CRAN (R 4.4.1)
    sessioninfo            1.2.3   2025-02-05 [1] CRAN (R 4.4.1)
    SingleCellExperiment * 1.28.1  2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    SparseArray            1.6.2   2025-02-20 [1] Bioconductor 3.20 (R 4.4.2)
    SpatialExperiment    * 1.16.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SummarizedExperiment * 1.36.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    tibble                 3.2.1   2023-03-20 [1] CRAN (R 4.4.0)
    tidyr                  1.3.1   2024-01-24 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1   2024-03-11 [1] CRAN (R 4.4.0)
    UCSC.utils             1.2.0   2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    vctrs                  0.6.5   2023-12-01 [1] CRAN (R 4.4.0)
    viridisLite            0.4.2   2023-05-02 [1] CRAN (R 4.4.0)
    WeberDivechaLCdata   * 1.8.0   2024-10-31 [1] Bioconductor 3.20 (R 4.4.1)
    withr                  3.0.2   2024-10-28 [1] CRAN (R 4.4.1)
    xfun                   0.52    2025-04-02 [1] CRAN (R 4.4.1)
    XVector                0.46.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    yaml                   2.3.10  2024-07-26 [1] CRAN (R 4.4.0)
    zlibbioc               1.52.0  2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * ── Packages attached to the search path.
 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

──────────────────────────────────────────────────────────────────────────────────────────
